{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene and sample analysis\n",
    "\n",
    "This notebook performs by gene and sample precomputations needed for the regression of H3K36me3 and nucleosomes (Figure 4 and Supplementary Figure S6).\n",
    "\n",
    "---\n",
    "\n",
    "## Output\n",
    "\n",
    "Two tabulated files *germinal_gene_dependent_results.txt* and *germinal_sample_dependent_results.txt*).\n",
    "\n",
    "One of the files file contains information at the gene level: number of bp in exons and introns, mean replication time, number of synonymous and non-synonymous variants, and sumed probabilities of exon, intron, synonymous and non-synonymous mutations to occur.\n",
    "\n",
    "The other file contains information at sample level: number of observed mutations in exons and introns and number of  observed synonymous and non-synonymous mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from os import path\n",
    "\n",
    "import bgdata\n",
    "import tabix  # package pytabix\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "from collections import Counter\n",
    "from bgreference import hg19"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input\n",
    "\n",
    "Files in the **data** directory.\n",
    "\n",
    "- *middle_exons_coords*: file with the coordinates of the middle exons\n",
    "- *consequence_type_file*: tabix file with the consequence types of exonic mutations\n",
    "- *replication_time_path*: tab separated file containing information about replication time (processed but not used in our analysis unlike Frigola et al. 2017).\n",
    "\n",
    "### Other inputs\n",
    "\n",
    "- mutations_folder: base directory where the files with the mutations are located\n",
    "- tumor_type: id of the tumor (germinal in our case)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "consequence_type_file = 'data/consequence/consequence_ranking.tsv.bgz'\n",
    "\n",
    "mutations_folder = 'non_provided_data/mutations/'\n",
    "coordinates_folder = 'data/coordinates/'\n",
    "replication_time_path = 'non_provided_data/replication_timing.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions\n",
    "\n",
    "All the functions needed for this notebook are coded below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def remove_version_ensembl(x, colname):\n",
    "    \"\"\"\n",
    "    Remove version from the ensembl ID\n",
    "\n",
    "    Args:\n",
    "        x: dataframe row\n",
    "        colname: column name\n",
    "\n",
    "    Returns:\n",
    "        str. Text before a '.'\n",
    "\n",
    "    \"\"\"\n",
    "    ensembl_id = x[colname]\n",
    "    ensembl_id_only = ensembl_id.split('.')[0]\n",
    "\n",
    "    return ensembl_id_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def synonymous_or_not(x):\n",
    "    \"\"\"\n",
    "    Classifies consequence types between synonymous and non synonymous\n",
    "    using the RANK column\n",
    "\n",
    "    Args:\n",
    "        x: dataframe row\n",
    "\n",
    "    Returns:\n",
    "        str. *synonymous* or *non_synonymous*\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    my_rank = x['RANK']\n",
    "\n",
    "    if my_rank < 15:\n",
    "        my_consequence = 'non_synonymous'\n",
    "\n",
    "    else:\n",
    "        my_consequence = 'synonymous'\n",
    "\n",
    "    return my_consequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_rep_time_data(replication_time_path, whole_gene_coords):\n",
    "    \"\"\"\n",
    "    Compute the mean replication time of each gene (using the replication time of each of regions)\n",
    "\n",
    "    Args:\n",
    "        replication_time_path (str): path to a tabulated file with 3 columns (chomosome, position and replication time)\n",
    "        whole_gene_coords (:class:`~pandas.DataFrame`): table with a genomic coordinates table\n",
    "\n",
    "    Returns:\n",
    "        :class:`~pandas.DataFrame`.\n",
    "\n",
    "    \"\"\"\n",
    "    replication_time_df = pd.read_csv(replication_time_path, sep='\\t', header=None)\n",
    "    replication_time_df[0] = 'chr' + replication_time_df[0].astype(str)\n",
    "\n",
    "    replication_time_df.columns = ['chr', 'position', 'rep_time']\n",
    "    replication_time_df['position'] = replication_time_df['position'].astype(np.int64)\n",
    "\n",
    "    replication_time_df['start'] = replication_time_df['position'] - 1\n",
    "    replication_time_df_tointersect = replication_time_df[['chr', 'start', 'position']]\n",
    "\n",
    "    replication_time_bed = pybedtools.BedTool.from_dataframe(replication_time_df_tointersect)\n",
    "    whole_gene_coords_bed = pybedtools.BedTool.from_dataframe(whole_gene_coords)\n",
    "\n",
    "    rep_time_gene_bed = whole_gene_coords_bed.intersect(replication_time_bed)\n",
    "    rep_time_gene_df = pd.read_table(rep_time_gene_bed.fn, names=['chr', 'start', 'position', 'gene'])\n",
    "    rep_time_gene_df = rep_time_gene_df[['chr', 'position', 'gene']]\n",
    "\n",
    "    final_gene_rep_time_df = pd.merge(rep_time_gene_df, replication_time_df, on=['chr', 'position'])\n",
    "    final_gene_rep_time_df = final_gene_rep_time_df[['chr', 'position', 'gene', 'rep_time']]\n",
    "    sub_final_gene_rep_time_df = final_gene_rep_time_df[['gene', 'rep_time']]\n",
    "\n",
    "    rep_time_mean_per_gene_df = sub_final_gene_rep_time_df.groupby('gene').mean()\n",
    "    rep_time_mean_per_gene_df = rep_time_mean_per_gene_df.sort_values(by='rep_time', ascending=False)\n",
    "    rep_time_mean_per_gene_df = rep_time_mean_per_gene_df[np.isfinite(rep_time_mean_per_gene_df['rep_time'])]\n",
    "\n",
    "    whole_gene_coords_bed.delete_temporary_history(ask=False)\n",
    "    replication_time_bed.delete_temporary_history(ask=False)\n",
    "    rep_time_gene_bed.delete_temporary_history(ask=False)\n",
    "\n",
    "    return rep_time_mean_per_gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def precompute_counts(my_genes_list, exons_coords_symbol_df, introns_coords_symbol_df, nucleotides):\n",
    "    \"\"\"\n",
    "    For each gene, count all trinucleotides seen. Set a counter where the key is the\n",
    "    combination of the reference triplet and each of the possible alterations\n",
    "    (taken from the remaining nuclotides).\n",
    "    As the count is only done for the reference, each combination of reference triplet\n",
    "    and possible alteration has the same count as the reference triplet.\n",
    "\n",
    "    .. note::\n",
    "\n",
    "       The first and last nucleotide of each exon or intron are do not contribute to the counts.\n",
    "\n",
    "\n",
    "    Args:\n",
    "        my_genes_list (list): list of gene Ensembl identifiers\n",
    "        exons_coords_symbol_df (:class:`~pandas.DataFrame`): table with the genomic coordinates of the exons\n",
    "        introns_coords_symbol_df (:class:`~pandas.DataFrame`): table with the genomic coordinates of the introns\n",
    "        nucleotides (set): nucleotides to consider (typically A, C, G and T)\n",
    "\n",
    "    Returns:\n",
    "        dict. For each gene, the counts on exons and introns are retrieved.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    counters_dict = dict()\n",
    "\n",
    "    for my_gene in my_genes_list:\n",
    "\n",
    "        introns_counter = Counter()\n",
    "        exons_counter = Counter()\n",
    "\n",
    "        # Exons\n",
    "        try:\n",
    "            exons_gene = exons_coords_symbol_df[exons_coords_symbol_df['ensembl'] == my_gene]\n",
    "            exons_gene_lol = exons_gene.values.tolist()\n",
    "\n",
    "            for my_exon in exons_gene_lol:\n",
    "\n",
    "                my_chr = my_exon[0][3:]\n",
    "                my_start = int(my_exon[1])\n",
    "                my_end = int(my_exon[2])\n",
    "\n",
    "                n_bases = my_end - my_start\n",
    "\n",
    "                my_exon_bases = hg19(my_chr, my_start+1, n_bases)\n",
    "\n",
    "                for i in range(len(my_exon_bases)):\n",
    "\n",
    "                    my_relative_position = i + my_start\n",
    "\n",
    "                    signature = hg19(my_chr, my_relative_position, 3)\n",
    "\n",
    "                    my_ref_base = signature[1].upper()\n",
    "\n",
    "                    my_alt_bases = nucleotides - set(my_ref_base)\n",
    "\n",
    "                    for alt_base in my_alt_bases:\n",
    "\n",
    "                        tri_ref = signature[0] + str(my_ref_base) + signature[-1]\n",
    "                        tri_alt = signature[0] + str(alt_base) + signature[-1]\n",
    "\n",
    "                        my_key = tuple([tri_ref, tri_alt])\n",
    "\n",
    "                        exons_counter[my_key] += 1\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"problem with exons from gene: \", my_gene)\n",
    "\n",
    "        # Introns\n",
    "        try:\n",
    "            introns_gene = introns_coords_symbol_df[introns_coords_symbol_df['ensembl'] == my_gene]\n",
    "            introns_gene_lol = introns_gene.values.tolist()\n",
    "\n",
    "            for my_intron in introns_gene_lol:\n",
    "\n",
    "                my_chr = my_intron[0][3:]\n",
    "                my_start = int(my_intron[1])\n",
    "                my_end = int(my_intron[2])\n",
    "\n",
    "                n_bases = my_end - my_start\n",
    "\n",
    "                my_intron_bases = hg19(my_chr, my_start+1, n_bases)\n",
    "\n",
    "                for i in range(len(my_intron_bases)):\n",
    "\n",
    "                    my_relative_position = i + my_start\n",
    "\n",
    "                    signature = hg19(my_chr, my_relative_position, 3)\n",
    "\n",
    "                    my_ref_base = signature[1].upper()\n",
    "\n",
    "                    my_alt_bases = nucleotides - set(my_ref_base)\n",
    "\n",
    "                    for alt_base in my_alt_bases:\n",
    "\n",
    "                        tri_ref = signature[0] + str(my_ref_base) + signature[-1]\n",
    "                        tri_alt = signature[0] + str(alt_base) + signature[-1]\n",
    "\n",
    "                        my_key = tuple([tri_ref, tri_alt])\n",
    "\n",
    "                        introns_counter[my_key] += 1\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"problem with introns from gene: \", my_gene)\n",
    "\n",
    "        counters_dict[(my_gene, 'exons_count')] = exons_counter\n",
    "        counters_dict[(my_gene, 'introns_count')] = introns_counter\n",
    "\n",
    "    return counters_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_muts_per_gene(mutations_df, exons_coords_df, introns_coords_df):\n",
    "    \"\"\"\n",
    "    Intersects the mutations and the coordinates file getting mutations in introns and exons\n",
    "\n",
    "    Args:\n",
    "        mutations_df (:class:`~pandas.DataFrame`): table with the mutations\n",
    "        exons_coords_df (:class:`~pandas.DataFrame`): table with the genomic coordinates of the exons\n",
    "        introns_coords_df (:class:`~pandas.DataFrame`): table with the genomic coordinates of the introns\n",
    "\n",
    "    Returns:\n",
    "        :obj:`tuple` of :class:`~pandas.DataFrame`. One DataFrame with the mutations that fall in any of the exons\n",
    "        and the other with the ones falling in introns.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    sub_mutations_df = mutations_df[['CHROMOSOME', 'POSITION', 'SAMPLE', 'ID']].copy()\n",
    "    sub_mutations_df['START'] = sub_mutations_df['POSITION'] - 1  # for pybedtools\n",
    "\n",
    "    sub_mutations_df = sub_mutations_df[['CHROMOSOME', 'START', 'POSITION', 'SAMPLE', 'ID']]\n",
    "    sub_mutations_df['CHROMOSOME'] = sub_mutations_df['CHROMOSOME'].astype(str)\n",
    "\n",
    "    sub_mutations_df.columns = ['chr', 'start', 'end', 'sample', 'ID']\n",
    "    exons_coords_df.columns = ['chr', 'start', 'end', 'ensembl']\n",
    "    introns_coords_df.columns = ['chr', 'start', 'end', 'ensembl']\n",
    "\n",
    "    sub_exons_coords_df = exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "    sub_introns_coords_df = introns_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "\n",
    "    exons_bed = pybedtools.BedTool.from_dataframe(sub_exons_coords_df)\n",
    "    introns_bed = pybedtools.BedTool.from_dataframe(sub_introns_coords_df)\n",
    "    sub_mutations_bed = pybedtools.BedTool.from_dataframe(sub_mutations_df)\n",
    "\n",
    "    mutations_in_exons_bed = exons_bed.intersect(sub_mutations_bed, wao=True)\n",
    "    mutations_in_introns_bed = introns_bed.intersect(sub_mutations_bed, wao=True)\n",
    "    \n",
    "    mutations_in_exons_symbol_df = pd.read_table(mutations_in_exons_bed.fn, names = ['range_chr', 'range_start',\n",
    "                        'range_end', 'ensembl', 'mut_chr', 'mut_start', 'mut_end', 'mut_sample', 'mut_ID',\n",
    "                        'overlap_bp'], sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_exons_symbol_df = mutations_in_exons_symbol_df[mutations_in_exons_symbol_df['overlap_bp'] != 0]\n",
    "    mutations_in_exons_symbol_df = mutations_in_exons_symbol_df[['mut_chr', 'mut_start', 'mut_end', 'ensembl',\n",
    "                                                                 'mut_sample', 'mut_ID']]\n",
    "    mutations_in_exons_symbol_df.columns = ['chr', 'start', 'end', 'ensembl', 'sample', 'ID']\n",
    "    \n",
    "    mutations_in_introns_symbol_df = pd.read_table(mutations_in_introns_bed.fn, names = ['range_chr', 'range_start',\n",
    "                        'range_end', 'ensembl', 'mut_chr', 'mut_start', 'mut_end', 'mut_sample', 'mut_ID',\n",
    "                        'overlap_bp'], sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_introns_symbol_df = mutations_in_introns_symbol_df[mutations_in_introns_symbol_df['overlap_bp'] != 0]\n",
    "    mutations_in_introns_symbol_df = mutations_in_introns_symbol_df[['mut_chr', 'mut_start', 'mut_end', 'ensembl',\n",
    "                                                                 'mut_sample', 'mut_ID']]\n",
    "    mutations_in_introns_symbol_df.columns = ['chr', 'start', 'end', 'ensembl', 'sample', 'ID']\n",
    "\n",
    "    exons_bed.delete_temporary_history(ask=False)\n",
    "    introns_bed.delete_temporary_history(ask=False)\n",
    "    sub_mutations_bed.delete_temporary_history(ask=False)\n",
    "    mutations_in_exons_bed.delete_temporary_history(ask=False)\n",
    "    mutations_in_introns_bed.delete_temporary_history(ask=False)\n",
    "\n",
    "    return mutations_in_exons_symbol_df, mutations_in_introns_symbol_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def build_my_positions_result2(genes_list, exons_coords_symbol_df, tb_consequence_type, CONSEQUENCE_RANK):\n",
    "    \"\"\"\n",
    "    For a list of genes, get the consequence type associated with each combination of reference and\n",
    "    alternate triplets.\n",
    "\n",
    "    Args:\n",
    "        genes_list (list): list of genes Ensembl identifiers\n",
    "        exons_coords_symbol_df (:class:`~pandas.DataFrame`): table with the exons genomic coordinates\n",
    "        tb_consequence_type: tabix with the consequence types\n",
    "        CONSEQUENCE_RANK (:class:`~pandas.DataFrame`): table that maps each consequence type with its type\n",
    "        (synonymous or non-synonymous)\n",
    "\n",
    "    Returns:\n",
    "        :class:`~pandas.DataFrame`. Table with the following fiels:\n",
    "        chromosome, position, ref, alt, ensembl ID, consequence type, type of the consequence,\n",
    "        reference triplet, alternate triplet\n",
    "\n",
    "    \"\"\"\n",
    "    my_position_results = list()\n",
    "\n",
    "    for my_gene in genes_list:\n",
    "\n",
    "        try:\n",
    "            exons_gene = exons_coords_symbol_df[exons_coords_symbol_df['ensembl'] == my_gene]\n",
    "            exons_gene_lol = exons_gene.values.tolist()\n",
    "\n",
    "            for my_exon in exons_gene_lol:\n",
    "\n",
    "                my_chr = my_exon[0][3:]\n",
    "                my_start = int(my_exon[1])\n",
    "                my_end = int(my_exon[2])\n",
    "\n",
    "                res_exons = tb_consequence_type.querys(\"{}:{}-{}\".format(my_chr, my_start, my_end))\n",
    "\n",
    "                for my_result in res_exons:\n",
    "                    if len(my_result) < 6:\n",
    "                        continue\n",
    "\n",
    "                    my_result[5] = my_result[5].rstrip('\\r')\n",
    "\n",
    "                    signature = hg19(my_result[0], int(my_result[1])-1, 3)\n",
    "\n",
    "                    tri_ref = signature[0] + my_result[2] + signature[-1]\n",
    "                    tri_alt = signature[0] + my_result[3] + signature[-1]\n",
    "\n",
    "                    my_result.append(tri_ref)\n",
    "                    my_result.append(tri_alt)\n",
    "                    my_result.append(my_gene)\n",
    "\n",
    "                    my_position_results.append(my_result)\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"problem with gene: \", my_gene)\n",
    "\n",
    "    my_position_results_df = pd.DataFrame(my_position_results)\n",
    "    my_position_results_df.columns = ['CHROMOSOME', 'POSITION', 'REF', 'ALT', 'ENSEMBL_NOVER', 'CONSEQUENCE', 'TRI_REF',\n",
    "                                      'TRI_ALT', 'ENSEMBL']\n",
    "\n",
    "    my_position_results_df = my_position_results_df.drop('ENSEMBL_NOVER', 1)\n",
    "    my_position_results_df = my_position_results_df[\n",
    "        ['CHROMOSOME', 'POSITION', 'REF', 'ALT', 'ENSEMBL', 'CONSEQUENCE', 'TRI_REF',\n",
    "         'TRI_ALT']]\n",
    "\n",
    "    my_position_results_df = pd.merge(my_position_results_df, CONSEQUENCE_RANK, on='CONSEQUENCE')\n",
    "    my_position_results_df['POSITION'] = my_position_results_df['POSITION'].astype(int)\n",
    "    my_position_results_df['CHROMOSOME'] = my_position_results_df['CHROMOSOME'].astype(str)\n",
    "\n",
    "    return my_position_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_syn_nonsyn_muts2(genes_list, mutations_in_exons_symbol_df, tb_consequence_type, mutations_df, CONSEQUENCE_RANK):\n",
    "    \"\"\"\n",
    "    For each gene in the list, get the consequence type associated with each combination of reference and\n",
    "    alternate triplets and merge it with the mutations.\n",
    "\n",
    "    Args:\n",
    "        genes_list (list): list of genes Ensembl identifiers\n",
    "        mutations_in_exons_symbol_df (:class:`~pandas.DataFrame`): table with the exons genomic coordinates\n",
    "        tb_consequence_type: tabix with the consequence types\n",
    "        mutations_df :class:`~pandas.DataFrame`): table with the mutations\n",
    "        CONSEQUENCE_RANK (:class:`~pandas.DataFrame`): table that maps each consequence type with its type\n",
    "        (synonimous or non-synonimous)\n",
    "\n",
    "    Returns:\n",
    "        :class:`~pandas.DataFrame`. Table\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    my_position_results = list()\n",
    "\n",
    "    for symbol_gene in genes_list:\n",
    "\n",
    "        gene_mutations = mutations_in_exons_symbol_df[mutations_in_exons_symbol_df['ensembl'] == symbol_gene]\n",
    "        gene_mutations_list = gene_mutations.values.tolist()\n",
    "\n",
    "        # For every exonic mutation in every gene, know if it is synonymous or nonsynonymous\n",
    "\n",
    "        if len(gene_mutations_list) > 0:\n",
    "\n",
    "            for gene_mutation in gene_mutations_list:\n",
    "\n",
    "                my_chr = gene_mutation[0][3:]\n",
    "                my_start = int(gene_mutation[1])\n",
    "                my_end = int(gene_mutation[2])\n",
    "\n",
    "                res_exons = tb_consequence_type.querys(\"{}:{}-{}\".format(my_chr, my_end, my_end))  # mutations are expected to have only 1 position so start and end are equal\n",
    "\n",
    "                for my_result in res_exons:\n",
    "                    my_result[5] = my_result[5].rstrip('\\r')\n",
    "\n",
    "                    signature = hg19(my_result[0], int(my_result[1])-1, 3)\n",
    "\n",
    "                    tri_ref = signature[0] + my_result[2] + signature[-1]\n",
    "                    tri_alt = signature[0] + my_result[3] + signature[-1]\n",
    "\n",
    "                    my_result.append(tri_ref)\n",
    "                    my_result.append(tri_alt)\n",
    "                    my_result.append(symbol_gene)\n",
    "\n",
    "                    my_position_results.append(my_result)\n",
    "\n",
    "    my_position_results_df = pd.DataFrame(my_position_results)\n",
    "    my_position_results_df.columns = ['CHROMOSOME', 'POSITION', 'REF', 'ALT', 'ENSEMBL_ONLY',\n",
    "                                      'CONSEQUENCE', 'TRI_REF', 'TRI_ALT', 'ENSEMBL']\n",
    "\n",
    "    my_position_results_df = my_position_results_df.drop('ENSEMBL_ONLY', 1)\n",
    "\n",
    "    my_position_results_df = pd.merge(my_position_results_df, CONSEQUENCE_RANK, on='CONSEQUENCE')\n",
    "    my_position_results_df['POSITION'] = my_position_results_df['POSITION'].astype(int)\n",
    "    my_position_results_df['CHROMOSOME'] = my_position_results_df['CHROMOSOME'].astype(str)\n",
    "    my_position_results_df.drop_duplicates(inplace=True)\n",
    "    my_position_results_df['CHROMOSOME'] = 'chr' + my_position_results_df['CHROMOSOME'].astype(str)\n",
    "\n",
    "    mutations_in_exons = pd.merge(my_position_results_df, mutations_df, on=['CHROMOSOME', 'POSITION', 'REF', 'ALT'])\n",
    "\n",
    "    return mutations_in_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_results_per_gene(sub_mutations_in_exons, my_gene, sub_exon, sub_intron, my_sample):\n",
    "    \"\"\"\n",
    "    Extract the number of synonymous and non-synonymous mutations\n",
    "    and their proportion and the exon and intron proportion.\n",
    "    Additionally, the number of mutation in exons and introns is computed.\n",
    "\n",
    "    Args:\n",
    "        sub_mutations_in_exons (:class:`~pandas.DataFrame`): table with the exonic mutations and their type (syn vs. non-syn)\n",
    "        my_gene (str): Ensembl identifier\n",
    "        sub_exon (:class:`~pandas.DataFrame`): table with the mutations in exons\n",
    "        sub_intron (:class:`~pandas.DataFrame`): table with the mutations in introns\n",
    "        my_sample (str): sample identifier\n",
    "\n",
    "    Returns:\n",
    "        list. Gene identifier, number of mutations in exons and introns, number of mutations synonymous and\n",
    "        non-synonymous mutations and sample identifier.\n",
    "\n",
    "    \"\"\"\n",
    "    sub_gene_mutations = sub_mutations_in_exons[sub_mutations_in_exons['ENSEMBL'] == my_gene]\n",
    "\n",
    "    n_syn_muts = len(sub_gene_mutations[sub_gene_mutations['TYPE_x'] == 'synonymous'])\n",
    "    n_nonsyn_muts = len(sub_gene_mutations[sub_gene_mutations['TYPE_x'] == 'non_synonymous'])\n",
    "\n",
    "    n_exon_muts = len(sub_exon[sub_exon['ensembl'] == my_gene])\n",
    "    n_intron_muts = len(sub_intron[sub_intron['ensembl'] == my_gene])\n",
    "\n",
    "    my_row = [my_gene, n_exon_muts, n_intron_muts, n_syn_muts,\n",
    "              n_nonsyn_muts, my_sample]\n",
    "\n",
    "    return my_row"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## By gene analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "\n",
    "# Read common files\n",
    "## Load consequences\n",
    "tb_consequence_type = tabix.open(consequence_type_file)\n",
    "\n",
    "## Obtain consequence rank\n",
    "CONSEQUENCE_RANK = pd.read_csv(path.join('~/.bgdata/intogen/ensembl/1.0-20150729', 'consequence_ranking.tsv'), delimiter='\\t')\n",
    "\n",
    "## Classify each consequence according to the rank\n",
    "CONSEQUENCE_RANK['TYPE'] = CONSEQUENCE_RANK.apply(lambda x:synonymous_or_not(x) ,1)\n",
    "CONSEQUENCE_RANK = CONSEQUENCE_RANK[['CONSEQUENCE', 'TYPE']]\n",
    "\n",
    "## Load signatures\n",
    "signatures_path = path.join('results', '3mer_signatures_DNM_new.txt')\n",
    "all_signatures = pd.read_csv(signatures_path, sep='\\t')\n",
    "\n",
    "## Load mutations\n",
    "mutations_path = mutations_folder + 'germinal_ultimate_dataset.bed.gz'\n",
    "mutations_df = pd.read_csv(mutations_path, sep=\"\\t\", header=None) \n",
    "mutations_df.columns = ['CHROMOSOME', 'START', 'POSITION', 'REF', 'ALT', 'SAMPLE', 'TYPE', 'CLASS']\n",
    "mutations_df = mutations_df[mutations_df['TYPE'] == 'subs']\n",
    "mutations_df['ID'] = mutations_df.index + 1\n",
    "\n",
    "## Load coordinates\n",
    "introns_coords = path.join(coordinates_folder, 'germinal' + '_filtered_introns_coords.txt') # Output from other notebook\n",
    "middle_exons_coords = path.join(coordinates_folder, 'germinal' + '_filtered_exons_coords.txt') # Output from other notebook\n",
    "exons_coords_df = pd.read_csv(middle_exons_coords, sep=\"\\t\", low_memory=False)\n",
    "introns_coords_df = pd.read_csv(introns_coords, sep=\"\\t\", low_memory=False)\n",
    "exons_coords_df['ensembl'] = exons_coords_df.apply(lambda x: remove_version_ensembl(x, 'ensembl'), axis=1)\n",
    "introns_coords_df['ensembl'] = introns_coords_df.apply(lambda x: remove_version_ensembl(x, 'ensembl'), axis=1)\n",
    "exons_coords_symbol_df = exons_coords_df[['chr', 'start', 'position', 'ensembl']]\n",
    "introns_coords_symbol_df = introns_coords_df[['chr', 'start', 'position', 'ensembl']]\n",
    "exons_coords_symbol_df.columns = ['chr', 'start', 'end', 'ensembl']\n",
    "introns_coords_symbol_df.columns = ['chr', 'start', 'end', 'ensembl']\n",
    "    \n",
    "## Find mean replication time for all genes\n",
    "whole_gene_coords = pd.concat([exons_coords_symbol_df, introns_coords_symbol_df])\n",
    "rep_time_mean_per_gene_df = parse_rep_time_data(replication_time_path, whole_gene_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Gene by gene\n",
    "my_genes_list = list(set(exons_coords_df['ensembl'].tolist()))\n",
    "print('{} genes'.format(len(my_genes_list)))\n",
    "    \n",
    "# For each gene, get the counts of each trinucleotide. Useful for expected exon-intron muts.\n",
    "# note that the counts are returned for pairs ref-alt triplets, where all possible alt have the same counts.\n",
    "# This is done to make our life easier later on when doing some merging.\n",
    "counters_dict = precompute_counts(my_genes_list, exons_coords_symbol_df, introns_coords_symbol_df, nucleotides)\n",
    "    \n",
    "# Assing syn/non-syn to all possible alterations in all exonic coordines\n",
    "gene_position_results_df = build_my_positions_result2(my_genes_list,   #USING ANOTHER IMPORTED FUNCTION\n",
    "                                                   exons_coords_symbol_df,\n",
    "                                                   tb_consequence_type,\n",
    "                                                   CONSEQUENCE_RANK)\n",
    "\n",
    "# Choose signature (useless, only one signature)\n",
    "probability_name = 'Probability_germline'\n",
    "sub_signatures_df = all_signatures[['mutation', probability_name]]\n",
    "\n",
    "prob_name = 'Probability_germline'\n",
    "\n",
    "# Lists to store results\n",
    "final_results_lol = list() \n",
    "        \n",
    "genes_other_info = list()\n",
    "        \n",
    "# Expected proportions at gene level (using the signature of the cluster)\n",
    "for my_gene in my_genes_list:\n",
    "\n",
    "    try:\n",
    "\n",
    "        sub_positions_results_df = gene_position_results_df[gene_position_results_df['ENSEMBL'] == my_gene]\n",
    "\n",
    "        # Count possible synonymous and non-synonymous mutations in the gene                \n",
    "        n_synonimous_variants = sub_positions_results_df[sub_positions_results_df['TYPE'] == 'synonymous'].shape[0]\n",
    "        n_non_synonimous_variants = sub_positions_results_df[sub_positions_results_df['TYPE'] == 'non_synonymous'].shape[0]\n",
    "                \n",
    "        # Exclude genes with 0 synonimous and non synomimous variants\n",
    "        if n_synonimous_variants == 0 and n_non_synonimous_variants == 0:\n",
    "#               print('Error with gene: ', my_gene)\n",
    "            pb.value += 1\n",
    "            continue\n",
    "\n",
    "        # Compute probability of a syn/non-syn mutation to occur given its context\n",
    "        # (and the probability associated from the signature)\n",
    "        non_syn_sub_df = sub_positions_results_df[sub_positions_results_df['TYPE'] == 'non_synonymous'].copy()\n",
    "        non_syn_sub_df['MY_KEY'] = list(zip(non_syn_sub_df.TRI_REF, non_syn_sub_df.TRI_ALT))  # identify by ref triplet - alt triplet\n",
    "        non_syn_sub_df['MY_KEY'] = non_syn_sub_df['MY_KEY'].astype(str)\n",
    "        nonsyn_cnt = non_syn_sub_df['MY_KEY'].value_counts()  # counts of each (ref triplet, alt triplet)\n",
    "        nonsyn_df = pd.DataFrame(nonsyn_cnt)\n",
    "        nonsyn_df['mutation'] = nonsyn_df.index\n",
    "        nonsyn_df = pd.merge(nonsyn_df, sub_signatures_df, on='mutation')\n",
    "        nonsyn_df['product'] = nonsyn_df['MY_KEY']*nonsyn_df[prob_name]  # multiply counts by its probability\n",
    "        nonsyn_products_sum = sum(nonsyn_df['product'])\n",
    "\n",
    "        syn_sub_df = sub_positions_results_df[sub_positions_results_df['TYPE'] == 'synonymous'].copy()\n",
    "        syn_sub_df['MY_KEY'] = list(zip(syn_sub_df.TRI_REF, syn_sub_df.TRI_ALT))\n",
    "        syn_sub_df['MY_KEY'] = syn_sub_df['MY_KEY'].astype(str)\n",
    "        syn_cnt = syn_sub_df['MY_KEY'].value_counts()\n",
    "        syn_df = pd.DataFrame(syn_cnt)\n",
    "        syn_df['mutation'] = syn_df.index\n",
    "        syn_df = pd.merge(syn_df, sub_signatures_df, on='mutation')\n",
    "        syn_df['product'] = syn_df['MY_KEY']*syn_df[prob_name]\n",
    "        syn_products_sum = sum(syn_df['product'])\n",
    "\n",
    "        # Compute probability of \n",
    "        # any exon and intron mutation\n",
    "        exons_counter = counters_dict[(my_gene, 'exons_count')]\n",
    "        exons_counter_df = pd.DataFrame.from_dict(exons_counter, orient='index').reset_index()\n",
    "        exons_counter_df.columns = ['mutation', 'count']\n",
    "        exons_counter_df['mutation'] = exons_counter_df['mutation'].astype(str)\n",
    "        exons_counter_df = pd.merge(exons_counter_df, sub_signatures_df, on='mutation')\n",
    "        exons_counter_df['product'] = exons_counter_df['count']*exons_counter_df[prob_name]\n",
    "        exons_products_sum = sum(exons_counter_df['product'])\n",
    "                \n",
    "        introns_counter = counters_dict[(my_gene, 'introns_count')]\n",
    "        introns_counter_df = pd.DataFrame.from_dict(introns_counter, orient='index').reset_index()\n",
    "        introns_counter_df.columns = ['mutation', 'count']\n",
    "        introns_counter_df['mutation'] = introns_counter_df['mutation'].astype(str)\n",
    "        introns_counter_df = pd.merge(introns_counter_df, sub_signatures_df, on='mutation')\n",
    "        introns_counter_df['product'] = introns_counter_df['count']*introns_counter_df[prob_name]\n",
    "        introns_products_sum = sum(introns_counter_df['product'])\n",
    "\n",
    "        my_row = [my_gene,n_synonimous_variants, n_non_synonimous_variants, \n",
    "                          exons_products_sum, introns_products_sum, syn_products_sum, nonsyn_products_sum]\n",
    "\n",
    "        final_results_lol.append(my_row)\n",
    "                \n",
    "                \n",
    "        # Other gene information (number of base pairs, replicaton time)\n",
    "                \n",
    "        gene_exons_coords = exons_coords_symbol_df[exons_coords_symbol_df['ensembl'] == my_gene]\n",
    "        exons_bp = sum(gene_exons_coords['end'] - gene_exons_coords['start'])\n",
    "\n",
    "        gene_introns_coords = introns_coords_symbol_df[introns_coords_symbol_df['ensembl'] == my_gene]\n",
    "        introns_bp = sum(gene_introns_coords['end'] - gene_introns_coords['start'])\n",
    "\n",
    "        sub_gene_reptime = rep_time_mean_per_gene_df[rep_time_mean_per_gene_df.index == my_gene]['rep_time']\n",
    "\n",
    "        if len(sub_gene_reptime) > 0:\n",
    "            gene_reptime = sub_gene_reptime.values[0]\n",
    "\n",
    "        elif len(sub_gene_reptime) == 0:\n",
    "            gene_reptime = np.nan\n",
    "\n",
    "        genes_other_info.append([my_gene, exons_bp, introns_bp, gene_reptime])\n",
    "\n",
    "\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    pb.value += 1\n",
    "\n",
    "print('{} valid genes'.format(len(final_results_lol)))\n",
    "\n",
    "final_results_df = pd.DataFrame(final_results_lol)\n",
    "final_results_df.columns = ['ensembl', 'n_synonimous_variants', 'n_non_synonimous_variants',\n",
    "                                    'exons_products_sum', 'introns_products_sum', 'syn_products_sum',\n",
    "                                    'nonsyn_products_sum']\n",
    "\n",
    "genelevel_results = final_results_df\n",
    "        \n",
    "new_genes_list = list(set(genelevel_results['ensembl'].tolist()))\n",
    "        \n",
    "genes_other_info_df = pd.DataFrame(genes_other_info)\n",
    "genes_other_info_df.columns = ['ensembl', 'exons_bp', 'introns_bp', 'reptime']\n",
    "        \n",
    "genelevel_results_toprint = genelevel_results\n",
    "genelevel_results_toprint = pd.merge(genelevel_results_toprint, genes_other_info_df, on='ensembl')\n",
    "\n",
    "genelevel_results_toprint.to_csv(path.join('results', 'germinal' + '_gene_dependent_results.txt'), sep='\\t', header=True, index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## By sample analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 samples\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ensembl            ENSG00000161573ENSG00000137941ENSG00000163508E...\n",
       "obs_exon_muts                                                   4452\n",
       "obs_intron_muts                                               158492\n",
       "obs_syn_muts                                                    1116\n",
       "obs_nonsyn_muts                                                 3331\n",
       "dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "by_gene_path = path.join('results', 'germinal' + '_gene_dependent_results.txt')\n",
    "genelevel_results = pd.read_csv(by_gene_path, sep=\"\\t\", header='infer', low_memory=False)\n",
    "new_genes_list = list(set(genelevel_results['ensembl'].tolist()))\n",
    "\n",
    "# Get exon and intron mutations in the samples of the cluster (without sample identifier)    \n",
    "mutations_in_exons_symbol_df, mutations_in_introns_symbol_df = get_muts_per_gene(mutations_df=mutations_df,\n",
    "                                                                                 exons_coords_df=exons_coords_df,\n",
    "                                                                                 introns_coords_df=introns_coords_df)\n",
    "\n",
    "# For every exonic mutation in every gene, know if it is synonymous or nonsynonymous\n",
    "mutations_in_exons_cons = compute_syn_nonsyn_muts2(new_genes_list,\n",
    "                                                          mutations_in_exons_symbol_df,\n",
    "                                                          tb_consequence_type,\n",
    "                                                          mutations_df,\n",
    "                                                          CONSEQUENCE_RANK)\n",
    "\n",
    "# All samples together\n",
    "\n",
    "common_samples = ['subs']\n",
    "\n",
    "# Analysis at sample level (per cluster)\n",
    "allgenes_results = list()\n",
    "print('{} samples'.format(len(common_samples)))\n",
    "for my_sample in common_samples:\n",
    "\n",
    "    sub_mutations_in_exons = mutations_in_exons_cons[mutations_in_exons_cons['TYPE_y'] == my_sample]\n",
    "\n",
    "    sub_exon = mutations_in_exons_symbol_df.copy()\n",
    "    sub_intron = mutations_in_introns_symbol_df.copy()\n",
    "\n",
    "    for my_gene in new_genes_list:\n",
    "\n",
    "        # Per each gene in each sample, compute the number of exon, intron, syn,\n",
    "        # nonsyn mutations and their expected proportions\n",
    "                \n",
    "        allgenes_results.append(compute_results_per_gene(sub_mutations_in_exons,\n",
    "                                         my_gene, sub_exon, sub_intron, my_sample))\n",
    "\n",
    "allgenes_results_df = pd.DataFrame(allgenes_results)\n",
    "\n",
    "allgenes_results_df.columns = ['ensembl', 'obs_exon_muts', 'obs_intron_muts',\n",
    "                                       'obs_syn_muts', 'obs_nonsyn_muts', 'sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 samples\n"
     ]
    }
   ],
   "source": [
    "# Add sample to mutations in introns and exons\n",
    "formated_mutations_df = mutations_df[['CHROMOSOME', 'POSITION', 'CLASS']]\n",
    "formated_mutations_df.columns = ['chr', 'end', 'class']\n",
    "\n",
    "sample_mutations_in_exons_symbol_df = pd.merge(mutations_in_exons_symbol_df,\n",
    "                                                      formated_mutations_df, on=['chr', 'end'])\n",
    "\n",
    "sample_mutations_in_introns_symbol_df = pd.merge(mutations_in_introns_symbol_df,\n",
    "                                                         formated_mutations_df, on=['chr',\n",
    "                                                            'end'])\n",
    "\n",
    "# All samples together\n",
    "\n",
    "common_samples = ['mixed', 'autism', 'healthy']\n",
    "\n",
    "# Analysis at sample level (per cluster)\n",
    "allgenes_results = list()\n",
    "print('{} samples'.format(len(common_samples)))\n",
    "for my_sample in common_samples:\n",
    "\n",
    "    sub_mutations_in_exons = mutations_in_exons_cons[mutations_in_exons_cons['CLASS'] == my_sample]\n",
    "\n",
    "    sub_exon = sample_mutations_in_exons_symbol_df[sample_mutations_in_exons_symbol_df['class'] == my_sample]\n",
    "    sub_intron = sample_mutations_in_introns_symbol_df[sample_mutations_in_introns_symbol_df['class'] == my_sample]\n",
    "\n",
    "    for my_gene in new_genes_list:\n",
    "\n",
    "        # Per each gene in each sample, compute the number of exon, intron, syn,\n",
    "        # nonsyn mutations and their expected proportions\n",
    "                \n",
    "        allgenes_results.append(compute_results_per_gene(sub_mutations_in_exons,\n",
    "                                         my_gene, sub_exon, sub_intron, my_sample))\n",
    "\n",
    "allgenes_results_df = pd.DataFrame(allgenes_results)\n",
    "\n",
    "allgenes_results_df.columns = ['ensembl', 'obs_exon_muts', 'obs_intron_muts',\n",
    "                                       'obs_syn_muts', 'obs_nonsyn_muts', 'sample']\n",
    "            \n",
    "\n",
    "allgenes_results_df_toprint = allgenes_results_df\n",
    "\n",
    "allgenes_results_df_toprint.to_csv(path.join('results', 'germinal_sample_dependent_results.txt'),\n",
    "                                          sep='\\t', header=True, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
