{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## H3K36me3 and nucleosomes coverage introns vs exons: Supplementary Figure S5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "from os import path\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import statsmodels.stats.multitest as smm\n",
    "from scipy.stats import mannwhitneyu\n",
    "from tqdm import tqdm\n",
    "from bgreference import hg19\n",
    "\n",
    "from python_scripts.plot_functions import config_plot_params\n",
    "%matplotlib inline\n",
    "config_plot_params()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Included\n",
    "intron_coords = 'data/coordinates/genes_intron_coords.bed.gz'\n",
    "middle_exons_coords = 'data/coordinates/genes_middle_exon_coords.bed.gz'\n",
    "exons_coords = 'data/coordinates/genes_exon_coords_filtered.bed.gz'\n",
    "nucl_coords_path = 'data/nucleosomes/nucleosome_coords.bed'\n",
    "\n",
    "# Not included\n",
    "histone_data_folder = 'non_provided_data/histonmarks/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_exons_per_gene(exons_position_df):\n",
    "    \"\"\"Given a df numbered exons per gene, get the total number of exons per gene\"\"\"\n",
    "    \n",
    "    my_exons_per_gene = list()\n",
    "\n",
    "    for index, my_df in exons_position_df.groupby(3):\n",
    "        my_exons_per_gene.append([index, my_df[8].max()])\n",
    "        \n",
    "    my_exons_per_gene_df = pd.DataFrame(my_exons_per_gene)\n",
    "    my_exons_per_gene_df.columns = ['gene', 'n_exons']\n",
    "\n",
    "    return my_exons_per_gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def overlap_with_his_marks(my_coords_df, histon_mark_bed):\n",
    "    \"\"\"\n",
    "    The size is computed as the value of the 25th percentile.\n",
    "    Segments below are discarded.\n",
    "    For the rest, they are centered and only positions around the centre an within \n",
    "    a distance of half the size are considered.\n",
    "    \"\"\"\n",
    "\n",
    "    my_coords_df['size'] = (my_coords_df['end'] - my_coords_df['start'])\n",
    "\n",
    "    lower_quart = my_coords_df['size'].describe()['25%']\n",
    "\n",
    "    filtered_coords_df = my_coords_df[my_coords_df['size'] >= lower_quart].copy()  # Remove short segments\n",
    "    \n",
    "    filtered_coords_df['middle_start'] = (filtered_coords_df['start'] + filtered_coords_df['size']/2)\n",
    "    filtered_coords_df['middle_start'] = filtered_coords_df.apply(lambda x: math.floor(x['middle_start']), axis=1)\n",
    "    filtered_coords_df['middle_end'] = filtered_coords_df['middle_start'] + 1\n",
    "\n",
    "    filtered_coords_df['region_start'] = filtered_coords_df['middle_start'] - math.floor(lower_quart/2)\n",
    "    filtered_coords_df['region_end'] = filtered_coords_df['middle_end'] + math.floor(lower_quart/2)\n",
    "    filtered_coords_bed = pybedtools.BedTool.from_dataframe(filtered_coords_df[\n",
    "            ['chr', 'region_start', 'region_end', 'ensembl', 'size']\n",
    "        ])\n",
    "    \n",
    "    my_histon_intersec_bed = filtered_coords_bed.intersect(histon_mark_bed)\n",
    "\n",
    "    my_histon_intersec_df = pd.read_table(my_histon_intersec_bed.fn, names=['chr', 'start', 'end', 'ensembl', 'size'])\n",
    "\n",
    "    total_bases = sum(filtered_coords_df['region_end'] - filtered_coords_df['region_start'])\n",
    "    overlapping_bases = sum(my_histon_intersec_df['end'] - my_histon_intersec_df['start'])\n",
    "\n",
    "    n_seqs = len(filtered_coords_df)\n",
    "    \n",
    "    return overlapping_bases, total_bases, lower_quart, n_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_values_barplot(coverage_df, scores_df, histone):\n",
    "    '''function to return the values for boxplot'''\n",
    "\n",
    "    x_values = []\n",
    "    x_labels = []\n",
    "    x_colors = []\n",
    "    box_plot_values = {}\n",
    "    box_plot_values['exon'] = []\n",
    "    box_plot_values['intron'] = []\n",
    "\n",
    "    for x in x_order:\n",
    "        for region in ['exon', 'intron']:\n",
    "\n",
    "            row = coverage_df[(coverage_df['hist_cell']==histone) & (coverage_df['position']==x) & \n",
    "                                 (coverage_df['region']==region)][['covered_bp', 'total_bp']]\n",
    "\n",
    "            if len(row)==0:\n",
    "                print(x, region)\n",
    "                break\n",
    "                \n",
    "            per_cov = row['covered_bp'].values[0]/row['total_bp'].values[0]\n",
    "\n",
    "            x_values.append(per_cov)\n",
    "            x_labels.append(\"%s %s\" % (region, x))\n",
    "            \n",
    "            x_colors.append(rcolors[region])\n",
    "            \n",
    "            box_plot_values[region].append(per_cov)\n",
    "\n",
    "    # Special cases:\n",
    "    ## Second last exons\n",
    "    row = coverage_df[(coverage_df['hist_cell']==histone) & (coverage_df['position']=='sec_last') & \n",
    "                         (coverage_df['region']=='exon')][['covered_bp', 'total_bp']]\n",
    "\n",
    "    print(len(row))\n",
    "    per_cov = row['covered_bp'].values[0]/row['total_bp'].values[0]\n",
    "\n",
    "    x_values.append(per_cov)\n",
    "    x_labels.append(\"%s second last\" % ('exon'))\n",
    "    box_plot_values['exon'].append(per_cov)\n",
    "    x_colors.append(rcolors['exon'])\n",
    "\n",
    "    ## Last intron and then exon\n",
    "    for region in ['intron', 'exon']:\n",
    "\n",
    "        row = coverage_df[(coverage_df['hist_cell']==histone) & (coverage_df['position']=='last') & \n",
    "                             (coverage_df['region']==region)][['covered_bp', 'total_bp']]\n",
    "\n",
    "        if len(row)==0:\n",
    "            print(x, region)\n",
    "            break\n",
    "\n",
    "        per_cov = row['covered_bp'].values[0]/row['total_bp'].values[0]\n",
    "\n",
    "        x_values.append(per_cov)\n",
    "        x_labels.append(\"%s %s\" % (region, 'last'))\n",
    "        \n",
    "        box_plot_values[region].append(per_cov)\n",
    "        x_colors.append(rcolors[region])\n",
    "        \n",
    "    # get the p-value to mention in the boxplot\n",
    "    wpvalue = scores_df[scores_df['index']==histone]['p_value'].values[0]\n",
    "        \n",
    "    return x_values, x_labels, x_colors, box_plot_values, wpvalue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "exons_position_df = pd.read_csv(exons_coords, sep='\\t', header=None)\n",
    "my_exons_per_gene_df = compute_exons_per_gene(exons_position_df)\n",
    "max_n_exons = my_exons_per_gene_df['n_exons'].max()\n",
    "\n",
    "exons_position_bed = pybedtools.BedTool.from_dataframe(exons_position_df)\n",
    "\n",
    "## Get last and second last exons\n",
    "last_bed = pybedtools.BedTool.from_dataframe(exons_position_df[exons_position_df.loc[:, 9] == 'last'])\n",
    "others_than_last_bed = pybedtools.BedTool.from_dataframe(exons_position_df[exons_position_df.loc[:, 9]!='last'])\n",
    "last_bed = last_bed.merge(c='6,8', o='distinct')\n",
    "last_bed = last_bed.intersect(others_than_last_bed, wo=True, sorted=True, v=True)  # drop all segment that overlap and are classified as last and any other category \n",
    "second_last_bed = pybedtools.BedTool.from_dataframe(exons_position_df[exons_position_df.loc[:, 9] == 'secondlast'])\n",
    "others_than_second_last_bed = pybedtools.BedTool.from_dataframe(exons_position_df[exons_position_df.loc[:, 9]!='secondlast'])\n",
    "second_last_bed = second_last_bed.merge(c='6,8', o='distinct')\n",
    "second_last_bed = second_last_bed.intersect(others_than_second_last_bed, wo=True, sorted=True, v=True)  # drop all segment that overlap and are classified as last and any other category \n",
    "second_and_last_bed = pybedtools.BedTool.from_dataframe(pd.concat([\n",
    "        pd.read_table(last_bed.fn, names=['chr', 'start', 'end', 'strand', 'ensembl']),\n",
    "        pd.read_table(second_last_bed.fn, names=['chr', 'start', 'end', 'strand', 'ensembl'])\n",
    "    ])).sort().merge()\n",
    "\n",
    "\n",
    "# Get the nucleosomes coordinates\n",
    "nucl_coords_df = pd.read_csv(nucl_coords_path, sep='\\t', header=None)\n",
    "nucl_coords_df = nucl_coords_df[nucl_coords_df[3] >= 2.5]\n",
    "nucl_coords_df = nucl_coords_df[[0,1,2]]\n",
    "nucl_coords_df.columns = ['chr', 'start', 'end']\n",
    "# expand to cover the 147 bases of the nucleosome\n",
    "nucl_coords_df['start'] = nucl_coords_df['start'] - 73\n",
    "nucl_coords_df['end'] = nucl_coords_df['end'] + 73\n",
    "nucl_coords_df = nucl_coords_df[~nucl_coords_df['start'] < 0]\n",
    "\n",
    "nucleosomes_bed = pybedtools.BedTool.from_dataframe(nucl_coords_df)\n",
    "   \n",
    "    \n",
    "my_bins = [[1],[2],[3],[4],[5],[6],[7],[8],[9],[10],[11,15],[16,20],[21,40],[41,100]]\n",
    "# Bins of the exons positions to be grouped\n",
    "\n",
    "exon_position_results = list()\n",
    "\n",
    "\n",
    "for my_exon_position in my_bins:\n",
    "\n",
    "    print(my_exon_position)\n",
    "    \n",
    "    if len(my_exon_position) == 1:\n",
    "        current_and_next_pos = [my_exon_position[0], my_exon_position[0]+1]\n",
    "    else:\n",
    "        current_and_next_pos = list(range(min(my_exon_position), max(my_exon_position)+1)) \n",
    "    \n",
    "    conc_exons_df = pd.DataFrame()\n",
    "    \n",
    "    # Take current position exonic coordinates plus the following exon coordinates\n",
    "    # In order to be able to define the coordinates of the intron following the current exon\n",
    "    \n",
    "    for my_pos in current_and_next_pos:  \n",
    "        \n",
    "        df = exons_position_df[exons_position_df.loc[:, 8] == my_pos]\n",
    "        \n",
    "        bed = pybedtools.BedTool.from_dataframe(df)\n",
    "        bed = bed.merge(c='6,8', o='distinct')\n",
    "        bed = bed.intersect(exons_position_bed, wo=True, sorted=True)\n",
    "        bed = bed.merge(c='4,5,14', o='distinct')\n",
    "        \n",
    "        my_position_df = pd.read_table(bed.fn, names=['chr', 'start', 'end', 'strand', 'ensembl', 'position'])\n",
    "        \n",
    "        my_position_df = my_position_df[my_position_df['position'] == str(my_pos)]\n",
    "\n",
    "        conc_exons_df = pd.concat([conc_exons_df, my_position_df])\n",
    "        \n",
    "    my_position_bed = pybedtools.BedTool.from_dataframe(conc_exons_df)\n",
    "    \n",
    "    # Remove segmets that overlap with last or second last exons\n",
    "    inter_my_position_bed = my_position_bed.intersect(second_and_last_bed, wao=True)\n",
    "    inter_my_position_df = pd.read_table(inter_my_position_bed.fn, names = \n",
    "                                         ['chr', 'start', 'end', 'strand', 'ensembl', 'position', 'chr_last', 'start_last', 'end_last', 'overlap']\n",
    "                                        )\n",
    "    inter_my_position_df = inter_my_position_df[inter_my_position_df['overlap'] == 0][['chr', 'start', 'end', 'strand', 'ensembl', 'position']]\n",
    "    \n",
    "    conc_exons_df = inter_my_position_df.copy()\n",
    "\n",
    "    # Get the coordinates of the introns as the coordinates between the exons\n",
    "    intron_coords_lol = list()\n",
    "    for index, my_df in conc_exons_df.groupby('ensembl'):\n",
    "\n",
    "        if len(set(my_df.position)) > 1:\n",
    "\n",
    "            intron_chr = my_df[0:1].chr.values[0]\n",
    "            intron_start = min(my_df.end)\n",
    "            intron_end = max(my_df.start)\n",
    "            intron_ensembl = my_df[0:1].ensembl.values[0]\n",
    "\n",
    "            my_intron_row = [intron_chr, intron_start, intron_end, intron_ensembl]\n",
    "            intron_coords_lol.append(my_intron_row)\n",
    "            \n",
    "    intron_coords_df = pd.DataFrame(intron_coords_lol)\n",
    "    intron_coords_bed = pybedtools.BedTool.from_dataframe(intron_coords_df)\n",
    "\n",
    "    # Remove the exons that are in the middle\n",
    "    filtered_intron_coords_bed = intron_coords_bed.subtract(exons_position_bed)  # Ensure that the positions do not belong to any exon\n",
    "\n",
    "    filtered_intron_coords_df = pd.read_table(filtered_intron_coords_bed.fn, names=['chr', 'start', 'end', 'ensembl'])\n",
    "\n",
    "    # Get rid of the extra exon we included to be able to compute the corresponding intron\n",
    "    filtered_exons_coords_df = conc_exons_df[conc_exons_df['position'].isin(my_exon_position)]\n",
    "    filtered_exons_coords_df = filtered_exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "    \n",
    "    pos_str = str(my_exon_position).replace('[', '').replace(']', '').replace(', ', '-').replace(',', '-')\n",
    "\n",
    "    # For each histon mark and cell line compute the total bases and overlapping bases in intron and exons\n",
    "    # according to the criteria in overlap_with_his_marks\n",
    "    for file in os.listdir(histone_data_folder):\n",
    "        if file.endswith(\".narrowPeak.gz\"):\n",
    "            try:\n",
    "                hist_mark_path = path.join(histone_data_folder, file)\n",
    "                cell_line = file.split('.')[0].split('-')[0]\n",
    "                hist_mark = file.split('.')[0].split('-')[1]                \n",
    "                my_histon_cell = cell_line + '-' + hist_mark\n",
    "                hist_mark_df = pd.read_csv(hist_mark_path, sep='\\t', header=None)\n",
    "                histon_mark_bed = pybedtools.BedTool.from_dataframe(hist_mark_df[[0,1,2]])\n",
    "   \n",
    "            except:\n",
    "                print(\"Problem with \", file)\n",
    "                continue\n",
    "\n",
    "            overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "                overlap_with_his_marks(filtered_exons_coords_df, histon_mark_bed)\n",
    "\n",
    "            ex_row = ['exon', pos_str, my_histon_cell, overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "            overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "                overlap_with_his_marks(filtered_intron_coords_df, histon_mark_bed)\n",
    "\n",
    "            intr_row = ['intron', pos_str, my_histon_cell, overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "            exon_position_results.append(ex_row)\n",
    "            exon_position_results.append(intr_row)\n",
    "            \n",
    "    # Nucleosomes\n",
    "    overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "                overlap_with_his_marks(filtered_exons_coords_df, nucleosomes_bed)\n",
    "\n",
    "    ex_row = ['exon', pos_str, 'unk-nucleosomes', overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "    overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "        overlap_with_his_marks(filtered_intron_coords_df, nucleosomes_bed)\n",
    "\n",
    "    intr_row = ['intron', pos_str, 'unk-nucleosomes', overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "    exon_position_results.append(ex_row)\n",
    "    exon_position_results.append(intr_row)\n",
    "            \n",
    "\n",
    "# Repeat for last and second last exons\n",
    "current_and_next_pos = ['secondlast', 'last']\n",
    "conc_exons_df = pd.DataFrame()\n",
    "\n",
    "print('Second last and last')\n",
    "\n",
    "for my_pos in current_and_next_pos:\n",
    "    \n",
    "    df = exons_position_df[exons_position_df.loc[:, 9] == my_pos]\n",
    "        \n",
    "    bed = pybedtools.BedTool.from_dataframe(df)\n",
    "    bed = bed.merge(c='6,8', o='distinct')\n",
    "    bed = bed.intersect(exons_position_bed, wo=True, sorted=True)\n",
    "    bed = bed.merge(c='4,5,15', o='distinct')\n",
    "\n",
    "    my_position_df = pd.read_table(bed.fn, names=['chr', 'start', 'end', 'strand', 'ensembl', 'position'])\n",
    "\n",
    "    my_position_df = my_position_df[my_position_df['position'] == str(my_pos)]\n",
    "\n",
    "    conc_exons_df = pd.concat([conc_exons_df, my_position_df])\n",
    "\n",
    "intron_coords_lol = list()\n",
    "\n",
    "for index, my_df in conc_exons_df.groupby('ensembl'):\n",
    "\n",
    "    if len(set(my_df.position)) > 1:\n",
    "\n",
    "        intron_chr = my_df[0:1].chr.values[0]\n",
    "        intron_start = min(my_df.end)\n",
    "        intron_end = max(my_df.start)\n",
    "        intron_ensembl = my_df[0:1].ensembl.values[0]\n",
    "\n",
    "        if intron_start > intron_end:\n",
    "            print(my_df)\n",
    "\n",
    "        my_intron_row = [intron_chr, intron_start, intron_end, intron_ensembl]\n",
    "        intron_coords_lol.append(my_intron_row)\n",
    "\n",
    "intron_coords_df = pd.DataFrame(intron_coords_lol)\n",
    "intron_coords_bed = pybedtools.BedTool.from_dataframe(intron_coords_df)\n",
    "\n",
    "filtered_intron_coords_bed = intron_coords_bed.subtract(exons_position_bed)\n",
    "filtered_intron_coords_df = pd.read_table(filtered_intron_coords_bed.fn, names=['chr', 'start', 'end', 'ensembl'])\n",
    "\n",
    "filtered_last_exons_coords_df = conc_exons_df[conc_exons_df['position'] == 'last']\n",
    "filtered_last_exons_coords_df = filtered_last_exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "\n",
    "filtered_seclast_exons_coords_df = conc_exons_df[conc_exons_df['position'] == 'secondlast']\n",
    "filtered_seclast_exons_coords_df = filtered_seclast_exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "\n",
    "for file in os.listdir(histone_data_folder):\n",
    "    if file.endswith(\".narrowPeak.gz\"):\n",
    "        try:\n",
    "            hist_mark_path = path.join(histone_data_folder, file)\n",
    "            cell_line = file.split('.')[0].split('-')[0]\n",
    "            hist_mark = file.split('.')[0].split('-')[1]\n",
    "            my_histon_cell = cell_line + '-' + hist_mark\n",
    "            hist_mark_df = pd.read_csv(hist_mark_path, sep='\\t', header=None)\n",
    "            histon_mark_bed = pybedtools.BedTool.from_dataframe(hist_mark_df[[0,1,2]])\n",
    "        except:\n",
    "            print(\"Problem with \", file)\n",
    "            continue\n",
    "            \n",
    "\n",
    "        overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "            overlap_with_his_marks(filtered_seclast_exons_coords_df, histon_mark_bed)\n",
    "\n",
    "        seclast_ex_row = ['exon', 'sec_last', my_histon_cell, overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "\n",
    "        overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "            overlap_with_his_marks(filtered_last_exons_coords_df, histon_mark_bed)\n",
    "\n",
    "        last_ex_row = ['exon', 'last', my_histon_cell, overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "        overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "            overlap_with_his_marks(filtered_intron_coords_df, histon_mark_bed)\n",
    "\n",
    "        intr_row = ['intron', 'last', my_histon_cell, overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "        exon_position_results.append(seclast_ex_row)\n",
    "        exon_position_results.append(last_ex_row)\n",
    "        exon_position_results.append(intr_row)\n",
    "        \n",
    "# Nucleosomes\n",
    "overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "            overlap_with_his_marks(filtered_seclast_exons_coords_df, nucleosomes_bed)\n",
    "    \n",
    "ex_row = ['exon', 'sec_last', 'unk-nucleosomes', overlapping_bases, total_bases, lower_quart, n_seqs] \n",
    "exon_position_results.append(ex_row)\n",
    "\n",
    "overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "            overlap_with_his_marks(filtered_last_exons_coords_df, nucleosomes_bed)\n",
    "\n",
    "ex_row = ['exon', 'last', 'unk-nucleosomes', overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "overlapping_bases, total_bases, lower_quart, n_seqs = \\\n",
    "    overlap_with_his_marks(filtered_intron_coords_df, nucleosomes_bed)\n",
    "\n",
    "intr_row = ['intron', 'last', 'unk-nucleosomes', overlapping_bases, total_bases, lower_quart, n_seqs]\n",
    "\n",
    "exon_position_results.append(ex_row)\n",
    "exon_position_results.append(intr_row)\n",
    "        \n",
    "    \n",
    "exon_position_results_df = pd.DataFrame(exon_position_results)\n",
    "exon_position_results_df.columns = ['region', 'position', 'hist_cell', 'covered_bp', 'total_bp', 'lower_quartile', 'n_elements']\n",
    "exon_position_results_df['rate'] = exon_position_results_df['covered_bp']/exon_position_results_df['total_bp']\n",
    "\n",
    "exon_position_results_df.to_csv(path.join('results', 'histon_mark_and_nucleosomes_per_position.tsv'), \n",
    "                                sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scores\n",
    "\n",
    "Compute the mean, median and standard deviation of the coverage of exons and introns. Additionally, the Mann–Whitney U test is also used to compute a p-value of the comparison of both distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def return_cell(x):    \n",
    "    return x['index'].split('-')[0]\n",
    "\n",
    "def return_hist(x):    \n",
    "    return x['index'].split('-')[1]\n",
    "\n",
    "global_results_lol = list()\n",
    "\n",
    "stratified_df = pd.read_csv(path.join('results', 'histon_mark_and_nucleosomes_per_position.tsv'), sep='\\t')\n",
    "\n",
    "stratified_df['cov'] = stratified_df['covered_bp']/stratified_df['total_bp']\n",
    "\n",
    "# Analyse per pair of histon mark and cell line\n",
    "for index, my_df in stratified_df.groupby('hist_cell'):\n",
    "    \n",
    "    exons_df = my_df[my_df['region'] == 'exon']\n",
    "    introns_df = my_df[my_df['region'] == 'intron']\n",
    "    \n",
    "    exons_cov_mean = exons_df['cov'].mean()\n",
    "    introns_cov_mean = introns_df['cov'].mean()\n",
    "    total_cov_mean = my_df['cov'].mean()\n",
    "    \n",
    "    exons_cov_median = exons_df['cov'].median()\n",
    "    introns_cov_median = introns_df['cov'].median()\n",
    "    total_cov_median = my_df['cov'].median()\n",
    "    \n",
    "    exons_cov_std = exons_df['cov'].std()\n",
    "    introns_cov_std = introns_df['cov'].std()\n",
    "    total_cov_std = my_df['cov'].std()\n",
    "    \n",
    "    u, p_value = mannwhitneyu(exons_df['cov'], introns_df['cov'])\n",
    "    \n",
    "    results_row = [index, exons_cov_mean, introns_cov_mean, total_cov_mean,\n",
    "                   exons_cov_median, introns_cov_median, total_cov_median, \n",
    "                   exons_cov_std, introns_cov_std, total_cov_std, u, p_value]\n",
    "\n",
    "    global_results_lol.append(results_row)\n",
    "\n",
    "global_results_df = pd.DataFrame(global_results_lol)\n",
    "global_results_df.columns = ['index', 'exons_cov_mean', 'introns_cov_mean', 'total_cov_mean',\n",
    "                   'exons_cov_median', 'introns_cov_median', 'total_cov_median', \n",
    "                   'exons_cov_std', 'introns_cov_std', 'total_cov_std', 'u', 'p_value']\n",
    "\n",
    "global_results_df['cell_line'] = global_results_df.apply(lambda x: return_cell(x), axis=1)\n",
    "global_results_df['hist_mark'] = global_results_df.apply(lambda x: return_hist(x), axis=1)\n",
    "\n",
    "global_results_df['ex_in_mean_diff'] = global_results_df['exons_cov_mean'] - global_results_df['introns_cov_mean']\n",
    "global_results_df['ex_in_mean_reat'] = global_results_df['exons_cov_mean']/global_results_df['introns_cov_mean']\n",
    "\n",
    "global_results_df['ex_in_median_diff'] = global_results_df['exons_cov_median'] - global_results_df['introns_cov_median']\n",
    "global_results_df['ex_in_median_reat'] = global_results_df['exons_cov_median']/global_results_df['introns_cov_median']\n",
    "\n",
    "global_results_df['q_value'] = smm.multipletests(global_results_df['p_value'].tolist(), alpha=0.05, method='fdr_i', is_sorted=False, returnsorted=False)[1]\n",
    "\n",
    "global_results_df.to_csv(path.join('results', 'score_per_histon_mark_and_nucleosome.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure S5 A\n",
    "\n",
    "Coverage of H3K36me3 peaks in exons and introns of H1 stem cell (E003) and nucleosome covered regions in lymphoblastoid cell line (GM12878)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "x_order = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11-15', '16-20', '21-40', '41-100']\n",
    "\n",
    "# Read data\n",
    "coverage_df = pd.read_csv(path.join('results', \"histon_mark_and_nucleosomes_per_position.tsv\"), sep=\"\\t\")\n",
    "coverage_df.drop_duplicates(inplace=True)\n",
    "scores_df = pd.read_csv(path.join('results', \"score_per_histon_mark_and_nucleosome.tsv\"), sep=\"\\t\")\n",
    "scores_df.drop_duplicates(inplace=True)\n",
    "\n",
    "fig = plt.figure(figsize=(7, 4.25))\n",
    "\n",
    "rcolors={'exon':'#52a7f9', 'intron':'#6fc041'}\n",
    "\n",
    "axs=[]\n",
    "NROW=3\n",
    "NCOL=5\n",
    "\n",
    "for row in range(0, NROW):\n",
    "    axs.append(plt.subplot2grid((NROW, NCOL), (row, 0), colspan=4))\n",
    "    axs.append(plt.subplot2grid((NROW, NCOL), (row, 4)))  \n",
    "\n",
    "ymax=0.45 # ylimit\n",
    "ymin=0\n",
    "\n",
    "\n",
    "# first panel bar plot\n",
    "count=0\n",
    "for histone in ['E003-H3K36me3', 'Nucleosomes']:\n",
    "\n",
    "    if histone != 'Nucleosomes':\n",
    "        x_values, x_labels, x_colors, box_plot, wpvalue = get_values_barplot(coverage_df, scores_df, histone)\n",
    "    else:\n",
    "        x_values, x_labels, x_colors, box_plot, wpvalue = get_values_barplot(coverage_df, scores_df, 'unk-nucleosomes')\n",
    "\n",
    "    # generate x tick count\n",
    "    tmp_x = []\n",
    "    for i in range(0, len(x_values)):\n",
    "        tmp_x.append(i)\n",
    "    \n",
    "    # do the bar plot    \n",
    "    axs[count].bar(tmp_x, x_values, color=x_colors)\n",
    "    axs[count].set_ylim(ymin, ymax)\n",
    "    # show the left spline\n",
    "    for axis in ['right', 'top', 'bottom']:\n",
    "        axs[count].spines[axis].set_visible(False)\n",
    "    \n",
    "    axs[count].yaxis.major.locator.set_params(nbins=5)\n",
    "    axs[count].yaxis.set_ticks_position('left')\n",
    "    axs[count].xaxis.set_ticks_position('none')\n",
    "    axs[count].set_ylabel(histone)\n",
    "    \n",
    "    if histone == 'Nucleosomes':\n",
    "        print(x_labels)\n",
    "        axs[count].set_xticklabels(['', ''] + x_labels, rotation=90)\n",
    "        axs[count].xaxis.major.locator.set_params(nbins=len(x_labels)*2)\n",
    "        \n",
    "    else:\n",
    "        axs[count].axes.xaxis.set_ticklabels([])\n",
    "\n",
    "    \n",
    "    count+=1\n",
    "\n",
    "    # first panel box plot\n",
    "    flierprops = dict(marker='o', markersize=3, linestyle='none', mew=0.5) # outlier size\n",
    "    medianprops = dict(linewidth=0.7)\n",
    "    boxprops = capprops = whiskerprops = dict(linewidth=0.7) # outer linewidth\n",
    "    \n",
    "    #axs[count].boxplot([box_plot['exon'], box_plot['intron']], widths=(0.3, 0.3), flierprops=flierprops,\n",
    "    #                  medianprops = medianprops, boxprops=boxprops, capprops=capprops, whiskerprops=whiskerprops)\n",
    "    axs[count].violinplot([box_plot['exon'], box_plot['intron']],\n",
    "                   showmeans=False,\n",
    "                   showmedians=True, showextrema=True)\n",
    "    axs[count].set_ylim(ymin, ymax)\n",
    "    \n",
    "    # show the left spline\n",
    "    for axis in ['right', 'top', 'bottom', 'left']:\n",
    "        axs[count].spines[axis].set_visible(False)\n",
    "\n",
    "    axs[count].margins(0.05)\n",
    "    axs[count].axhline(0, color='gray', lw=0.5)\n",
    "\n",
    "    axs[count].yaxis.major.locator.set_params(nbins=8)\n",
    "    axs[count].xaxis.set_ticks_position('none')\n",
    "    axs[count].yaxis.set_ticks_position('none')\n",
    "    axs[count].axes.yaxis.set_ticklabels([])\n",
    "    axs[count].axes.xaxis.set_ticklabels([])\n",
    "    \n",
    "    if histone == 'Nucleosomes':\n",
    "        axs[count].set_xticklabels(['', 'exons', 'introns'], rotation=90)\n",
    "\n",
    "    axs[count].set_title(\"%.5f\" % wpvalue)\n",
    "    count+=1\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(path.join('figures', 'figure1A.svg'), bbox_inches='tight' )\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute sequences coverage\n",
    "\n",
    "For each position in the exons (centered) compute the ratio of sequences covered by peaks of H3k36me3 and nucleosome feature in stem cell H1 vs. the total number of sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "middle_distance_threshold = 2500\n",
    "\n",
    "# Read exon coordinates and their middle values and usable ranges\n",
    "\n",
    "#Read coordinates\n",
    "exons_coords_df = pd.read_csv(middle_exons_coords, sep=\"\\t\", low_memory=False)\n",
    "exons_coords_df.columns = ['chr', 'start', 'end', 'ensembl', 'gene', 'symbol']\n",
    "exons_coords_df = exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "exons_coords_symbol_df = exons_coords_df\n",
    "\n",
    "#Define middle of the exon\n",
    "exons_coords_symbol_df['exon_size'] = (exons_coords_symbol_df['end']\n",
    "                                   - exons_coords_symbol_df['start'])\n",
    "\n",
    "exons_coords_symbol_df['exon_middle_start'] = (exons_coords_symbol_df['start']\n",
    "                                             + exons_coords_symbol_df['exon_size']/2)\n",
    "\n",
    "exons_coords_symbol_df['exon_middle_start'] = exons_coords_symbol_df.apply(lambda x:\n",
    "                                                   math.floor(x['exon_middle_start']),\n",
    "                                                   axis=1)\n",
    "\n",
    "exons_coords_symbol_df['exon_middle_end'] = exons_coords_symbol_df['exon_middle_start'] + 1\n",
    "\n",
    "#Expand 2500bp at each side\n",
    "exons_coords_symbol_df['region_start'] = exons_coords_symbol_df['exon_middle_start'] - middle_distance_threshold\n",
    "exons_coords_symbol_df['region_end'] = exons_coords_symbol_df['exon_middle_end'] + middle_distance_threshold\n",
    "\n",
    "sub_exons_coords = exons_coords_symbol_df[['chr', 'region_start', 'region_end', 'ensembl',\n",
    "                                      'exon_size', 'exon_middle_start',\n",
    "                                      'exon_middle_end']]\n",
    "\n",
    "sub_exons_coords_bed = pybedtools.BedTool.from_dataframe(sub_exons_coords)\n",
    "n_exon_coords = len(sub_exons_coords_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Compute histone coverage per bp at the coordinates defined above. \n",
    "\n",
    "histon_cell = 'E003-H3K36me3'\n",
    "\n",
    "histone_data_path = histone_data_folder + histon_cell + '.narrowPeak.gz'\n",
    "histone_data_df = pd.read_csv(histone_data_path, sep='\\t', header=None)\n",
    "\n",
    "histone_data_df = histone_data_df[[0,1,2]]\n",
    "\n",
    "histon_mark_bed = pybedtools.BedTool.from_dataframe(histone_data_df)\n",
    "\n",
    "my_histon_intersec_bed = sub_exons_coords_bed.intersect(histon_mark_bed, wao=True)\n",
    "\n",
    "histon_marks_in_range = pd.read_table(my_histon_intersec_bed.fn, names = ['range_chr', 'range_start', 'range_end',\n",
    "                        'ensembl','region_size', 'region_middle_start', 'region_middle_end',\n",
    "                        'hist_chr', 'hist_start', 'hist_end', 'overlap_bp'])\n",
    "\n",
    "histon_marks_in_range['relative_start'] = histon_marks_in_range['hist_start'] - histon_marks_in_range['region_middle_start']\n",
    "histon_marks_in_range['relative_end'] = histon_marks_in_range['hist_end'] - histon_marks_in_range['region_middle_start']\n",
    "\n",
    "overlapping_histon_marks = histon_marks_in_range[histon_marks_in_range['overlap_bp'] != 0]\n",
    "\n",
    "all_reads_in_range = (middle_distance_threshold*2+1)*[0]\n",
    "\n",
    "total_region = len(histon_marks_in_range)\n",
    "\n",
    "#Collapse all exons together by their middle position\n",
    "for index, my_histon_row in tqdm(overlapping_histon_marks.iterrows()):\n",
    "\n",
    "    range_start = my_histon_row.range_start\n",
    "    range_end = my_histon_row.range_end\n",
    "\n",
    "    hist_start = my_histon_row.hist_start\n",
    "    hist_end = my_histon_row.hist_end\n",
    "\n",
    "    middle_start = my_histon_row.region_middle_start\n",
    "    middle_end = my_histon_row.region_middle_end\n",
    "\n",
    "    overlap_start = max([range_start, hist_start])\n",
    "    overlap_end = min([range_end, hist_end])\n",
    "\n",
    "    relative_start = overlap_start - middle_start\n",
    "    relative_end = overlap_end + 1 - middle_end\n",
    "\n",
    "    initial_bases = relative_start - ( - middle_distance_threshold)\n",
    "    middle_bases = relative_end - relative_start\n",
    "    final_bases = middle_distance_threshold + 1 - relative_end\n",
    "\n",
    "    initial_reads = [0]*initial_bases\n",
    "    midle_reads = [1]*middle_bases\n",
    "    final_reads = [0]*final_bases\n",
    "\n",
    "    my_reads_in_range = initial_reads + midle_reads + final_reads\n",
    "\n",
    "    a=all_reads_in_range\n",
    "    b=my_reads_in_range\n",
    "\n",
    "    all_reads_in_range = [x+y for x,y in zip(a,b)]\n",
    "\n",
    "all_reads_in_range_df = pd.DataFrame(all_reads_in_range)\n",
    "\n",
    "all_reads_in_range_df['position'] = list(range(- middle_distance_threshold,\n",
    "                                           middle_distance_threshold + 1))\n",
    "\n",
    "all_reads_in_range_df.columns = ['reads_count', 'position']\n",
    "all_reads_in_range_df = all_reads_in_range_df[['position', 'reads_count']]\n",
    "\n",
    "all_reads_in_range_df['reads_count'] = all_reads_in_range_df['reads_count']/n_exon_coords\n",
    "\n",
    "exon_reads_in_range = all_reads_in_range_df.copy()\n",
    "exon_reads_in_range.to_csv(path.join('results', \"%s_exon.csv\" % histon_cell), \n",
    "                           sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Compute nucleosome coverage per bp at the coordinates defined above. \n",
    "\n",
    "histon_cell = 'Nucleosome'\n",
    "\n",
    "# Get the nucleosomes coordinates\n",
    "nucl_coords_df = pd.read_csv(nucl_coords_path, sep='\\t', header=None)\n",
    "nucl_coords_df = nucl_coords_df[nucl_coords_df[3] >= 2.5]\n",
    "nucl_coords_df = nucl_coords_df[[0,1,2]]\n",
    "nucl_coords_df.columns = ['chr', 'start', 'end']\n",
    "# expand to cover the 147 bases of the nucleosome\n",
    "nucl_coords_df['start'] = nucl_coords_df['start'] - 73\n",
    "nucl_coords_df['end'] = nucl_coords_df['end'] + 73\n",
    "nucl_coords_df = nucl_coords_df[~nucl_coords_df['start'] < 0]\n",
    "\n",
    "histon_mark_bed = pybedtools.BedTool.from_dataframe(nucl_coords_df)\n",
    "\n",
    "my_histon_intersec_bed = sub_exons_coords_bed.intersect(histon_mark_bed, wao=True)\n",
    "\n",
    "histon_marks_in_range = pd.read_table(my_histon_intersec_bed.fn, names = ['range_chr', 'range_start', 'range_end',\n",
    "                        'ensembl','region_size', 'region_middle_start', 'region_middle_end',\n",
    "                        'hist_chr', 'hist_start', 'hist_end', 'overlap_bp'])\n",
    "\n",
    "histon_marks_in_range['relative_start'] = histon_marks_in_range['hist_start'] - histon_marks_in_range['region_middle_start']\n",
    "histon_marks_in_range['relative_end'] = histon_marks_in_range['hist_end'] - histon_marks_in_range['region_middle_start']\n",
    "\n",
    "overlapping_histon_marks = histon_marks_in_range[histon_marks_in_range['overlap_bp'] != 0]\n",
    "\n",
    "all_reads_in_range = (middle_distance_threshold*2+1)*[0]\n",
    "\n",
    "total_region = len(histon_marks_in_range)\n",
    "\n",
    "#Collapse all exons together by their middle position\n",
    "for index, my_histon_row in tqdm(overlapping_histon_marks.iterrows()):\n",
    "\n",
    "    range_start = my_histon_row.range_start\n",
    "    range_end = my_histon_row.range_end\n",
    "\n",
    "    hist_start = my_histon_row.hist_start\n",
    "    hist_end = my_histon_row.hist_end\n",
    "\n",
    "    middle_start = my_histon_row.region_middle_start\n",
    "    middle_end = my_histon_row.region_middle_end\n",
    "\n",
    "    overlap_start = max([range_start, hist_start])\n",
    "    overlap_end = min([range_end, hist_end])\n",
    "\n",
    "    relative_start = overlap_start - middle_start\n",
    "    relative_end = overlap_end + 1 - middle_end\n",
    "\n",
    "    initial_bases = relative_start - ( - middle_distance_threshold)\n",
    "    middle_bases = relative_end - relative_start\n",
    "    final_bases = middle_distance_threshold + 1 - relative_end\n",
    "\n",
    "    initial_reads = [0]*initial_bases\n",
    "    midle_reads = [1]*middle_bases\n",
    "    final_reads = [0]*final_bases\n",
    "\n",
    "    my_reads_in_range = initial_reads + midle_reads + final_reads\n",
    "\n",
    "    a=all_reads_in_range\n",
    "    b=my_reads_in_range\n",
    "\n",
    "    all_reads_in_range = [x+y for x,y in zip(a,b)]\n",
    "\n",
    "all_reads_in_range_df = pd.DataFrame(all_reads_in_range)\n",
    "histon_cell\n",
    "all_reads_in_range_df['position'] = list(range(- middle_distance_threshold,\n",
    "                                           middle_distance_threshold + 1))\n",
    "\n",
    "all_reads_in_range_df.columns = ['reads_count', 'position']\n",
    "all_reads_in_range_df = all_reads_in_range_df[['position', 'reads_count']]\n",
    "\n",
    "all_reads_in_range_df['reads_count'] = all_reads_in_range_df['reads_count']/n_exon_coords\n",
    "\n",
    "exon_reads_in_range = all_reads_in_range_df.copy()\n",
    "exon_reads_in_range.to_csv(path.join('results', \"%s_exon.csv\" % histon_cell), \n",
    "                           sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Supplementary Figure S5 B\n",
    "\n",
    "Internal exons-centered coverage of H3K36me3 peaks in H1 stem cells (E003)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read the data from the computed files and make the plot\n",
    "\n",
    "middle_distance_threshold = 2000\n",
    "\n",
    "histon_cell = 'E003-H3K36me3'\n",
    "\n",
    "exon_reads_in_range = pd.read_csv(path.join('results', \"%s_exon.csv\" % histon_cell), sep=\"\\t\", header=0)\n",
    "\n",
    "my_max = max(exon_reads_in_range['reads_count'])\n",
    "my_min = min(exon_reads_in_range['reads_count'])\n",
    "my_diff = my_max-my_min\n",
    "my_upper_lim  = my_max + my_diff/10\n",
    "my_lower_lim = my_min - my_diff/10\n",
    "\n",
    "fig = plt.figure(figsize=(3.25, 2.25))\n",
    "\n",
    "# exon centered plot\n",
    "ax = fig.add_subplot(111)\n",
    "\n",
    "ax.margins(0.05)\n",
    "\n",
    "ax.plot(exon_reads_in_range['position'], exon_reads_in_range['reads_count'], linewidth=1.5, color='#8da0cb')\n",
    "ax.set_ylim(my_lower_lim, my_upper_lim)\n",
    "\n",
    "ax.set_facecolor('none')\n",
    "\n",
    "ax.set_frame_on(True)\n",
    "# show the left spline\n",
    "for axis in ['right', 'top']:\n",
    "    ax.spines[axis].set_visible(False)\n",
    "\n",
    "for axis in ['bottom', 'left']:\n",
    "    ax.spines[axis].set_position(('axes', -0.03))\n",
    "\n",
    "ax.xaxis.set_ticks_position('none') \n",
    "ax.tick_params(direction='out', pad=3)\n",
    "ax.yaxis.set_ticks_position('left')\n",
    "ax.yaxis.major.locator.set_params(nbins=4)\n",
    "ax.xaxis.major.locator.set_params(nbins=4)\n",
    "ax.xaxis.set_ticks_position('bottom')\n",
    "ax.set_xlim(-middle_distance_threshold, middle_distance_threshold)\n",
    "plt.title('')\n",
    "plt.xlabel('Distance from exon center (bp)')\n",
    "plt.ylabel('Proportion of bases covered')\n",
    "\n",
    "\n",
    "plt.xlabel('Distance from exon center (bp)')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(path.join('figures', 'figure2A.svg'))\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Supplementary Figure S5 C\n",
    "\n",
    "Internal exons-centered coverage of nucleosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read the data from the computed files and make the plot\n",
    "\n",
    "middle_distance_threshold = 2000\n",
    "\n",
    "histon_cell = 'Nucleosome'\n",
    "\n",
    "exon_reads_in_range = pd.read_csv(path.join('results', \"%s_exon.csv\" % histon_cell), sep=\"\\t\", header=0)\n",
    "\n",
    "my_max = max(exon_reads_in_range['reads_count'])\n",
    "my_min = min(exon_reads_in_range['reads_count'])\n",
    "my_diff = my_max-my_min\n",
    "my_upper_lim  = my_max + my_diff/10\n",
    "my_lower_lim = my_min - my_diff/10\n",
    "\n",
    "fig = plt.figure(figsize=(3.25, 2.25))\n",
    "\n",
    "# exon centered plot\n",
    "ax = fig.add_subplot(111)\n",
    "\n",
    "ax.margins(0.05)\n",
    "\n",
    "ax.plot(exon_reads_in_range['position'], exon_reads_in_range['reads_count'], linewidth=1.5, color='#8da0cb')\n",
    "ax.set_ylim(my_lower_lim, my_upper_lim)\n",
    "\n",
    "ax.set_facecolor('none')\n",
    "\n",
    "ax.set_frame_on(True)\n",
    "# show the left spline\n",
    "for axis in ['right', 'top']:\n",
    "    ax.spines[axis].set_visible(False)\n",
    "\n",
    "for axis in ['bottom', 'left']:\n",
    "    ax.spines[axis].set_position(('axes', -0.03))\n",
    "\n",
    "ax.xaxis.set_ticks_position('none') \n",
    "ax.tick_params(direction='out', pad=3)\n",
    "ax.yaxis.set_ticks_position('left')\n",
    "ax.yaxis.major.locator.set_params(nbins=4)\n",
    "ax.xaxis.major.locator.set_params(nbins=4)\n",
    "ax.xaxis.set_ticks_position('bottom')\n",
    "ax.set_xlim(-middle_distance_threshold, middle_distance_threshold)\n",
    "plt.title('')\n",
    "plt.xlabel('Distance from exon center (bp)')\n",
    "plt.ylabel('Proportion of bases covered')\n",
    "\n",
    "\n",
    "plt.xlabel('Distance from exon center (bp)')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(path.join('figures', 'figure2B.svg'))\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
