{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Synonymous vs non-synonymous internal exon-centered analysis across context models\n",
    "\n",
    "The aim of this notebook is to find the mixed effects of extending the sequence context dependency within the internal exon-analysis by computing the expected under a 12-way spectrum model (1-mer model), 18-way spectrum model (1-mer considering CpG sites), and extended context 3-mer, 5-mer and 7-mer, separating mutation in synonymous and nonsynonymous. This information is used to build Supplementary Tables S4 and S5.\n",
    "\n",
    "## Output\n",
    "\n",
    "The output of the notebook is multiple intermediate files containing numbers of observed and expected mutations and several plots.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "from os import path\n",
    "import math\n",
    "import pybedtools\n",
    "import tabix  # package pytabix\n",
    "from bgreference import hg19\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import numpy.polynomial.polynomial as poly\n",
    "import matplotlib.ticker as plticker\n",
    "import gzip, pickle\n",
    "import statistics as st\n",
    "\n",
    "from python_scripts.plot_functions import config_plot_params\n",
    "config_plot_params()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input files\n",
    "\n",
    "Files in **data** directory.\n",
    "\n",
    "- *middle_exons_coords*: file with the coordinates of the middle exons\n",
    "- *consequence_type_file*: tabix file with the consequence types of exonic mutations\n",
    "- *consequence_ranking_file*: file with information about the ensembl hierarchy in predicted functional consequences\n",
    "\n",
    "Files in **results** directory.\n",
    "\n",
    "- *kmer_signatures.txt*: files with the mutational signatures computed from whole genome mutations all DNMs or dataset specific, given a kmer model.\n",
    "\n",
    "### Other inputs\n",
    "\n",
    "- mutations_folder: base directory where the files with the mutations are located\n",
    "- tumor_type: germinal in our case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Path to data\n",
    "middle_exons_coords = 'data/coordinates/genes_middle_exon_coords.bed.gz'\n",
    "consequence_type_file = 'data/consequence/consequence_ranking.tsv.bgz'\n",
    "consequence_ranking_file = 'data/consequence/consequence_ranking_info.tsv'\n",
    "\n",
    "# Other\n",
    "mutations_folder = 'non_provided_data/mutations/'\n",
    "tumor_type = 'germinal'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions\n",
    "\n",
    "All the functions needed for this notebook are coded below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def remove_version_ensembl(x, colname):\n",
    "    \"\"\"\n",
    "    Remove version from the ensembl ID\n",
    "\n",
    "    Args:\n",
    "        x: dataframe row\n",
    "        colname: column name\n",
    "\n",
    "    Returns:\n",
    "        str. Text before a '.'\n",
    "\n",
    "    \"\"\"\n",
    "    ensembl_id = x[colname]\n",
    "    ensembl_id_only = ensembl_id.split('.')[0]\n",
    "\n",
    "    return ensembl_id_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def synonymous_or_not(x):\n",
    "    \"\"\"\n",
    "    Classifies consequence types between synonymous and non synonymous\n",
    "    using the RANK column\n",
    "\n",
    "    Args:\n",
    "        x: dataframe row\n",
    "\n",
    "    Returns:\n",
    "        str. *synonymous* or *non_synonymous*\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    my_rank = x['RANK']\n",
    "\n",
    "    if my_rank < 15:\n",
    "        my_consequence = 'non_synonymous'\n",
    "\n",
    "    else:\n",
    "        my_consequence = 'synonymous'\n",
    "\n",
    "    return my_consequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This function was made for testing purposes:\n",
    "def check_consequence_TEST(chromosome, start_coord, alt, tabix):\n",
    "    \"\"\"\n",
    "    Function that classifies a given exonic mutation into the VEP classification given\n",
    "    tabix file with predicted effect.\n",
    "\n",
    "    Args:\n",
    "        chromosome: chromosome coordinate\n",
    "        start_coord: coordinate of the mutation\n",
    "        alt: alternative nucleotide\n",
    "        tabix: tabix indexed file containing the information\n",
    "        rank_info: dictionary with all the types of predicted effect and if we classify them as synonymous or not.\n",
    "    Returns:\n",
    "        str. Class of mutation on VEP classification\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    my_chr = chromosome[3:]\n",
    "    effects = tabix.querys(\"{}:{}-{}\".format(my_chr, start_coord, start_coord))\n",
    "    for effect in effects:\n",
    "        if len(effect) < 6:\n",
    "            continue\n",
    "        if effect[3] == alt:\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            return(consequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def check_consequence(chromosome, start_coord, alt, tabix, rank_info):\n",
    "    \"\"\"\n",
    "    Function that classifies a given exonic mutation into synonymous \n",
    "    or not given a rank info and a tabix file with predicted effect.\n",
    "\n",
    "    Args:\n",
    "        chromosome: chromosome coordinate\n",
    "        start_coord: coordinate of the mutation\n",
    "        alt: alternative nucleotide\n",
    "        tabix: tabix indexed file containing the predicted VEP classification effect for a given position and change.\n",
    "        rank_info: dictionary with all the types of predicted effect and if we classify them as synonymous or not.\n",
    "    Returns:\n",
    "        str. *synonymous* or *non_synonymous*\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    my_chr = chromosome[3:]\n",
    "    effects = tabix.querys(\"{}:{}-{}\".format(my_chr, start_coord, start_coord))\n",
    "    for effect in effects:\n",
    "        if len(effect) < 6:\n",
    "            continue\n",
    "        if effect[3] == alt:\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            try:\n",
    "                con = rank_info[consequence]\n",
    "                return(con)\n",
    "            except:\n",
    "                return('None')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This function was made for testing purposes:\n",
    "def countX(lst, x): \n",
    "    count = 0\n",
    "    for ele in lst: \n",
    "        if (ele == x): \n",
    "            count = count + 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def remove_muts(mut_data, IDs_list):\n",
    "    \"\"\"\n",
    "    Function that removes mutations from a data frame, only if the mutation matches and ID\n",
    "    and it is enclosed in the exon coordinates of the specific 2001-nt window.\n",
    "\n",
    "    Args:\n",
    "        mut_data: dataframe with columns \"range_chr range_start range_end ensembl region_size\n",
    "                    region_middle_start region_middle_end exon_start exon_end mut_chr mut_start\n",
    "                    mut_end mut_ref mut_alt mut_sample mut_type mut_ID overlap_bp\n",
    "        IDs_list: list of IDs that must be removed\n",
    "    Returns:\n",
    "        pandas data frame with mutations with the ID removed, they sould be removed only once if the IDs are repeated.\n",
    "\n",
    "    \"\"\"\n",
    "    rows_to_remove = []\n",
    "    for ID in IDs_list:\n",
    "        sub_frame = mut_data[mut_data['mut_ID'] == ID]\n",
    "        for index, row in sub_frame.iterrows():\n",
    "            if (row['exon_start'] <= row['mut_end'] and row['exon_end'] >= row['mut_end']):\n",
    "                rows_to_remove.append(index)\n",
    "    return(mut_data.drop(rows_to_remove))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_intron_probs(sequence, signatures_dict, rel_start, middle_distance_threshold, tumor_type, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given intronic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        sequence: intronic sequence. It should contain one extra nucleotide on each side to compute all the\n",
    "        trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and\n",
    "        associated probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    position_prob = []\n",
    "    \n",
    "    ## Divide sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_alt_bases = nucleotides - set(my_ref_base) #The rest are alternative ones\n",
    "        my_base_probs = 0\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "            \n",
    "        ### Foreach alternative, each of the three possible changes given a kmer.\n",
    "        for alt_base in my_alt_bases:\n",
    "            #### Built reference and alternate kmers\n",
    "            tri_ref = previous_base + str(my_ref_base) + next_base\n",
    "            tri_alt = previous_base + str(alt_base) + next_base\n",
    "            #### Create key for the signature dictionary\n",
    "            my_key = str((tri_ref, tri_alt))\n",
    "            try:\n",
    "                ##### Add the probability of the three possible changes\n",
    "                my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                my_base_probs = my_base_probs + my_prob\n",
    "            except:\n",
    "                None\n",
    "                \n",
    "        ### Add the position in the 2001-nt window with the correspondent probability to a list\n",
    "        normalized_pos = rel_start - middle_distance_threshold\n",
    "        my_row = [normalized_pos, my_base_probs]\n",
    "        position_prob.append(my_row)\n",
    "        rel_start += 1\n",
    "        \n",
    "    return(position_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_exon_probs(chromosome, abs_start, sequence, signatures_dict, rel_start, middle_distance_threshold,\n",
    "                       tumor_type, rank_info, tabix, include, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given exonic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        chromosome: chromosome where the sequence is located in hg19 genome.\n",
    "        abs_start: absolute start of the exonic sequence in hg19 coordinates.\n",
    "        sequence: exonic sequence. It should contain one extra nucleotide on each side to compute all the trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "        rank_info: dictionary containing the rank classification (syn or non-syn) of an specific VEP class.\n",
    "        tabix: tabix indexed file containing the predicted VEP classification effect for a given position and change.\n",
    "        inlcude(str): *synonymous* or *non_syonymous* . Which type od change consider to add probabilities.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and associated\n",
    "        probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    position_prob = []\n",
    "    my_chr = chromosome[3:]\n",
    "    \n",
    "    ## Divide sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_base_probs = 0\n",
    "        #Get previous and next base\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        effects = tabix.querys(\"{}:{}-{}\".format(my_chr, abs_start + 1, abs_start + 1))\n",
    "        for effect in effects:\n",
    "            if len(effect) < 6:\n",
    "                print('At chromosome ' + str(my_chr) + ' position ' + str(abs_start + 1) + ' no effect is reported.')\n",
    "                continue\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            try:\n",
    "                synonymous_or_not = rank_info[consequence]\n",
    "                if synonymous_or_not == include:\n",
    "                    #Built reference and alternative trinucleotide\n",
    "                    tri_ref = previous_base + str(my_ref_base) + next_base\n",
    "                    tri_alt = previous_base + str(effect[3]) + next_base\n",
    "                    #Create key for the signatue dictionary.\n",
    "                    my_key = str((tri_ref, tri_alt))\n",
    "                    try:\n",
    "                        #Get the probability\n",
    "                        my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                        #Sum the probability of the three possible changes.\n",
    "                        my_base_probs = my_base_probs + my_prob\n",
    "                    except:\n",
    "                        None\n",
    "            except:\n",
    "                None\n",
    "                \n",
    "        ### Add the position in the 2001-nt window with the correspondent probability in a list\n",
    "        normalized_pos = rel_start - middle_distance_threshold\n",
    "        my_row = [normalized_pos, my_base_probs]\n",
    "        position_prob.append(my_row)\n",
    "        abs_start += 1\n",
    "        rel_start += 1\n",
    "    \n",
    "    return(position_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                           k, include, tb_consequence_type, consequence_rank_dict):\n",
    "    \"\"\"\n",
    "    Get positions of the exons that correspond to the central position and a number of base pair\n",
    "    equal to middle_distance_threshold on each side. Compute the number of mutations per position under analysis.\n",
    "    For each of the relative positions, compute the sum of the probabilities of each possible alteration\n",
    "    taking into account its context (using the values of the signatures_file) wheighted by the number of observed \n",
    "    mutations (MODIFIED FROM FRIGOLA ET AL.).\n",
    "    You can exclude synonymous or non-synonymous mutations by specifying it.\n",
    "    \n",
    "    Args:\n",
    "        mutations_df:\n",
    "        exons_coords_symbol_df:\n",
    "        signatures_dict:\n",
    "        tumor_type:\n",
    "        cluster_id: \n",
    "        middle_distance_threshold:\n",
    "        include:\n",
    "        tb_consequence_type:\n",
    "        consequence_rank_dict: \n",
    "\n",
    "    Returns:\n",
    "        :class:`~pandas.DataFrame`. Table with the relative position (to the exon center)\n",
    "        the number of mutations observed and expected in that position.\n",
    "\n",
    "    \"\"\"\n",
    "    exons_coords_symbol_df = exons_coords_df.copy(deep=True)\n",
    "    \n",
    "    # Get signatures\n",
    "    all_signatures = pd.read_csv(signatures_file, sep='\\t')\n",
    "    probability_name = 'Probability_' + tumor_type\n",
    "    sub_signatures_df = all_signatures[['mutation', probability_name]]\n",
    "    signatures_dict = sub_signatures_df.set_index('mutation').T.to_dict()\n",
    "    \n",
    "    # Convert mutations into bed to intersect values\n",
    "    mutations_bed = pybedtools.BedTool.from_dataframe(mutations_df)\n",
    "    \n",
    "    # Convert exonic coordinates into bed to intersect values\n",
    "    exons_coords_symbol_df = exons_coords_symbol_df[['chr', 'start', 'end', 'ensembl']]\n",
    "    exon_coords_bed = pybedtools.BedTool.from_dataframe(exons_coords_symbol_df)\n",
    "\n",
    "    # Process exonic coordinates to get the 2001-nt region coords and convert into bed\n",
    "    ## Positions that correspond to the center of the exons\n",
    "    exons_coords_symbol_df['exon_size'] = (exons_coords_symbol_df['end'] - exons_coords_symbol_df['start'])\n",
    "    exons_coords_symbol_df['exon_middle_start'] = (exons_coords_symbol_df['start'] + exons_coords_symbol_df['exon_size']/2)\n",
    "    exons_coords_symbol_df['exon_middle_start'] = exons_coords_symbol_df.apply(lambda x:\n",
    "                                                                       math.floor(x['exon_middle_start']), axis=1)\n",
    "    \n",
    "    ## Positions around the middle point\n",
    "    exons_coords_symbol_df['exon_middle_end'] = exons_coords_symbol_df['exon_middle_start'] + 1\n",
    "    exons_coords_symbol_df['region_start'] = exons_coords_symbol_df['exon_middle_start'] - middle_distance_threshold\n",
    "    exons_coords_symbol_df['region_end'] = exons_coords_symbol_df['exon_middle_end'] + middle_distance_threshold\n",
    "    region_coords = exons_coords_symbol_df[['chr', 'region_start', 'region_end', 'ensembl',\n",
    "                                          'exon_size', 'exon_middle_start', 'exon_middle_end', 'start', 'end']]\n",
    "    region_coords_bed = pybedtools.BedTool.from_dataframe(region_coords)\n",
    "\n",
    "    # Filter mutations by the full region of interest\n",
    "    my_bed = region_coords_bed.intersect(mutations_bed, wao=True)\n",
    "\n",
    "    mutations_in_range = pd.read_table(my_bed.fn, names = ['range_chr', 'range_start', 'range_end',\n",
    "                        'ensembl','region_size', 'region_middle_start', 'region_middle_end',\n",
    "                        'exon_start', 'exon_end', 'mut_chr', 'mut_start', 'mut_end', 'mut_ref', 'mut_alt',\n",
    "                        'mut_sample', 'mut_type', 'mut_ID', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_range = mutations_in_range[mutations_in_range['overlap_bp'] != 0]\n",
    "    print(\"The total amount of mutations at the middle exon-centered sequences is \" + str(len(mutations_in_range.index)))\n",
    "\n",
    "    # Filter mutations by the exonic regions\n",
    "    my_bed2 = exon_coords_bed.intersect(mutations_bed, wao=True)\n",
    "    mutations_in_range2 = pd.read_table(my_bed2.fn, names = ['exon_chr', 'exon_start', 'exon_end',\n",
    "                        'ensembl', 'mut_chr', 'mut_start', 'mut_end', 'mut_ref', 'mut_alt',\n",
    "                        'mut_sample', 'mut_type', 'mut_ID', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_range2 = mutations_in_range2[mutations_in_range2['overlap_bp'] != 0]\n",
    "    mutations_in_range2 = mutations_in_range2[['mut_chr', 'mut_start', 'mut_end', 'mut_ref', \n",
    "                        'mut_alt', 'mut_sample', 'mut_type', 'mut_ID']]\n",
    "    mutations_in_range2.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'ID']\n",
    "    print(\"The total amount of mutations at the exonic sequences are \" + str(len(mutations_in_range2.index)))\n",
    "\n",
    "    #Find the amount of mutations that fall in each VEP category (TESTING PURPOSES)\n",
    "\n",
    "    mutations_in_range2['category'] = mutations_in_range2.apply(lambda x: \n",
    "                           check_consequence_TEST(x['chr'], x['end'], x['alt'], tb_consequence_type),\n",
    "                           axis=1)\n",
    "    all_elements = mutations_in_range2['category'].tolist()\n",
    "    unique_elements = set(all_elements)\n",
    "    for n in unique_elements:\n",
    "        count = countX(all_elements, n)\n",
    "        print (\"The category \" + str(n) + \" is found \" + str(count) + \" times.\")\n",
    "\n",
    "    # Find if mutations at exonic regions are synonymous or not.\n",
    "    mutations_in_range2['conseq'] = mutations_in_range2.apply(lambda x: \n",
    "                       check_consequence(x['chr'], x['end'], x['alt'], tb_consequence_type, consequence_rank_dict),\n",
    "                       axis=1)\n",
    "    \n",
    "    # Check the number of mutations that were not classified (they would not be included,\n",
    "    # but is important to know the amount of drops):\n",
    "\n",
    "    #mutations_in_range2[(mutations_in_range2['conseq'] != 'synonymous')\n",
    "    #& (mutations_in_range2['conseq'] != 'non_synonymous')]\n",
    "\n",
    "    # Find the amount of mutations that fall in both synonymous and non-synonymous categories.\n",
    "    # Additionally, check the number of mutations that were not classified (they would not be included,\n",
    "    # but is important to know the amount of drops):\n",
    "    #all_elements = mutations_in_range2['conseq'].tolist()\n",
    "    #unique_elements = set(all_elements)\n",
    "    #for n in unique_elements:\n",
    "    #    count = countX(all_elements, n)\n",
    "    #    print (\"The category \" + str(n) + \" is found \" + str(count) + \" times.\")\n",
    "\n",
    "\n",
    "    # Get the index of exonic non synonymous mutations and mutations without category \n",
    "    # to remove them from the mutation dataset at the 2001-nt sequences\n",
    "    muts_to_drop = mutations_in_range2[mutations_in_range2['conseq'] != include]\n",
    "    IDs_to_drop = muts_to_drop['ID'].tolist()\n",
    "    mutations_in_range = remove_muts(mutations_in_range, IDs_to_drop)\n",
    "\n",
    "    # Count observed mutations\n",
    "    ## Create a dictionary of all observed positions at the range of the middle exon-centered window\n",
    "    mutations_in_range['relative_start'] = mutations_in_range['mut_start'] - mutations_in_range['region_middle_start']\n",
    "    exon_counts = mutations_in_range['relative_start'].value_counts().to_dict()\n",
    "\n",
    "    ## Set count to 0 were no mutation has been found\n",
    "    for i in range(-middle_distance_threshold,middle_distance_threshold):\n",
    "        if i not in exon_counts.keys():\n",
    "            exon_counts[i] = 0\n",
    "    \n",
    "    ## Count mutations for every position with reported counts\n",
    "    my_exon_results_lol = list()\n",
    "    for my_position in exon_counts.keys():\n",
    "        my_count = exon_counts[my_position]\n",
    "        my_exon_results_lol.append([my_position, my_count])\n",
    "\n",
    "    my_exon_results_df = pd.DataFrame(my_exon_results_lol)\n",
    "    my_exon_results_df.columns = ['position', 'muts_count']\n",
    "    my_exon_results_df = my_exon_results_df.sort_values(by='position')\n",
    "\n",
    "    # Compute expected mutations\n",
    "    ## For each exonic region under analysis get its relative position to the center of the exon\n",
    "    ## and add the probability that is the sum of the probabilities of the changes\n",
    "    final_position_prob_df = pd.DataFrame()\n",
    "    for my_row in region_coords.values.tolist(): ###For each middle exon-centered sequence:\n",
    "        position_prob_lol = list()\n",
    "        \n",
    "        ## Get coordinates\n",
    "        my_chr = my_row[0]\n",
    "        my_start = int(my_row[1])\n",
    "        my_start_exon = int(my_row[7])\n",
    "        my_end = int(my_row[2])\n",
    "        my_end_exon = int(my_row[8])\n",
    "        my_exon_size = int(my_row[4])\n",
    "        n_bases_fullregion = my_end - my_start\n",
    "        n_bases_first_flank = my_start_exon - my_start\n",
    "        n_bases_second_flank = my_end - my_end_exon\n",
    "        \n",
    "        ## Delimit mutations in the range of that specific middle exon-centered sequence\n",
    "        my_region_muts = len(mutations_in_range[(mutations_in_range['range_start'] >= my_start) &\n",
    "                               (mutations_in_range['range_end'] <= my_end) &\n",
    "                               (mutations_in_range['range_chr'] == my_chr)])\n",
    "        \n",
    "        ## If no mutations are reported for that specific middle exon-centered sequence, skip to the next one.\n",
    "        if my_region_muts == 0:\n",
    "            continue\n",
    "        ## If mutations are reported, get the exonic sequence and the two intronic flanking sequences:\n",
    "        if my_start_exon <= my_start:\n",
    "            my_exon_region = hg19(my_chr, my_start+1-(k//2), size=n_bases_fullregion+2*(k//2))\n",
    "            position_prob_lol = compute_exon_probs(my_chr, my_start, my_exon_region, signatures_dict, 0,\n",
    "                                    middle_distance_threshold, tumor_type, consequence_rank_dict,\n",
    "                                    tb_consequence_type, include, k)\n",
    "\n",
    "        else:\n",
    "            first_flank_region = hg19(my_chr, my_start+1-(k//2), size=n_bases_first_flank+2*(k//2))\n",
    "            first_flank_region_list = compute_intron_probs(first_flank_region, signatures_dict, 0, \n",
    "                                    middle_distance_threshold, tumor_type, k)\n",
    "            \n",
    "            second_flank_region = hg19(my_chr, my_end_exon+1-(k//2), size=n_bases_second_flank+2*(k//2))\n",
    "            second_flank_region_list = compute_intron_probs(second_flank_region, signatures_dict,\n",
    "                                    n_bases_first_flank + my_exon_size, middle_distance_threshold, tumor_type, k)\n",
    "            \n",
    "            my_exon_region = hg19(my_chr, my_start_exon+1-(k//2), size=my_exon_size+2*(k//2))\n",
    "            exon_region_list = compute_exon_probs(my_chr, my_start_exon, my_exon_region, signatures_dict,\n",
    "                                    n_bases_first_flank, middle_distance_threshold, tumor_type, consequence_rank_dict,\n",
    "                                    tb_consequence_type, include, k)\n",
    "            \n",
    "            position_prob_lol = first_flank_region_list + exon_region_list + second_flank_region_list\n",
    "    \n",
    "        ## Convert the list to a dataframe\n",
    "        position_prob_df = pd.DataFrame(position_prob_lol)\n",
    "        position_prob_df.columns = ['position', 'probability']\n",
    "        \n",
    "        ## Reescaling probabilities to one\n",
    "        my_total = sum(position_prob_df['probability'])\n",
    "        position_prob_df['probability'] = position_prob_df['probability']/my_total\n",
    "        \n",
    "        ## Compute the expected by redistributing mutations acoording to probabilities\n",
    "        position_prob_df['expected_muts'] = position_prob_df['probability']*my_region_muts\n",
    "        position_prob_df = position_prob_df[['position', 'expected_muts']]\n",
    "        \n",
    "        # Sum the expected results\n",
    "        if len(final_position_prob_df) == 0:\n",
    "            final_position_prob_df = position_prob_df\n",
    "        else:\n",
    "            final_position_prob_df['expected_muts'] = (final_position_prob_df['expected_muts'] +\n",
    "                                                       position_prob_df['expected_muts'])\n",
    "\n",
    "\n",
    "    return pd.merge(my_exon_results_df, final_position_prob_df, on='position')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_CpG_site(my_ref, my_pre, my_post):\n",
    "    \"\"\"\n",
    "    Compute if mutation falls on CpG site\n",
    "    \"\"\"\n",
    "    if my_ref == 'C' and my_post == 'G':\n",
    "        my_site = 'CpG'\n",
    "    elif my_ref == 'G' and my_pre == 'C':\n",
    "        my_site = 'GpC'\n",
    "    elif my_ref == 'C' and my_post != 'G':\n",
    "        my_site = 'nonCpG'\n",
    "    elif my_ref == 'G' and my_pre != 'C':\n",
    "        my_site = 'nonGpC'\n",
    "    elif my_ref == 'A':\n",
    "        my_site = 'A'\n",
    "    elif my_ref == 'T':\n",
    "        my_site = 'T'\n",
    "    return my_site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_intron_probs_CpG(sequence, signatures_dict, rel_start, middle_distance_threshold, tumor_type, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given intronic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        sequence: intronic sequence. It should contain one extra nucleotide on each side to compute all the\n",
    "        trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and\n",
    "        associated probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    position_prob = []\n",
    "    \n",
    "    ## Divide sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_alt_bases = nucleotides - set(my_ref_base) #The rest are alternative ones\n",
    "        my_base_probs = 0\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        tri_ref = compute_CpG_site(my_ref_base, previous_base, next_base)\n",
    "            \n",
    "        ### Foreach alternative, each of the three possible changes given a kmer.\n",
    "        for alt_base in my_alt_bases:\n",
    "            #### Built reference and alternate kmers\n",
    "            tri_alt = alt_base\n",
    "            #### Create key for the signature dictionary\n",
    "            my_key = str((tri_ref, tri_alt))\n",
    "            try:\n",
    "                ##### Add the probability of the three possible changes\n",
    "                my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                my_base_probs = my_base_probs + my_prob\n",
    "            except:\n",
    "                None\n",
    "                \n",
    "        ### Add the position in the 2001-nt window with the correspondent probability to a list\n",
    "        normalized_pos = rel_start - middle_distance_threshold\n",
    "        my_row = [normalized_pos, my_base_probs]\n",
    "        position_prob.append(my_row)\n",
    "        rel_start += 1\n",
    "        \n",
    "    return(position_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_exon_probs_CpG(chromosome, abs_start, sequence, signatures_dict, rel_start, middle_distance_threshold,\n",
    "                       tumor_type, rank_info, tabix, include, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given exonic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        chromosome: chromosome where the sequence is located in hg19 genome.\n",
    "        abs_start: absolute start of the exonic sequence in hg19 coordinates.\n",
    "        sequence: exonic sequence. It should contain one extra nucleotide on each side to compute all the trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "        rank_info: dictionary containing the rank classification (syn or non-syn) of an specific VEP class.\n",
    "        tabix: tabix indexed file containing the predicted VEP classification effect for a given position and change.\n",
    "        inlcude(str): *synonymous* or *non_syonymous* . Which type od change consider to add probabilities.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and associated\n",
    "        probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    position_prob = []\n",
    "    my_chr = chromosome[3:]\n",
    "    \n",
    "    ## Divide sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_base_probs = 0\n",
    "        #Get previous and next base\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        tri_ref = compute_CpG_site(my_ref_base, previous_base, next_base)\n",
    "        effects = tabix.querys(\"{}:{}-{}\".format(my_chr, abs_start + 1, abs_start + 1))\n",
    "        for effect in effects:\n",
    "            if len(effect) < 6:\n",
    "                print('At chromosome ' + str(my_chr) + ' position ' + str(abs_start + 1) + ' no effect is reported.')\n",
    "                continue\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            try:\n",
    "                synonymous_or_not = rank_info[consequence]\n",
    "                if synonymous_or_not == include:\n",
    "                    #Built reference and alternative trinucleotide\n",
    "                    tri_alt = str(effect[3])\n",
    "                    #Create key for the signatue dictionary.\n",
    "                    my_key = str((tri_ref, tri_alt))\n",
    "                    try:\n",
    "                        #Get the probability\n",
    "                        my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                        #Sum the probability of the three possible changes.\n",
    "                        my_base_probs = my_base_probs + my_prob\n",
    "                    except:\n",
    "                        None\n",
    "            except:\n",
    "                None\n",
    "                \n",
    "        ### Add the position in the 2001-nt window with the correspondent probability in a list\n",
    "        normalized_pos = rel_start - middle_distance_threshold\n",
    "        my_row = [normalized_pos, my_base_probs]\n",
    "        position_prob.append(my_row)\n",
    "        abs_start += 1\n",
    "        rel_start += 1\n",
    "    \n",
    "    return(position_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                           k, include, tb_consequence_type, consequence_rank_dict):\n",
    "    \"\"\"\n",
    "    Get positions of the exons that correspond to the central position and a number of base pair\n",
    "    equal to middle_distance_threshold on each side. Compute the number of mutations per position under analysis.\n",
    "    For each of the relative positions, compute the sum of the probabilities of each possible alteration\n",
    "    taking into account its context (using the values of the signatures_file) wheighted by the number of observed \n",
    "    mutations (MODIFIED FROM FRIGOLA ET AL.).\n",
    "    You can exclude synonymous or non-synonymous mutations by specifying it.\n",
    "    \n",
    "    Args:\n",
    "        mutations_df:\n",
    "        exons_coords_symbol_df:\n",
    "        signatures_dict:\n",
    "        tumor_type:\n",
    "        cluster_id: \n",
    "        middle_distance_threshold:\n",
    "        include:\n",
    "        tb_consequence_type:\n",
    "        consequence_rank_dict: \n",
    "\n",
    "    Returns:\n",
    "        :class:`~pandas.DataFrame`. Table with the relative position (to the exon center)\n",
    "        the number of mutations observed and expected in that position.\n",
    "\n",
    "    \"\"\"\n",
    "    exons_coords_symbol_df = exons_coords_df.copy(deep=True)\n",
    "    \n",
    "    # Get signatures\n",
    "    all_signatures = pd.read_csv(signatures_file, sep='\\t')\n",
    "    probability_name = 'Probability_' + tumor_type\n",
    "    sub_signatures_df = all_signatures[['mutation', probability_name]]\n",
    "    signatures_dict = sub_signatures_df.set_index('mutation').T.to_dict()\n",
    "    \n",
    "    # Convert mutations into bed to intersect values\n",
    "    mutations_bed = pybedtools.BedTool.from_dataframe(mutations_df)\n",
    "\n",
    "    # Convert exonic coordinates into bed to intersect values\n",
    "    exons_coords_symbol_df = exons_coords_symbol_df[['chr', 'start', 'end', 'ensembl']]\n",
    "    exon_coords_bed = pybedtools.BedTool.from_dataframe(exons_coords_symbol_df)\n",
    "\n",
    "    # Process exonic coordinates to get the 2001-nt region coords and convert into bed\n",
    "    ## Positions that correspond to the center of the exons\n",
    "    exons_coords_symbol_df['exon_size'] = (exons_coords_symbol_df['end'] - exons_coords_symbol_df['start'])\n",
    "    exons_coords_symbol_df['exon_middle_start'] = (exons_coords_symbol_df['start'] + exons_coords_symbol_df['exon_size']/2)\n",
    "    exons_coords_symbol_df['exon_middle_start'] = exons_coords_symbol_df.apply(lambda x:\n",
    "                                                                       math.floor(x['exon_middle_start']), axis=1)\n",
    "    \n",
    "    ## Positions around the middle point\n",
    "    exons_coords_symbol_df['exon_middle_end'] = exons_coords_symbol_df['exon_middle_start'] + 1\n",
    "    exons_coords_symbol_df['region_start'] = exons_coords_symbol_df['exon_middle_start'] - middle_distance_threshold\n",
    "    exons_coords_symbol_df['region_end'] = exons_coords_symbol_df['exon_middle_end'] + middle_distance_threshold\n",
    "    region_coords = exons_coords_symbol_df[['chr', 'region_start', 'region_end', 'ensembl',\n",
    "                                          'exon_size', 'exon_middle_start', 'exon_middle_end', 'start', 'end']]\n",
    "    region_coords_bed = pybedtools.BedTool.from_dataframe(region_coords)\n",
    "\n",
    "    # Filter mutations by the full region of interest\n",
    "    my_bed = region_coords_bed.intersect(mutations_bed, wao=True)\n",
    "\n",
    "    mutations_in_range = pd.read_table(my_bed.fn, names = ['range_chr', 'range_start', 'range_end',\n",
    "                        'ensembl','region_size', 'region_middle_start', 'region_middle_end',\n",
    "                        'exon_start', 'exon_end', 'mut_chr', 'mut_start', 'mut_end', 'mut_ref', 'mut_alt',\n",
    "                        'mut_sample', 'mut_type', 'mut_ID', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_range = mutations_in_range[mutations_in_range['overlap_bp'] != 0]\n",
    "    print(\"The total amount of mutations at the middle exon-centered sequences is \" + str(len(mutations_in_range.index)))\n",
    "\n",
    "    # Filter mutations by the exonic regions\n",
    "    my_bed2 = exon_coords_bed.intersect(mutations_bed, wao=True)\n",
    "    mutations_in_range2 = pd.read_table(my_bed2.fn, names = ['exon_chr', 'exon_start', 'exon_end',\n",
    "                        'ensembl', 'mut_chr', 'mut_start', 'mut_end', 'mut_ref', 'mut_alt',\n",
    "                        'mut_sample', 'mut_type', 'mut_ID', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    mutations_in_range2 = mutations_in_range2[mutations_in_range2['overlap_bp'] != 0]\n",
    "    mutations_in_range2 = mutations_in_range2[['mut_chr', 'mut_start', 'mut_end', 'mut_ref', \n",
    "                        'mut_alt', 'mut_sample', 'mut_type', 'mut_ID']]\n",
    "    mutations_in_range2.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'ID']\n",
    "    print(\"The total amount of mutations at the exonic sequences are \" + str(len(mutations_in_range2.index)))\n",
    "\n",
    "    #Find the amount of mutations that fall in each VEP category (TESTING PURPOSES)\n",
    "\n",
    "    mutations_in_range2['category'] = mutations_in_range2.apply(lambda x: \n",
    "                           check_consequence_TEST(x['chr'], x['end'], x['alt'], tb_consequence_type),\n",
    "                           axis=1)\n",
    "    all_elements = mutations_in_range2['category'].tolist()\n",
    "    unique_elements = set(all_elements)\n",
    "    for n in unique_elements:\n",
    "        count = countX(all_elements, n)\n",
    "        print (\"The category \" + str(n) + \" is found \" + str(count) + \" times.\")\n",
    "\n",
    "    # Find if mutations at exonic regions are synonymous or not.\n",
    "    mutations_in_range2['conseq'] = mutations_in_range2.apply(lambda x: \n",
    "                       check_consequence(x['chr'], x['end'], x['alt'], tb_consequence_type, consequence_rank_dict),\n",
    "                       axis=1)\n",
    "    \n",
    "    # Check the number of mutations that were not classified (they would not be included,\n",
    "    # but is important to know the amount of drops):\n",
    "\n",
    "    #mutations_in_range2[(mutations_in_range2['conseq'] != 'synonymous')\n",
    "    #& (mutations_in_range2['conseq'] != 'non_synonymous')]\n",
    "\n",
    "    # Find the amount of mutations that fall in both synonymous and non-synonymous categories.\n",
    "    # Additionally, check the number of mutations that were not classified (they would not be included,\n",
    "    # but is important to know the amount of drops):\n",
    "    #all_elements = mutations_in_range2['conseq'].tolist()\n",
    "    #unique_elements = set(all_elements)\n",
    "    #for n in unique_elements:\n",
    "    #    count = countX(all_elements, n)\n",
    "    #    print (\"The category \" + str(n) + \" is found \" + str(count) + \" times.\")\n",
    "\n",
    "\n",
    "    # Get the index of exonic non synonymous mutations and mutations without category \n",
    "    # to remove them from the mutation dataset at the 2001-nt sequences\n",
    "    muts_to_drop = mutations_in_range2[mutations_in_range2['conseq'] != include]\n",
    "    IDs_to_drop = muts_to_drop['ID'].tolist()\n",
    "    mutations_in_range = remove_muts(mutations_in_range, IDs_to_drop)\n",
    "\n",
    "    # Count observed mutations\n",
    "    ## Create a dictionary of all observed positions at the range of the middle exon-centered window\n",
    "    mutations_in_range['relative_start'] = mutations_in_range['mut_start'] - mutations_in_range['region_middle_start']\n",
    "    exon_counts = mutations_in_range['relative_start'].value_counts().to_dict()\n",
    "\n",
    "    ## Set count to 0 were no mutation has been found\n",
    "    for i in range(-middle_distance_threshold,middle_distance_threshold):\n",
    "        if i not in exon_counts.keys():\n",
    "            exon_counts[i] = 0\n",
    "    \n",
    "    ## Count mutations for every position with reported counts\n",
    "    my_exon_results_lol = list()\n",
    "    for my_position in exon_counts.keys():\n",
    "        my_count = exon_counts[my_position]\n",
    "        my_exon_results_lol.append([my_position, my_count])\n",
    "\n",
    "    my_exon_results_df = pd.DataFrame(my_exon_results_lol)\n",
    "    my_exon_results_df.columns = ['position', 'muts_count']\n",
    "    my_exon_results_df = my_exon_results_df.sort_values(by='position')\n",
    "\n",
    "    # Compute expected mutations\n",
    "    ## For each exonic region under analysis get its relative position to the center of the exon\n",
    "    ## and add the probability that is the sum of the probabilities of the changes\n",
    "    final_position_prob_df = pd.DataFrame()\n",
    "    for my_row in region_coords.values.tolist(): ###For each middle exon-centered sequence:\n",
    "        position_prob_lol = list()\n",
    "        \n",
    "        ## Get coordinates\n",
    "        my_chr = my_row[0]\n",
    "        my_start = int(my_row[1])\n",
    "        my_start_exon = int(my_row[7])\n",
    "        my_end = int(my_row[2])\n",
    "        my_end_exon = int(my_row[8])\n",
    "        my_exon_size = int(my_row[4])\n",
    "        n_bases_fullregion = my_end - my_start\n",
    "        n_bases_first_flank = my_start_exon - my_start\n",
    "        n_bases_second_flank = my_end - my_end_exon\n",
    "        \n",
    "        ## Delimit mutations in the range of that specific middle exon-centered sequence\n",
    "        my_region_muts = len(mutations_in_range[(mutations_in_range['range_start'] >= my_start) &\n",
    "                               (mutations_in_range['range_end'] <= my_end) &\n",
    "                               (mutations_in_range['range_chr'] == my_chr)])\n",
    "        \n",
    "        ## If no mutations are reported for that specific middle exon-centered sequence, skip to the next one.\n",
    "        if my_region_muts == 0:\n",
    "            continue\n",
    "        ## If mutations are reported, get the exonic sequence and the two intronic flanking sequences:\n",
    "        if my_start_exon <= my_start:\n",
    "            my_exon_region = hg19(my_chr, my_start+1-(k//2), size=n_bases_fullregion+2*(k//2))\n",
    "            position_prob_lol = compute_exon_probs(my_chr, my_start, my_exon_region, signatures_dict, 0,\n",
    "                                    middle_distance_threshold, tumor_type, consequence_rank_dict,\n",
    "                                    tb_consequence_type, include, k)\n",
    "\n",
    "        else:\n",
    "            first_flank_region = hg19(my_chr, my_start+1-(k//2), size=n_bases_first_flank+2*(k//2))\n",
    "            first_flank_region_list = compute_intron_probs(first_flank_region, signatures_dict, 0, \n",
    "                                    middle_distance_threshold, tumor_type, k)\n",
    "            \n",
    "            second_flank_region = hg19(my_chr, my_end_exon+1-(k//2), size=n_bases_second_flank+2*(k//2))\n",
    "            second_flank_region_list = compute_intron_probs(second_flank_region, signatures_dict,\n",
    "                                    n_bases_first_flank + my_exon_size, middle_distance_threshold, tumor_type, k)\n",
    "            \n",
    "            my_exon_region = hg19(my_chr, my_start_exon+1-(k//2), size=my_exon_size+2*(k//2))\n",
    "            exon_region_list = compute_exon_probs(my_chr, my_start_exon, my_exon_region, signatures_dict,\n",
    "                                    n_bases_first_flank, middle_distance_threshold, tumor_type, consequence_rank_dict,\n",
    "                                    tb_consequence_type, include, k)\n",
    "            \n",
    "            position_prob_lol = first_flank_region_list + exon_region_list + second_flank_region_list\n",
    "    \n",
    "        ## Convert the list to a dataframe\n",
    "        position_prob_df = pd.DataFrame(position_prob_lol)\n",
    "        position_prob_df.columns = ['position', 'probability']\n",
    "        \n",
    "        ## Reescaling probabilities to one\n",
    "        my_total = sum(position_prob_df['probability'])\n",
    "        position_prob_df['probability'] = position_prob_df['probability']/my_total\n",
    "        \n",
    "        ## Compute the expected by redistributing mutations acoording to probabilities\n",
    "        position_prob_df['expected_muts'] = position_prob_df['probability']*my_region_muts\n",
    "        position_prob_df = position_prob_df[['position', 'expected_muts']]\n",
    "        \n",
    "        # Sum the expected results\n",
    "        if len(final_position_prob_df) == 0:\n",
    "            final_position_prob_df = position_prob_df\n",
    "        else:\n",
    "            final_position_prob_df['expected_muts'] = (final_position_prob_df['expected_muts'] +\n",
    "                                                       position_prob_df['expected_muts'])\n",
    "\n",
    "\n",
    "    return pd.merge(my_exon_results_df, final_position_prob_df, on='position')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bin_average(count, size):\n",
    "    averaged = []\n",
    "    for i in range(0,len(count),size):\n",
    "        mean = np.mean(count[i:i+size])\n",
    "        averaged.extend([mean for i in range(len(count[i:i+size]))])\n",
    "    return(pd.Series(averaged))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_HD(cluster, ax, cnt):\n",
    "    \n",
    "    # read observed and backgruond data directly\n",
    "    filename = \"results/exons_centered_%s.tsv\" % (cluster)\n",
    "    \n",
    "    if path.isfile(filename):\n",
    "    \n",
    "        data = pd.read_csv(filename, sep=\"\\t\")\n",
    "    \n",
    "        coefs = poly.polyfit(data['position'], data['muts_count'], 50)\n",
    "        ffit = poly.polyval(data['position'], coefs)\n",
    "\n",
    "        ax.plot(data['position'],\n",
    "                 data['muts_count'], '#e41a1c', alpha=0.5, linewidth=0.7)\n",
    "        ax.plot(data['position'], ffit, '#e41a1c', linewidth=1)\n",
    "\n",
    "        coefs2 = poly.polyfit(data['position'], data['expected_muts'], 50)\n",
    "        ffit2 = poly.polyval(data['position'], coefs2)\n",
    "        ax.plot(data['position'], ffit2, 'black', linewidth=1, ls=\"--\")\n",
    "\n",
    "        ax.set_axis_bgcolor('none')\n",
    "\n",
    "        # show the left spline\n",
    "        for axis in ['right', 'top']:\n",
    "            ax.spines[axis].set_visible(False)\n",
    "\n",
    "        # add gap between x- and y-axis\n",
    "        for axis in ['bottom', 'left']:\n",
    "            ax.spines[axis].set_position(('axes', -0.03))\n",
    "\n",
    "        ax.yaxis.major.locator.set_params(nbins=4)\n",
    "        ax.yaxis.set_ticks_position('left')\n",
    "        ax.xaxis.set_ticks_position('bottom')\n",
    "        ax.set_xlim(-1000, 1000)\n",
    "        # add axis label\n",
    "        if cnt in [0, 3]:\n",
    "            ax.set_ylabel('Number of mutations')\n",
    "        if cnt >=3:\n",
    "            ax.set_xlabel('Distance from exon center (bp)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_HD_BINNED(cluster, ax, cnt, size):\n",
    "    \n",
    "    # read observed and backgruond data directly\n",
    "    filename = \"results/exons_centered_%s.tsv\" % (cluster)\n",
    "    \n",
    "    if path.isfile(filename):\n",
    "    \n",
    "        data = pd.read_csv(filename, sep=\"\\t\")\n",
    "        \n",
    "        bin_obs = bin_average(data['muts_count'], size)\n",
    "\n",
    "        ax.plot(data['position'],\n",
    "                 data['muts_count'], '#e41a1c', alpha=0.5, linewidth=0.7)\n",
    "        ax.plot(data['position'], bin_obs, '#e41a1c', linewidth=1)\n",
    "\n",
    "        bin_exp = bin_average(data['expected_muts'], size)\n",
    "\n",
    "        ax.plot(data['position'], bin_exp, 'black', linewidth=1)\n",
    "\n",
    "        ax.set_axis_bgcolor('none')\n",
    "\n",
    "        # show the left spline\n",
    "        for axis in ['right', 'top']:\n",
    "            ax.spines[axis].set_visible(False)\n",
    "\n",
    "        # add gap between x- and y-axis\n",
    "        for axis in ['bottom', 'left']:\n",
    "            ax.spines[axis].set_position(('axes', -0.03))\n",
    "\n",
    "        ax.yaxis.major.locator.set_params(nbins=4)\n",
    "        ax.yaxis.set_ticks_position('left')\n",
    "        ax.xaxis.set_ticks_position('bottom')\n",
    "        ax.set_xlim(-1000, 1000)\n",
    "        # add axis label\n",
    "        if cnt in [0, 3]:\n",
    "            ax.set_ylabel('Number of mutations')\n",
    "        if cnt >=3:\n",
    "            ax.set_xlabel('Distance from exon center (bp)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_intron_probs_permut(sequence,signatures_dict, tumor_type, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given intronic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        sequence: intronic sequence. It should contain one extra nucleotide on each side to compute all the\n",
    "        trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and\n",
    "        associated probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    position_prob_lol = list()\n",
    "    \n",
    "    ## Divide intronic sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "        \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_alt_bases = nucleotides - set(my_ref_base) #The rest are alternative ones\n",
    "        my_base_probs = 0\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "            \n",
    "        ### Foreach alternative, each of the three possible changes given a kmer.\n",
    "        for alt_base in my_alt_bases:\n",
    "            #### Built reference and alternate kmers\n",
    "            tri_ref = previous_base + str(my_ref_base) + next_base\n",
    "            tri_alt = previous_base + str(alt_base) + next_base\n",
    "            #### Create key for the signature dictionary\n",
    "            my_key = str((tri_ref, tri_alt))\n",
    "            try:\n",
    "                ##### Add the probability of the three possible changes\n",
    "                my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                my_base_probs = my_base_probs + my_prob\n",
    "            except:\n",
    "                None      \n",
    "                \n",
    "        position_prob_lol.append(my_base_probs)\n",
    "        \n",
    "    return position_prob_lol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_exon_probs_permut(chromosome, abs_start, sequence, signatures_dict, tumor_type, rank_info, tabix, include, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given exonic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and associated\n",
    "        probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    position_prob_lol = list()\n",
    "    \n",
    "    ## Divide intronic sequences into trinucleotides\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    #Foreach trinucleotide\n",
    "    for my_kmer in my_kmers:\n",
    "        #Get the reference\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper()\n",
    "        my_base_probs = 0\n",
    "        #Get previous and next base\n",
    "        previous_base = my_kmer[0:(k//2)]\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        effects = tabix.querys(\"{}:{}-{}\".format(chromosome, abs_start + 1, abs_start + 1))\n",
    "        for effect in effects:\n",
    "            if len(effect) < 6:\n",
    "                print('At chromosome ' + str(my_chr) + ' position ' + str(abs_start + 1) + ' no effect is reported.')\n",
    "                continue\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            try:\n",
    "                synonymous_or_not = rank_info[consequence]\n",
    "                if synonymous_or_not == include:\n",
    "                    #Built reference and alternative trinucleotide\n",
    "                    tri_ref = previous_base + str(my_ref_base) + next_base\n",
    "                    tri_alt = previous_base + str(effect[3]) + next_base\n",
    "                    #Create key for the signatue dictionary.\n",
    "                    my_key = str((tri_ref, tri_alt))\n",
    "                    try:\n",
    "                        #Get the probability\n",
    "                        my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                        #Sum the probability of the three possible changes.\n",
    "                        my_base_probs = my_base_probs + my_prob\n",
    "                    except:\n",
    "                        None\n",
    "            except:\n",
    "                None\n",
    "        \n",
    "        position_prob_lol.append(my_base_probs)\n",
    "        abs_start += 1\n",
    "    \n",
    "    return position_prob_lol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def stacked_sequence_level_permutation_excluding(mutations_df, exons_coords_df, signatures_file, tumor_type,\n",
    "                                middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict):\n",
    "    \"\"\"\n",
    "    Get the co-ordinates of exons and flanking intronic sequences and perform the random sampling of mutations\n",
    "    (1000 times) with the size equal of the number of mutation observed in the 2001-nt window (exon and intron\n",
    "    together).\n",
    "    \n",
    "    Args:\n",
    "        \n",
    "    Returns:\n",
    "        \n",
    "    \"\"\"\n",
    "    exons_coords_symbol_df = exons_coords_df.copy(deep=True)\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    \n",
    "    # Get signatures\n",
    "    all_signatures = pd.read_csv(signatures_file, sep='\\t')\n",
    "    probability_name = 'Probability_' + tumor_type\n",
    "    sub_signatures_df = all_signatures[['mutation', probability_name]]\n",
    "    signatures_dict = sub_signatures_df.set_index('mutation').T.to_dict()\n",
    "    \n",
    "    # Create bed of the selected mutations\n",
    "    sub_mutations_df = mutations_df[['chr', 'start', 'end', 'alt', 'sample']]\n",
    "    mutations_bed = pybedtools.BedTool.from_dataframe(sub_mutations_df)\n",
    "    \n",
    "    exons_coords_symbol_df['exon_size'] = (exons_coords_symbol_df['end'] - exons_coords_symbol_df['start'])\n",
    "    exons_coords_symbol_df['exon_middle_start'] = (exons_coords_symbol_df['start'] +\n",
    "                                        exons_coords_symbol_df['exon_size']/2)\n",
    "    exons_coords_symbol_df['exon_middle_start'] = exons_coords_symbol_df.apply(lambda x:\n",
    "                                        math.floor(x['exon_middle_start']), axis=1)\n",
    "\n",
    "    exons_coords_symbol_df['exon_middle_end'] = exons_coords_symbol_df['exon_middle_start'] + 1\n",
    "    exons_coords_symbol_df['region_start'] = exons_coords_symbol_df['exon_middle_start'] - middle_distance_threshold\n",
    "    exons_coords_symbol_df['region_end'] = exons_coords_symbol_df['exon_middle_end'] + middle_distance_threshold\n",
    "    \n",
    "    # Left flank \"intronic\" sequences => region_start and start from coordinates.\n",
    "    intron1_df = exons_coords_symbol_df[['chr', 'region_start', 'start', 'ID']]\n",
    "    ## The region may span less than the centered exonic sequence. Only take left flank intronic coordinates with sense.\n",
    "    intron1_df = intron1_df[intron1_df['region_start'] <= intron1_df['start']]\n",
    "    intron1_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    \n",
    "    # Right flank \"intronic\" sequences\n",
    "    intron2_df = exons_coords_symbol_df[['chr', 'end', 'region_end', 'ID']]\n",
    "    ## The region may span less than the centered exonic sequence. Only take right flank intronic coordinates with sense.\n",
    "    intron2_df = intron2_df[intron2_df['end'] <= intron2_df['region_end']]\n",
    "    intron2_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## Take them together\n",
    "    intron_df = pd.concat([intron1_df, intron2_df])\n",
    "    intron_df['label'] = 'intron'\n",
    "\n",
    "    # Exonic sequences\n",
    "    all_exon_df = exons_coords_symbol_df[['chr', 'start', 'end', 'ID', 'region_start', 'region_end']]\n",
    "    ## The region may span less than the centered exonic sequence. Process exons that surpass the region size.\n",
    "    large_exon_df = all_exon_df[(all_exon_df['end']-all_exon_df['start']) > (2*middle_distance_threshold+1)]\n",
    "    large_exon_df = large_exon_df[['chr', 'region_start', 'region_end', 'ID']]\n",
    "    large_exon_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## The region may span less than the centered exonic sequence. Process exons that do not surpass the region size.\n",
    "    short_exon_df = all_exon_df[(all_exon_df['end']-all_exon_df['start']) <= (2*middle_distance_threshold+1)]\n",
    "    short_exon_df = short_exon_df[['chr', 'start', 'end', 'ID']]\n",
    "    short_exon_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## Take them together\n",
    "    exon_df = pd.concat([short_exon_df, large_exon_df])\n",
    "    exon_df['label'] = 'exon'\n",
    "    \n",
    "    # Merge exons and introns\n",
    "    coords_symbol_df = pd.concat([exon_df, intron_df])\n",
    "    coords_symbol_bed = pybedtools.BedTool.from_dataframe(coords_symbol_df)\n",
    "\n",
    "    # Filter mutations by the position of interest\n",
    "    my_bed = coords_symbol_bed.intersect(mutations_bed, wo=True)\n",
    "\n",
    "    mutations_mapped = pd.read_table(my_bed.fn, names = ['chr', 'start', 'end', 'ID','label', 'mut_chr', 'mut_start',\n",
    "                                    'mut_end', 'mut_alt', 'sample', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    \n",
    "    mutations_mapped['conseq'] = mutations_mapped.apply(lambda x: \n",
    "                                    check_consequence(x['mut_chr'], x['mut_end'], x['mut_alt'], tb_consequence_type,\n",
    "                                    consequence_rank_dict), axis=1)\n",
    "    \n",
    "    print(\"Total amount of mutations at the middle exon-centered sequences is \"+ str(len(mutations_mapped)))\n",
    "    exonic_obs = mutations_mapped[(mutations_mapped['label'] == 'exon') & \n",
    "                                  (mutations_mapped['conseq'] == include)].shape[0]\n",
    "    intronic_obs = mutations_mapped[mutations_mapped['label'] == 'intron'].shape[0]\n",
    "    print(\"Observed at middle exons: \" + str(exonic_obs))\n",
    "    print(\"Observed at flanking introns: \" + str(intronic_obs))\n",
    "    \n",
    "    # Compute expected mutation by randomization approach\n",
    "    RANDOMIZATION = 1000\n",
    "\n",
    "    collect_rand = {}\n",
    "    # Initialize dictionary\n",
    "    for sam in range(0, RANDOMIZATION):\n",
    "        collect_rand[sam] = {}\n",
    "        collect_rand[sam]['exon_Exp'] = 0\n",
    "        collect_rand[sam]['intron_Exp'] = 0\n",
    "\n",
    "    # For each unique staked exon-centered sequence with one or more mutations\n",
    "    for gene in mutations_mapped['ID'].unique():\n",
    "        position_prob_lol = []\n",
    "        label_flag_lol = []\n",
    "        \n",
    "        # Take the exonic and the flanking intronic sequences.\n",
    "        for dx, row in coords_symbol_df[coords_symbol_df['ID']==gene].iterrows():\n",
    "            my_chr = row['chr'][3:]\n",
    "            my_start = int(row['start'])\n",
    "            my_end = int(row['end'])\n",
    "            n_bases = my_end - my_start\n",
    "\n",
    "            ## Get sequence.\n",
    "            my_sequence = hg19(my_chr, my_start+1-(k//2), size=n_bases+2*(k//2))\n",
    "\n",
    "            if row['label'] == 'intron':\n",
    "                label_flag = 0\n",
    "                probs = compute_intron_probs_permut(my_sequence, signatures_dict, tumor_type, k)\n",
    "                flags = [label_flag]*len(probs)\n",
    "            else:\n",
    "                label_flag = 1\n",
    "                probs = compute_exon_probs_permut(my_chr, my_start, my_sequence, signatures_dict, tumor_type,\n",
    "                                           consequence_rank_dict, tb_consequence_type, include, k)\n",
    "                flags = [label_flag]*len(probs)\n",
    "                \n",
    "            position_prob_lol = position_prob_lol + probs\n",
    "            label_flag_lol = label_flag_lol + flags\n",
    "        \n",
    "        # Total number of mutations falling in the stacked sequence.\n",
    "        mutcnt = len(mutations_mapped[(mutations_mapped['ID']==gene) &\n",
    "            ((mutations_mapped['label'] == 'intron') | ((mutations_mapped['label'] == 'exon') & \n",
    "                                  (mutations_mapped['conseq'] == include)))])\n",
    "        # Normalize the probability vector and perform sampling\n",
    "        prb_vector = np.array(position_prob_lol)\n",
    "        prb_vector = prb_vector/prb_vector.sum()\n",
    "        label_vector = np.array(label_flag_lol)\n",
    "        mutation_rand_label = np.random.choice(label_vector, size=mutcnt*RANDOMIZATION, replace=True, p=prb_vector)\n",
    "        \n",
    "        # Save the mutation count for each randomization\n",
    "        for sam in range(0, RANDOMIZATION):\n",
    "            start = int(mutcnt*sam)\n",
    "            # count for one random sampling\n",
    "            count = np.unique(mutation_rand_label[start:(start+mutcnt)], return_counts=True)\n",
    "            for i in range(0, len(count[0])):\n",
    "                if count[0][i] == 1: # count of exons\n",
    "                    collect_rand[sam]['exon_Exp'] += count[1][i]\n",
    "                else: # count of introns\n",
    "                    collect_rand[sam]['intron_Exp'] += count[1][i]\n",
    "        \n",
    "    collect_df = pd.DataFrame.from_dict(collect_rand)\n",
    "    \n",
    "    return(collect_df, exonic_obs, intronic_obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_intron_probs_permut_CpG(sequence, signatures_dict, tumor_type, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given intronic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "        sequence: intronic sequence. It should contain one extra nucleotide on each side to compute all the\n",
    "        trinucleotides.\n",
    "        signatures_dict: Dictionary containing probabilities of each trinucleotide change.\n",
    "        rel_start: Relative start of that sequence within the middle exon-centered window.\n",
    "        middle_distance_threshold: middle distance of the exon-centered window.\n",
    "        tumor_type: First type of class to find the specific mutation signature class.\n",
    "        cluster_id: Second type of class to find the specific mutation signature class.\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and\n",
    "        associated probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    position_prob_lol = list()\n",
    "    \n",
    "    ## Divide intronic sequences into kmers\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "        \n",
    "    ## Foreach kmer\n",
    "    for my_kmer in my_kmers:\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper() #Get the reference\n",
    "        my_alt_bases = nucleotides - set(my_ref_base) #The rest are alternative ones\n",
    "        my_base_probs = 0\n",
    "        previous_base = my_kmer[0:(k//2)] #Get previous and next base\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        tri_ref = compute_CpG_site(my_ref_base, previous_base, next_base)\n",
    "        \n",
    "        ### Foreach alternative, each of the three possible changes given a kmer.\n",
    "        for alt_base in my_alt_bases:\n",
    "            #### Built reference and alternate kmers\n",
    "            tri_alt = alt_base\n",
    "            #### Create key for the signature dictionary\n",
    "            my_key = str((tri_ref, tri_alt))\n",
    "            try:\n",
    "                ##### Add the probability of the three possible changes\n",
    "                my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                my_base_probs = my_base_probs + my_prob\n",
    "            except:\n",
    "                None\n",
    "                \n",
    "        position_prob_lol.append(my_base_probs)\n",
    "        \n",
    "    return position_prob_lol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_exon_probs_permut_CpG(chromosome, abs_start, sequence, signatures_dict, tumor_type, rank_info, tabix, include, k):\n",
    "    \"\"\"\n",
    "    Function that computes the summed probabilities of the three posible trinucleotide changes for each trinucleotide\n",
    "    in a given exonic sequence of a middle exon-centered window.\n",
    "    \n",
    "    Args:\n",
    "      \n",
    "    Returns:\n",
    "        list containing subslists with paired normalized position within the middle exon-centered window and associated\n",
    "        probability.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    position_prob_lol = list()\n",
    "    \n",
    "    ## Divide intronic sequences into trinucleotides\n",
    "    my_kmers = [sequence[i:i+k] for i in range(len(sequence)-(k-1))]\n",
    "    \n",
    "    #Foreach trinucleotide\n",
    "    for my_kmer in my_kmers:\n",
    "        #Get the reference\n",
    "        my_ref_base = my_kmer[(k//2):(k//2)+1].upper()\n",
    "        my_base_probs = 0\n",
    "        #Get previous and next base\n",
    "        previous_base = my_kmer[0:(k//2)]\n",
    "        next_base = my_kmer[(k//2)+1:]\n",
    "        tri_ref = compute_CpG_site(my_ref_base, previous_base, next_base)\n",
    "        effects = tabix.querys(\"{}:{}-{}\".format(chromosome, abs_start + 1, abs_start + 1))\n",
    "        for effect in effects:\n",
    "            if len(effect) < 6:\n",
    "                print('At chromosome ' + str(my_chr) + ' position ' + str(abs_start + 1) + ' no effect is reported.')\n",
    "                continue\n",
    "            consequence = effect[5].rstrip('\\r')\n",
    "            try:\n",
    "                synonymous_or_not = rank_info[consequence]\n",
    "                if synonymous_or_not == include:\n",
    "                    #Built reference and alternative trinucleotide\n",
    "                    tri_alt = str(effect[3])\n",
    "                    #Create key for the signatue dictionary.\n",
    "                    my_key = str((tri_ref, tri_alt))\n",
    "                    try:\n",
    "                        #Get the probability\n",
    "                        my_prob = signatures_dict[my_key]['Probability_' + tumor_type]\n",
    "                        #Sum the probability of the three possible changes.\n",
    "                        my_base_probs = my_base_probs + my_prob\n",
    "                    except:\n",
    "                        None\n",
    "            except:\n",
    "                None\n",
    "        \n",
    "        position_prob_lol.append(my_base_probs)\n",
    "        abs_start += 1\n",
    "    \n",
    "    return position_prob_lol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def stacked_sequence_level_permutation_excluding_CpG(mutations_df, exons_coords_df, signatures_file,\n",
    "                    tumor_type, middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict):\n",
    "    \"\"\"\n",
    "    Get the co-ordinates of exons and flanking intronic sequences and perform the random sampling of mutations\n",
    "    (1000 times) with the size equal of the number of mutation observed in the 2001-nt window (exon and intron\n",
    "    together).\n",
    "    \n",
    "    Args:\n",
    "        \n",
    "    Returns:\n",
    "        \n",
    "    \"\"\"\n",
    "    exons_coords_symbol_df = exons_coords_df.copy(deep=True)\n",
    "    \n",
    "    nucleotides = set(['A', 'T', 'C', 'G'])\n",
    "    \n",
    "    # Get signatures\n",
    "    all_signatures = pd.read_csv(signatures_file, sep='\\t')\n",
    "    probability_name = 'Probability_' + tumor_type\n",
    "    sub_signatures_df = all_signatures[['mutation', probability_name]]\n",
    "    signatures_dict = sub_signatures_df.set_index('mutation').T.to_dict()\n",
    "    \n",
    "    # Create bed of the selected mutations\n",
    "    sub_mutations_df = mutations_df[['chr', 'start', 'end', 'alt', 'sample']]\n",
    "    mutations_bed = pybedtools.BedTool.from_dataframe(sub_mutations_df)\n",
    "    \n",
    "    exons_coords_symbol_df['exon_size'] = (exons_coords_symbol_df['end'] - exons_coords_symbol_df['start'])\n",
    "    exons_coords_symbol_df['exon_middle_start'] = (exons_coords_symbol_df['start'] +\n",
    "                                        exons_coords_symbol_df['exon_size']/2)\n",
    "    exons_coords_symbol_df['exon_middle_start'] = exons_coords_symbol_df.apply(lambda x:\n",
    "                                        math.floor(x['exon_middle_start']), axis=1)\n",
    "\n",
    "    exons_coords_symbol_df['exon_middle_end'] = exons_coords_symbol_df['exon_middle_start'] + 1\n",
    "    exons_coords_symbol_df['region_start'] = exons_coords_symbol_df['exon_middle_start'] - middle_distance_threshold\n",
    "    exons_coords_symbol_df['region_end'] = exons_coords_symbol_df['exon_middle_end'] + middle_distance_threshold\n",
    "    \n",
    "    # Left flank \"intronic\" sequences => region_start and start from coordinates.\n",
    "    intron1_df = exons_coords_symbol_df[['chr', 'region_start', 'start', 'ID']]\n",
    "    ## The region may span less than the centered exonic sequence. Only take left flank intronic coordinates with sense.\n",
    "    intron1_df = intron1_df[intron1_df['region_start'] <= intron1_df['start']]\n",
    "    intron1_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    \n",
    "    # Right flank \"intronic\" sequences\n",
    "    intron2_df = exons_coords_symbol_df[['chr', 'end', 'region_end', 'ID']]\n",
    "    ## The region may span less than the centered exonic sequence. Only take right flank intronic coordinates with sense.\n",
    "    intron2_df = intron2_df[intron2_df['end'] <= intron2_df['region_end']]\n",
    "    intron2_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## Take them together\n",
    "    intron_df = pd.concat([intron1_df, intron2_df])\n",
    "    intron_df['label'] = 'intron'\n",
    "\n",
    "    # Exonic sequences\n",
    "    all_exon_df = exons_coords_symbol_df[['chr', 'start', 'end', 'ID', 'region_start', 'region_end']]\n",
    "    ## The region may span less than the centered exonic sequence. Process exons that surpass the region size.\n",
    "    large_exon_df = all_exon_df[(all_exon_df['end']-all_exon_df['start']) > (2*middle_distance_threshold+1)]\n",
    "    large_exon_df = large_exon_df[['chr', 'region_start', 'region_end', 'ID']]\n",
    "    large_exon_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## The region may span less than the centered exonic sequence. Process exons that do not surpass the region size.\n",
    "    short_exon_df = all_exon_df[(all_exon_df['end']-all_exon_df['start']) <= (2*middle_distance_threshold+1)]\n",
    "    short_exon_df = short_exon_df[['chr', 'start', 'end', 'ID']]\n",
    "    short_exon_df.columns = ['chr', 'start', 'end', 'ID']\n",
    "    ## Take them together\n",
    "    exon_df = pd.concat([short_exon_df, large_exon_df])\n",
    "    exon_df['label'] = 'exon'\n",
    "    \n",
    "    # Merge exons and introns\n",
    "    coords_symbol_df = pd.concat([exon_df, intron_df])\n",
    "    coords_symbol_bed = pybedtools.BedTool.from_dataframe(coords_symbol_df)\n",
    "\n",
    "    # Filter mutations by the position of interest\n",
    "    my_bed = coords_symbol_bed.intersect(mutations_bed, wo=True)\n",
    "\n",
    "    mutations_mapped = pd.read_table(my_bed.fn, names = ['chr', 'start', 'end', 'ID','label', 'mut_chr', 'mut_start',\n",
    "                                    'mut_end', 'mut_alt', 'sample', 'overlap_bp'],  sep=\"\\s+\", index_col=False)\n",
    "    \n",
    "    mutations_mapped['conseq'] = mutations_mapped.apply(lambda x: \n",
    "                                    check_consequence(x['mut_chr'], x['mut_end'], x['mut_alt'], tb_consequence_type,\n",
    "                                    consequence_rank_dict), axis=1)\n",
    "    \n",
    "    print(\"Total amount of mutations at the middle exon-centered sequences is \"+ str(len(mutations_mapped)))\n",
    "    exonic_obs = mutations_mapped[(mutations_mapped['label'] == 'exon') & \n",
    "                                  (mutations_mapped['conseq'] == include)].shape[0]\n",
    "    intronic_obs = mutations_mapped[mutations_mapped['label'] == 'intron'].shape[0]\n",
    "    print(\"Observed at middle exons: \" + str(exonic_obs))\n",
    "    print(\"Observed at flanking introns: \" + str(intronic_obs))\n",
    "    \n",
    "    # Compute expected mutation by randomization approach\n",
    "    RANDOMIZATION = 1000\n",
    "\n",
    "    collect_rand = {}\n",
    "    # Initialize dictionary\n",
    "    for sam in range(0, RANDOMIZATION):\n",
    "        collect_rand[sam] = {}\n",
    "        collect_rand[sam]['exon_Exp'] = 0\n",
    "        collect_rand[sam]['intron_Exp'] = 0\n",
    "\n",
    "    # For each unique staked exon-centered sequence with one or more mutations\n",
    "    for gene in mutations_mapped['ID'].unique():\n",
    "        position_prob_lol = []\n",
    "        label_flag_lol = []\n",
    "        \n",
    "        # Take the exonic and the flanking intronic sequences.\n",
    "        for dx, row in coords_symbol_df[coords_symbol_df['ID']==gene].iterrows():\n",
    "            my_chr = row['chr'][3:]\n",
    "            my_start = int(row['start'])\n",
    "            my_end = int(row['end'])\n",
    "            n_bases = my_end - my_start\n",
    "\n",
    "            ## Get sequence.\n",
    "            my_sequence = hg19(my_chr, my_start+1-(k//2), size=n_bases+2*(k//2))\n",
    "\n",
    "            if row['label'] == 'intron':\n",
    "                label_flag = 0\n",
    "                probs = compute_intron_probs_permut_CpG(my_sequence, signatures_dict, tumor_type, k)\n",
    "                flags = [label_flag]*len(probs)\n",
    "            else:\n",
    "                label_flag = 1\n",
    "                probs = compute_exon_probs_permut_CpG(my_chr, my_start, my_sequence, signatures_dict, tumor_type,\n",
    "                                           consequence_rank_dict, tb_consequence_type, include, k)\n",
    "                flags = [label_flag]*len(probs)\n",
    "                \n",
    "            position_prob_lol = position_prob_lol + probs\n",
    "            label_flag_lol = label_flag_lol + flags\n",
    "        \n",
    "        # Total number of mutations falling in the stacked sequence.\n",
    "        mutcnt = len(mutations_mapped[(mutations_mapped['ID']==gene) &\n",
    "            ((mutations_mapped['label'] == 'intron') | ((mutations_mapped['label'] == 'exon') & \n",
    "                                  (mutations_mapped['conseq'] == include)))])\n",
    "        # Normalize the probability vector and perform sampling\n",
    "        prb_vector = np.array(position_prob_lol)\n",
    "        prb_vector = prb_vector/prb_vector.sum()\n",
    "        label_vector = np.array(label_flag_lol)\n",
    "        mutation_rand_label = np.random.choice(label_vector, size=mutcnt*RANDOMIZATION, replace=True, p=prb_vector)\n",
    "        \n",
    "        # Save the mutation count for each randomization\n",
    "        for sam in range(0, RANDOMIZATION):\n",
    "            start = int(mutcnt*sam)\n",
    "            # count for one random sampling\n",
    "            count = np.unique(mutation_rand_label[start:(start+mutcnt)], return_counts=True)\n",
    "            for i in range(0, len(count[0])):\n",
    "                if count[0][i] == 1: # count of exons\n",
    "                    collect_rand[sam]['exon_Exp'] += count[1][i]\n",
    "                else: # count of introns\n",
    "                    collect_rand[sam]['intron_Exp'] += count[1][i]\n",
    "        \n",
    "    collect_df = pd.DataFrame.from_dict(collect_rand)\n",
    "    \n",
    "    return(collect_df, exonic_obs, intronic_obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get exon coordinates\n",
    "exons_coords_df = pd.read_csv(middle_exons_coords, sep=\"\\t\", header=None, low_memory=False)\n",
    "exons_coords_df.columns = ['chr', 'start', 'end', 'ensembl', 'symbol', 'strand']\n",
    "exons_coords_df = exons_coords_df[['chr', 'start', 'end', 'ensembl']]\n",
    "exons_coords_df['ensembl'] = exons_coords_df.apply(lambda x: remove_version_ensembl(x, 'ensembl'), axis=1)\n",
    "exons_coords_df.columns = ['chr', 'start', 'end', 'ensembl']\n",
    "\n",
    "## Get consequence tabix file\n",
    "tb_consequence_type = tabix.open(consequence_type_file)\n",
    "\n",
    "## Get the consequence rank info and classify separate syn from other types of non-syn (missense, non-sense,...)\n",
    "consequence_rank_info = pd.read_csv(consequence_ranking_file, delimiter='\\t')\n",
    "consequence_rank_info['TYPE'] = consequence_rank_info.apply(lambda x:synonymous_or_not(x) ,1)\n",
    "consequence_rank_info = consequence_rank_info[['CONSEQUENCE', 'TYPE']]\n",
    "consequence_rank_dict = dict(zip(consequence_rank_info['CONSEQUENCE'], consequence_rank_info['TYPE']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exons centered analysis, Goldmann 2018\n",
    "\n",
    "Mutation probabilities were computed genome-wide using DNM. They were used to compute the expected number of mutations based on a generalized k-mer model, with the method described in the **Exon-centered and intron-centered mutational analyses** section at Frigola et al. for the 3-mer case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get mutations\n",
    "mutations_file = mutations_folder + 'germinal_ultimate_dataset.bed.gz'\n",
    "mutations_df = pd.read_csv(mutations_file, sep=\"\\t\", header=None)\n",
    "mutations_df.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'class']\n",
    "mutations_df = mutations_df[['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type']]\n",
    "mutations_df = mutations_df[mutations_df['type'] == 'subs']\n",
    "mutations_df = mutations_df[mutations_df['sample'] == 'Goldmann2018']\n",
    "mutations_df['ID'] = mutations_df.index + 1\n",
    "\n",
    "tumor_type = 'germinal'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>115.149</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.851</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       115.149            5113        5124.851   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.13  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    10.291883\n",
       "dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    9.966747\n",
       " dtype: float64, 0    10.617018\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>299.403</td>\n",
       "      <td>5113</td>\n",
       "      <td>5171.597</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       299.403            5113        5171.597   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    19.57128\n",
       "dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    19.362319\n",
       " dtype: float64, 0    19.780241\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>131.129</td>\n",
       "      <td>5113</td>\n",
       "      <td>5108.871</td>\n",
       "      <td>0.679</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       131.129            5113        5108.871   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.679  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -3.148808\n",
       "dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.399596\n",
       " dtype: float64, 0   -2.898019\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>346.288</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.712</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       346.288            5113        5124.712   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.26  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.382156\n",
       "dtype: float64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.215567\n",
       " dtype: float64, 0    3.548745\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>130.631</td>\n",
       "      <td>5113</td>\n",
       "      <td>5109.369</td>\n",
       "      <td>0.644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       130.631            5113        5109.369   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.644  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.779585\n",
       "dtype: float64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.05383\n",
       " dtype: float64, 0   -2.50534\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>343.417</td>\n",
       "      <td>5113</td>\n",
       "      <td>5127.583</td>\n",
       "      <td>0.201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       343.417            5113        5127.583   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.201  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.246441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.080358\n",
       " dtype: float64, 0    4.412524\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>129.603</td>\n",
       "      <td>5113</td>\n",
       "      <td>5110.397</td>\n",
       "      <td>0.595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       129.603            5113        5110.397   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.595  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.008441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -2.276826\n",
       " dtype: float64, 0   -1.740057\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>342.903</td>\n",
       "      <td>5113</td>\n",
       "      <td>5128.097</td>\n",
       "      <td>0.193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       342.903            5113        5128.097   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.193  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.402703\n",
       "dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.233248\n",
       " dtype: float64, 0    4.572158\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>126.673</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113.327</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       126.673            5113        5113.327   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0               0.5  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.258145\n",
       "dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -0.018615\n",
       " dtype: float64, 0    0.534905\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>345.76</td>\n",
       "      <td>5113</td>\n",
       "      <td>5125.24</td>\n",
       "      <td>0.248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358        345.76            5113         5125.24   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.248  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.540028\n",
       "dtype: float64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.370396\n",
       " dtype: float64, 0    3.70966\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot results\n",
    "\n",
    "Binned average approach. Bin size is 25."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/ipykernel_launcher.py:20: MatplotlibDeprecationWarning: The set_axis_bgcolor function was deprecated in version 2.0. Use set_facecolor instead.\n",
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/matplotlib/font_manager.py:1297: UserWarning: findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfcAAAEqCAYAAAAI1RnWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXecZFWV+L/nVXV1nOkJ5EcBgqIEA6tgWAQU1wSKIoIi\nmBDUVRFFQUGYGYyoa1gVMbu4ivx0XVcxrBFXXHdddVkVEwaY9iEMTGBCx6p3fn+c+6Zuva6uqu6u\n6unpud/Ppz71wr33nBvOTe+++0RVCQQCgUAgsHSIdrUCgUAgEAgEOkto3AOBQCAQWGKExj0QCAQC\ngSVGaNwDgUAgEFhihMY9EAgEAoElRmjcA4FAIBBYYoTGPRAIBAKBJUbTxl1Eetz/YSKyYmFUCgQC\ngUAgMB9ajdzfIiIx8BbggwugTyAQCAQCgXnSqnFfAZwGvB24s/vqBAKBQCAQmC+tGvebgJWq+gvg\ntu6rEwgEAoFAYL4UW9z/PqAi8nxgYgH0CQQCgUAgME9ajdzfDWwD7na/QCAQCAQCi5xWI/efqurX\nFkSTQCAQCAQCHUGaffJVRH4IbAVGAVXVMxdKsUAgEAgEAnOjaeMOICLLAFR124JoFAgEAoFAYF60\nGrm/CTjYnY6o6lULolUgEAgEAoE502pB3SpVPV9VzweWLYRCgUAgEAgE5kerBXUrROQF7nh1t5UJ\nBAKBQCAwf1pNy/cAT3Sn31LVqQXRKhAIBAKBwJyZsXEXkYuAAwBxl1RVL1koxQKBQCAQCMyNZo37\n/sAqYLu7NKSqty6UYoFAIBAIBObGjAvqVPWvwLNU9Q5VvQN43sKpFQgEAoFAYK7MuKBORL4AHCki\nR7tLYfvZQCAQCAR2A1otqDsqTMUHAoFAILB70epVuEeLyHucu4KqntR9lQKBQCAQCMyHVpvYPBT4\nsaqeDHx9AfQJBAKBQCAwT1o17luAkoicA/zNAugTCAQCgUBgnrR65r4MmAKeCvxeVX+1UIoFAoFA\nIBCYG61G7q9R1XFV/RJw9kIoFAgEAoFAYH6EV+ECgUAgEFhihFfhAoFAIBBYYrRq3K8HFBgChlX1\nxIVSLBAIBAKBwNxo+p67qj43O3YfkgkEAoFAILDIadq4i8hTPXfhVbhAIBAIBHYDWu1Qtzc2LT8B\nhM+9BgKBQCCwG9DqVbj/xj77egjw/G4pISIf7Ua4SVxelcTl+3Uj7EAg0B5dtO8oicvHdCPsQGB3\np1Xj/nbgv4DvAN9t5lBEDhWRT4jIF9352SLyMRG5TkQGW8g5oG2NZ8fDgGe2cpTE5ft3SX4gsCRY\npPbdC5zWylESl++fxGXpkg6BwKKkVeP+E+BXqvozVf1ZM4eq+idVPc+79ExVPR+4ATh9nnrOFQXa\nMepzuq1IILA7s4jtux3Oob16IBBYMrRq3PcHrhOR60Xkc7MMOzO89cCB+Zuu53+jiNxI93r27Tbu\ngV1AEpeHW9wvJXF5oI1wBpO43NM5zQJtsBjsezYUdrUCexpt2Le0cuPcFZO4PNQ5zfYMWjXuo6p6\nuqo+V1Xnuv3sQcBf8hdV9XOqeqqqngrcOcewAUjiciGJy6XctV5ax282Mvr9qb0kLvcmcblpheH8\n9HdKh6WEy6/XNLjup9fjgee1Edx5wGNayOvfHadmXQW4mMtQ1+0bppULnO31zSKIVrba43cQXZ3S\n28JPn+tY7nblai7MwYYa2rcXxopGbhrwCODlLXQrJXG51QLxRUm37LtV4/dIEblWRN4pIu9s5lBE\nVovItcAxIvJG4Msi8mHg2cCX5qpgmyOyE4EX5a6dDxwLSLujOrdAZ6ZK4FLqRyjnA8dlOs4g41L3\ny8tpqk/+fq7SmebX9WylmZtWfjrBLEfPM8n206sA9Ph5kunt0jwLo5h3l7nNhbuXdy/Khds0jRvR\nIJ9mDLOdsDL3OX8H0KAMLTS70r6TuLyc6WlwHPBSd7+QxOVWdVnk3M7UAJwGnOqdHwtckOk4g71c\nCLwe2Cenbyv7rtN3tmWvTTfShcbuUuDQDoQRu2OB+jxxNlTM2VKE2Xfe3vz4nQn8Xe6+n647828W\n9j0tz+dST2Rus3xvUF67Yt+tDOKF2KK6D7nfjKjqRlV9maoepqpvx3rrRwPjmKHMGpewl89w3Wdf\n7BGCzwBW6QtwuT+t4wp+PtMEGymemd1v4M5Prz5sQQ9Ox8tn0K2RzGlxynF5rofrh91IzquAI3z3\nXkFuFA+AFzPHfGmhd9MOg3c/3xA30hGs4vRnjc4HjsEM4iB3TbGydpQX3/2BN+XCKng6nIRVCL7u\npXwat4hPPr4nA2c4f3vh5XOTMrczLBfeoPtf5uucD2ehR4vdtO9G6dLE7jJ6gGzE8zzgb5v4Bygk\ncXkV8Cb/vueun5o9g9Ud2ULBy7ER5gNy4Wby83lyeZKbSczxAuBRM9l3Ym/5+J31RuH3eH7Iuwce\njs1oNXSTD3um6w38lpqVwZnKeM59fgDl2+mx2EDtKcDT3TUFSi7efji+vz6g19NtBfX17IuoldM6\nu8zHweMVwFGe29U5HWZr3+diM4xnA4+dSWin7Ltp466qd/i/WYatwHYs0adN27XJmhmuX+kSOqMI\nkMTlk71rQv0z99flws2H/WjgeOBg7/6jgDVJ4+dCKRDleuDSROdGMptxqZPtZ/KaGY6HsYbqmQ3u\nrwGegc1uZHquxUaELZ9nz4GHzHQjicsPwToVUBtJHZbE5X2By5g5fVZ4xyuxMlUETk/i8pOpbZHc\n78I4FcufPC9L4vIDgCuAZc6PzznUVzx/gxlklmaN8Hvvy7AOxxrgcTl3F+HFL4nLh7hrM3Gx+88v\nGjvBhbOrt4LupH3n66HTmJ5+JHH54iQuPzmJy0+hfuZnGXByUnvrZQ1mO364z8FG2v79wzwdDqM+\nrdOcXoNYXH2b9MvYM9y9rDw0ewwwCDwReHUSlw91cSt4uvS4sLIwnsb0+iNr4NckcflBWdySuJyV\nv1OpDXhOS2ztypokLq9I4vKZSVzOOkMvAx7UQMdM3kWJDYwy2Wdhnak12HT5TnJxyMgWW67BOubQ\nfB3UALDc/R+RxOU3UJ8vdengGnGcm4dhnfY11HfUAPbDPl2e+csed61x56925cFnObUOHlgHI9Mh\nb7uPYHrc8+d7Ob1W0KB8Oz1WOn8XNro/G5p9FW5YVe+bR9g/VNUfiMi+wHvIPTsVkbOpjcjaWnCT\nxOVHYj06sJ7Yu7Pg3P9jk7j8WOBad55/Dl+Mk5GKd77Wu73K/fvP8Q52/9lzoRclcRng/VjGn0B9\nGmYFZS3woxZxeX2cjLwrictXYpsEvR84EttTwOcReV2z4yQuXwJ82HP7ENeA5uP3UOAnSVxeQ71h\nneQqwBOArwNPjJORtzTTOxeHNcBb4mSk6horgNWe3Ktp/FhirXd6sJPv338S1tnKyML8EJan2eOR\nYawDlpEZ9M6On+sEvspzcxZWaT8I6Hfh7nD3DnLhk8TlJ2I9/bryEicja91x9gpWLzDpjv2OzVE5\nP8MN4p6Fe4Z3mm/o/BGZ7/ekJC4vi5ORr+bDWyA6ad8RUE1stuJi7/r3cu6WYfl9GzDqXd/b/S/P\nuT8/sRkcmD6zB24a10vXI9zxl7HGN5/mOxsqVw/4+j/UHWeNwaXA2iQuvxIYA34K/B54LbDJuVlB\nbf+QK7zwsufLZyZx+Y/UyhdJXPYblWzk+Bx370ysDsFzvxb7omfWkT8Ds59tXtyW+e6zMu4Ypn5g\nBJB1opYlcfko7NFMXmaGXx9k5eDJSVzer4Gfm7BO61a8ETPwZGZGcvKyTu+zXLhPw2Yxqu48k3sg\nNoOSsRI4N4nL6+JkRF1HsUB9/f5C79i357fhBgoz2PcR7nAIq0/8qX1/cOj7XZnE5UvjZOTqhrFu\ngxk/HCMiVwObVHXOgbtwSsDnVPWMJm5udAtv6miQUP4imnEsw4RaYk1RG63nr/kswzItBebTgckj\n1Ixo0sluxji1+PwZyDbc6aHWMdmR99Qlfo6NVO+hVlmui5ORaQUkicsXA+8FrsQ6BU/Nu3GMUZu6\nbMRyLE8rwLbcvRK1Efs2F9YU9SNlsIo1kzHG9BH7HdQ6ae2wGTP0hNqzwR3U9+D9eH0oTkbuaTKy\nb8UvaDLb4bGd6TMN5CriBafD9g21dJ6i1qj9DngwZu9Vdy2zD/9aRj81u9pC+6/MtcMAZucVrFMO\nVj6Xzeij5mYUe4TYqB7rNvfirTlpgV9/VjCbKlFbW7AdS9cqzWcollOL573U22aBWodrlFqHp5Eu\nWX00xvR6Yrb8Dngg8FVsRgTq6948N8XJyE1JXL6Q2gBwNnwIm95vxVamd07nZd/NGvd/wHpMt+Km\nuFX1zIaOG/s/HXgSVkF/WFVvauK2ofE/tFRaX0AKgyIrBKQo0hthllpBJ7anunkokhUlpB9gXHX7\nxjS9c1UU7Tsgkk2la8UqCR1X3Q7o8kj2riqVglBUpZpCWoWpCjpZQnpLIgOKpgoqoApaUSYVNJPb\nK9I/KLJSQFOoVtFKr0h/CelPoSoQVdBJQRjTdGtVmRqIZDhCIoBRTe+rKJN9IgORM5AKTCqaLpdo\nb0sXIic/dXmAgEwoO6roVBHpiaxxkzHVrZPouDiHAoWSSJ83DBSAXpG0hGhR0CKiPSJ6VE/P6H5R\nYWpVFI0+uFT6mpcFV8fJyFh2ksTlU6lNxWWN4C3Aw/5SqRx/e7VyvzHV6OaJicnNaboNV8mmaDFC\nVqaoREjBxTeKoFCFqQiK6irgKjpVVSolkYEUqgpaRas7UvWNXyIoCtAnMlwU6VU0BZhQxjA/UwAV\npVJFK0Wkpyg2so+QHgHJypEqqVo6o5Bu13RLVan0igxGEPWI9JbEOm2pUq1ARdlpOJk/LSKlXmEw\ntbgo7vrWNN1YUSZXRtF+BaFHQSaV0RStplCtKJXI8qtfd3ozply5S9HqmOr2CCn2inUsqlD9ycRE\n3WKuhaIT9n1kTykBGIpkhUBUQIoFKwspoPdpem8v0jcQyTCWwdW70+r6QZGh4SjaB9eRrygTAkyo\njlVgalkkqwCNIFLQVK0cVdDxCCn0CgPOtsHlX1WZUkgV0q2abixAcZlEqyJBVEkrMFWEnl6RQYVq\nBAVXrzClOjaJjveJDBaREsCk6ui46o4ekd4i9IhYWVSo9ossLwo9zk6j1DV67lwqykTFld8IEUCm\n0PEJ1VHJkhSkJNIndpxds9R24bq6gYFIqn0iaYTo4cXiWEmEQ4vFfe9XKEokyH5R4c+RyAQwAtwy\noXrixrR6nCrpLVOT0XbViW1p2ndbpTI1pVpV6NmYpndVUUmhoErUI7IMiBQVV3iliJSsDrS8dTaW\nTqmOO1uMqjBZREpb0/TeWhx21n1RD1IqCn0pVItIaVJ1DKyeyMKrKtVIiHqQXgUpQk8kVqdOKRMp\nWtFaXuuE6uikMtUr9BWQoghRCUtLV052Dgj91lEg6kX6vbAAmEB3jKvuGBRZURLpd2VysopOKlBR\npiy/6BNE1Au1CpWsfI6r7lCgT6Q/q59+Pjmxcia7akXL77lj77GiqrPqMbldq67BDOAmVf1sE7cN\njf9jq1bfA8gO1fUTqpuqkO5I9d6CUNq/UDgWKCikd1Wr/1OAnnKheEoKlQiKSbXy7R2qW/pFhiMo\n9ImsLImsBBhTvXdzmt4uUCyJDApIv8gqsfB0a5omqfVKI4GoKFIqwSBIcWUUHSlmNLopTX9TQcd6\nRIYiM1K2pOn6qupUSaQ/sspg+YDIPgAVmNiepnf1iiwflGh/Bami4+OqmyIoFt1OX5Oq2zam6Z8V\npGQNZNaAUBTpGRTZG6xyn1IdLYkMDYkckNbKomKFeFvqZg8y/+Oq0aRqVEFlSpH70rTnV1NTw4rK\nbyuVFceXeneURKrPGRj401P7B27B5P8W+By1adaHAQcklcrQ1du2Hj6lyjfHxwr7FArbh0QqR/T0\n9B5UKFYLVlluK4qkAvdVVLcqpFM2WqGiOp5CGoEURIoC0isy5Ap2dUeabo6Ewn5R4UjZafeqzqin\n1Br/dFOa/jmF6qDIyggKBaRYcA15yeWNQjquugWoVFWnqjBZgJ6SyIDLT7FyEO3VL7I/kFbRsUll\nEyg7VP9ahfESDInTNatEcf8Cul31rorqpN0T+kVWDkfyAEGooNvuqlb/ryhSGhRZDdCDDEZiHZVx\n1U0VZULEKvkIiiVkSAQZEDkgcrMFk8qGCjq6Iooe+sKN9y7owrqMTtj3dav32iAgFdixNU1/p8AO\n1XtUSfcrFI4rCAOqsFnTX42m6cZDij2nRzYSlE1p+l+b0uofe0WGCkhfjzDg8o2KMr4hrf5CQPpE\nlgP0iqwsuMc2o6r3TKqOuWSWSCiUkGWgsjyKDu9Blivodk3/NKZ6TxHpz8rT1jT964Tq9qJIqQf6\nCiJ9QyJl10hXtqZ6eyT0Dkt0mFi5q45qeg+gPSJDmLupDdXqbxTSAmQdzazzHg1F0d64hkaVVITi\nkMiBuM4Kzs4ryugUOoprtLCBhmInaRWdBNicpn1ViDanad+d1cpACtFvpqaWT6gyBb2TqqXIdYbA\nhuoDImMKleEoGj2wUNgskJYLxWQwkslyoXhkD1IoCmkEaQFJFd0xpnp3hKSIDYoqquNTbnYjG8AM\nWD6UFNiuepdCdZ8oOrpgjaI1qNZzrqSug3Nfmv55Csb6RJb1QL8ghZLYTEmE9EZiswyTqluqyoTC\n1Ljqlkgo9ovs5WxTXF0+NCCSzeRVxlQTIJ1Q3TKl7CgK/Vk5kZ15IjsHVuOqmybR7S7MqEdkcIVE\nR7i8qW5Iqz9WSIdE9sPk9RXc7MWksnUS3erCwtV3KwWkF9m7R1gJMAX3TapuWi7R0WdvvKfZwsym\ntGrcr6C2InlEVa9qO2CRc4EtqvpVEblBVc9q4rbVtN3tTH8W7VPBptmzd9tTalNmc2GU1ovNvknj\nZ0FXYdPV7dJsams78P+oLULrFpNA6eaJ8aN+P1WJk2pl+OM7th+pTrHH9vbtKNhIRxD49dTUYFKt\nIqCPLPX+6dG9vb8eFhl93uDQ6/pEXtKGvPtwz6xa4D8iaEV+ARTAb4AjGrjtFB8EXjlHv1+hthp4\nriyPk5HXzjOMOdFh+26HX2CdyiyPK03ctuJ72JsxzdhC/ULOjHXMbmFsU+JkZG2DdNhA7vW6bpGq\nslW1bu1VBCyPogrwDuANwG1xMvJZaDvP/EdaX8YWG3aCduuNjIZT3Tma1b8/w57XZ9xO83bIZ7aP\nAxsxFCcj+fUObdPqPciVbotJRORdswz7QOCX7ng+z5T+k9qzxsyw1mELR56NLVzbB7iR+g0RfCP8\nX6wyBShji/H8jsq+2KrRP2AF8yiscf9XaivQx7EFYkdgK9OngHdiX8t7M7YgJluMAbZAbos778EW\nv/wA67Sd4MlXL05/j/e8G8D5943lW1ihe6Nz8zxsgctn3fEXsUUzABudzNOBt2L5cCi17XazNPo/\noHR8b9+643tZA9z8sFLpx3/X13/pzRPjR/65Utk/mypV0JN6+0af2T/w0wjoF6lG1rH9zzgZ+Yu3\n0OifXf70Ys/bPuTJ+hq2Oh4sf7MNaLYA3wZ+7eXNK7HXrh7szm93+fQEz80vsDLwEWwh0xeBP2Er\nZLPG/fPY8zacHtuwRU7ZqtUvY52606h/vj4C/DpORn7sFr+oF8YUtWe62fTZNXEyssFbUZ2Vwa8D\n/0M9WeN+E/Yq19tyfta5/3OxfLsJS5u/d3G8lV1Hp+z7C9QWZH0Ee3f9WmwR2BrgZuD4OBn5klss\nmj23fS9Wbh7pzrO0inC2mJOzxsl6ErWOQVYBZ4+Y1nn+t2M2vIZah9237/8Avp8Lf6vTK/MD9fb9\nKCcf4NNOPtQ//yVORq5xjehV2MJP/3GoX699B1vl34etmcm2CPcHF9dhddcFzu/FWF35OxcXWSGS\nhXeV0zcCroiTkfEkLl8VJyON3jzJ0uSv2LNzX6+Pe8d3YjN/D8Lycn+sbsj4W6z+OsS79i5s/4CM\nX2GvXb4PW+i4DKubU0/O74HrPT9rnB5Z5/cOLM3XYHaYvRb3H1hbci31PBcrAzc52Xtha29+HCcj\n/97AvqG+zhasLO9H7Zl+Viaf4OL9QayOPgqrs+9zcVwD/DGXTrOm1cj9k1jjoMBJqtr2CNL17Der\n6o0i8nlVfU7uvr+atl9VT54WSCAQWJQE+w4EFjetGvcerKckwLdUdXJGx9P9DmI9k3Hg5mbP5JqE\n8V1shWQ3OYAObI8ZZAQZu0AG7KKGM9h3kLGIZSyUnIWQMXf7VtVF+wNuDDKCjCBj18vZXXUPMvY8\nGUspLvOR0XSHOpGdbyYEAoFAIBDYTWi1t/yu/mDFbD8zG2QEGXuSjIWU0w2WSj4EGYtLxkLJWdQy\nWj1zvxFbjX4fgKpeM1dBgUAgEAgEFoZWr8L5r7/N3AsIBAKBQCCwaGg1LX8gcKqq/oDae8aBQCAQ\nCAQWMa1G7o/GdkuC9nfmmTUichy2ucKIqr7OXbsYe/G/B9tgZn9sJqEKfEpVv593o82eMdRkPQPb\nWW4Ftl/+c0XkJ9gmEHeo6ttF5IC8rFnG54XYxhPrgQ+p6i/bic8sZbwEOA77mMFbVPWWTsfDk9X2\nVqOzCPMZwCnYDlKfwDYgugPYrqqv64RMETkJ22DoVmwTmwOwTWt6qX15a74yHottHlTENsO4r1Px\nEJFDsc2PhlX1DPfueFP9825UdaE+PDQjwb4Xt307eR218aVi307O7mnjLZbhvw/bhWgY+EyXl/wf\nArzbHZeAz7rjV2Aftr8Cmz2IsEUG09zMUt5FwJPc8feBTwIvcOd1suYQl+djO4h9DNs5rWV85pFu\nxwBXdCMenoxzgae54xs6nO8rMeP/F+CjwOs7JRP7/OM3sJ2p7g98wV0/xYXfsXhhW2y+tEvx+KL7\nb6l/3k0n82qe6RPse27p1nX77lQ5nSHcJWHfLozdysZbjdw/jW3fdy221eq8EZEHA2/PXc7vfLca\n21ccrHd8oPuNqGrq3tBr5KYtWaq6AfuW8z+6aye7cG8Qka82kDXr+KjqdSLyEOyNg3e3EZ9Zy8A+\nk3ghte0m5xyPFnRqq9FGvAnbmvYWp+d7XHw7ITP/zfFs5Lee2mdWOxWvs4HzgI91IR4Z7ejfyM2C\nEex7t7Rv6J6NLxX7ht3Mxls17ndT2+f7rtnp2BhV/SUw7SMSInKId7qR2reHD8L2Dv8LcKCIbG3i\npl1ZxwP/rWqfCc3+sT2m+xvImnV8HBuwfZDbic+sZLjdA68B3qeqI/ONRwv+ghXgW2i9TqMt3B4K\n7wC+oao/925laTZvmbn06KW2L/lBLnzmKwNARA4C7tP6Lyd2LB4NaEd/301DRKRHVadE5DBgo9pX\n8+ZFsO/d0r6hw+V0Kdk37J423qpxfy+1FfPvwTbT7zgicji2Wf5RInKBqn5URH4uIu/HMu0abCP9\nd2Afffi4qk42cNMuL8b1hsU+A/t+bBvNTaqaiMgnfFlziM8FwN9go491M+haF5/ZysA+YnM48HK3\njed3Oh0Pjy8BHxSRU7CPXHSCV2EfUBgWkftj6ztGsTL5TuwDM/OSKfXfHP8gcICIfBirGF/hnHUi\nXucBn3Iy/6lT8RD7LOxbgWNE5I3Al9vQv5GbmXiLiPwj8BZsVHBOC/dzItj3ordv6LyNLyX7ht3Q\nxlu9575OVde446tUdTafMg0EAosYEfkIVjHdDJyjqpfsYpUCgUCHmHHkLiJfAPZxvUahc9MNgUBg\ncXATcKiq/kJEbtvVygQCgc7RdOQeCASWLiKyH3ASttobVb1ulyoUCAQ6RtNn7iLyD0A5O1fVM7uu\nUSAQWCjeDVyPPbMNBAJLiFYL6qqhQQ8Eliw/VdWv7WolAoFA52m1oO6V2OYTWwFU9aqFUSsQCHQb\nEfkhZtuj2G5uoSMfCCwRWo3cT8Bek5lYAF0CgcACoqqPFZFl7nhbK/eBQGD3oVXj/hvsRf1ObJIQ\nCAQWESLyJuBgdzwSZuYCgaVDq8b9IOBt2KtwyvRtJAOBwO7LKlU9H0BE3tXKcSAQ2H1o1biv9Y7D\nO3OBwNJihYi8wB2v3qWaBAKBjtJqQd3bsUZ9CHiwqj5uoRQLBALdxe1f/kR3+i1VndqV+gQCgc7R\n9iY2IvIaVX1vl/UJBAILgIhchH3/OvucmIbtZwOBpUOrTWzehY3ce4DlC6JRIBBYCG4AVgHb3fnQ\nLtQlEAh0mFbT8ge7wwlV7cgnXwOBwOJARK7MVsiLyNtU9bJdrVMgEOgMzT4cswZvEZ2IhE1sAoEl\ngvsw1JEicrS7dPeu1CcQCHSWGUfuIvJAd9gHXAisUNVnLZRigUCgu4jIUap6667WIxAIdJ5W0/Ln\nAqcAH1DVHy2YVoFAoOuIyEuAZ2MzeAVVPWnXahQIBDrFjN9oF5GfAo8DPgMMi8hTF0yrQCCwEDwU\n+LGqngx8fVcrEwgEOkez1fIfcP97LYQigUBgwdkClETkHOBvdrUygUCgc7T9nnsgEFhauI/GTAFP\nBX6vqr/axSoFAoEOMeO0fCAQWPK8RlXHVfVLwNm7WplAINA5Wu0tHwgEliDhVbhAYGkTpuUDgT2U\n8CpcILB0CY17ILCHIiLXU/sw1LCqnriLVQoEAh0iTMsHAnsoqvrc7Nh9SCYQCCwRQuMeCOyheHtX\nFAmvwgUCS4rQuAcCey57Y9PyE0D43GsgsIRYFK/CichHuxFuEpdXJ3H5sG6EHQgsAf4b++zrIcDz\nuyWki/YdJXH54d0IOxDY3elY4y4ih4rIJ0Tki+78bBH5mIhcJyKDLbwf0Ck9cjwUOK2VoyQuP7CV\nm0BgCfJ24L+A7wDfbeZwkdp3L/C0Vo6SuPzAJC5Ll3QIBBYlHWvcVfVPqnqed+mZqno+cANweqfk\nzFYtoB2jfm5rJ4HAkuMnwK9U9Weq+rNmDhexfbfDc2mvHggElgzdnJbPDG89cGD+puv53ygiN9K9\nnn27jXtgF5DE5RUt7peSuDzQRjiDSVzu6Zxmewz7A9eJyPUi8rlZ+l0M9j0bCrtagT2NNuxbWrlx\n7opJXB5foi9kAAAgAElEQVTqnGZ7BgvxzP0g4C/5i6r6OVU9VVVPBe6cjwCX+b25a310MH5JXB7w\np/aSuNyXxOWmFYbz07Jx2hNJ4nIJmPb6VS69Hg+c00ZwLwH+toW8gd1xatZVgN0qQ6OqerqqPldV\n57r9bNftG6aVC5zt9c8iiFa2WvI7iElcLrg6pJmf/iQuD+2O5WouzMGGGtq3F8aKRm4acCzw9y10\nKyVxebdcIN4t++7kM/fVInItcIyIvBH4soh8GPte9JfmGm6bI7ITgBfnrp2PFQppd1TnFujMVAlc\nQv0I5SXAIzMdZ5BxCQ1WIbfSJ38/V+lM8+vkSzM3DfwUk7jc0c7dLEfPM1USfnoVgB7faJ3ekotz\n5q7gucvn+yV4XzjM53WrNG5Eg3yaMcx2wsrc5/zFdG8l+yNF5FoReaeIvLOZw11p30lcXs70NDgO\neKm7305ZjlrIOw041Ts/Frgg85O3MceFwOuAfXL6trLvOn1nW/badBN1obG7BJjvAuVLsDINrg7I\nxT9y6VPwbEmAYrN6ETgLeOJM9/00n4195/N8LvWEF1bRi59fXrti35185r5RVV+mqoep6tux3vrR\nwDhmKLPGJcDlDa7njWxfYN/c9QHsVT8BLk/i8jLff4NME+AJwFkuA8T9/DTyj/uAkju+PNNzpp5t\nTua0OOW4PNfD9cNuJOdC4EjffXbsxcNv/AU4jznmSwu9Z+zZ5/QoeLo0zBPH3tR/1OQC7J3sNwAH\nOz+KlbWjvXD2Y3o6Fz2Zj8MqBF/3Uj6Nff3aiO/JwLNd+dnLl9+kzO0My4U36P6Xu+tR3k+TtJot\nL8QW1X3I/WakS/ad2ea0dMldi/zr7l4Js0GA5wHHN/EP1jisdvIiL1+z9O3DFujtdI/VIWB58xrg\n8Fy4mfx8nlye2MzUTDwfeMxM9p3YWz519urr7IffqHy6OD0cG4CQj2u+XDcqS3mZnpueZmVwpjLe\nyPY9fDs9DhuoPQU4zbPvkov3TPVoL9Dr6bYyd//F1MppnV020tvxCqyMZ27z5Sdv39NsNRfHc7AZ\nxrOxAWlDOmXf3X7mvh0zgGnTdm1y5QzX17iEzsh6qCc30CFLpIt9/+7n8xj3O8jJXQM8GrgyicvD\nDXRIgSip74FLg3CbyWzGJVg8/UxeM8PxMmwEdXqD+2uAZwIn5u7tz+ymNdvloU3uPQTrVECt7B2a\nxOV9scI/U/r46b8CM+QCFt8nYfk8iJW1NdgK6rRBOC9N4vLhzs2Q8+NzLvUVz8PdNZro5vfel1Er\nP4/LuXutH0YSl++HNRgz8doZrp/owjmpid+2UNU7/N9svTN/+87SI18PPYPp6Qdmw0/GKn6fIeDx\nSVx+gBfumlxl+xzgVe74Snf/MGp1zGHUL9BLc3oNYOXOt0m/jD3T3cvKQ7PHAIPYQOKiJC4fCjsf\nM2TpUXRhZWE8nVqdlHGZa+DXJHH5QcCrXThZ+TsF6+CCNZIDWF22AjiT2mOslwMPaqDjGhfOaxN7\n3p3JPgtrpNYAj/A95OKQ8SwvvAvccbOGqx+zoz7gCKwT7+dLXT2a1J7ZK1b3PM7dz3eu9sUrN64B\n7ffCujCJy/fP+VlGrYOHF+aVTLfdRzC9vcqnxV4ujGFmsF/XKVmDy8/50M1nFD9U1R+IyL7Ae7De\n9U5E5GxqI7K2FtwkcflRmHGDNRLZVGJWWI5P4vLxwLXuvC6Dk7hcjJORine+1ru90v37z9kOcv9Z\nRr4oicsA7weWY70vv3Jf44X74xZxuTRORq52bieA9wJHYe8c+xyb1zU7TuLyG6gfcT04icsPbhC/\nhwA/ycUX4CRnkI8FvgE8KU5Grmqmdy4Oa4E3x8lI1TVWAKs8OVcDl87gL+Ngcr3YJC4/GXiUd2m1\n83MNlqfZ45HlOXfZyGuVF9ZewCs9N9lo/UFAvwt3zF0ru/BJ4vKTsLSf8vWOk5G17vgZ7nIfMOmO\nH+LJOSrnZ1mDuGfhnumd5hs6f0Tm+z0xicvL4mTkK/nw2kFEhlX1vrn4dXTSviOg6mYr/E7N93Lu\nhrBHYbcBO7zre7v/ZTn3L3UdR6g1dD5PhLp0PcIdf7nBPah1nte4esDXP8v7rMN4KbA2icsXAqPA\nT4HfYdP49zo3w9T2F7jCC+/l7v85SVz+AzY7gtPHb1Quy9y5e2dhDSKe+7XABmqN1LOxafGtXtz8\nWc2dZdy79zrqyabmlyVx+SgXZl5mhl8f7O/+n+Lli+/nB1jndSveiJland8IycnL6pIzXLhPx2b6\nqu4806GMzVxlrATOSeLyujgZUddRLFBfv7/IO17u6f02rGzOZN9HusMhoOKHmRsc+n5XJHH5DXEy\n8o6GsW6Drn84RkRKwOdU9Ywmbm50C2/qaJBQBWqFdAxLqIhaYk1hvels2s6/5rMc69ikwJZ249IG\ngstkrMGebOIWdz/T8w6socNdyxqqbR3Urxk/w0aqG4FsVmRdnIxMKyBJXL4EeDfWU/0G00dSGRPU\nT3XmGcbytArkG5peah2urVgFWWF6hzSiVqFmZcJnBDPkdrnP6ZVQezY4Tn2nz4/Xh+Jk5J5GRt0m\nvwQe3NKVxW3aTEuuIm4bEbka2KSqV8/FvxdOJ+0bag3NJJbOAH/AOktVavmbpb9/LWOAWn5tofEs\nzlwZxMrcFLVGdwfTZ4HybMfycG/q67FRXMPTZTbhdXpb4NefFUy/Ejb6BYvLZqbPbuQZxsqsAvdQ\nnw8Fap2+UazemUmXfdz/KFYXzIffA4cDX6W2R4Jf9+a5KU5Gbkri8kXYrOFsuYYWiwEd26m1HTuZ\nq31DF0fuIpJNma4APjjHYA7CjOYRmK5ZJqdYYflvrEHKKu/bgC9gC2MeghWqFCvYKZaxVWxKZCtW\nkYxjFclWrIDt4+ROOj8pVsA3uuMNWENYBo5xcsewRvhAzAAyY9/k7v/WhX8EtUrp1+7+wd61TU7W\nY6g9Z1KsIhH3KwK3u/CWU6vEbnO6Ze56vfTKRn/5f//4FJcW/mYmfdRGtSRx+WlYekNt2ijrCT+S\nmoGsxwxmGEvjXuBQzKB7MYMvYp2ybdQKderOx1y4WcU3CfwvVpFk8el354djldakO7/Lpdl97n8U\nM5zlWKdFnE4Fd/9eahVYlte3OPf3w/JgL6xMqNNtK7WydUoSl1N3vsL5qbjzzM1/YuXnFE/2HU63\n05yuA1j50dxvE7WG5NdYucpmSfzR62wpAs8XkYe7NFFVPbOFn5100L4jbPalH0vzZdTK3Dewmaxs\n5DWBzRIcg6VlBcuf7Fv0f8YanROx8tDn4rbDnSdY+h1KzbazdN7iwp8A/sO5O8HpV8HyYQVwf6w8\nDTpZYOX9TndvLyfzT+63L7XHSlnZPtpdU6fjztkhJ+9eJ6/gziPMttdTK/8FbDaiSM0u/P+iCzuL\no/+f2U0W1u0u3iOYnZ2AzVqlTs4OzEY3O3fLgJs83YouvD4vzAizt7sxm1/hwht3abWPC3sz1uH5\nXs5vrwt3b6wu2OHScsTFY7NLtymXH73UOgursfIkWH2wjVp+V12+rMbK37Dze5Bzn+V1Nqh5ehKX\nszLSR33ZyTort2GzMo/CymvBxXsT9lhik4vLQUx/XLPV6VdxYaRYOSq447XMka6N3N2uVddgRnWT\nqn62iduZevZ3YAl+K5ZJU1jB6MOMO0uAr2EZ9Crnpgf4KJbAe7nz/ak1RCPAL7AEX+FkHEhtNP8b\nrBBkhXcAKwQ9WMWRufshljkr3b0U+BU1A+h18rNprG1YQdgXMwawCn7E+c9GqhuBn1NriPxKaAAr\nQOLSdgvWuGXTcb7bezFjyjLZz+z8NcWM+q8u7nc5vfqwRuV6apv9HIONavuwgohLk6+6OD/UpVsW\njmKV0wanX9Y47vDSudfFaTW1kdGdLl0e7675jd6Ec1NxaVXB8rfk/Ay68FZRKyd3Umsox6lVOlk+\nFzADfICn553u+A9O35VORr4yzSqm27A8zc73xxYJgVVI/45VPNkjn6xyAWt8fL891GZRDqfWCVrv\ndHtCnIzMeeGNiBzlwkJVZzVD1GH73gr8D5avCVYmnoyb+gRuxvLhIndNgH/DOmHDmE0sx+xMsMr+\nB9TyXzCby0bL6zFbbJTOj3THqQt/BEv3zO8fsHzs864fiZWdSaxe6cc6wQWsXN7u4pZ1LieAH7l4\n9lI/+i1Qs++sw9njyfAb6q1Yx7BRA55iZalRwy9YHXOXS7f93fXHUxtpfw5bS5EtTFZqs4hPcnpn\njWVmK3+m3kZ3OB2yeIF1SLIZqNtd+hzv9PDrrilqMze/dLJXUpt1zWYhBqjNMmyg1pG7x6XbgdR3\nklbgHpk5Gbc5uXe5tPSfs4v3y87/itW5BS+8R1HL66+4NLkftRmazG43Unskg/O/r3MXU5sZudfJ\neWycjMx5XVQ3G/dzgS2q+lURuUFVz2rittW03e1Mfxbtk03XlrAES2k9Jd6MUeoXUjTimzR+FnQV\nMy8EbESjqeaMrcAXmf6aX6fJHg/shxmQ35N9ILX3lLMGcH/g+5hxJFhlVwU+jFuh24LN1DoyzdhA\n7jWjJjSaIvwNuWeQHeaD1D/Pnw3/RhtbI7dgKE5G8s9D20ZErqDWyRhR1bbXW3TYvtvh/4CHUato\n5zPV/j2sIWvGFhpPw65jdgtjmxInI2sbpMNdNF4jsNC8A1vQ9rs4Gbke2s6zv1BbF/NlbIFkJ5gp\nT2Yie8TWjCozL37MHlVm3E7zdshnNm5nYjBORl4/V8/dXC1/INbrhfk9U/oR7nls9vzB/X/B3b8Z\n6zG+F2ukxoHJ3LOKn7vzdcAngNQ7X0dtAd5t2HRTNt35r14YY87PDe58ktqCvqsyveJkJKt03oc9\ns14LvMVduwkbUeDJfqt3viET5sJ6T5yMrKd+NfK3sNeXsnS4zV3PRk5f8NzeC/yLO36rk/EZX4Y7\nvAWbjns51iB+AHtT4N+Af8K2Kf0vbKbie8AnsQWDP8V6mOPYsylfz89QeyZ5T07WNZ67H3nHm53+\n67C0u8bF4Zeemz8D36ae/wPudjLGnOx12PRqxvVenoN1nL7v3f9XLA2358JeD3wz85sLY8rpvMlz\n/6Fc2cr4mh9GnIz8r3fv+9TK7E4/zt1a4I+euyztvjCfht2xUlXPd9vI5hejtaJT9u2X14+4/2up\npcMPAeJk5F+pH5W+ByuTAOql3U5bpFaO1jp3N2B1SfaM/nb3v8nz82Z3bZvnb11231uD8gOml4f7\ncn78MrAOGxBkfNpz+1U/QeJk5FrPzw3U45epb2NTzHdis5f5MgdwHbV0XYeNgK/3ypbvPotPlobj\n7tr1TCfzd6dzm8WFOBn5uOcuweoUsLz8A/X58m3Mpn3yey780oX7PuwR16+Aq3J1/O8a2N0nvPu3\ne+5/4l3/AdaZWpf7/Q6rC76PzQLgzv9zBvsmZ9+fxupGsvh5/m521z/grn3RnW/xwvwD83z/vdsj\n982qeqOIfF5Vn5O776+m7VfV/GtsgUCgi4jIJ7HKTYGTVLXtGaJg34HA4qbbz9w/iI3ebm72TK5J\nGN/FW9DVJQ6gA9tjBhlBxi6QAfNoOEWkB3vlS4BvqWrbj7KCfQcZi1jGQslZCBlz7xir6qL9ATcG\nGUFGkLHr5eyuugcZe56MpRSX+chYiA/HBAKBRYiIzHmlfSAQWNws9sZ9tp+hDDKCjD1JxnzlTNs9\ncIFZKvkQZCwuGQslZ1HL6PoOdYFAYHHivrW+820UVb2muY9AILC7sFt+/zYQCHSEd3nHoZcfCCwh\nFvu0fCAQ6B4HAqeq6g9ob3/7QCCwm7AoRu4ichz2OccRVX2du3YxtoVfD/AybFe0d2EbZnxKVb+f\nd6NtPGMQkWdgO8utwPbTfq6I/ATbwvQOVX27iByQlzXL+LwQ+6zieuBDqvrLduIzSxkvwbY1XQW8\nRVVv6XQ8PFltbzU6izCfgW0hvBzbbOI12H7r21X1dZ2QKSInYZuS3Ap8Hnt15XHYDnzZl7fmK+Ox\n2BfRitgWofd1Kh4icij2KdxhVT3DvTveVP+8G1Vttv/8o6ltnHRIu3rNlmDfi9u+nbyO2vhSsW8n\nZ/e08W4v5Z/Fkv9DgHe74xLwWXf8CuyTpFdgo4sIW2Qwzc0s5V0EPMkdfx/bde0F7rxO1hzi8nxs\n16GPYduntozPPNLtGOCKbsTDk3Eu8DR3fEOH830lZvz/gn0P4PWdkol9B+AbwKexPfC/4K6f4sLv\nWLywLTZf2qV4fNH9t9Q/76ZFuO/DtkoeBj7TyXxtICvY99zSrev23alyOkO4S8K+XRi7lY0v+Mhd\nRB6M2z7VI78z1mpqW/6tx6YPD8R6/ql7g6eRm7ZkqeoG4O+Af3TXTnbh3iAiX20ga9bxUdXrROQh\n2Irkd7cRn1nLwLbNvJDaXvZzjkcLDqS2BWynP0/5Juyb9Lc4Pd/j4tsJmflvjmcjv/XUvr/dqXid\nDZwHfKwL8choR/9Gbmbi09j3069l+pafcyLY925p39A9G18q9g27mY0veOOuqr/EPslah4gc4p1u\nxL6mBvZhi1/gPkYgIlubuGlX1vHAf6tq6txl+8Fvxr5YlJc16/g4NmB7drcTn1nJcLuLXQO8T1VH\n5huPFmQfgriFDq3TcO9YvwP4hqr+3LuVpdm8ZebSI/uKFVgeZPvgzzteInIQcJ/Wf1mtY/FoQDv6\n+25m4m7si39g+2vPm2Dfu6V9Q4fL6VKyb9g9bXyxPHM/HPvS0lEicoGqflREfi4i78cy7Rrs4xnv\nwD768HFVnWzgpl1ejOsNi8hK4P3YNpqbVDURkU/4suYQnwuwb1CvBtbNoGtdfGYrA7ga+wzoy902\nnt/pdDw8vgR8UEROIfeRi3nwKuAJwLCI3B97/juKlcl3Yh+DmZdMmf7N8QNE5MNYxfgK56wT8ToP\n+JST+U+dioeIrMY++HOMiLwR+HIb+jdyMxPvpbZi/j3UPunbUYJ9L3r7hs7b+FKyb9gNbTy85x4I\n7KGIyDpVXeOOr1LV2XyqOBAILGIWxcg9EAgsLCLyBWAfNyoUwmuxgcCSIozcA4FAIBBYYoSReyCw\nhyIi/wCUs3NVPXMXqhMIBDpIaNwDgT2XamjQA4GlSXjOFgjsuawXkXeLyJUiEhbTtYGIvFBEbhSR\n94vIJ0XkUBE5QERe08DtWhE5ugs6PEdEPiUiD/eu7Zaf7xWRL7bhRrzjB4jIJSJyiIi8u4W/h4vI\nyzqh5+5IGLkHAnsuJwBvASZ2tSK7Gdeq6o3uNab3Yq/dxSJyBPbK3x3AdcDx2KLFz2E79B2DvRf9\nCmzL0WXYa2y/VdVPici7sDp5PfAB7BWpXqAAXKSqVREZwra3/TXQKyI/wl6R+pRr7LYCv1DVj4jI\nLdgObvtjr2s9yum+c5taEbkce0d/GbZF7AdcfN6I7WD4IOBpQB+w1sXpROBPQKqqb/XCehLwHOy9\n86uBk5zM5S7cZ/hxBm4DjhSRN7n767DFndtU9U0icivwGew1vd87MS/DXj/rBx4lIm918XsJ8BVs\nF7/7Ax9W1Z+5Tte1zbNzaRJG7oHAnstvsI04RtwvMAtUdSO2l3zGPljDdr2q3grcDFyjqje7+1NA\njDXyAP9PVS8FnigiRwKTqvoaVX0vtsPeIcAWYMj5Q1W3Azdh79LfCfxaVd8BPBv4pKq+Anu/HGCD\nqr7R6fRdrPF+Rqas64ycgO2VPgUcAbwOuAH4X1X9HfB8VX0J1hnJRsHfVNU3A/lZiZdiu/e9VlXv\nBl7pwt6A7ZNfF2eXLr9W1bdgu7/1O10PFZEScKeqvkNVf+/JeIALG+A2Vb0cK8ePxDpB7wcuo/b+\n97jbwW6PI4zcA4E9l4OAt2GjJWX6NrGBJriR+2R27rZC/RNwnog8lNqOaQBnqurTRWQNMOCuZR/8\nEPfz3UfAj1T1H2nOfV4Y+Vefsp3rJtzxJDYT4Mu4VVXXenF6ILAN2xjGR52MvN55Nz5jubDXNvCb\n+YmAr6nqVzz3Wdx80gbX/DAK1He4NmMzB3fnPS11QuMeCOy5rPWOwzux7fMyEfk7bIp5bXZRRB6H\nTWGvAr6NjVhfLyKfBP4qIpdgI9gf5ANU1VtFpN9NzY9gI/Nr3fkK4FWqOt5Ep+uB97pd29r6OpyT\nmYrIe7BR89uAq4BnAZeLyLHAP4vIR7AOyZuBxzQJ8uPAR0VkM7bf/j+LyEeBMeBrM/iZEpE3O/cf\nFPsCW0lVX93EfdYxOFxE3gbsh+2wOIntZf8AFxewGY/1TRNiiRLecw8E9lBE5O1Yoz4EPFhVH7eL\nVQoEmiIizwY2q+p3Gtz7oqqe4Z0PYV8i3CMX1YXGPRAIICLZs95AYNEiIhH2dbxvt+H2cOx1zz92\nX7PFR2jcA4E9FDflq9gzyuWqet4uVikQCHSI0LgHAnsoInKwO5xQ1Y588jUQCCwOQuMeCOyBuFXb\ndcavqlftInUCgUCHCavlA4E9k8+7/z7gQqa/+hQIBHZjwsg9ENhDEZFzgVOAD6jqj3a1PoFAoHOE\nxj0Q2AMRkZ8CvwD+BTc9r6pf36VKBQKBjhGm5QOBPZMPuP+9dqkWgUCgK4SReyAQCAQCS4zw4ZhA\nIBAIBJYYoXEPBAKBQGCJERr3QCAQCASWGKFxDwQCgUBgiREa90AgEAgElhihcQ8EAoFAYIkRGvdA\nIBAIBJYYi6JxF5GPdiPcJC6vTuLy/bsRdiAQaI8u2neUxOVjuxF2ILC707HGXUQOFZFPiMgX3fnZ\nIvIxEblORAZbeD+gU3rkeCjw9FaOkrh8RJfkBwJLgkVq373Y3vhNSeLyEUlcli7pEAgsSjrWuKvq\nn1T1PO/SM1X1fOAG4PROyZmtWkA7Rn1WtxUJBHZnFrF9t8NZLJJZykBgoehmgc8Mbz1wYP6m6/nf\nKCI30r2efbuNe2CBSeKyJHF5ZQs3vUlcbjUqJInLQ0lcLnVOu0AbLAb7ng2hcV9g2rDvlnWAc9eT\nxOVlndNsz2AhCvxBwF/yF1X1c6p6qqqeCtw5HwFJXC4mcbkvd62PDsYvicuD/tReEpf7krhcaMNP\ny8ZpD6UEvDp/MZdejwfOaSOslwCPaeYgn3+7C64CXMxlqOv2DdPKBc72BmYRRCtbLfkdxCQuF/J1\nSgM//UlcXrY7lqu5kMTlgVnGtaF9e2GsaOSmAY8A/r6Fbr1JXO6ZhW6Lhm7Zdyefua8WkWuBY0Tk\njcCXReTDwLOBL8013DZHZCcCL85duwA4DpB2R3XOoGf6Ut7rqR+hnA88MtNxBhmvd7+8nKb65O/n\nKp1pfl3PVpq5mcFPRzt3HRo9++kVAUU/TzK9XZqL567H72y5RtHX5/XA3t79KBdu0zRuRIN8KswU\nZjthZe5z/mIalKGFZlfadxKXlzM9DR6J2XjWuW9VlqMW8p4BnOqdHwe8NPOTK28ZrwYuBvbN6dvK\nvuv0nW3Za9NN1IXG7hLgsHmG8XqsTIObVc3FP3I2XvDsWbB6YMZ6EXv08ncz3ffru1nYd08+z+dS\nT2Rus3xvUF67Yt+dfOa+UVVfpqqHqerbsd760cA4MKcVrS5zL2twPa/3PsA+uYzox3rrAlzmKoid\n/vNhOL9PAM50BStyDYTf4/f99GKjT5yOl82gWyOZ0+KU4zLXS65z784byXk1cLQX/8u8gpzFw2/8\nI2y02+mVxpc169k7PTK9C9k1T89Gabc3cLZ3fgHwN8AbgEOcf8XK2tFZfIH9mJ7OmcwIeBz1ay0u\nc71/oZbG4qdjG/E9GXi2M969ffkzlDn//DIX3qD7H3bXp5VTL44LRpfsO8Irq951ycWvUVnuAbKR\n9TnACS3KUjGJy3s5eQUvXzP77nNhZhSwOgQsb14DHJ4Ltzen8057bVHxvwD425nsO7G3fPzGX3yd\nnZtezw859wUsT16Sc1PwwvPTalpZctcLvjt3q6dZGZyhXs3nSX4WxbfT47CB2lOA05wfxfLmsib1\naAno9XRbmbv/Yhd2nb98PZnT65VYGc/c5suPH06+rWhk3+cCxwPPA05iBprUhbOi28/ct2NGM23a\nrk2umOH6lS6hM7LR0skNdMgy7LW+f/fz+Vvg0cDBTu6V7vyKJC6vaKBDCtRlgisc+XCbyWzGJVg8\n/QJ35QzHQ8Cz3C9//0pswdNJuXv7Uqu8OsnDmtx7CJAtysrS7bAkLu8HXM7M6TOcOy45/6cDT8by\neRAra1cCT8PyJ89Lk7h8uHMzxPRp3XOpr3ge7q7RRDe/QRgCysCbsM6Dz2v9MJK4fD+swZiJ7F5+\n0diJLpyTmvhdCDph31l65CvVZ2KPZPJcjFX6T81dH8TSI3vt9UrMdvz67TlYZQ3OvhN7TTarYw7N\nhZlSXz/2Y425b5NZGVOsLF6Z1EbLzR4DDGB11UVJXD4Mdja8WXoUXVhZGKdRq5My3ug6EFcm9rbP\nq104Wf31FGozCqclcXmAWl12FtbIALwcaPS20JUunIuTuDzkyT4Ls4kryXXqcnHIOMML7wJ33Kxj\n2o/ZUZ/T6w3U20BdPZrUntkr9nbU4939fOdqH6yuyPxJEpf7vbAuZPqsxCD1dUSWt1cw3XaPZXp7\nlU+L1S6M5cAJNMDF50rae1zRlJmmoDvBD1X1ByKyL/AerLeyExE5m9qIrK0FN0lcfjTwJHd6HnB1\nFpz7Pz6Jy8cD17rz/DROMU5GKt75Wu921oD3etcOdv8Xuf8XJXEZ4B+pZZAvY40X7n+1iMsb4mTk\nHc7tFPAPWC/xkJzTY/O6ZsdJXL4M+IDn9ugkLh+ddwc8GPhJLr4AJ7nK6G+BbwJPjpORvJtmcVgL\nvDlORqpJXM4qx5WenKuBS2fwl3EQuYKexOWn4B55OFY7P9dg6V1215fl3GV5t8oLay9qlTpYJQ9W\ncfS5cCfctQNd+CRx+clY2k/6emfpk8TlbIV4n+fmIZ6cI3N+hhrEPQvXn0HId7j9GRff74lJXB6O\nkyTbIC0AACAASURBVJEv58NbIDpp3xFQTWx2ze+EfzfnbhAbfd2GdSwyssct+UVXL0vi8j7ueL8G\ncp8Idel6hDv+twb3oPZWwBpXD2QUMBvLdAQr92uTuHwRMAr8D/BbrNO+wblZTq3z6DcML3f/z03i\n8h+AsexGEpcv9txlI8ez3L3nAg/0FXP6b6DWSJ2J5cd9XtyGfPe5OmAIeB31ZLY+5OqbM/ybuTTz\n64MsD57q5Yvv5z+wumAr3ogZr1FuRE7eY93/me7eacAxQNWd7+/uHwS80PO3AjgnictXxclI6gYB\nReo77/7bIss82W/D5fsM9n2kOxwCKn6YucGh73c4icuXxcnI2xpGug1Etd23SeYoQKQEfE5Vz2ji\n5ka38KaOBglVpFZIx7BGMaJWqU9Q63Hnr/mscGGlwKZ249IGQq2CGcdrGGagSq2Hvx4rcGC6Z/pv\no/1XfubDz7CR6mYg6w2vi5ORabKTuPwG4J1YD/ObzGx8U9QbR56VWPwrwJbcvT5qjfR9wA6mj6Zw\n/rMKddSF5fMXGqzmbsJWrNJNqD0bnKS+E1eh1jH+UJyM3NPIqNvkVuCoNtxNUN/xBGA2nbFu0GH7\n9u1nErMhgNuxztgUtfzNZp38axmD3v1NNJ7FmStDWBmsYOUN999qcd8OrM7ai/p6rFGZ7Qa+XbdC\nqD3ymMTqqV5qjfM2LF1bvY20AounYh2MqnevSM2+dgD3NtFlP/e/A6sL5sNtwAOAG6mtsRihNmDI\nc1OcjNyUxOXXYvXCbLmGFosBHQ3L0Hzsu2uNu4icjo2yVwAfVtWbmrhtaPxH9pTujKAwEMnyCKSI\n9GSlqYJOjqpuHRRZ1oP0AUyioxvT9K6VUbTPgMhQFrNUmQJ0XHUHoEORrEqhUkAKiqpCmkKlokz2\nCL0lpE+tQCqgavImBJhSJnZouqVXpH9QomEARatVqJSQ/h6hN4VKBMVUmUSQMdWtVaUyILJMoCAC\nY6pbK8pEnzAgSCGLk0K6XKK9FFKBSCy6iou3CNGE6vaqMlUUejK/Y6pbp1QnxJlaBFFJpD+zvBKi\nBUGLCAXQgogWQYsiPLBYHNsrKlRWF6LRR5R6v0KtM/HOOBnJKjC/Fwxm4MuAW4CHra9UTry9Urnf\nmGr048mJ8S1puj2CPoFeheIkuipVRISCmGELIBVlKoIexBK6CpUqWiki/VV0Si3/dIfqFgVxiSER\nFBSkV2SoAMXUNfxTqhMuvyqAVqFaVdLI0r3k0qaYpWsVplR35jUK6ajq9hSt9Fn6SREplcTKmEK1\namHjpZMCFJCeXqHfC0sF2J7qpgo6NRxF+xSRoiur46pUFNIKWikghZLsrFB3GmVFmVRIFdIx1W0R\nFPpFBgCqUP3PifG6EdBC0Qn7fmBPz10CDEi0PIKoIBQjiFIr+2xLdVNJ6OuXaAhQRfXuajoyFMnQ\nMolWY+kbVdFJgEllrIpWhiRamUI1goIAKVQVtIKOR1DotfTbmXeujE2pHafbNN0YQWGZRKuyclJB\np4pIsVdkoGr5VayiU4IwqTo2iY73iQz2ICVAJlV3jKvu6BHpLbryXbX8rvaLLCtYp1cj2TnIsLQy\nm5ioolMgiJV5mULHJ1RHnT4IQknoc3UEYqaRuUcgEtnZEVaACNKDCsXJXhGWR9FePUCviAxKtCFF\np0ZVN25M0ztWRdERAyL3SyHdUK0WxlUrU1D6a7U6nqJphPRsSdO7LUiz435Xt6nTT0EKUJpEx4Go\niJQUSNF0UnWsgPQ6O50qIr1bNd2Y+UMRxOq+ApQE6a2i1SJSmkInYGdapoqqy+uoxzqaFKAnQgpY\n+JNVpeKqFwCdQMemVCd7RfoiKBaQqEfoFUQU1Yp1GLMSoiJ2GEFUQrJOo7o2gkl0dFx1dFBkOKtz\nK8pkBZ1yuk4iSK/5resMVdFKap07HXPtU5+VTwH4ycREux2yaXRzWv7fsWdFW6n10GbFG5cvF0C3\nq/7fhOpmheqOVDcVhNK+UeE4tcK14+60+j8R9JQLxdPURr+FO6uVr4+qbukVWR5BoU9kZUlkFcC4\n6m83pekdERRKIgMKUb/Ialex6H1p+hdrZ4gEoiL09ogMChRXRtFR2LMT3Zym/zuljPcIgxEUUtAt\naTqSQqUIfZFIsQ9Z3h/JfgBV1Q3bVO/qFRkeENnfrrF5TNONEVLsEQYVmFT9y6Y0vV2BHmtUNLWi\nqQWR0qDIPgApVCZVR0siQ0MisUJJvYZlUvXeFKZSq2ykClRUpQJSdf/b0rT42dEdwwryy6mpw48t\nlR5YQtJnDQzcflJv35Fb94+Ly6PoF8DncQ37hOrDx1QPvKtaHXrvtq1XTKry3Ynx6IBCYesyiSoP\n6ulZFRcKaQRpFbYVQHtF/pRarzutqI4BVNysSgQU3OLHfpEhERGU6pimmyORwr5R4ajI6jOVrMNl\nlXEVqG7R9I+pajoQRcNFKBagWPz/7J15mCVVebjfr+7W+6wMMsUdAdkRDS5o3ElckqAJigZC4goi\nhhgx4jKyzQgoLo+IomwKioqAuEQwapRNJYnmJ6AERBGEuRbb7Etv996q7/fHOTX3dPXt7tvd9/b0\ntOd9npm+VXXqfGf7vnPqrCIlAYrG6AQCWlM2gNZjqMYwGkCpgPQKBGobUl0iK7pEwgQ0QQerqhsT\n0CHVqK6MFIV+TOMgNZwB1gADbFd9tK5apWH0liwKgoMAasqDjyfxXTko9ojsgcnf3gDyCjKiurEO\no6rGv0DIF5E+BekRWZkT+tWk/4aaMrQ4CCab39BpZq3f5wwsVgFieGxLktwPMKzJxhiSPXO55+bM\nl8zg5iS5fUh10z65/OsCMcNWm5Lkjs1J8mABenMipQL0dEuwEiBGK+vj5B5FpSTSD1ASWZJDSgCD\nmqyvwXCjoiRfRPoFZCCQg/IiSwDdkST3D6luyIt054xusSNJHh+FwZwpZ90FkVKfBKtARWHj1kQf\nDoTiIgmeZsvV6JAm6wEKIn2AJMrj65P4ftuAL6TW3jZeg74g2AMQVVTRJBDJ94jsLUZH1LZEta66\nqQ5D9jqx/6UfJEkMVW2UTRlULT4Wxz1bkiR4qF5/bIcmI3Wlt4ouCoAlQXBwHjlMgI1J8odR1cFF\nQVAfCKQqoEuD3MYcJCtzucMCJC+QWJ1M6qpPDKquT/VTTMN5NFatiogGNq17RBYVkCKCDibJeiBe\nkcsdkoduEBX7rkJdoSagQ5o8FMNwUaS3gPkYsnpIznzYFQCq6OZYdTSB2ojqlgDyPRLsYXU7ECHI\nQ1+PBKuAkqL1IdV1QDKqurkGQ3noyiFdNk/ENgjsdxUyrLqxhg7aRlSQF+lbLMEhNq/XP57E/5Mo\ncV8gTxEIctCVsx8GVdVtVdXtafkXkKLIEjEfKcsLyGKAOrp1VNk0EIg7NDFtOvnl/kZgi6reKCLX\nqeqEu8C10G33MOPHol3S7t88xtgqaetrZrTSzTZRd/SHmd7EuabdrZatwA2MHevpBKNA6Rejowf+\nvl4LH4vjRV8c3HFYbAxC7pB8oWoVTgBZF9cLddVYRPSFxdJvn1cs3r84yA2/tqf7g90SvL0FeZtw\nxsUn4XGaj5U2o1mX/X3Yse8OcTFjx/Onw79jKsfZ0BtGlV2yTK7N+t0Kd2Mmazp14Yy5heYT9ly2\n0JiH47IWO7emHYRRZU2TdHgM2KuJ87nmAsyEtvvDqHIttJxn7hDjdzDLDNtBq3YjpZWhCHdoNEs6\nVJnyMJPXQy4PMX6S5nTpCaPK+2f6cidny++NGcuAsWMt0+Vn2PHYdPzB/r3ePv8p8AfgQkz3RhWo\nWTepAfilvV4LfAGIneu1wCXW3W+B22hM1nHX7w7Zd66111XMuDPWD8KosiaMKmkX26cxY9ZrgPPs\nvVuB25131mIUKL1+PBVm/bowjCoVGukI5ovpo046/M7e/6r9e73jdj3wTfv7fCvjK64M+/NXwF1H\nlkr/eEJv363vHVj0b/ftFfb9dq/wvJv32PPq9w8M3PL+gYGbPzAw8J8fHBj4/leWLv/6gyv3Pu/3\ne4Xnf3nZ8hv+uX/g/07o7f3S/o9GkSP7ahpjkk84su6ikd4Adzi/N9nwr7Xxv9TG4deOm4dsGrjc\nDTxuZQxZ2WsxZSPl606eg2k43eI8/zbwNcxwg8sjwPfTdzN+1GyY3fHCz2XKVsr3XD/CqHKX8+wW\nYDTjd1oG1gC/d9x9zv6+fldV7JZ26ff1NPQ0nQh7CY10+AmAM3EwHUL5FPDf9l6STXP32il712Fs\nSdrw/4P9u9F558P23jbnPVe/07DexvjysCXzztrM9fedeH/JefZdN0HCqHKZ8851jMUtU/8JPIiZ\nH3IT48scwJeByzBpthbT03JNWrYyY7pr3TQIo8qIvXct40nlRNbtTn/CqHKl4y7CNLLB5OUDjE2b\n/8TotMvHM9e/tv5+Bvgv4B7Gpi2YBsgaxqbBVc7zPzjuf+7cvx3TmFqb+Xc/xhbcAjxh3W4H7sjY\ngp1k9PtqGps3PeQ+p2GXPmvvfcNeb3bC/gBNJiNPh05/uW9W1ZtE5FpVPT7z3J1N262q2WVsHo9n\nnuL12+OZ33Sycu/FdFuOAD9T1a/NwI+bcZaBdIiVtGF7TC/Dy9gFMmAXVZxev72MeSxjruTMhYwZ\n63fHJtSpmfn31ll6M9xsrK6dTDQe6GV4GfNdRiqn0zKa4fXby5ivMuZKzlzJmOm7nRxz93g8Ho/H\nswuY75X7NV6Gl+FlzAs5nWCh5IOXMb9kzJWceS2j4zvUeTwej8fjmVvm+5e7x+PxeDyeaeIrd4/H\n4/F4Fhid3H62ZUTkSMxxjhVVPd3eey+wL2bnuVMwOzZ9ArNhxlWqemvWjbYwxiAix2B2llsMqKr+\ng4j8ArgTeERVPyoiK7Oyphmft2BOJVoHfE5V72klPtOUcRLmhKylwHmqene74+HI6sUcgFAFbpvJ\nsqcmfh4DHI05jOGLmCMUHwF2qOrp7ZApIi8DzsUcznItZunKUZgdAdOTt2Yr48WYE9HymN3wtrYr\nHiKyH+Yo3EWq+nq7dnzS8Gfd2FntuxSv3/Nbv628tur4QtFvK2f31HFVnRf/MNv6fdL+LgJfs79P\nxRzjdxbmWMUAM8lgnJtpyjsNeJX9fStwJfBmez1G1gzi8ibMtrFXYM4RnjI+s0i3I4CzOhEPR8Yb\ngdfY39e1Od+XYJT/m8DlwPvaJRNz9vn3gS9hzvr+hr1/tPW/bfHCbLH5jg7F4wb7d8rwZ920M69m\nmT5ev2eWbh3X73aV0wn8XRD6bf3YrXR8zr/cReRw7PapDm/LXC/DbDsKpnW8t/1XUdVEzAFIzdy0\nJEtVnwRegTmXHeAvrb/XiciNTWRNOz6qerWIPAOzheAnW4jPtGVgtj79Vxp72c84HlOwN2bLR5jd\nVqPNOBOzperdNpyfsvFth8zsmePpl986GmevtyteJ2DOALiiA/FIaSX8zdzMGV6/d0v9hs7p+ELR\nb9jNdHzOK3dVvYfGObo7EZF9nMuNmHOPwRxA8Gvsudwism0SN63KehHwc1VNrLt0P/jNmHOgs7Km\nHR/Lk5hjUVuJz7RkiEgB013zaVWtzDYeU5CeiX43bZqnIcYaXQB8X1XvdB6laTZrmZn0KNE4WnOV\n9Z/ZygAQkVXAVnVOfKKN8WhCK+F33cwZXr93S/2GNpfThaTfsHvq+HwZcz8Qc9LSYSJysqpeLiJ3\nishFmEz7POaAhAswh8N8Qc1Rglk3rfI2bGtYzPGOF2G20dykqpGIfNGVNYP4nAw8C/P1sXaCsI6J\nz3RlAB8DDgTeabfx/HG74+HwLeBiETkauHEW/ri8C3g5sEhE9gf+HHPoSx5zcMSvZitTxp45fjGw\nUkQuwRjGU62zdsTrROwhFSLy5XbFQ0SWYQ78OUJEVgPfaSH8zdzsUrx+z3v9hvbr+ELSb9gNddyv\nc/d4PB6PZ4Hhl8J5PB6Px7PA8JW7x+PxeDwLDF+5ezwej8ezwPCVu8fj8Xg8CwxfubcZEXmLiNwk\nIheJyJUisp+IrBSR9zRxu0ZEnt6BMBwvIleJyLOde7NeCLsrEJEbWnAjzu8DROT9IrKPiHxyivee\nLSKntCOcnj8NvH63F6/fnWNeLIVbgFyqqjfZZQ4XYpblhCJyCGZJ0CPA1cCLgBUicg1mB68jMOsm\nT8VsSdiPWeZyv6peJSKfwOTZOuCzmCUUJSAHnKaqsYj0Yba/vA8oicgdmCUUV1ll2Ab8WlUvE5G7\nMTs87YVZzvF8G/ad21iKyBmYNbz9mC0kP2vjsxr4NHAw8BqgC1hj4/RS4CEgUdXzHb9eBRyPWZf6\nMeBlVuaA9fcYN87AA8ChInKmfb4WEGC7qp4pIvcCX8Es4/mdFXMKZnlKN/B8ETnfxu8k4LuYXb72\nBy5R1V9ao3zp5Nnp8YzB67fX73mP/3LvIKq6EbPXdMoKTMH/uqreC/wM+Lyq/sw+rwEhxggAXK+q\nHwBeKSKHAlVVfY+qXojZgWsfYAvQZ99DVXcAt2HW2j4K3KeqFwBvAK5U1VMx608BnlTV1TZMN2OU\n+5g0sNZYvQSzl3INOAQ4HbgOuEtVfwu8SVVPwhirtJX8A1U9F8h+tbwDs7vXv6nqE8C/WL+fxOyj\nPSbONl3uU9XzMLtDdduw7iciReBRVb1AVX/nyDjA+g3wgKqeAfwGeB7GSF4EfIjG+tARMTtceTzT\nwuu31+/5jP9y7yC2ZV9Nr+1WiQ8BJ4rIM2nsqATw96r6tyJyDtBj76UHAoj957oPgDtU9TNMzlbH\nj+ymBunOVqP2dxXzpeDKuFdV1zhxOgjYjtk4wkWtjGy4s25chjN+r2nybvpOAHxPVb/ruE/j5pI0\nuef6kWOsQd6M+bJ4IvuSxzMZXr+9fs9nfOXeGU4RkVdguqDWpDdF5ChMF9dS4EeYFu37RORK4DER\neT+mhXt71kNVvVdEum3XXQXTcr/UXi8G3qWqI5OE6evAhXZXp5ZOj7IyExH5FKZV/RHgw8CxwBki\n8lzgqyJyGcZgnQu8YBIvvwBcLiKbMftxf1VELgeGge9N8E5NRM617i8Wc0JTUVXfPYn71HAcKCIf\nAZ6C2Xu6itnr+gAbFzBfROsmTQiPZyxev5vj9Xse4Xeo8ywoROQNwGZV/XGTZzeo6uud6z7MSWV/\nspNuPJ7dCa/freMrd8+CQkQCzOlZP2rB7YFArKoPdj5kHo9ntnj9bh1fuXs8Ho/Hs8Dws+U9Ho/H\n41lg+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3D0ej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H\n41lg+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3D0ej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H\n41lg+Mrd4/F4PJ4Fxryo3EXk8k74G4Xl5VFYPqATfns8ntbooH4HUVh+Xif89nh2d9pWuYvIfiLy\nRRG5wV6fICJXiMjVItI7xesr2xWODM8AXjOVoygsH9Yh+R7PgmCe6ncJ+OupHEVh+bAoLEuHwuDx\nzEvaVrmr6kOqeqJz67Wq+nbgOuB17ZIz3WABrSj1GzodEI9nd2Ye63crvIF50kvp8cwVnSzwqeKt\nA/bOPrQt/5tE5CY617JvtXL3zDFRWJYoLC+dwk0pCstTfRUSheX+KCwX2xc6TwvMB/2eDr5yn2Na\n0O8pbYB1V4jC8kD7QvanwVwU+FXAH7M3VfUaVX21qr4aeHQ2Amzmd2XudQO52fib8a/P7dqLwnJ3\nFJbzLbzT164wLDAKwL9mb2bS6y+Af2rBrxOBF07mIJt/uwvWAM7nMtRx/YZx5QKrez3T8GJSWxCF\n5aLbQIzCcs7akMne6YnC8sDuWK5mQhSWe6cZ16b67fixuJmbJjwH+OcpwlaKwnJhGmGbN3RKv9s5\n5r5MRC4FjhCR1cB3ROQSTJfYt2bqbxSWSy04ewlwUubeycBzAWnRj1ShJ6qwTwfKzvXbgeelYZxA\nxun2X1bOpOHJPnevm71rDZNM5qbJO4UoLLe1cddqOlsmMhJuegVA3lXaNNw2zVM/BCi4eWcrRTc8\npwN7OM9zGfeTpnEzmuTThH624lfqPvNeSJMyNNfsSv22X23ZNDgSeId93kpZDqaQdwzw6gn8L2V1\nzPJu4N+APTPhnUq/x4R3umWvRTdBByq79wH7z9KP02n09AiMi39g0ydri3OT2UXgOOCVEz1303wa\n+j0uz2diJ1K3TrwKUVh2G5sd0e92jrlvVNVTVPVpqvpRTGv96cAIppKdNjZzVze5nw33CmB5JiO6\nMK11AVZHYXmR+37WD/vuK4HjorCct24kkwmu/0XM1yc2jKsnCFszmePilGG1bSWPcW+vm8l5N3C4\n694pyEEaFzcsmMbJkVOEY7qsnqxlb9MzDWMuveeGs8lrewAnONfvAJ4FfBDYx5F3GHCYVR4BnsL4\ndM6nsoCjgOMzYU8bDGkai5uOLcT35cAbrPLu4cq38cs515Lxc7X1r8/+XWzvjyunThznjA7pdw6n\nrDr3g0z83LIs9lkBM6EOTO/OS6coS/k0T1L9tm7TCqSLhj6DKZ9pb+BqTCV+UMZf90vfzd/V0eTD\nRG8GXuTEc4x+R2F5WabyFzfM1k3JeYeM+zxGt9+ecbOz/GfSalxZcirZrP9p43rC95pdZ/Ikmzeu\nnj4PeBvwN8DfuXJtvCeyo0WglOpVFJaXZJ6fCDw/+14mfNn4/AtwuON2OWPLj+uPZBoj2bRYDbwR\n08P4j8DLmIBJbOG06PSY+w6MgozrtmuRMye4f7ZN6JQ0UV/eJAxphr3Hfd/+c3kRpmCtsnLPBl4A\nnOUYWpcEGJMJtnCcPYHhbSYzLRTN3L+viV9nT/C7FzOp6XVNnp9t7x+VebYCa7zaXFH82STPnkGj\nhyVNt/2jsPwUGmnejIHM7yImX4/FzJZWTBp0AWcBf4vJnywnR2H5ICunF8h2u76Jsd23z7H3mCRs\nrhHvxfTunIEZUnB5rw1byr6YCiNLOpZ9WuY65WXWn6PYtbRDv9P0yJa/Y4C/bOL+dIzRPzpzvwd4\nKZAuez0bozuufTsBONX+PtM+35+Gjdkv42eq32nYujB57epkWsYUo2NnRY2v5Z2N1yb61Y0pH+9J\n5dqGQVrG8lZOWhaPYbx+pA2Is6OwfCi2vERh+anW7V9hdBxMJdkDnGlt2XEYewemu/tQxnM2phy/\n1zY4U9l/j6mkzibzcZCJQ8qxjn/vsL8nszfdQB+m4XYwplJ0dWCMHbWVONbNMzDpejZj9RLMR8Ir\nnfckMsMuqV+nMb5XooexQz9p3p7JeN09kvH1VTYtllo/+oEX0wQbn7MZW1/NiEnHjGfJT1X1dhHZ\nE/gUprWyExE5gcYXWUsTbqKw/AIaGXQScEHqnf37wigsvxC4zF6PyeAoLOfDqFJ3rtc4j9Mve7er\nZZX9mxrat0ZhGeAzmErmJRn356R/o7D88yni8qEwqnzEvhNHYfmTmK/PfTNOj8yGNf0dheUzgYsc\nt0+PwvLTm8Tv6cAvMvEFeJk1EC+IwvIPgVeFUSXrZrI4rAHODaNKHIXl1DguceR8DPjABO+lrCJT\n0KOw/DeMNRzL7Dufx+RpOjzSl3GX5sUSGl1+e9Aw6gD/YP8eimnprwFq9l5o/ScKy3+N+SIddcOd\npk8UllOjVXLcPMORc0jmnd4mcU/9dXsQsg1ut8fFffclUVheFEaVb2f9myPaqd8BRgcGGGs0f5xx\n14PJkweA7c79dLglO3b5z85HwArG8yoYk66H2N/fpWFnznHcp43nc6wdSMlhdAxsPmPK/RqMkR6y\n9uB+e/8J66YfU1HC2IbfO+3fE6Kw/HtgMH0QheX3Oe4+ZP/+vX12AnCgGzAbn/U0KqnjMb1aW5x4\n97ruMzagl/HdxqmN6o3C8uE0KnBXZoprD9Lhi6OtXmbf+SnGFmyjkZ5gGioTkpGX2pLj7LPXAs8E\n6vY6LYurgLc67w0A/xiF5Q+HUSWJwvKBmPrRrSPdod8+R/ZHmFy/D3PeqeP0EGU+Dt13+6OwfGYY\nVc5rHuupEdVWV5PMUIBIEbhGVV8/iZub7MSbMTRJqDwNBR7EGOUcDaM+CsQYY9GVueey1PqVABta\njUsLCI1GwjBOxTABbs/CH2mMRXXRCP9WWl/yMxvuxHR3b6URh7VhVBknOwrLH8I0rM4Gfog1kk2I\nmXwi0zL7vA5syjzrxuQTwBYcA5chhzGS0CgTLo8yvdnaOzBlLMJU9tjwuUqe0KiEPxdGlfXNlLpF\n7qP5l1OWGmO7jQGYTmOsE7RZv139GcXokGBm5B+ESYM0f9PKyr2X0k+jV2YDzXtxZsoAJu9rNMrk\nCA19nYhBTJxS25O1Y53G1eupcO1nFVP+u4C97L1twMYW/FlKIx8eZ6wdztNopA8CT07gh9DQ30GM\nLZgNv8d8od9EY46Fq+tZbgujym1RWD6d8Y3HVriERmNtMpqWodnod8cqdxF5HcboLwYuUdXbJnHb\nVPkPKRQeD5CgR2RABPJIUVEFNIH6kOq2bpH+IlIEqKIjG+PkiaVBsLxLpE9BA5AYagpUVQcVtC+Q\nJbFSywl5QFWNf3W0mkdKRZESoIoiiCqqdaUKUEdHB1W3FpHuvkAGABIlTqBeFLpySDGBOAf5xCit\njKhuj9FatwT9gVEcGVbdVkerXSI9YiuKWKna8C0FEoFAkABjnMS2AqSmOliHeh7ygZjKc0R1e01N\nY0LMv1zJxAMByQtJDkGAnEAJSfqDIAbYP58fWR7k4qVBMPS8Uuk7NBoTHw+jylCaH04rGIyi9QJ3\nAUc8Uq8f9Yd6fd9hVfmv6kh1c5JsD5AugZJCPkaX1BURyAWmQg4UqEMtgLzsvNZaosR5oStWagkm\nwwc12ZpGTYEAySkqJZHeAMknaCIQVG0axmhdTd4kMRoHSD4nFMy7Zi6GmjSvqYmvAiiqJr+IL2ja\npgAAIABJREFUSyJdAQR5pFAQukyZIEmUum2V2XdQm675okgPYysTHUx0Sx2tLg6CFVY2dWUkgVgh\nidF6ALmiSJfY+Np807oJX6KQDGuyPUDyXSLdAAka//fo6JgvoLmiHfp9YKHwBCA9Iv1i0jknQpAo\nSSDI9iTZVBTp6hbpVZMuuj6O/9grQV9/IEuNfksQozWAqjIco/VekcVGByUQCBI0BpKaMhoIuRIm\n/Wy+I6Ax1BVNFJLtiW4OhNyABGkPkFrbUCgK3QnUc0g+Ruti5dZUR0oivQUxPT9VZWhEdagoFHNI\nASCBeoLG3SL9OSQHEJhKTnd2z4ixA3W0KiBi1b6uOjqKDjszSKUoUkpth3Vrktte27KWFicNBC3n\n8tWiCAMiy4siFEXoE1lv7enGjXH8yNJc7uAekafGSvJkEhdGVOs11cLjSTySKHEgFLcmyXp2mhqk\nW6RfkJyCqE3WAClUVYcFcjmhmOpjFR3JISUFravWCiKlbUmySXdGDcHaybx5r2htQjG1w8amq6qp\nC2Krp0WAvFCwtpMaWk2UGEdXq6rDNbRaFOnKQT5ApGj0GwWNlbqTbikaCLki0q1Ompq81qERdKhX\nZFER6QJj12K0pqCJmnqgKNKdHY+I0bq1AzqiOghoCekRW+7+tzq6hBnSycq9F9ONWgVuU9WvTeK2\nqfJ/ddkejwIMafLwiOrmBOJh1c0BFFbkckcaxUXXx/H/BlDYO58/RhUVCKI4vmlYk60lU+iCkrC0\nJLIMYFR1w6YkeSSAoGCMsXSLLBfIKejWJFmn9utMIMiJFIvQK5BfHARPN4VYdUuS3FuD4QL0iVFW\n3apJpa5az4uUclAoivT3iuwFUFOGd2jyWElkUZ9IqEAMoyOqG4B8AekDGEW3b06SRwDypnLUtEDl\noNgbBCsASVTrVRgsQl9vEOwNoGoqFgWtotsTpZag1Ix7YpAE2JEk+WHVYFg1/7t6bUCBX9Vqezyr\nUBwpiSSv6+555Kiurt92i8iiILgbs1nJcTb9njOoyarH47j3wu3b9q0q3Do6Invnclv7JIgPKRSK\n5VwuDkSSWHV7TkQLsCUxre6krjqMMahVTCOGnBhj1y3SH5iKtz6iuiWA3J653GHGeKG28lM1Dac6\nEG9Jkj8kEPeILMpBPieSK5hWsBRF+myDSqvKhsRU+tUEqnkoFEwjUMTmdUlkzxKyEqCODlaVjYAO\naRLVYKSA9FuDvNMApf9EkG1JEtVhVEwcpFtk6SIJDrINlx2Px/EvnTwkD72B/SIfUt0Ym687UWOc\n80WRflXoDSQMoEtBR1U31GF4sQRHnLhpwy5ZhtUO/b5u+R4RQKwMbdXkfoBh1Y2JoitywXNySI+C\nbEmS/xvSZONT8/ljA9tLszlJfrYpSR4siPTkoZhDentEVmIa88Prk/j/AIpGp6QksjgndKkiQ6pP\n1NFhnHQuifQryCIJDsoLAwq6I9EHh1U35IWuwH5VDao+NqI6lIN83lQOpX4JVmEaEfWtqg/nobgo\nCPZXCFSJh1WfVNCimLkjMdQ3xPFvFZIAyZlvEIOC9AXBCmvbEvuBkusTKWMawIrVgboyXEeHEuce\npjJRUK2bspRW/sEO1cLjcdxTR4OhREd3qI7UVHtr6OIAYUkQrMhDLhDYECd/GEUHBySoLQqCqoAu\nDYJNOUj2yuUOy4nkMbqbYCrEoUFN1mN1FEhiGE2gJk7jpUdkUUGkKCYt16PEe+SCg/NIt31PRVCg\nrkpNBB1K9A8xOlwU6S2Yii9XFFMO7AdEHqCKbjGNZq2Nqm4JkHx3IHtgvmcECApIX7fIU8WkdTyk\nSQQko8qmGjqYh56c/SjC0e00f0ZUN1ZVB8X6WRDpXRQEh4rN68eT+L8V4j6RvQSCAOnKi+m9qKpu\nHVW20fA8KAlLQIIuYXlBZDFATdlWVd00EMgz3rhxw3RWII2hk5X7G4EtqnqjiFynqsdN4naqbruH\ngX0mEZd2Weawhhw7xjJDhph6De0PaD4W9GEmnnzVjMm69DZjlhmdOMHzdjECdN1ZHd3/gXo9fDSO\nF31hx/bDY5AdqrmD8/mqIPbzVqUSxwXbgtcXFku/eV6p9JslQTB8XE/vGdgZulOwAVg+pSt4jEZX\n4FS4XeUp92LmMXSKizEzamfCdzATpWZDTxhV3j9LP2ZEm/W7FdJho3ZwC+MnPGbZgumVyLKWsePw\nsyKMKmuapMNk3cRzyQWYVSm/CaPKddBynj0CPNX+/jbw2jaFp1W7kbKJxtDeRGSH3Fx+CTzbuX6Y\nyeshlweBp7XodiK6w6gybs5Sq3RytvzeQMX+zo55T4efYsdZ0vEH+/d6+/wnwEPAhVZODahbN6nc\n/2ev1wJXOM/X2n+XWHf3A7fSmKzzTSccg/adr9vrUcyEMawfhFFlTRhV0i7ZT2PGrNcA6aSIW4Hb\nnHfWZvx4LBVm/boojCoVzJhjyg+Ajzrp8Ft7/6v27/WO2yeBG+zv862Mq10Z9uevgDufVSy98bie\n3tvf0z/w3nv3Cvvu3ys875Y99vzSBwcW3fzBgYEffWhg4AdnDiz+3jXLln/t9yvD8363V3j+VcuW\nf+uUvv7fHNfT+5UwqkSO7C/TGJN83JF1J40JjwA/c35vtOFfa+N/mY3Drxw3D2LG+V3uAh6zMgZt\nHNdm/L7GyXMwZeoW5/m3MGm4jbE8DPxH+m7Gj5oN83rH/cWZspVyk+tHGFXudp7dDIxk/E7LwBrM\nBLLU3cX29/W7qmK3tEu/r8N+/QGX2r+X0EiH2wHCqPLdzHufAv4rlZ9Nc/faKXvXYhrM6Rj3Q/bv\nBuedD9t7W533XP1Ov4ZuY3x52Jx5Z23m+j+c8F/lPPt3N2JhVLnCeefaTLzdMvVDzBjyH4EbGV/m\nwOjhZZg0XosZe78mLVuZMd21bhqEUWXE3ruO8aRyKtbtTn/CqHKV4+5RTCMbTF7+jrFp858YnXb5\nWOb619bfi4E77PXaTNh/00Tv3HA85Lj/H+f+7TaMazP/7sfYgpsx8wXA1As/y9iCnWT0+yuYRhqY\nPHLL5E/t/c/ae9+w1xudsP8O07CaMZ3+ct+sqjeJyLWqenzmuTubtltVmy198Xg88xCv3x7P/KbT\nY+4XY7p8fzbZmNwkftyMmTHbSVbShu0xvQwvYxfIgF1UcXr99jLmsYy5kjMXMmas3x1b565m5t9b\nZ+nNcLOxunYy0Xigl+FlzHcZqZxOy2iG128vY77KmCs5cyVjpu/6k5I8Ho/H41lgzPfK/Rovw8vw\nMuaFnE6wUPLBy5hfMuZKzryW0fEd6jwej8fj8cwt8/3L3ePxeDwezzTxlbvH4/F4PAuMTp4K1zIi\nciTmSMyKqp5u770Xc/pQATgFs1PZJzAbZlylqrdm3WgLYwwicgxmZ7nFgKrqP4jILzCbqzyiqh8V\ns43lGFnTjM9bMCc1rQM+p6r3tBKfaco4CXMi2lLgPFW9u93xcGS1vNXoNPw8BnN05wDwRczpWY8A\nO1T19HbIFJGXAediNtG4FrN05SjMQUPpYQ6zlfFizIloecwBMFvbFQ8R2Q9z7OYiVX29XTs+afiz\nbuys9l2K1+/5rd9WXlt1fKHot5Wze+q4qs6Lf5ht/T5pfxeBr9nfp2KO8TsLOBzT23BNMzfTlHca\n8Cr7+1bgSuDN9nqMrBnE5U2YneGuwBw1OWV8ZpFuRwBndSIejow3Aq+xv69rc74vwSj/N4HLgfe1\nSybmjO/vA1/CnAT1DXv/aOt/2+KF2Ur2HR2Kxw3275Thz7ppZ17NMn28fs8s3Tqu3+0qpxP4uyD0\n2/qxW+n4nH+5i8jh2O1THd6WuV5GY0vPdZitLvfGtPwTMQclNXPTkixVfRJ4BeZcdoC/tP5eJyI3\nNpE17fio6tUi8gzM+c2fbCE+05aB2Tv5X2nsZT/jeEzB3sA99vdsthptxpnA54C7bTg/ZePbDpnZ\nM8fTL791NM5eb1e8TsCcAXBFB+KR0kr4m7mZM7x+75b6DZ3T8YWi37Cb6ficV+6qeg+Nc3R3IiL7\nOJcbaRwQsAqzl/Afgb1FZNskblqV9SLg56qaWHfpfvCbMecPZ2VNOz6WJzGnWLUSn2nJEJECprvm\n06pamW08piA9a/5u2jRPQ4w1ugD4vqre6TxK02zWMjPpUaJxFOsq6z+zlQEgIquAraq63bndtng0\noZXwu27mDK/fu6V+Q5vL6ULSb9g9dXy+jLkfiDlp6TAROVlVLxeRO0XkIkymfR5zuMAFmFN8vqCq\n1SZuWuVt2NawiCwBLsJso7lJVSMR+aIrawbxORlzgtUyYO0EYR0Tn+nKwByucCDwTruN54/bHQ+H\nbwEXi8jRmAMq2sG7gJcDi0Rkf+DPMafx5YGPYw6LmZVMGXvm+MXAShG5BGMYT7XO2hGvE7GHVIjI\nl9sVDzFHFJ8PHCEiq4HvtBD+Zm52KV6/571+Q/t1fCHpN+yGOu7XuXs8Ho/Hs8DwS+E8Ho/H41lg\n+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3NuMiLxFRG4SkYtE5EoR2U9EVorIe5q4XSMi\nT+9AGI4XkatE5NnOvVkvhN0ViMgNLbgR5/cBIvJ+EdlHRD45xXvPFpFT2hFOz58GXr/bi9fvzjEv\nlsItQC5V1ZvsMocLMctyQhE5BLMk6BHgauBFwAoRuQazg9cRmHWTp2K2JOzHLHO5X1WvEpFPYPJs\nHfBZzBKKEpADTlPVWET6MNtf3geUROQOzBKKq6wybAN+raqXicjdmB2e9sIs53i+DfvObSxF5AzM\nGt5+zBaSn7XxWQ18GjgYeA3QBayxcXop8BCQqOr5jl+vAo7HrEv9GPAyK3PA+nuMG2fgAeBQETnT\nPl8LCLBdVc8UkXuBr2CW8fzOijkFszylG3i+iJxv43cS8F3MLl/7A5eo6i+tUb508uz0eMbg9dvr\n97zHf7l3EFXdiNlrOmUFpuB/XVXvBX4GfF5Vf2af14AQYwQArlfVDwCvFJFDgaqqvkdVL8TswLUP\nsAXos++hqjuA2zBrbR8F7lPVC4A3AFeq6qmY9acAT6rqahummzHKfUwaWGusXoLZS7kGHAKcDlwH\n3KWqvwXepKonYYxV2kr+gaqeC2S/Wt6B2d3r31T1CeBfrN9PYvbRHhNnmy73qep5mN2hum1Y9xOR\nIvCoql6gqr9zZBxg/QZ4QFXPAH4DPA9jJC8CPkRjfeiImB2uPJ5p4fXb6/d8xn+5dxDbsq+m13ar\nxIeAE0XkmTR2VAL4e1X9WxE5B+ix99IDAcT+c90HwB2q+hkmZ6vjR3ZTg3Rnq1H7u4r5UnBl3Kuq\na5w4HQRsx2wc4aJWRjbcWTcuwxm/1zR5N30nAL6nqt913Kdxc0ma3HP9yDHWIG/GfFk8kX3J45kM\nr99ev+czvnLvDKeIyCswXVBr0psichSmi2sp8CNMi/Z9InIl8JiIvB/Twr0966Gq3isi3bbrroJp\nuV9qrxcD71LVkUnC9HXgQrurU0unR1mZiYh8CtOq/gjwYeBY4AwReS7wVRG5DGOwzgVeMImXXwAu\nF5HNmP24vyoilwPDwPcmeKcmIuda9xeLOaGpqKrvnsR9ajgOFJGPAE/B7D1dxex1fYCNC5gvonWT\nJoTHMxav383x+j2P8DvUeRYUIvIGYLOq/rjJsxtU9fXOdR/mpLI/2Uk3Hs/uhNfv1vGVu2dBISIB\n5vSsH7Xg9kAgVtUHOx8yj8czW7x+t46v3D0ej8fjWWD42fIej8fj8SwwfOXu8Xg8Hs8Cw1fuHo/H\n4/EsMHzl7vF4PB7PAsNX7h6Px+PxLDB85e7xeDwezwLDV+4ej8fj8SwwfOXu8Xg8Hs8Cw1fuHo/H\n4/EsMHzl7vF4PB7PAsNX7h6Px+PxLDB85e7xeDwezwLDV+4ej8fj8SwwfOXu8Xg8Hs8CY15U7iJy\neSf8jcLyHlFYPqgTfns8ntbooH4HUVj+80747fHs7rStcheR/UTkiyJyg70+QUSuEJGrRaR3itdX\ntiscGQ4Hjp7KURSWD++QfI9nQTBP9bsEvGoqR1FYPjwKy9KhMHg885K2Ve6q+pCqnujceq2qvh24\nDnhdu+RMN1hAK0p9bKcD4vHszsxj/W6FY5knvZQez1zRyQKfKt46YO/sQ9vyv0lEbqJzLftWK3fP\nHBOFZYnC8rIp3HRFYbmvBb/6o7BcbF/oPC0wH/R7OvjKfY5pQb+ntAHWXSEKywPtC9mfBnNR4FcB\nf8zeVNVrVPXVqvpq4NHZCLCZ35251wPkZuNvxr8+t2svCss9UVjOT/FOfxSW+9sVhgVGAXhX9mYm\nvf4C+KcW/DoJeOFkDmxe7HYNPWsAp2zg7ELmQr8lq0dW96YaDnCZ1BZEYbkUheWS67+1IZO90xuF\n5UW7Y7maCVkb2ALN9Nv1Y0kzN014LvDPU4Sta3dt4HeqjmjnmPsyEbkUOEJEVgPfEZFLgDcA35qp\nv67CTcJLMQbe5WTgSEBa9IMoLOeisFyY4PHpQNm5fjvwvDSME8h4r/2XlTNpeLLPM0Zn3LtWvkzm\npsk7xSgst7Vx12o6WyYyEu/NuBmTJ2m4s3EGCm5jy1YIbnjeC+zhPM/6O2kaN6NJPk3oZyt+pe4z\n7+2NKXu7lF2s3wOM16MjgXdYPwpRWJ6qIR9MIe8Y4NUZ/09O32lS3gD+FXgPsKd7swX9HqN70y17\nLboJOlDZnQ7s3wY/xvT0ZOIf2PTJO/qsGDuQ1bcu5/I44BWT+Lszzaej39k8n4mdcPwqpDYiU17H\n1RHtoJ1j7htV9RRVfZqqfhTTWn86MIJpeU0bayhXN7mfVeQVwLJMZVXCxE+A1VFYXuS+n/XDZuJf\nAcc5mSCZr/MxlYn9hw3j6gnC1kzmuDhlWG1buGPc2+tmct4NPMOJ/+r0uZUbZIxJjkbjp52snqxl\nb9MzDUcaPnHC2Szt9gD+0bl+B/BsTDrs6/h3GPB0axQCYC/Gp3MhlYXpFTg+E/YuG540jXeGd4Kw\nZeP7cuANVpFXuPKz8bN+u36utv712b9L7P2scREbxzn9WuyQfucxcQ0y94PMPbeMpM8KQFp5vRF4\n6RRlqZDmiWtcncZYF+DqemDvgcmbfwMOyvi7s/K04UrfXz1Fxfpm4MWpjSGj31FYXtakrLgVwuq0\nYnPslOu+gPnweLvj78642rC6aTWuLNn7boWY+p82rid8L3O9833nWbbecfX0+cDbgL8GjnHeKdh4\nu/580HmvAJTStIjC8tKMvycC6cqKMXo5STr8C2ZSdup2OWPLj+uPRJkPwyb6/SbgRRh7dhQTMIkt\nnBadHnPfgVGQcd12LXLGBPfPsgmdkibEy5uEIc2w97jv238uL8YYqVVW7lmYrt4zo7C8uEkYFMhW\nmmLD1szwNpOZFopm7k9v4tdZE/zuAV7L2ImBZzl/j2VsYToLWI41Xm2uKI6Y5NkzsQaHRtnbPwrL\nTwHOpEn6WNzxtn4ajapjMUZAgT5Mg+5M4G+BuIk/b4/C8sFWTg8N453yJsZ23z4XY4iZJGyuEe/B\n9O6sZrzyvjfjx76YCiNLOpb97sx1ylGYOP7FBOGZK9qh32fav9ny91rgL5u4Px2T39kVMN3AS4AD\n7PVZGN1x7dsJNLp2z7DPD6BhY/bN+Jnqdxq2EiavXZ1MHPfHYmxFWjZ3Nl6b6FcXJh/fAzzNusvR\nKB95Kycti6+lYZNSPmgbEGcBhwCnWX/2sW5fRaOn6u8iM8RwhrVlx2PsHTZNDmU8ZwEfAt4XmW7j\nVPYbMI2ps8h8HGTigOM+9e8d9vdk9qYLM9xSAg7G6JKrA2PsqNMIVuAZmHJzFg0bkbIc58ve5ku3\n49dpjO+V6LH/UlI/z2C87h7J+PoqmxZLMHnbj6nkx2EbJWcxtr6aEZOOGc+Sn6rq7SKyJ/Apxn59\nISInYBQOWpxwE4XlFwCvtJcnAx9JvbN/X2DdXGavx7Seo7CcD6NK3ble4zweaPLOKvv3NPv3rVFY\nBvgMJoNewtgK4pz0bxSWfzFFXM4Ko8q59h2NwvLHMV9C+2Wcpl3/a5x319i/ZwOfdtweFoXlw5rE\n7zDgF5n4ArwsMl1Lfx6F5f8EXhlGlaybyeKwBjg3jCpxFJafZm8vduR8DPjABO+lrKJhaNLnrwae\n49xaat+5BJM/ab70MvarMe0mW0Lja28P4FTHTfq1fiimpb8GSMvESus/UVj+G+v3iBvuNH2isPx6\nR+ao/f1MR84hmXd6msQ99fcE5zLb4HaHW9x3XxyF5UVhVJlxl/gsaad+B0AcmUlTrtH8Ueqd/duN\nyZMHgG2Ou7QSy85NODVqTNjag/G8Csak6yH29400KoJzHPfpqoBzrB1ww5/qXToP4APAGkyjbjAK\nyz8HfmPvP+GEN51T4lYE77R//ykKyw9gGlHYsLr69CH79+/ts38iU0HZ+KynUUkdDzwF2OzEu9d1\nn7EB3YzvNk4bQr2RWUY8ZrVRppwucq5X2L+vznycpe/cgfmg2oaxhSl/xSRk5KWV5vH22eswlX7d\nXqdlcRXwVue9fuAfo7D84TCqJJHZHyXH2DryZOd3ryP7I5nrbPjSuPTZcLhDd+7HoftuXxSWzw6j\nyoebx3pqRLXV1SQzFCBSBK5R1ddP4uYmO/FmDE0SqoDJBDAFvorJgHQy3Qgm8Zrdc1mGybQEU/Db\nRQCkX/lDOBVDC0RAaH930wj/Zlpf8jMb7gSehVGstKGzNowq42RHYflMTIE+G/ghE681nmq1wnJM\nXsXAhsyzbvscTBrsoDl5J7xpmXB5DNNF3ypDGEPo5kfCxL1cnwujyvpmSj0BaXqk6Xo/5gtlKmKa\nTAqbTmOsE7RZvwXTMAOjO0P2XgQcCNQwDSmhUSFVGZ/nAzT0ZwPNe3FmymJMWajSKJOjNBqWEzFk\n31nMWDu2HROvTuPqNYzVy6yOBzQaA6OY8HXRaKRtY7y+NmOZ9UeBxzH5kMrK02ikD9Jo8GQJrFyx\n7ja1IHcyHsT0ltxEY47FZDbitjCq3BaF5fcz9it+IrL6fQmNxtpkVMl8jMLs9LtjlbuIvA5j9BcD\nl6jqbZO4bar8BxUKTwRIrkekD5C8UFDbFaZKfUh1e7dIf0FMS6iqjGxOkicXB8HyLpEeAAGJ0bqA\njipDimqfBItjqAWQF0AhSSCuo9UCUizITkVVWztpXakqaIxWB1W3FpHu3kAGABIlVqgXhK4cUlC0\nnkMKMdQFGFHdHqP1bpH+AAmw9+potYj05MRUHHWlCmhfIEsSJQmEIIBAQcWUGRGQqupgDLUcFOy7\nMqy6va47vyDJQVAQ6bJpsJMeES2IJHnQvAg50IMKhZHlQVBfGgTDLyh1fYtGwfx4GFWG0nedVjDA\nMMaA3gUc8Yd67S8erNf3HVEN/mt0tLo5SXYEQkmglEA+VhbHqARITiAnIApaR2s5pCCQU9AErcdK\nXBAp1dGaqgnLkOpWtVERkw45gC6RPvtuEgi5muqoArFSN/4Rx2icQ/J5oagmbfIY+dh8TXRnfqsO\nqm6Plbgk0i0Q5IVCAekGsP7V0wRKw6egOaFQRLoVtIZSMwHWIZP/tQEJluSEnDTkxqbsaT1AckWT\nXmleI6a81dSUw3hEdUcOyXUJ3YKQoPX/GR1t9kXacdqh3wcUCk8C0iPSF0AuQHIBSAKxQLBDk81F\nka4ukR6bzro+iaMeCfr6A1mikASQi5UaglRVhxOo94osSv2w+hMraE0ZzUGuKCaPwKQxQIzWE2Nb\nku2JbslBri+QxdY+aKzUc0K+JNIVK/VAyCdKXUCq6HBNGS0JPQXT2KGqOjSiOlQUKdnyRmw+MpIu\nE99AgTyS14a+EYDU0WpdqQkgYnS+poyOqg4LO/VZCiKlwNqERlKbC4HA/BPUmDECRMNcrloSoV9k\naVGEogh9IhtjqA+pbtqUJJUlQXBgj8jesWryZJIUdmgSx0p+SHV4VDWpo6VNSbLJBM/I7RbpE2Pb\ndupoDvJVdCRAgpytuKxujOSRotX/ahHp2q7Jzkrb6ncASAEp5oRiAvUC0lVDR7H5YfVWY7QeQFBA\nuhTIC0VrZ6WGjibWFmDt+ajqUA0dLSLdOSEvEBTF6DeK1huNrZ35oigBEpREehRTa6b5NmryerBX\ngkVFMY3KGGp1pYa1cWLi0u3kFam7GI0BHVYdBLQk0pO6+8XoaLMh4ZboZLf8D4G/w7TywincNmXN\nwOIaUB9U/c2I6hZF42HVzQFSXJELjsQY6C1Pxsn/5oRCmMu9DtMalEfj+MZh1S1FpE+EfBeytCSy\nVFFGlHu2JskjQK4gdGEK5x5pJbE1SSoJJDaBc3mRYhHpA3JLguDpAv0JJFuS5Od1dLhgZOQU2JIk\n6xIlzgtFgUKXyECPyFMUqMMfB5Pk8aLIQK9IiLm3fUR1QwD5vJFBFX1wkwkfeSjagT21FVOhLwj2\nBIhV61UYKkFfbxDsvbOCMj/qVfSxWLUGpiAqMKiaq6tKDaSuGmxXzX9raGiRgtxVqz7tzwqF/Usi\nyeu7e9cd1dV1xNa9QhYFwV3A9diKfUT1uYNJsurxJO791PZt59QUvWV0RPbJ5bb0BUH90Hyh/6BC\nYXFgGk3bAxgtiNyfqG4BkioMiwl/1TRcIBCjjN22IQfEw6rbAghWBLlDRQhQ1NgqNEFrQJxAvDlJ\nHsYYzf4A8nmRfA66BLQoMmANRTKiukmhnkC1rlrNiRRLSI9AgLFM0iWyoktkLzEGZLiKPimgg4k+\nGqPDRWOYC+LYUTFGVQAGk+SPdagWRZIiol3CsoEgOFSBmurjT8TJ/xOh0COyp4LkoU8gr6aBtqFm\nGmiixt9CUehXky5hgHSZ8quP1WBosQTPnoletYlZ6/f5ixZXAYmVdds0uV+BYZNHukeQe05gv5S2\navJ/w4luKudzx+aQMsDmJPnppiR+qCDSnYdSXqSn2+pUrDy4IYnvVaBod8/rElkS2CE5XfxHAAAN\nbUlEQVS0IdUn6qojtrEoAeRKIgMKMhAEBxWQRYomO1R/Pay6Pg/dOaRk9Cd5vKo6FJgyVsohpf5A\nnioQJFDfmiQP5ZDSokAOwFTg24dVnwQ0j/Qbd1rbkCT3J6bsB6khtx8SQZ/IniBiK5E4gEJvEOyN\nadBq2qCsoZvqyhBGt1VR7MePKiR10+jBln8Gk6T4RJJ0V1Vz65LkoR2ajFSV3prqkkBgWRDsm0f2\n3xDHbEiSB0ZVBxcFQXV5EIwUROJFIiNFiPfN5w/uFgls5ZaYSpqHB5PkibQSBajDaKy6s9dUEO0W\nWZQXbFrqelV0eS44KA/dqf2y4Y8TqAno1iRZV4PRInTnRboD00DrwyRIKbDd3KPKlhgdTaA2oro1\ngFyPBMsFAhHT2ClAb48EZaCoUB/WZJ2CVpXNNXQwbyr8UqYS3qnnw6obq6qDAkEgEhShZ3EQHCIm\nr594LI5/rhD3SbCnCEEeunKYj6wqur2mus3xVEoiS+zfZQWRJQB11Y1VZdOiQJ7BLOjkl/sbgS2q\neqOIXKeqx03idqpuu4eBfSYRl3Zp7Gw5MruuuLRrdjJ+QPOxoLWMHaebikEmXq+7Efg245f5tZsR\noOvuavVpD9RrK6M4XnTFju3PjEF2qOYOzheqaQtdUdbFcSGBegD6wmLpN88vle5bGgQjb+jpPYOx\n41ITsZ7mY6BZ3K7xqWjWdX4PzmzXDnAxZkbtTPg2ZqLUbOgKo8oHp3bWftqs363wS8wqiXZwC1NP\nSNxCY4jNZbr6PSlhVFnTJB3+SJONgXYBF2BmpN8bRpVvQMt59jANe92Ocp7Sqt1I2UBjaG8iaoyf\nfJeSLXMPM3k95PIAjQmeM2VW+t3JL/e9McYVZlfR/gQzEWGfVBHs30Mwaxt/gpmo8R84k3GsmzMx\ncfzfMKp8DyAKyyHw5jCqpJPxiMwSmXdixj8fw0y26gG+SWOyyI4wqnwyCssHYiYKjWAnjLnhsv4B\nXBhGla32Oo+ZGXwLpvFxlONWgHOsHydjx7XcsZYoLD8CPNVe/gDTFb7avnM8Zsz2K5hZrNfZdAEz\njvUTzIzV88KoUo/C8r7YGeCOYbkbKBy9/on0+vbT+gfuAD7021rtqevi+goxPRkJkAxIMPjcUun3\njB1T/0kYVR51Jhp9CXi9zbvHwqhymfX7lzYO6czSn9KYULcBuDmMKr9x4v5OzLKrdBb+7+0/t2F1\nJ7AijCpfiMLy6cA3w6jyBzsLP63crwmjyu+sn2sw4/h30piV/S1Mo+41wCLH7z8A94VR5X+de6kf\nVYzBSWisdb44jCobrBuhURHcGEaVX2b8SI3ej4EXhlHlY47fOGXkBMyY848xZfRfgOvcdNoFtEu/\nr8PomABXYBqHlwFPYtLuNuBlYVS5MQrLrqH9FGbZ1AuAWhhVzk8fNJkUlqbp1zFDCekYfTr+uh7Y\nw+pDgJlLsiWMKp927E1Wv28No8rtGf83hVHlM1n5jh/PpTHT/6owqjxif/87phcEAFuO03cOAv7B\niYrbuPghZgJuF3B3Wr4yFfCXMLbq7WFUOTcKy6cB3wujygOZsLvlLQDODqPKCGZS4Dgc27EOWOU2\nUsKo8iUnDBHwf5gJcrcBe4VR5euO7D/H5IE7ETA7EfdXwDPDqPK5KCy/HOgPo8q3M3G9L4wq19t7\nqd59mcZkwAfDqPIV6/6/aSyLux04JIwqV7jxi8LyccBW4GbM5NiVmF6qn4ZR5eZs2rnp5zw7CaMn\nDwAHOOn7F5jJ2J8Jo8qmKCwfipkUuSGMKhdbP3/rptNM6PSX+2ZVvUlErlXV4zPP3dm03arabOmL\nx+OZh3j99njmN52s3Hsx3ZYjwM9U9Wsz8ONmzMStTrKSWW6P6WV4GbtIBuyiitPrt5cxj2XMlZy5\nkDFj/e5Yt7yamX9vndLh5Aw3G6trJxONB3oZXsZ8l5HK6bSMZnj99jLmq4y5kjNXMmb6rj8pyePx\neDyeBcZ8r9yv8TK8DC9jXsjpBAslH7yM+SVjruTMaxkd36HO4/F4PB7P3DLfv9w9Ho/H4/FME1+5\nezwej8ezwOjkJjYtIyJHYjYbqKjq6fbeezGnDxWAUzAb+38Cs2HGVap6a9aNtjDGICLHYDZAWQyo\nqv6DiPwCs6HJI6r6URFZmZU1zfi8BbMpwTrgc6p6TyvxmaaMkzDHDC4FzlPVu9sdD0dWL/B5zOYf\nt81k2VMTP4/BbOgxAHwRc8ThI8AOVT29HTJF5GXAucC9wLWYpStHYQ75SA9zmK2MF2NORMtjNj/a\n2q54iMh+mM1+Fqnq6+3a8UnDn3VjZ7XvUrx+z2/9tvLaquMLRb+tnN1Tx1V1XvzDbOv3Sfu7CHzN\n/j4Vs4PZWZjdxgLMJINxbqYp7zTgVfb3rcCVwJvt9RhZM4jLm4AbMDturWglPrNItyOAszoRD0fG\nG4HX2N/XtTnfl2CU/5vA5cD72iUTeCnwfcwuXfsD37D3j7b+ty1ewDGY86o7EY8b7N8pw5910868\nmmX6eP2eWbp1XL/bVU4n8HdB6Lf1Y7fS8Tn/cheRw4GPZm6/LXO9jMZRrOswW/jtjWn5J2LOImrm\npiVZag5yeAXmXHaAv7T+XiciNzaRNe34qOrVIvIMzDaKn2whPtOWgTn+8F8x22XOKh5T0K6tRptx\nJvA54G4bzk/Z+LZDZvbM8fTLbx2N0+3aFa8TgBOBKzoQj5RWwt/MzZzh9Xu31G/onI4vFP2G3UzH\n57xyV9V7aJyjuxMR2ce53Ehjw/9VwK+xhymIyLZJ3LQq60XAz1XVHh+r9uA1NmOOMc3KmnZ8LE9i\nzm1uJT7TkiEiBUx3zadVtTLbeExBepDF3bRpnoYYa3QB8H1VvdN5lKbZrGVm0qOEPS4Ykwd/tL9n\nHS8RWQVsVdXtzu22xaMJrYTfdTNneP3eLfUb2lxOF5J+w+6p4/NlzP1AzEb/h4nIyap6uYjcKSIX\nYTLt85hDHi7AnIn8BVWtNnHTKm/DtobFHLN3EWYbzU2qGonIF11ZM4jPycCzMF8faycI65j4TFcG\n5nCFA4F32m08f9zueDh8C7hYRI4GbpyFPy7vAl4OLBKR/TEHOQxhyuTHMYdFzEqmjD1z/GJgpYhc\ngjGMp1pn7YjXicBVVuaX2xUPEVkGnA8cISKrge+0EP5mbnYpXr/nvX5D+3V8Iek37IY67te5ezwe\nj8ezwPBL4Twej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H41lg+Mrd4/F4PJ4Fhq/c24yIvEVEbhKRi0Tk\nShHZT0RWish7mrhdIyJP70AYjheRq0Tk2c69WS+E3RWIyA0tuBHn9wEi8n4R2UdEPjnFe88WkVPa\nEU7PnwZev9uL1+/OMS+Wwi1ALlXVm+wyhwsxy3JCETkEsyToEeBq4EXAChG5BrOD1xGYdZOnYrYk\n7Mcsc7lfVa8SkU9g8mwd8FnMEooSkANOU9VYRPow21/eB5RE5A7MEoqrrDJsA36tqpeJyN2YHZ72\nwizneL4N+85tLEXkDMwa3n7MFpKftfFZDXwaOBh4DdAFrLFxeinwEJCo6vmOX68CjsesS/0Y8DIr\nc8D6e4wbZ+AB4FAROdM+XwsIsF1VzxSRe4GvYJbx/M6KOeX/t3c/ITaFYRzHv88oNbKQhZSNxIiN\nlchCWbCU8mdpQ1FI5E/MwmSQlYnEkKymRFZKKRTFUlZDiAVFpjREDTPxs3jekzO3cRtpaly/T03N\nufc973nPnXneP6d7nkPentIOLI+I4+X8tgI3yCxf84Hzkh6VTrm3+Z/TbBTHt+N70vPKfQJJ+kDm\nmq7MIv/xr0jqBx4A5yQ9KO+PAHPITgDgmqSDwJqIWAwMS9ojqYfMwDUX+AhML/sh6Qtwj7zX9i3w\nRNJJYCNwWdIO8v5TgAFJh0qb7pLBva5qbOmsVpK5lEeARcA+4CrwWNIzYLOkrWRnVc2Sb0nqBhpX\nLdvI7F57Jb0Hdpa6B8g82qPOuXwuTyQdI7NDtZe2zouIqcBbSSclPa8dY0GpG+CFpE7gKbCM7CRP\nA4f5dX/o18gMV2Z/xPHt+J7MvHKfQGVmP1xtl1SJr4AtEbGEXxmVADZJWhsRR4Bp5bXqgQBRfurl\n24CHks7Q3KdaHY1JDarMVt/K78PkSqF+jH5JXbVzWgh8JhNH1Kkco7HdjWXqhhrq7hpj32qfNuCm\npBu18tW51f0Y47V6HVMY3SEPkiuL9407mTXj+HZ8T2Ye3CfG9ohYTV6C6qpejIhV5CWumcBtcka7\nPyIuA+8i4gA5w73fWKGk/ohoL5fu3pAz996yPQPYJelrkzZdAXpKVqdxPT2qHPNHRJwiZ9UngKPA\neqAzIpYCfRFxgeywuoEVTaq8BFyMiEEyH3dfRFwEhoCbv9lnJCK6S/mzkU9omippd5PyVcfREREn\ngNlk7ulhMtf1gnIukCui100/CLPRHN9jc3xPIs5QZy0lIjYCg5LujPHedUkbatvTySeV/bdfujH7\nlzi+x8+Du7WUiGgjn551exxlO4Dvkl5OfMvM7G85vsfPg7uZmVmL8bflzczMWowHdzMzsxbjwd3M\nzKzFeHA3MzNrMR7czczMWsxP0fz66rnw5RIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f84c5cb8358>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "size = 25\n",
    "\n",
    "listt = ['1mer_DNM_context_golden_synonymous', '1mer_DNM_context_golden_non_synonymous',\n",
    "        'CpGmer_DNM_context_golden_synonymous', 'CpGmer_DNM_context_golden_non_synonymous',\n",
    "        '3mer_DNM_context_golden_synonymous', '3mer_DNM_context_golden_non_synonymous',\n",
    "        '5mer_DNM_context_golden_synonymous', '5mer_DNM_context_golden_non_synonymous',\n",
    "        '7mer_DNM_context_golden_synonymous', '7mer_DNM_context_golden_non_synonymous']\n",
    "\n",
    "NROW=5\n",
    "NCOL=2\n",
    "\n",
    "axs = []\n",
    "\n",
    "fig = plt.figure(figsize=(7,4))\n",
    "\n",
    "for row in range(0, NROW):\n",
    "    for col in range(0, NCOL):\n",
    "        axs.append(plt.subplot2grid((NROW, NCOL), (row, col)))\n",
    "\n",
    "cnt=0\n",
    "for ele in listt:\n",
    "    plot_HD_BINNED(ele, axs[cnt], cnt, size)\n",
    "    cnt+=1\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.savefig(path.join('figures', 'internal_exon_centered_DNM_context_binned_selection.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exons centered analysis, all DNMs\n",
    "\n",
    "Mutation probabilities were computed genome-wide using DNM. They were used to compute the expected number of mutations based on a generalized k-mer model, with the method described in the **Exon-centered and intron-centered mutational analyses** section at Frigola et al. for the 3-mer case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get mutations\n",
    "mutations_file = mutations_folder + 'germinal_ultimate_dataset.bed.gz'\n",
    "mutations_df = pd.read_csv(mutations_file, sep=\"\\t\", header=None)\n",
    "mutations_df.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'class']\n",
    "mutations_df = mutations_df[['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type']]\n",
    "mutations_df = mutations_df[mutations_df['type'] == 'subs']\n",
    "mutations_df['ID'] = mutations_df.index + 1\n",
    "\n",
    "tumor_type = 'germinal'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 1170\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1170</td>\n",
       "      <td>1035.458</td>\n",
       "      <td>46111</td>\n",
       "      <td>46245.542</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          1170      1035.458           46111       46245.542   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    12.993477\n",
       "dtype: float64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    12.880646\n",
       " dtype: float64, 0    13.106309\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 3488\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3488</td>\n",
       "      <td>2717.856</td>\n",
       "      <td>46111</td>\n",
       "      <td>46881.144</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          3488      2717.856           46111       46881.144   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    28.336453\n",
       "dtype: float64"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    28.263691\n",
       " dtype: float64, 0    28.409216\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 1170\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_new.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1170</td>\n",
       "      <td>1195.464</td>\n",
       "      <td>46111</td>\n",
       "      <td>46085.536</td>\n",
       "      <td>0.786</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          1170      1195.464           46111       46085.536   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.786  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.130052\n",
       "dtype: float64"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -2.214824\n",
       " dtype: float64, 0   -2.045279\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 3488\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_new.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3488</td>\n",
       "      <td>3175.426</td>\n",
       "      <td>46111</td>\n",
       "      <td>46423.574</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          3488      3175.426           46111       46423.574   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    9.84353\n",
       "dtype: float64"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    9.784135\n",
       " dtype: float64, 0    9.902925\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 1170\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1170</td>\n",
       "      <td>1181.696</td>\n",
       "      <td>46111</td>\n",
       "      <td>46099.304</td>\n",
       "      <td>0.651</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          1170      1181.696           46111       46099.304   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.651  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -0.989764\n",
       "dtype: float64"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -1.075904\n",
       " dtype: float64, 0   -0.903624\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 3488\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3488</td>\n",
       "      <td>3159.642</td>\n",
       "      <td>46111</td>\n",
       "      <td>46439.358</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          3488      3159.642           46111       46439.358   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    10.392253\n",
       "dtype: float64"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    10.335415\n",
       " dtype: float64, 0    10.449092\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 1170\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1170</td>\n",
       "      <td>1177.725</td>\n",
       "      <td>46111</td>\n",
       "      <td>46103.275</td>\n",
       "      <td>0.592</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          1170      1177.725           46111       46103.275   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.592  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -0.655926\n",
       "dtype: float64"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -0.747989\n",
       " dtype: float64, 0   -0.563862\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 3488\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_new.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3488</td>\n",
       "      <td>3178.357</td>\n",
       "      <td>46111</td>\n",
       "      <td>46420.643</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          3488      3178.357           46111       46420.643   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    9.742235\n",
       "dtype: float64"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    9.68534\n",
       " dtype: float64, 0    9.799129\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 1170\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1170</td>\n",
       "      <td>1175.267</td>\n",
       "      <td>46111</td>\n",
       "      <td>46105.733</td>\n",
       "      <td>0.561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          1170      1175.267           46111       46105.733   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.561  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -0.448153\n",
       "dtype: float64"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -0.535776\n",
       " dtype: float64, 0   -0.360531\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 50780\n",
      "The total amount of mutations at the exonic sequences are 4757\n",
      "The category initiator_codon_variant is found 6 times.\n",
      "The category intron_variant is found 1 times.\n",
      "The category missense_variant is found 3280 times.\n",
      "The category None is found 10 times.\n",
      "The category downstream_gene_variant is found 2 times.\n",
      "The category upstream_gene_variant is found 8 times.\n",
      "The category splice_region_variant is found 58 times.\n",
      "The category splice_donor_variant is found 3 times.\n",
      "The category 5 is found 2 times.\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category stop_gained is found 202 times.\n",
      "The category synonymous_variant is found 1184 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 50780\n",
      "Observed at middle exons: 3488\n",
      "Observed at flanking introns: 46111\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3488</td>\n",
       "      <td>3211.136</td>\n",
       "      <td>46111</td>\n",
       "      <td>46387.864</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0          3488      3211.136           46111       46387.864   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    8.621995\n",
       "dtype: float64"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    8.565944\n",
       " dtype: float64, 0    8.678047\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot results\n",
    "\n",
    "Binned average approach. Bin size is 25."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/ipykernel_launcher.py:20: MatplotlibDeprecationWarning: The set_axis_bgcolor function was deprecated in version 2.0. Use set_facecolor instead.\n",
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/matplotlib/font_manager.py:1297: UserWarning: findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfcAAAEqCAYAAAAI1RnWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXecZEW1+L+ne+LO5gRs7V1yRhBRBCNBRQmKPkVFMaJP\nfwrqMz0VZAlPRAUMCBJERVSCYiI9JD4kKTlHCZdaJGzendx9fn+cuvTd3pnpCT2zs7P1/Xz60933\n1q1wK5xTp5KoKpFIJBKJRCYOhXUdgUgkEolEIvUlCvdIJBKJRCYYUbhHIpFIJDLBiMI9EolEIpEJ\nRhTukUgkEolMMKJwj0QikUhkghGFeyQSiUQiE4wBhbuINIbvLUVk+thEKRKJRCKRyEio1XM/QUQc\ncAJw2hjEJxKJRCKRyAipJdynA+8CTgQWjX50IpFIJBKJjJRawv16YIaq3gs8NvrRiUQikUgkMlIa\naty/DlAR+QjQNQbxiUQikUgkMkJq9dx/AKwEng+fSCQSiUQi45xaPffbVfWyMYlJJBKJRCKRuiAD\nHfkqIjcCK4B2QFX1kLGKWCQSiUQikeExoHAHEJEpAKq6ckxiFIlEIpFIZETU6rkfBWwa/qaqetyY\nxCoSiUQikciwqTWhbqaqfkpVPwVMGYsIRSKRSCQSGRm1JtRNF5GPht+zRjsykUgkEolERk4ts3wj\n8Lbw9ypV7RmTWEUikUgkEhk2/Qp3EfkiMA+QcElV9WtjFbFIJBKJRCLDYyDhvgkwE1gVLk1W1QfG\nKmKRSCQSiUSGR78T6lT1OeA/VPVpVX0a+NDYRSsSiUQikchw6XdCnYhcDOwgIjuFS3H72UgkEolE\n1gNqTajbMZriI5FIJBJZv6i1FG5PETkluCuq6l6jH6VIJBKJRCIjodYmNrsAt6jqvsDlYxCfSCQS\niUQiI6SWcF8GNInIh4FXjUF8IpFIJBKJjJBaY+5TgB5gf+BRVb1/rCIWiUQikUhkeNTquX9JVTtV\n9RLg0LGIUCQSiUQikZERl8JFIpFIJDLBiEvhIpFIJBKZYNQS7r8DFJgMTFPVN49VxCKRSCQSiQyP\nAde5q+oHs9/hIJlIJBKJRCLjnAGFu4jsn3MXl8JFIpFIJLIeUGuHujmYWb4LiMe9RiKRSCSyHlBr\nKdxt2LGvmwEfqeWZiLSJyO0icqCIHCoiZ4vIeSLSVuO5swYd40HgXdLoXdJYTz8jkXriXbLeWcLG\nUf2W8Vq/x3PcImOHd8ku3iXFdRmHWsL9ROBW4GrgmkH493XgovD73ar6KeBC4D01nps3CL+HwsHA\ngXX2sy54l2zpXbJJnfx6Qz38mQh4l7R6l9SyRNU7zJG8/3fWLSJjx3ip3wuAb9TZz3qxLVCX+Une\nJbO9S5rq4dcI4iDeJRuvyzisK0b4/t8NTK1nfIZKLeH+D+B+Vb1DVe8YyKGIvBV4EHghXMqm4T8D\nzO/D/aEicqmIXEqNyh8K2Iwacc3TDLR4l2ziXfKKwTzgXTLDu6TZu6TPXoh3yXTvklrvazC8Fdit\nDv4AvAXAu2TautYS+2IY+TYSPgG8frgPe5c475JXepfsNYCbb1ddekvV/bcO8OyUwfTovEu28y45\nxLtk2xCn6fl36F2yu3fJ22v5U29Gs34Pg0ZC29VffV2HtAD1itPngUG1X0MhCK2vDtL5dOAzA/j1\nlv7u1RPvkh28Sz46Qj/avEsmD3D/Td4lzblLnwd2zN3f17tEhhDkWm2yd0kh+DPLu2TXPu43e5e8\ncQhh9Eutns4mwHki0gWoqg60S91eWKHeAejAxunBtOxnqx2r6m+B3wKEBmAgjgHwLjk2+D8NuAWY\n4Xy6JLzwyUArsApreATYDpsIeJ93yULgMufTfwa/DgMudj7tDGF8AbgHmOVdMh841vlUg9uFwc1f\nvUvuzK7XImh9bc6nS3OXBXi1d8nGzqfneJfMcj5dXPXcbGCp82kpV5h2Bd7pfLowuJmbc/t54A/e\nJa8Frnc+fTzck77iGnq4RwFnA4sGm56hEOLVC3zBu+QCLE8eBTYK11/vfPrnnPtdwvXHnU+7vEve\nBdyANSwn5fJiU2B/59MzqoJsDJ9+010Vv49jZegDwPHATsCe4V5WvpYCb3E+vSLkw1rKXS5/tgBe\n711yG1buFgOHZfmF5dG12FDXWn7k4jstPD8daAJmBzcnOZ92YILyld4li4D35PwfbfZilOp3yNMP\nO5/+T/hfdD4tDfBIObhrAb4KLPQu2QF4JHsuKLvlXLk5AvgV8D7gb8AS59NV4d4O2PvudT79x8Cv\noSYS/Ny9P7+yToLzaTl3TYAdnE8fyMUJQLxLZmIdltnAHOfTa2tFwrtkOvBa59P/Df93Aw4K5WUt\nBSSE34C9A/Uu2cH59MFcet4P3A1s7nx6Ze7RNwBXV+dZ7vns/yRgI+fTJ3PpewgoVOd19bOBydjw\n8KAI6W/D6uyycPnL4d7x+XKSC38f4GEqCixUlFiANwLXe5eUg7/bZvH0LtkeSJxPr8q576vDVQj+\nLAHe5V3yphDGT0JZnQ7s611yM7CN8+lDg01zNbWEe7uq1jK5AaCq3wIQkY8BLwFTReQMTOB+brgR\nrGIWVjnBXty+3iXnAx/ux/02YGPw4f8B3iWPAJ3AlsC+wGU5M/lOVDLkGO+Ss7A5BxkHAQd5l5yE\nCdvncoW1EPy7B2uAV2Kn6u3sXXJa8LcBE24A8zOlwbvkYefTC4IyMAf4VLj+HeCbwH1Ad7i2KSaE\ntgv+bB2+27Ae1G5BUP4JONq75Bzgo8CvnE+zRvio8P0p4HHgfO+SzZxPn+rnPfaLd8kUoNn59KWg\ncT4Zwvk8NqQDJkD7enYFJkAPDt8zgAdCwd4V05qbsLw4E6sEc4CNQuPtgDS4mQ5MD++0y7vkIuCp\nXCV2wJuA3zuf9gCbYo0LwZ89c1HbNXxOBl7rXfJciGOmGG3kfOqD2wJwNNarBfivPtL5BqxxLod4\nzw/XFwLnAu/wLnkMKxuPBT/nseaukB8HTqeiYAyqXtaR8zEFeAYWRxWRB7F3v+dADw6Cj2c/vEve\nB+wY6th+zqd/Cu9sU8zitRRTLsDeacYh4Xth+D4a+Dfws/B/FvBmIAnhXQ78s+pZvEueAFY6n3bX\ninQQiO91Pr04d3l2+N4/+LW38+nvqx79dnj+R5iSNBUTQO/Dyv/0XJwOBO4ENgaew8rFteH52ZgS\nuSNQcj59KLyrqVhe4V1yOyZIdgv/Xw+8OvzOdyy2AT4Yrp8KHBK+Xxfub48JvR2BK4NwzgT/F4Ep\n3iU/xaxnWwAzvEtOwJTVXmy4Yn/vkhucT68L6fsO8E3vklMwpaI9hHUIlXzM6MUUnVlAO6YUrKZ/\n3gHMxeTP34Enc/eO9i55IOTb0ZhyOCvcq1bgNSj7reH/RuE9NWNlf2EoB7uE8K7KKfw7A38LSqgE\n5TzzP1PsMqvcPtjQd3b/6PBd/R4GTa1NbK4DHgFWAKjqqMyYF5FLVXWtMfLQ+H2Poc/Ub6Yy3rGC\nSi+jmseA64BPV13PBHyZiuaWjb30huurMG1yMdZw5EmBEmtrms9hlUSB5VX3fgB8perak8Dm/cR9\nOByPVZxtq67/HngvcIXz6Vo9y4HwLjkAeA3wQ0yJuxZTNNb1fIDVWDn4A6ZoHT4MPx6mokRl/Bo4\nLPc/U0rAGrvM7NdFUMj6oJmKKXs1JoQyHsLKSF8chwn1bEtoxrDnDoCIFIDzgGZVfZ+IHADMVNVf\nD/BMn/U7I2cZO4W1laMbMKE8EOdiwzJgSu19hMbR+XShd8nu2OFX1dyFWXt2qrp+KzbcsBtwfujJ\ntgKvxBrwv2NnbdyMKfz3BL+6WbstAbOmXR/S+npsaA5MYco6Jj/F6s/VVA33BP83Z80x3BOoKOn/\nwnrjZ2PK/lDPAfk7VkfeMQi3PcBJwBGY4B4s+XryNyrvIONh4AmsbdoK+C5mFerEFLKdqXSMADzw\n68zymrNGnou1v8dQafc7MOW/muOxcvIAFfP7WZh14qZQLi/B6lzezcuE8rU/sDsmC87EOnGZzPon\n9p62Cf+/C/x3H3EBOBazlr9chkZSv2sJ903z/1X16QHcHgwcgBXAn2Ma7N5YQ/ZZVe1Xy+qv8u/Z\n3PxIA9Iwu1jYtIAUJHMO9KCdXaqrCkixGCwQJejuVe2aVJDpRaRBoCAgHarLAO1FewqItInMlKB1\n/rtcerxH6ZxTKMwvCA2NSEuTSKv5pz2ry/pSi8iUZpHJQLlbtWNRqfRIUSg0IS0ArSKTC0JBgS7V\nVQrlKVKYXQxKwirVxb1o1yaF4nZlKBWFooKqUi5D90otv8ia5h8EoQiNzSJtAjQKrQWkCLCiXH6u\nG+1oFZlcRBoByuhqoBekp1N1RYvQPKNQnBfUwJ6V5fJzZejatKFhTgNoi0j3do2NiwR0q4bGh6cV\nCith6IXJu+QgcnMIyqrc2dPtSkrx+XJphxdLpY17QbpUu3pUextFZhZgagm0PVwrQ3m16lLQ4oxC\ncQvB3k+XaoeCKqplRcug3aqdZdButKO9rKuKQnNDyIcGoQUolBRdXC4v6kVfNvdNkcKMyQWZCVBS\nujtVV4cyICLIJJFpAoVQLAjvtFRSegRErdxJCXq7VdsVdLWWl5ehvEmxuGUBaShCYwEKZSgXoKEb\nXRUyVUO4vR2qy1tEpjWLTOpV7W4SaVtSLqcFKGblroyWupWORqFlqhTmYuF2p6XeBwtQnCKF6YC0\nikz9S0f7UBrYESEi7wQ+iyk4B6vqISLyCuBAVT2xyu2hVITMPFVda3VAaDxPBb40jOhk1rgSlV5Q\nRjswKfw+ljCsN0Suw9ovsA7CVCrK9jn0rSw+jQkmsE5Avk7/GrPs5ecf3IsJrVoMpPDlySs5o0HW\nq+ymttV3rDgPy5fh5PEi1p4P8nPgk7n/Q+1grQKmYIpJAVPe78UsL9B3hyHjeEy4v1y2Rk24D8tD\nkRlYL3TqSDX7k2fMXAGwqlz2L5bLjymU1YbTZWahkGAFTLtNcSi0iswQa1z1xVLpsTL0TisU5hWs\n0W0oijQDdKmuXFkuvzS7WNxqisjmQSgvXq76tKrqctXnFdWZheKCIjQp6OJy+RlFy1s0NO4toWHp\nUV1ZhlJJ6epGVxWguVlkarjXsVL1uRaRaa0ic0M8Vz1bKt0LSJPQgiLTC4X5RZGmfC7k86RDdUUJ\nusvQ26na3izSNqNQWKBAWbW3Q3WZQLEoMqWkSqPIVIFGhcLKcvnFdtWVk0RmiUhzSZUXyqXuktLz\n73Jp8vJyubFdtWH7hkbZvbl5xWSRrsPaJu/eIrIUa0zuDCbsPgm9kNnArpd1tB/4aG/vpi+USi2X\ndLRPnlMolBtF2LTY8HxR6G0RaSoiWoByu+pLBSi1ibQ0iNAi0tYs0irA6nL5heWqLzRAsVmkOQx0\niwg0IU2NIs1FKE4uFDYqmvKmXaa8lcrQVYauNpG5bSIbYw2/AipIcVm59Eg3rG4TmWuCHNVwv0fp\naNfyGnMfmkyxEg3uBLTVlJNio8jUJpHpqpRKaMdzpdLtWX4p0CIyuTGUN4JiMFlkk0wZW1IuPd6u\numzjYnGHBkyod6Mryqo9TSJTC8Hd0nLpiXbV5Vs2NL5N0d6iSGuv6uou1SW90Hnc8mWDafTriohc\nBqwOwn3Y9futLa03AbhicbuGkNdUlKieFeXyIoFCdq+M9raXdfnkQmHWjEJhAaBl1Z6lWn5GgG6l\no4SWZhSKmzQEk/3ScjldrbpsisjMJpGWBqGlTQqzABRKS8qlJxuQ5hmFwqZYXvc81dt7BxXBrE0i\nLY1CswCdqqsXlUrPbNXQuFOrKfysLJdf6lRdvaChYZcWkekChV7V9m5YqVBeUi6lCqVMbczSWUCK\nk0VmFwQJymETQLvqiyvK5RebRFpbwjLDHtXOHuhUKPeodhWgMLdYnN8kUihAeVm5/FQBSnOKxc0b\nrT2gADSAbNfY2LhtQ2OpCC2thULW+38SuAoLc0q4tjx8XkNlcuozwCVl1Td2qO4D6K3dXUuf7i09\nDrQ2CNNLCl1o9+pyeXlRZEojzFagDN09pgjLci37bqVzXrG4c8HCLGf1qlRxpwrlMmiPasfycvnF\ngrXrUwEarO43KLBSdXGPahfAolLJL2gobjGjUJwPUFLtXaHlF0IeIyBTpDC7UaRVUXk5B1S1Gzpf\nVvJNdpSWlcvPEzqDqjC7WFzQavlabBaZnpWNleXyopK1MQBaRksvlcp+ksiUjYvFnXpgVSNMebZU\nurMM5RYrLwVFtUN1RRFpmFsobCuCqKK+VLq/B+1tQppFkDYpTP/N6lVZj3/I9CvcRWSaqlabjmt7\nKHIy8Bvgv+uk2ed5DtNs+qLa1HMh8P4BotpBZRwlo4T1tk8mTL4YAv/ATDN5nsE0uGryJqr871qc\nC7yLtYcBhs39Pd1Tz1i1ch9Arurs3PmaORv9dEFDw0vh9mlhLP1LwM+dT1fkn/U2wXEWUHj7i89/\nam6h+PyUgqx4Q1PLfR9saxvsRJC7sPHtjD9iZthuKtoumDnri5j5sT+ep2K6a8AEeDF8C6ZVD5Zu\nKkMx/ZHdLwFPYebDLD8H2ysbLNmEUbA5LQpjZ5YXkaOwuRMtwBXY+9yHMOauqi8O8Gyfwv1rU6et\nEvOoYVGpdI/mlK3phUKSCbsetFOVcovI9KJIY1lVF5fLT6zQ8uI5heKCRlPCCo3QFoRKz7JyOW0T\nmdUsMr1LdWVboTBnVbn8bwVWqy7pUm2fUSgkjcFasrxcXtSu5ZVJseGV3cGq0yQyDVPqGrrQFYIU\nGi0fFJAVWvYFpGGSyOxetLsBaXmqt/e2XrR3ihRmCRQmFwqzm2Byli619/Zyo9uFru5SVoNqj9Il\nUJxZKGwqmCWwU3WZmoIxGSgUoaEcBEqXas8KLS9rQqY2iEwuo3Sqdqwq6/IyUELpVgp393TPLKlK\nGeQ/WifdulGx2LF3S8vur2lqLoa4ZBN+52NWj2bgz8BzV3R0fGaFlks3d3U1/rmjXYHyrEKhtGNj\nU2fRBGC7IL2TRNoaTdHQLtWlZehuEloKSLFJmBQUqnKPaudL5fKTEpQ4gUKTSEsRGsQMliIgU6Ww\niZoFrNADHWUTtOUSdAcrZiFY0Mol6GkR2gQpdmh52SQpzGhYs+6qQnml6vMhCxSUBqSpYO2DZorF\nJJGZjSItAg29aHuXsrJNZPbicvnxkmpvJ7qqV+lqE5neIjIFKra+yYXC3EZoE2CF6rPPl0pPbFQs\nbjlZZA5AL3SWVLsaRVobQjvWrvrSMi37uYXiNl2qK8rQO71QWFBS7VLQU1euGPbKi4FMK98UkSWq\netJgPBIRwRrgK1T1TvsLjGw27enY+NMcrME8GxPuz2HrCMGE9GasObkGrLG9EBt3eSpc+zKVWaJX\nYLO3t8bG3v8bm+BxuPPpSu+S4zCTzeH0rQi8TG4Geyu2dOURbOwoBa4E9sAa+qexiUH/h03om4FV\nphnYngIlzOyZF95XYBPBHnM+fca75EFstiXYGGwWxzux/NyZiinykRD3FBvX3RGbIPOyIrJTY9N3\nfjpj1gqA/V98fru3v/j854oIH21re/ArU6fd711yLTZm1BzSmGBmqxNv7era5chlS95VUi0vLpcL\np8+defEWDY2ZQngRuYlKOZZi41fXYnmZmfSfCO/kOeCX2NjeoVjdaXY+7QyTrASbkJQpejeG9/FH\nTKAS3OyO9T6yXsmx4fo7sbHT44C3hby5Mrybe7Dxy3lUJgFtiymOd2MT7zw2S/0PuXE+Qth/w/J/\nvvPpJd4lfwz3PoaVsfswE/QKrE7MxSwfF2LmOsFmo4PNgQAzW78KU3huCHH9VriXTeobdVT1BOCE\nKsvcDqHn/nbM7DwkvjBl6nnhZwc25jsQF1OZTDtYWqmMdbZjeZDRX52+j4qy/TymEJbCs5tRMU0r\nFSE9AxMSPaw9l2Yo3EftpW8NVIYc2rFyOih+sWrVFn/pbH/FQ709089dvUr2bGq+qyiUP9o2+Z6k\n2NAu0CRQvLW7a9bVnZ3zu1U3vrGrs/eVTU2PF5DSz2bM+t7bW1u3rh3SGvSl5Gbt40AUqcx9qp67\n0kpl+GMSFUH+EKb41oMGzHpZCOG/vHJgiFw/SHePYe0fWPs87FnyGQP13E/GKu0DBG1TVftqrDP3\nR2Kzsv+JNYQrsEa3FfjccMbcM7wt+9o0W8ZWda8FM5NnPa0vY2Nivr/lUN4lrwMeyi9R62/5lHfJ\nhzHB+Rw2A7wI/Bh4bbj+Qj/PTQG6nU+7ctcEOCbf2woCYrLz6bLctdlY4V3ifPpoH36v4Y93yf7O\np5dXuTkMuC3/vHfJvtjyjdOzcELP/M1A449Xrrj9na2TFl7cvvotD/b0JEdNm3ZJq0j7vGLDi5jg\nUuBtHVp+XXtZN72ss2PupR0dcvTUaed+e8Wyk/84e+72WP53OJ92e1szuhM26eg2bGz05vx7CWP2\nM51Pf+VdsnAwPdFg0TkeWwr5knfJt1hzNUDmroApAj92Pl1S9W7Ox4T4vsCVzqf55S+DJmdd+qPz\n6T2hPDZky6z6cL8/JqQVUzyn5eNW5W9+Oeb+zqeXZ3kP/MX59M7hxHkk1Nky903W7GHdgSl792AC\nId+YVk+s/T9M6c1IMcsJmMBbgZW5d+T8vQbL798Ev6snnnWxdicB4NIQp2/1cW8oXMrAm2udhqXh\nXeF/X5PrwGbW9zVBLM+/sFnrYJPEPk2wTPaqylHLl72mQ7Xx9u6uGc+WSq9oFHl5fkq3ammPpuZH\nZhUKK7drbHzhC1OmZoLmB9j725WKUl3NjdjqictCOs7AOiwZy7H3cAA25p0psz/E3v97qKwAquZM\nTNFtZm2FYSVwP5WVG3djZeKgnJssfy9n7QmW2cS5bHLxxVgbnFkVM4tcppicFfzbEyvD+fhkk63z\nZfQZrAzl49MX12IWMWB0J9TtGCKFqtbSrIdNLeE+FLxtzvDDgcaKJwLeJdOcT4fUSwhL7RpyS076\ncrPwrFUr289YtfLEMsiqcll+OGPmM01IJ2a6Kv1w1YrtnujtLQOlg1tb//n96TMvcj79yTDT0Ywt\na+mo6Xh4/k8DVozGWv7gf6b8vbzGerQJimP7WIUHICJbYNaR5ar6JhG5FetpbgH8TlXPGeDZ/lbD\nCGbZ2xWzpnwfUwzLQTGbhymEnw73vkplcts1wE3YEqNsueNuWON5FzDX+fRs75IG59Ne75ItnE//\n5W2joeudT3tCGG+lsnfGEtZc+ppxkvNph3fJtiE+TVSsCMdhisL/w4Tx+8O1rbHe3ybYypHjw74V\nW7DmVt7HhWd/FuI5I/g9DxMgU7ClVxdik7M+EWZo74VZJc4mZxnGVhtcifX+vg1cgAmbw0NY2UZM\nL2Gm98WY0N8CU4CaMcvZVuH627F8fhGzLk3F6usSb5uwZIoIwa+bgb2cT08O73cWtgogC/t2TPAL\nJiSzPSH+nq37D4ptNkR6AWEpbVWnqBjecbbM9hTn0xUhjz4I/BWz2B6BCeYrsbJ2DdYZnBfyR4At\nw14WhVD2tsKW0fZmSzPD+zodE/BHVMVl25zf3w7u9g7/P491PPJLpsnlwx8wK94XgFudT6/MD0eP\npnA/msqYcaqqxw3aY5sIcjrWo75eVX8zgNu6CffIyPC241cHVkG2PmLpknc+1tuTLdwUQFpEen81\nc/bZUwuFo7GG+UFX2QwoMgHJWeamY4L2Y5gQ3Rq4YDjCPY93yT79bc7iXdLmfLo6lM3PYNbEv/Wn\n3IShoxbn08f6ul/ldmNMoD+F9ap3wITbNEwoLHI+Xcv0HRrp5mqltC/rUxb//H/MhN/an4IeGvgz\nnU+fC4L8hryCGix+xSrLYAOwp/PpjeH/JKwuLwA+HpSCNkD7UvD78rMWwb+9gNcE/4tAU/ZegvL+\nKufTW7xL9gTurqXIe5d8AfgLZhVd7V2yH6ZQXNGH288CqfPppblrszDrRgHroT863M1gvEsOxhTP\nu5xP/xwU0tb+OkjeJXuHdfzZ/zXyPnf9q8A5mfXYu2QP4J6gRC4EXnI+PW04cc6oJdxPUdX/Cr+/\nr6qD3bIQETkMWKaqfxWRC1W138ltUbiPT7xLtsRMnM9jJsxM0fsuQBToGx4i8ntVfa+IXDSQWb7q\nmXpa5g4H7nc+vbWm4/UY75I5wGKX28VuBH4VCUNYI4/Z+CJYsXpH0fK3LTDF+fT20fC/nzBbgcbq\nCcxDpdZaxeki8lFsfHCoM7TnY+YcqCwX6I9FQ/Q7MgY4nz6R+3vuOotIZDzT54TZqjH3fiejDhXn\n034tBBMJ59N+Vx8Mw68S9ZtoNq5wPh214eLg/yOj6X8/YXZQ2YVx2NTquTdiM4oFuMqWBQ7SY+u5\nL1XVS0XkAlX9QNX9fOX/bZg9X+3HNdQhkTWYx+grFzGMGMZo0aqq+452ICIyC/gfbIz6HGzlx6Am\nzA7gZ6zfMYz1OZyxCGP49VtVR+WDLTn7BTZb8kPD9OPS0YpfDCOGsb6HMZbhrK9xj2FseGFMpLSM\nJIwBzfIiIhpCGIbSsJrcgRCRSCQSiUTGhlrnk399TGLRP2uZ6mMYMYwYxjoJZzSYKPkQwxhfYYxV\nOOM6jFpj7pdia0mXA6jq6cMNKBKJRCKRyNhQa7b893O/R2UTkEgkEolEIvWllll+PraG9QZq73kc\niUQikUhkHFCr574nkO25vdloRUJEdsf2hE9V9Svh2pexc3QbsV2pNsEsCSXgF6p6XbWbwUz+C+fO\nvx3baUtV9YMi8g9sC9GnVfVEEZlXHdYQ0/Mx7NCUZ4Cfqup9g0nPEMM4HDscZSZwgqreXe905MIa\n9G6DQ/DzYGx/6anYGcpfwpZXrVLVr9QjTBHZC9uD/gFsG8t52LaQzVT2ux5pGG8EPoTVpR2wIay6\npCNs+fotYJraxjGH1op/tRsdxhK1ehPr9/iu3yG8utbxiVK/QzjrZx2vMQ3/h9geuNOAX4/ylP/N\ngB+E303Ab8Lvz2HraY/GrAcFbJLBWm6GGN4Xgf3C7+uwTVo+Gv6vEdYw0vIR7ACCs7GTv2qmZwTv\nbVfg6NGZoTMSAAAgAElEQVRIRy6Mw4CDwu8L65zvM7DK/wdsL+2v1itM7CS1K6jslX1xuH5A8L9u\n6QIOBv5zlNLx+/BdM/7VbuqZVyN8P7F+D++9jXr9rlc57cffCVG/gx/rVR2v1XP/JXYQwc+wU5lG\nTNiusnqryk9U/Z+FHVQAph3PD59UVcvhONm+3AwqLFV9AduM48fh2r7B3wtF5K99hDXk9KjqeSKy\nM7bi4AeDSM+Qw8AOujiSyiEEw05HDYay2+BQOQr4KXB3iOcpIb31CPNGVb1BRDYCTqEyb+QZKqc4\n1Stdh2JH4Z49CunIGEz8+3IzZsT6vV7Wbxi9Oj5R6jesZ3W8lnB/HjvHFuw88BGjqvfRx7GHIrJZ\n7u9i7JxrsO0t78W2uJwvIisGcDPYsN4A3Kaq5eAu2795KbbjVnVYQ05P4AXsVKfBpGdIYYTdA08H\nfqiq6UjTUYNnsQJ8N7XnaQwKsdbou8AVqpo/ujR7ZyMOs+p9NGMnesGaW6aOOF0isgA7LS2/FWbd\n0tEHg4l/n9vC5hGRRlXtEZEtgcWqumwg94Mh1u/1sn5DncvpRKrfsH7W8VrC/VQqM+ZPwU4Kqzsi\nsg12RvWOIvJpVT1LRO4UkR9hmXY68ARWWHqBc1S1uw83g+UTBG1YRGYAPwI6gSWq6kXk5/mwhpGe\nT2PnDs8Cju0nrmukZ6hhACdhRz5+NmzjeXW905HjEuA0ETkAO0qxHhyBnVU9TUS2wuZ3tGNl8nvY\n2ccjClNE3gPsh42/ngbME5EzCFumBmf1SNcnsd0YEZFf1SsduS1fdxWRbwB/GkT8+3LTHyeIyI+B\nE7BewYeHEr/BEuv3uK/fUP86PpHqN6yHdbzWOvdjVfWY8Ps4Vf12v44jkch6hYiciTVMfwc+rKpf\nW8dRikQidaLfnruIXAzMDVqjUD9zQyQSGR9cD2yhqveKSM2zzyORyPrDgD33SCQycRGRjYG9sNne\nqOp56zRCkUikbtQ6OOZkIMn+q+ohox6jSCQyVvwA+B02ZhuJRCYQtSbUlaJAj0QmLLer6mXrOhKR\nSKT+1JpQ93ls84kVAKp63NhEKxKJjDYiciNWt9ux3dyiIh+JTBBq9dzfhC2T6RqDuEQikTFEVd8o\nIlPC75W13EcikfWHWsL9IWyhfj02SYhEIuMIETkK2DT8TqNlLhKZONQS7guA72BL4ZS1t5GMRCLr\nLzNV9VMAIvL9Wo4jkcj6Qy3hvjD3O66Zi0QmFtNF5KPh96x1GpNIJFJXak2oOxET6pOBV6jq3mMV\nsUgkMrqE/cvfFv5epao96zI+kUikfgx6ExsR+ZKqnjrK8YlEImOAiHwRO/86O05M4/azkcjEodYm\nNt/Heu6NwNQxiVEkEhkLLgRmAqvC/8nrMC6RSKTO1DLLbxp+dqlqXY58jUQi4wMR+XY2Q15EvqOq\n31zXcYpEIvVhoINjjiE3iU5E4iY2kcgEIRwMtYOI7BQuPb8u4xOJROpLvz13Edk2/GwBjgSmq+p/\njFXEIpHI6CIiO6rqA+s6HpFIpP7UMssfBhwA/ERVbxqzWEUikVFHRA4H3odZ8Iqqute6jVEkEqkX\n/Z7RLiK3A3sDvwamicj+YxarSCQyFuwC3KKq+wKXr+vIRCKR+jHQbPmfhO/ZYxGRSCQy5iwDmkTk\nw8Cr1nVkIpFI/Rj0OvdIJDKxCIfG9AD7A4+q6v3rOEqRSKRO9GuWj0QiE54vqWqnql4CHLquIxOJ\nROpHrb3lI5HIBCQuhYtEJjbRLB+JbKDEpXCRyMQlCvdIZANFRH5H5WCoaar65nUcpUgkUieiWT4S\n2UBR1Q9mv8NBMpFIZIIQhXsksoGS27uigbgULhKZUEThHolsuMzBzPJdQDzuNRKZQNR1KZyItInI\n7SJyoIgcKiJni8h5ItJW47mz6hkP75Im75LGevoZidQT75JXr+s4ALdhx75uBnykluNxVL/Fu6Sp\nnn7WixC32PZs4HiX7Opdsk47z/Ve5/514KLw+92q+ins3Oj31HhuXp3jcTBwYJ39rAveJVt5l9Ql\nvd4lb6qHPxMB75JJY92ojvD9j4fyeSJwK3A1cM0g3I+X+r0A+Ead/awX2wFfqodH3iVzvEua6+HX\nCOJQqFd7tb4R3v9wlch3AVPqGZ+hUjfhLiJvBR4EXgiXsmn4zwDz+3B/qIhcKiKXUqPyhwI2YwjR\naQJavEs28S7ZZTAPeJfM9C5p8S7psxfiXTLDu6Qe7+st1G98cx8A75Jp61pL7IvQi5k5RsF9HHj9\ncB/2LpnvXfIq75J9BnCzsOrSPlX39xvg2amDUT68S7b3Lnl/+Hah3M3M3d/Du+QdtfwZJP8A7lfV\nO1T1joEcjmb9HgaNgAB4l0yus98jpRmYVCe/PgfsVNPVEPEume1d8vVBOp8GfHoAv95Wn1gNjHfJ\nDt4lHxuhH5MHKi/eJXtVKVOfA3bM3X+rd4kMIchiH2EUgj+zvEt26+N+s3dJXVat1FMg7AW0ATsA\nHdg4HpiW/Wy1Y1X9LfBbgNAADMS3AbxLjsUK+zTgZmCG8+ni8MKnYJVqVXimgGnRrwLuCQ3z5c6n\n/wh+fQy40Pm0I7g/ErgXmOldMh841vlUg9uFwc2l3iV3ZNdrEQpKm/PpktxlAV7tXTLP+fQs75LZ\nzqcvVT03F1jifNqbK0yvAg5yPl0Y3GwSvmcDnwcu8S7ZA7jW+fTxcE+cTzX7zvnfABwFnAP4waZn\nKIQ09ABHepdcGC4/jDX03cAbnE//mMUTeCXQCzzmfNrpXfJu4Dqsgn0nlxebAfs7n55eFWQj0JC9\nr1pp8i75JHAT8AHgeKxc7RHuTQ/3lgD7OZ9e2l+lzl3fEtjTu+RWrA68CHzY+XRhcPN5rHd8W19+\n5OI7BdgWmB7SNDu4+Z7zaTuwMfBK75JFwLuz8jBMNgHOE5EuQFV1oF3q9mKU6rd3yebAR5xPjw3/\ni86npQEeKQd3LcBXgIXeJTsBD2XPeZcUgXKu3HwB+AXwfuBvwGLn05Xh3k7Y++51Pr11oLgOgkzp\n2NP59JZ+0lsAcD4t564J8Arn03tzcQIoeJfMwpSG2cAc59OaVpbQGdrT+fTy8P/VwIGhvLQArVXu\nBStvPaHN2Mn59P5cej4I3AVs7nx6Re7R1wFXVedZ7vnsfxuwkfPpv3LpewAoVOd19bOBycCmtdJd\nlf42TMAuxZTRL4d7x+fKSYPzaW94bC/WVGChosSCdR6u9S4pY/Jl+yye3iU7AAucT6/MuV9LuIfn\nXg+8BBwUrH8K/CiU1enA3t4lfwe2cz4d9j4UdRPuqvotABH5GBbxqSJyBlaIPlenYOYA2ZnyRewl\n/A74YD/utwYbgw//9/cueRRox8YZ9wEu8y5x4f6OVDLkGO+Sc4BZOf8OBA70LjkJ2A1YlCusBaxX\nfg8mwFZiAmsn75KfBn8bgI2CX/MypcG75FHn098GZWAOcHi4/h3gm1gl6ArXtsCE0DbBn+x7cgh3\nd++SVwF/AI72LjkX+Ih3yXnOp88Et0eF78OBJ4Bfe5dskaVlKHiXTAWanU9fDBrnE86nzwL/D+sZ\ngjWofT27Aqt47wzfM4AHvUtuwU4s2wl7b8d4l2TjtnOAuUFBmY/1HJuwSjEDOAbo8S65AHg6q7gh\nj/cGLnI+7QaS8Azh9x65qO0cPidjitiiEEdC73tj59M0uC0AR2ONAvRtkn1jiGM5F+9MafwVsJ93\nyeNY2Xg4+LkJa+4a9wngNCrWtnf39U6HSLuq1jKpZ5wPfAF7x48BKiIPYunac4Tx+Gj2w7vk/cD2\n3iXfA97ufHpJENSbAa/Gykl7cJ7vZb03fC8M30djjXSmBM7A8t8BHwMuBW6vepaQDyudTzPlpV+C\nQDzE+fTC3OU54Xs/75LHgH2cTy+qejTrrPwY6MQ6K0ux4Y17g2DK4nQAcCem1D2H1fFrwvOzgcXA\nKzDF5MHwrqZhnRW8S/4R3OwW/r8ReE32fK5jsQ2hHfUuORV4r3fJs8Abwv1tgX9jyt0V3iU7UhH8\nXwbaQjv3BkzRnepd8jBBaQrPv8O75MagnLwXeBT4ZgivJyivhHvVwr0XkJDmdkwpWEX/vAPLi1bg\n78BTuXtHe5c8FPLtKEw5zA5Iq7bOalD2M2VoI+DDmCLUCNwfysEuIbwrcwr/Lpji0xLi257zP1Ps\npoXvt2BKZ3b/6PA9bOE+LjaxEZFLVXWtMcjQ+H0f+Gr1I6ypUVXTAkwNv5dT6WVU8zhwPUGY5siU\nnlIIR6g0JD3h+ipMoGbfedLw3IKq689j1gSwE7nyaTiZoFnmeJY+TJ4j4HisAm9Vdf2PmLD43/56\nG/3hXXIA1lj8EFPirsXe/Ugb/JHSiVW+S7AyUJ3Hg+ERrFHKcz5WuTNWUhlbk9zvLvovdy2YQgFW\nfp7L3XuYShmp5ngsn1421Y6k5y4i12FpXAGgqjVnzItIATgPaFbV94nIAcBMVf31AM/0Wb8zcpax\nU1lbOboRU44G4ldUFIS/YBa4o8Dej3fJa7HGvpp7MOVk+6rrt2FK46uB80JPthVT1jcBbsDKwM2Y\n8L0P69V2AZ/qI5z/cz69NqT1jcC+4Xq+LJ2OKcTXUjXcE+K5OZU2DeAEKkr6vzDhcxbWoRnqOQE3\nY2Xg7YNw2wuchCkPQxlTXk5FkPWVxkewjsb2WFpPwnrqnVhd2ZmK4gTggfMzy2vOGvlL4GlMyc/q\nXxdrvruM7B3m69xZwJbOpzeGcnkJpnQ9xNrlJCtfB2JlZXF4voGKzLo9hJ11wr4L/HcfcQE4Fitf\nLw+DjKh+10u4i8jBWEGfCvwcMyHtjQnFz6rq6gGe7U+4n44VoEOwFybhA2YdeBJrKDPhugwzpW6P\nmeiLWOV9CBOky7AGfxdMQ1LgN1im7IdVkDlUCtEqrGJthmn83Zi59bwQ7swQn/nBX8UKXQ+wa0i7\nYpV/GdYAraKiBfaE/3f282omhXAJYbWE3/8I8Zif8+slrBB3YkrBzBAHCdfvwzTeTOPuDe8KTKtt\nh6EXJu+Sgwi9ghxTsHe/DVZYwdLZgVXYudh7WQGsxpSlf4Vn9qPyLheHe+XwvzekU0P6n8Z6Blk+\nzAp+9GK9m6wnANa7yeZfLMPyqRCea8AaxSYq5YsQ39XBXfZZjQljxXrr3cBhmAmwLcS9K7yDTMnL\nPquwXu9WWA9sKaYAXo6Vlcyq04UpgrOp9JxWAqdgZWCnEJetnE+zRmPIiMgaZk5VfbqG+3cCnwV+\nDRysqoeIyCuAA1X1xCq3h1IRMvNUda15JqHx/BFmERgqmbLdi5WRPJ1U6sqxWEM/VK7D2i+wcjQJ\na282x4az+lIWnyFYC7FykW9czwfeSiWPwcrPDoOIS5+CpQ9+gc09GS0yq2Yva9aTdcn5mFIwnDx+\njkr7lPFz4JO5/8+wdidtINqxdmBTrI62A3djlhfou8OQcXyIz8tla1wI95c9FJkB/ACYOlLN3rsk\nM0s+gI2BKtbQFzBNrjH8XxauOayhLgf33ZhG1hw+2WS5F7CK+kZg93DtKUzLKmMNcBkb524Nv+/A\nGpEjqTQsL2AFfVX4PYVK5V2BmZY2wTJTsQb7z1jFyDTJV+Xi1RceKyCdmDCeScWE3IUJkEZMIRFM\naEwK7+MBTNBvlwvv3yG+U0PaZmEmyxdCGG8J4ewI3BFM2H1HzHohmRLxeqzhm4QV5BXhff0Ty4dM\nAPeGeJUx4VUI7ywb/ng4vLfM+pIJVQnvaSqV3laW15lCtTp8tsYazRIV5aAJmxW+FBOu2bNZmVpG\nRRhnzMAatHLObRL8mouVt97w7O+pKB0a0tNGRSGVEOdM6Nwc8mJvKsMDz4c8mEtFabsluPtSCGta\nCCPFzMdZL3DMEJHLgNVBuI+kfi8MPw/F8lWwvC5i+Xhb+J29nw6ssd0GeBOWt11YTxpMGezGelHT\nsHy4GVMct8HycxqVXloP1oucEvwrhDDOoGI2JYSfz6NbMdNxNlHw4XD9w5ggaAxx+XcI4zYs72BN\nodgc4lVkzYl4j4TPLMKcC6znuyykeWV45rXhuV6sfSpjVrRmKnVGqJTjViq95yeB/w1usx74svDZ\nnYpS+TQ2xPcmbExasbbiVqxTlbV3q4FFWDu0WbjWgbUlillTlmO94DYqdbOEtW9LQvyz+rokvNeX\nh7FCeFnb+xjB4oQJzzdjnQzB6lA2TJa9723De5Dc9azuFqmUu94Q13JIk4Z3sQCr93mF+A4qSly2\nX8StWLtwCKY8OODckK6NQzglrF1uwWbVF8O132Ftc5ZfmzufHsAwGQ3hfjLWG/7vOmn2a1yi0pOt\n5m+YZpwx0Fg8WMZVC9VuLAP7Ggqoxa2sOW4LVjH6mgSSjS9X/67FOdgyv9m1HA4BwRQQwfLjDCqT\nEk9zPn0pjKmd7Xy6Iv+gtwmOmVJxCNaQLsMatX4tNVXchSkHGX/EKmoHa+b1iZhiNZAilNfEswqb\nVVoZQpxgzd5ff2Sz30uYcriAigJ2N2bGrRdtufhkDeawNXsRmaaqy4fg/ihs8mELcAX2PvchjLmr\n6osDPNufcH8x+NOETb7LK1uvpSLsMqG2gIrSczU2rLYHVh8yBRcs727BlLjNsfK4HVbWFBP2L4Uw\nMqXydqzB/SSmPBWoDJ00YUpFA2vO/L8FEzY7YmVrGja00IkJ7QJmZdmYinWhusH9d/iUMeHXjCl8\nmfXqyXBvXvCvDRNAGtw/EtKYtTOLMMGWtxitxgSohnQXMWtTFsay8GwS4t6E1cPngM+E8JqAC8I7\n3xcTtBrC66Ii/JSK4J0Trm2MKdu9mNJzeYhD9pmF5Wshd233kO6GkB/tWPuc+dsSrmdK/cbB7RPh\n3U9hTQWtG6uT2VCrYgplM2t2ArbC6m8z1jYvCv5dEeLwb0y52jTkSV553zHEo4wpdFcBb8OUegnv\neQXW3s8IcXgEK3v7B787sCHN5diE0LylZ0jU0ywv2HjC31T1ahG5qA6a/WysJzkXe+HHYi/v35iQ\nAyuMm2FaXn4JwUmYxteDVUwF/ouKlnoJ9mK3xArEN4DjgI87n/48TIrYGPhP+lYEXiY3g/3dmOk3\nM7fdhJnDd8d66P8CtgD+FL6nY5m5LfAdrIB9mjVNd5eGdD3mfPoXb0u1svXVxwW3/4kVkAZMoGSm\nyIeC/89iFfUV2OzWm8M3wE+AI8Lvj2MFVrEezc8x8/aRwE/DpLnNsAlJ3wXOxBr8peE9nxe+wRqC\nD/TxupZgPfdrsff7mvBuHsUq0U+xPO3BlDMBmpxPfxbesYS0pVhj9H/hfVwS3jXhmV0xy0zW4zo2\nXD8whHccVvH2wCruDpjG/gSVHvk24XMNNjyzSXiPBzuf/jFMXsomvlyGNQSvBFxYCZH1Ej6CNXj3\nY2VwOZYnc7GG6nch/ULFZJdNQjwlpGVvbDjmKipjrc86n57TxzuuiYicBCxR1ZOG+Fw9LXOnhZ+d\nVBTK/rgIUyCHQgtmxRGsYX48d6+/Or0JFWU7U1JLWP5tzpo976zxnI4Jpl4qvcnhcA+VoaP+KFJR\netpZe0hiJDRibYhieQImnIvh2jUMfblptfIOgxtmyAQ9VNqUjBYq84ZaQxzBZMDiIcavPxowhbCA\nKQYPUv99YfJsjckEsE7hwzAys3w9l8IdgQniaSKyFfCnkc6WDzM5L/AumQNsGpYKZJOPsmVUzYQl\nHJgJ+CvA2UCn8+ljVV6eHJ7ZA3gmzIh9MFzLlr79PIStwHNhxuudIdzPhbB+hAmluzHtMYvvH3Px\nmozNAO3CJvn8JRNMzqd3h2eziSCTc+bvM8LSly2BpSENt+fCuNa75LrgTznE8Rbn0/8NTv4U/H0M\nuMv59NHsWe+S7YHnnU+vwmZxzgpLCa8N6ToRm8m7M1bY9saUgOXAtt4l22J5/LoQ/2mYwPprePb1\nIa4dzqc9Ybb/jpjp6RZMo70V6M693zLwZFgtsND5NN8D/EVV/hEmN2X3pjufLvEu2R1bOpg1tgrc\n4V1yV0jPj8M9DfmQ9TruxXp9/3I+zS9Ry3oyj1QFvyh8/zHEpRTiQkiT9y5ZTGiUcvH5VS4PbsUU\nEsUajanOp0tz8X4ouIPKcswbvEtanE//N6cw/Nn59K7q9zMEGoCPiEhmylRVHYzwPApTwLJJQc9g\n5WUNqi1z/fiVmSYzbsdM6ncFP/NLiZ6qejZT6vL3Nwu/e7De1a1Yj+gfmIKdWfd+E9x9qMrPTqye\nP1d1/S+Y4vgt+p7Iu6yPaxnZhNzMn3cO4Pb/QjreFf5fjdW3PCWsN1/LevcE1oaATfL6NKZ8VLf5\nj2G91AXYe+vBNibKlMv8sNwtmEL0StZ+/xk3Yj3PS7EO2C2sKdyXU5n9v4jKOvJTQ1jvwoRqNhSW\n58wQdjMV5SNjGdaWZ52WO7Hy9U4qedAVnr0cKxd5fo8NtVwMvA+zJG1JZT7RckzpyybfnRnimw2N\n5JWybJLt9dhQBlie3Uf/+f9Y+PSV58NiXM+WHw7BhPxj59NqbW9C4V0ytdpMPohnGoHG3JKTvtws\nxHoQZ2AVYi4mcDqpmK92xhrKFCvgj/ex7nywcWoGxPm0urLWhbBUb+VorOUP/n8IG3t7rMba7HqG\nORlToEYUnojsiAlnVHVlDbdbYErNclV9k4jcijVWWwC/U9V+LQgD9NwFs768ErOmfA9Ll4Z7mRXp\n0+He17Ce8VSsF3kTVNaLe5fsigmHu4C5zqdn+7D+2rtkM+fTp7xL9gb+HpRPwYT9jpiimg2pVHOS\n82mHd8nWmMBpxKwIBcwCBGa+vhpTaI6lMq9jHmZBOs75tJyzfGUcF549M8RzBrbcdz4mlNswJeR3\nWE/9k2GG9l6YAp5/71Mwy9DlWEfnmPDcamyS1nGEZXjYuPmfMatbCVOMXoVZBX6J5WsJm0F/L9aJ\neSyEUXA+Xeptg7D8kswXMHP0m51PTwnvdw62CiAL+58hfmBWqm2xoYJbcvm4kIoi8tvwTtfoxXpb\nerwlFQXtFOfTFSGPPoR1OJ7COp1LgSuxsnZt8HeT8F3AZsdf6Sv7gmyBLaMteZe8DysfL2FK7Qzg\nyKq4bJPz+5jgbq/w/wjgl86nT+XircGdYvMZngW+CNzsfHpVfjh6XE2oe9lj22/6dKzxv15VfzOA\n27oJ98jI8C6ZhAnyQzANdWPM9J+NhWUTQm7HGrFdsM1Daq4LjowvRORoKjOBU1U9bgC3R2KrPaZj\nc1I+hgnRrYELhiPc83iX7OV8en0/9yY5n7aHsvlZbFjnb/0pN942oWpxYTOnGuFuhDXYz2CrfXbE\nrEJTsd7gor7CCY10U7VSGqxPC/uKf/4/1kNu7U9BDw38mc6nz3nb6ORGt/ZGVMV8vQvXXut8elP4\n34rV5QXYcOPCELa6yuZdVD1fGGgSbR/PTMJ68HsE/9d4L0F539X59NawJPHevsKu8vMITDi/EPL9\nrSGtV/bh9jNA6nx6We7aDKynna2+edz59OHBpqnK/4Mx5fMO59O/BoWltb8OknfJm51Pb8j9n9SX\n29AJPTez2uXfTcj7l5xPT6t+biiMpnA/DFimqn8VkQtVtc+NTILbKNzHId52DuvFNPL3UxkTOhEg\nCvT1GxE5RVX/K/z+vqrWnEQqIr9X1ffm5tT0OWG26pl6WuY+CdxfNYwy4QjzjZa43C52I/CrQNjN\nc+QxG194l0zBNvAZUGEYgf/bAFOcTwfcnrnOYbZgCtJI5m+M6pGv86lMcKplPlxU435kHeB8+mTu\n73nrLCKR0WK6iHyUytK94dDn9rNVY+6t1feHi/Ppz+vl13jGVW1JPUK/ytRvotm4woUthEfR/0dr\nu6p7mJ2sPadgyIx2z32pql4qIheo6geq7ucr/2/DXtTVflyDzfYeTeYx+spFDCOGMVq0quqw1rqL\nSCO2YkCAq1S1X3OsiMwC/gcboz4Hm9H7RsKE2YE2qRrAz1i/YxjrczhjEcaw6zeqOiofbCLIL7CJ\nWR8aph+Xjlb8YhgxjPU9jLEMZ32NewxjwwtjIqVlJGGMmlleTZMfza0QI5HICBAR0dCCRCKRicVo\nLsqvB2uZ6mMYMYwYRt3CGeyZ3qPFRMmHGMb4CmOswhnXYYyLde6RSGTsCees34QtG0JVh7VXQSQS\nGX+M5mz5SCQyvvl+7nfU8iORCcR4N8tHIpHRYz62Rv0G7NyBSCQyQRgXPXcR2R34MrZL1lfCtS9j\nBzU0YtszboL1NErAL1T1umo3g5kcFM6dfzu205aq6gdF5B/YXsRPq+qJIjKvOqwhpudj2A5vzwA/\nVdX7BpOeIYZxOLZf9kzgBFW9u97pyIU16N0Gh+DnwdiOYFOx/fy/hC2vWqWqX6lHmCKyF3ZG8gPY\nQTbzsP3ym7GdzqhDGG/EtrtswA6fWV6vdIQtX78FTFPbOObQWvGvdqMDL1HbE9ugCCr7stedWL/H\nd/0O4dW1jk+U+h3CWT/r+GhP5R/ClP/NgB+E303Ab8Lvz2HraY/GehcFbJLBWm6GGN4Xgf3C7+uw\nM3c/Gv6vEdYw0vIR7CCCs7G92WumZwTvbVfg6NFIRy6Mw4CDwu8L65zvM7DK/wdsL+2v1itM7DS9\nK7C9srcCLg7XDwj+1y1d2CEZ/zlK6fh9+K4Z/2o3Nfz9Ibbf9zTg1/XM1z7CivV7eO9t1Ot3vcpp\nP/5OiPod/Fiv6viY99zDdpXVW1V+our/LCqnrT2DmQ/nY5p/2U6X7dPNoMJS1RewzTh+HK7tG/y9\nUET+2kdYQ06Pqp4nIjtjM5J/MIj0DDkM7KCLI6kcBjHsdNRgKLsNDpXslLG7QzxPCemtR5g3quoN\nIrIRdnRq1vPLn2RWr3Qdip0FfvYopCNjMPHvy01//BI7aORn2MEsIybW7/WyfsPo1fGJUr9hPavj\nYy7cVfU+7EztNRCRzXJ/F1M50nABdirRs8B8EVkxgJvBhvUG4DZVLQd32f7NS7Edt6rDGnJ6Ai9g\np+OUoM4AACAASURBVCgNJj1DCiPsLnY68ENVTUeajho8ixXgu6nTPA2x1ui7wBWqemfuVvbORhxm\n1ftopnKEZH7L1BGnS0QWYKel5bfCrFs6+mAw8e9zW9gqnicceYydXT5iYv1eL+s31LmcTqT6Detn\nHR8vY+7bYEfg7Sgin1bVs0TkThH5EZZpp2PnE38XO8jkHFXt7sPNYPkEQRsWkRnY+eydwBJV9SLy\n83xYw0jPp7HjE2cBx/YT1zXSM9QwgJOAbYDPhm08r653OnJcApwmIgdgpzXVgyOwc4unichW2Phv\nO1Ymv4cdOzuiMEXkPdipUNOB04B5InIGYcvU4Kwe6fok4ex5EflVvdKR2/J1VxH5BvCnQcS/Lzf9\ncSqVGfOnAB8cSvwGS6zf475+Q/3r+ESq37Ae1vG4zj0S2UARkWNV9Zjw+zhV/XatZyKRyPrBuOi5\nRyKRsUVELgbmhl6hEJfFRiITithzj0QikUhkghF77pHIBoqInAwk2X9VPWQdRicSidSRKNwjkQ2X\nUhTokcjEJI6zRSIbLs+IyA9E5NsiEifTDQIR+ZiIXCoiPxKRc0VkCxGZJyJf6sPtQhHZaRTi8AER\n+YWI7Ja7NuKF7usCEfn9INxI7vfWIvI1EdlMRH5Q47ndROQz9Yjn+kjsuUciGy5vAk4AutZ1RNYz\nfqaql4ZlTKdiy+6ciGyPLfl7GjgPeAM2afG32A59u2Lroj+HbTk6BVvG9rCq/kJEvo+1yc8AP8GW\nSDUDReCLqloSkcnY9rYPAs0ichO2ROoXQditAO5V1TNF5G5sB7dNsOVae4S4v7xNrYh8C1ujPwXb\nIvYnIT3fwHYw3A44CGgBFoY0vRn4F1BW1f/J+bUf8AFs3flJwF4hzKnB34PzaQYeA3YQkaPC/WOx\nyZ0rVfUoEXkA+DW2TO/REMxnsOVnrcAeIvI/IX2HA3/BdvHbCjhDVe8IStfPBs7OiUnsuUciGy4P\nYRtxpOETGQKquhjbSz5jLibYfqeqDwB/B05X1b+H+z2Aw4Q8wEWq+nXgbSKyA9Ctql9S1VOxHfY2\nA5YBk8NzqOoq4HpsLf0i4EFV/S7wPuBcVf0ctr4c4AVV/UaI0zWY8D44i2xQRt6E7ZXeA2wPfAW4\nELhLVR8BPqKqh2PKSNYLvlJVjweqrRL/ie3e91+q+jzw+eD3C9g++WukObyXB1X1BGz3t9YQ1y1E\npAlYpKrfVdVHc2FsHfwGeExVv4WV49diStCPgG9SWf/dGXaw2+CIPfdIZMNlAfAdrLekrL1NbGQA\nQs+9O/sftkL9F/BJEdmFyo5pAIeo6jtF5BhgUriWHfgh4ZN3XwBuUtUfMzDLc35UL33Kdq7rCr+7\nMUtAPowHVHVhLk3bAiuxjWHyaAijOt7VbvJ0VPm9sI9ns2cKwGWq+pec+yxtecp9XMv7UWRNhWsp\nZjl4vvqhiU4U7pHIhsvC3O+4JnbwfEZE3oqZmBdmF0Vkb8yEPRP4G9Zj/aqInAs8JyJfw3qwN1R7\nqKoPiEhrMM2nWM/8Z+H/dOAIVe0cIE6/A04Nu7YN6nS4EGZZRE7Bes3fAY4D/gP4loi8BjhfRM7E\nFJLjgdcN4OU5wFkishTbb/98ETkL6AAu6+eZHhE5Prg/TewEtiZV/cIA7jPFYBsR+Q6wMbbDYje2\nl/3WIS1gFo9nBnwRE5S4zj0S2UARkRMxoT4ZeIWq7r2OoxSJDIiIvA9YqqpX93Hv96r63tz/ydhJ\nhBvkpLoo3CORCCKSjfVGIuMWESlgp+P9bRBut8GWez4x+jEbf0ThHolsoASTr2JjlFNV9ZPrOEqR\nSKROROEeiWygiMim4WeXqtblyNdIJDI+iMI9EtkACbO216j8qnrcOopOJBKpM3G2fCSyYXJB+G4B\njmTtpU+RSGQ9JvbcI5ENFBE5DDgA+Imq3rSu4xOJROpHFO6RyAaIiNwO3Av8gWCeV9XL12mkIpFI\n3Yhm+Uhkw+Qn4Xv2Oo1FJBIZFWLPPRKJRCKRCUY8OCYSiUQikQlGFO6RSCQSiUwwonCPRCKRSGSC\nEYV7JBKJRCITjCjcI5FIJBKZYEThHolEIpHIBCMK90gkEolEJhh1Fe4i0iYit4vIgSJyqIicLSLn\niUhbjefOqmc8vEuavEsa6+lnJFJPvEteu67jMFTGUf0W75KmevpZL8Zz3CJjh3fJbt4l63STuHr3\n3L8OXBR+v1tVPwVcCLynxnPz6hyPg4GD6uxnXfAu2ca7xNXJr73q4c9EwLukbawVuhG+/3fUKx5j\nyHip35sC/5+98w6XrCjz/+c93X3znRxgigNDVgQEUUwIKAYUUHTXhLq6yhqW1TXvuqAMiq6YMaBi\n4KcIZl0VxQDmBCJBJOc5FJOYdO/c2N3n/f3xVk+fafrmvsPM9Xyfp5/uPqdOvVWnwhvqrbf+p8V5\ntgqPAN7Sioy8i5d7F7e3Iq8ZlCHyLt7r4SzDw4UZvv9TgN5WlmeqaBlzF5FnADcD68OlWui71cBD\nOkeQ/C8TkcuYYPCHDrZoCsVpA9q9i1d4Fx8xmQe8i5d4F3d4F/eMcX+Rd3Er3tfTgCNbkA/A8QDe\nxQsebimxGYIWs3gnkftX4MnTfdi7OPYufqx38dPHSbOq4dLxDffHZNjexfMnI3x4Fx/iXfzS8B2H\nfrckc/+J3sXPmSifVmM2x/c0sL2vexc/rBNoE7QDnS3K6w3AoS3Kazu8i5d6F//3JJPPB04fJ68T\nW1Oq8eFd/Cjv4lfP4HnxLu4dr794Fz+1gZm/ATgkc/9E72KZAtlCExqRd/GzAr95bJP77d7FT50C\njTHRSoZwPNCNvYwhYCRc3xu4vzGxql4KXAoQJoDx8B4A7+JzgEcD84DfA4ucTx4ML3xeoN8PCCa4\nHAw8Brg+TMyXO59cFfI6HbjU+WQw0PgP4EZgUdCsz3E+0ZB2VUjzY+/ivzqfpJN5IaGj9DqfPJi5\nLMBjvYv3cj75nHfxUueTDQ3P7QFsdD4pZzrTUcDJzierQhoXvpcCZwDf8y5+EnCl88kd4V7kfJLW\nvjP5l4AzgS8DSa2erYR38XJgFHijd/G3MWZwC8YIRoFjnE++WysncARQAe5wPhnyLv4n4ErgjcD7\na+X3Ll4Z3sOnG0gWgWJNAJuojbyL/w3rQy8G3gc8CnhCuLcA+APwIHCi88mPxhLsQvsIsD/weO/i\nP2ET8nrgNOeTVSHNfwBXAFc1yyPTBr3AgVh/bgMWhzQfdj4ZAJYDR3gXrwGeV+sPOwHHM0vj27t4\nP+BfMn276HxSGeeRWl/oAN4GrPIuPhy4ufZcEHarmTH8FuBLwEuBn2Pjqy/cezTGxMrOJ38a9y1M\nDAl5HuN88vsx6vuQPhr6yBHOJ9dlygQQBQG5E+sLS51PrpioEEEZOsb55Ifh/+OAk8I7bseO+s2m\nF6y/jTqfqHfx4c4nf8vU5zTgOmA/55MfZx59AvDTxjbLPF/73w3s6XxyZ+0+Nt8WGtu68dmAbpoI\nkRPUvxtjsJuxPvO2cO/cbD/J0D8OuIm6AAt1IbZW1194F6ch30Nq5fQuPgTYx/nk8kz6hzB3jC89\nMdA42bv4uEDj46GvLgCO8y7+HfBI55MbJ1vnRrSMuavqmQAi8ipsUpwnIp/FOuUZLSKzDDO5g5X9\nWO/ibwEvGiP9gbCdyQI827v4dmAQ6yhPAy7LmJ0Ood4gZ3sXfwlYmsnvJOAk7+IPYcz2AeeTuwKN\nAvAM4AZMU+nDNPRDvIsvCOUtYJMzwB41ocG7+E7nk6+Fci4HXh2ufwAzP96CTah4Fx+AdbIDQj4H\nh+8eYA+MwTwWM5++27v4/wGv8C6+2Pnk3pD2zPD9auBu4KvexQfUBt5U4F08H2h3PlnvzUx9l/NJ\ngkm9fwnJXjjGs33YwDsZ2IJ17Ju9i68GDgsfgPd4F1+IDYylwJIwee8N3ItNVAuARZggWPUu/jpw\nb2YQx8AJwNedT0YAh03oYGbeJ2SKdmj4fBQ4KjDSk0M+JWCF88l9IW0EvBtrI4A3N6nqsUAJSEO5\n45DXKuBi4FnexXdhbX9TyHNPYF0mj9cAn6RubXtes3c6i/ga8J/AQuAOQEXkZowhPHGGef9L7Yd3\n8UuBg8MYe47zyXfCO1sJPA7rLwMheVbLekH4rAr/z8LmoZoQOB8b73sCrwQuA64J956foX8n0Bf6\nyLgIDPElzidfz1xeFr6f7l18G3CC88k3Gh6tKSufxISkBcBGrE2v8y5emCnTScC12Nheg80tV4Tn\nlwEbgMOBivPJTWEeWoAJxQRh80HgseH/scDR4XdWsTgIE3zwLv4E8ALv4gTru7X7a7Blhx97Fx9K\nnfG/E+j0Lv4M8BRs3u3yLr45lKUann+2d/Hvg3DyAuBW4H+8iz8eyl9r1xdgJxZmUcEEnaVY+xec\nT/oZG8/GDkXqxIT4ezP3zvIuvi2021mYcFhrt0YhXkN71IShPbD+WpvP/1YTzAK9yzMK2RHAz7yL\nO4Eo1K+Wf02wq1kSnoEJnbX7Z4XvaTP3XeLgGBG5TFVPbrweJr+PEiSuWvLwUXaUqrKoTfZgTGN4\njHR3Ab8hMNNM/jXzaQVrBKHeuOVwfRvGUEexCS6LJHzHDdcfxJixYpNUtvyN9QSb3JfTOpwLvByb\nKLP4IfBc4BfOJ1M619u7+CRs0j0f+Hfgl9i7n67DWNbsNZPOOYoNwO9jfeA108jjDoKAmMElwMsy\n/weBrvBbqAsMWe22ER2YYAI2UfnMvduoC2y1PGt4LzbpbzfV7kTN3QojEgFfBdpV9YUichKwSFUv\nHueZpuO7hoxl7BM8VDj6I/CkCYp1MfCK8Psy4HrC5BgsJ4+nuY/DjZiAcFDD9auB+7A+fFHQZLuw\nyXoF8Ctsgv9jyPfvmFY7QnMTdo2h1fw0jm9S7guw8fMbTIPM4gZgP3Zcw30/dSH9bkxL/RwZJj0F\nXIWNkWdNIm0FOA9rp3EdKRvQh1mjoHkdb8fq8UhM2P5Q+B7GxsrhBCtWwAPA12qW1yAAnoX1zXuA\ns7H2UEy4aFbWc8Mz2XF+IXCg88lvQr/8HnVB5BGNGYT+9TxMkdsEfB6bd94RklxL3RoH8EFgrCWR\nczDh87XZ/MdIOyFaxtxF5FRMypyHmb+WAE/FBs8bVHVgnGebDv5DS22rC0JxURTtIYjUZjkBKmi5\nrAwKRAUxbTtVyhW03CVRj0AhAhEhGlHt13BfQLojWWD5CBuq1fvK6OiiqLBHAQolkfYStAXJoTKo\nuqVDpLskdAKpwPCGanpHJBTakXZAO0R6CkIE6IjqgEK1V6KlBYgQGEh1YwUd2bNQeERVGS0IRUFE\nIVXVcp/qegHN1E8BikKpXaQHoA3pLASrQr/qmhHVwU6RnqJICaCsOqhQSaFSVt3SKdKxKIr2LIjQ\nITJShHsEhhdHhb2KAr0SDR9UKq4BdFlUuL0oMgBT70zexadgVoztuL9SWVaG0pa0+ojNqe6ZotxX\nqaTDqpV2kcUFZF6K6igMj6qWq5D2p+kmoNApcmAKUWp1GlLQKqCqWoW0jA6nio6gw4Op9ovQUVUb\nuEWhXUAUdFOarqkq2819vZEs6JFoEdZ3RkdUBwURQASkU2Re1CC1p1CtomVLY2lTtDIayjWgurWq\nVPcsFPYvQDESigKSQrWAlEZVa0s+ClCFyqCmfZ0ivW1IZwVG24XuzWnqI6QQ+hgK1VHVoRLS0RtF\nS60sWlldqd5cEAo9Ei0UoFOk97KhwXnsJIjIczGLzMXAqar6IhE5DDhZVf+3Ie1pwGnh7wpVfUxj\nfmHy/CTwpmkUp7auXRO2syhTF9DfS9CUp4hfYfNXNr97McH4izRn4gl1QWGYHYXTSzDmmT1i93Ye\nKlg0wy0Y05sIF2G+J7OF2jutUtc8H25cijHns6fx7FpME8/iS+yoCHjMylfDRMrHCCa874vNJ4PU\nLS/QXGGo4dyQbnvf2iWY+/YMRRYCHwHmzVSyv2jxkk2qMKi6en1avVlBNUzG8yVaGQmFVNEyug2I\nOkSWAJGCPlit3lqF6rwochEUBUpFpENRRpSt/ZpuWBIVHtETyYEAw6rrt6bpXQral6brUmBRFO0T\nIW2K6qY0vU8hdYXCM1MoKeio6jaFagVGRlS3FaC9JNIrIGXVoT7VNR0i8ztFlqnCKNrvq5WbQCia\n0CMLo2ivKAgTUO8tQbiQYdWtFdVyxSb8wTaRrgVRFANSVa0MoVsFCiVsO1JJZF4KpVRV+jTdOKi6\nraosHVZtr6AMq46OKtV1abW7P02LZYie2t7BY9ra+ntERl7e3XM0ZlU4FPiL88noWO0WtJD5wJE/\nGxo6+a5KeeXGNO34yuBAd5eIFkGXRIWtCunSQqQ9ElUF0gFNN0VI2iHSXgA6RLo6RDoi0ApsGNB0\nbREptIm0RUAEEgmUkLaSSHsRCr1RtDyCSECLyEaMEY9UYKRbZI8ukb2AVO1VpgKFTWn691HVgZ4o\nWi6hn2B9Siuqg4OqG7P1a7MtXgKoqqXtFFksQqENmdcmLMbK3L+mWr0aYEh1G6Dt0F0UaZdaFwfp\niaIVhWDl2Zimdwypbl0eRY8sinQCjKr2VaHcjsyLxCbSTWl615Dq1v2LxRMBLUBnWdkygm6qqA6f\nvXVLyx2uJoKI/BgYCMx92uP7uI6OqwD2KhQPKkKbWDuLCb5a3pLqmgikTWgHQdHykGpfr0SL5kfR\nXhiDSbek6WrBBNwUKgujyBVF2gTYkqart2m6ab5ES9pEOkpCR7dES4MAnW5J0ztLIm0Lo2j/YA6s\n3FOp/F6xiVEQbRO62pEOgGHV/geq1fsOLJUe3SkyH2Bbmq4dVO1bWSwe3S6yQKBQVR0cNf+f6qY0\nvSdVrUZBPalxhwIUQ1+UApQK0CYIA5qu7Vdd2w7d7Taf6CgMVlSHUkjLqoMiErlCYb92pCBCtS9N\n7xSoLo0KB5aEYoRIQaAAsiwqdM2LIkpC+4pCcZ5Vk3swM3AbdYvAFmzsPx44JlxbDXx7VPX4tdXq\nU4B0bbXaty6t3hpBZwFZpEAZHRpW3VxE5rcJyxWoKKNltD9VZLOm942oDi2NosdXoT2FNFU0RbUK\no6PWdmkKmqrqCAxuSdN1CsWq6jKAos0HRYBtmj5YVrOMPZhW1+9ZKMSLooIDE4T7U93uxyRATySL\nS9g4i0ITaFAWBJFwLVJIt6bpWlMqdFRBl0aFfTpEFhSg0B7aHKBf0wdQqmEeIYXKpjS9v0ukd3mh\ncOioan+7yLz7q9VrgWpH4A0KOqjp1iJS3KNQOCQUUe+vVq6rKKPtIl0C0i2y6HPb+h9iLZgsZoO5\nfxSTUv+7RZJ9FqupmzMbcQWQ9XRuNJ82omZWz6JmYv0Q8M5xnm2GZubDezAJrhGbsPVhsMG0cJI0\nLsR8DpZNlHCyuKVc7vlo/9ZnKkS/HRl51C+XLr8gLhZrA+MzzicbvIvfAVzofLI1+6w3B8dlQPTs\nDetevTCK1nVL1P+Mjo7rX9TVfdski3AdO+4e+D62zjfIjssaH8Cc0sbTVLNSds2pshC+heC3MA5S\n6tp71tw+Fmr+GSlmxt2betv+lQaLxgzRRX1paPuSzs4yy4vIWcBLQhkux97n0whr7qq6YZxnmzL3\n98xfULPmFe6vVq4JkpgGwXrvmtBbhuFUVTuMcRYBNqTpXf2abl4URXuVkG6BqBj221dVK5vT9P6u\nSBZ1iCwcUQY6RRb3pelagAHVzcOqQwuiaEUU2nhrmq4bUu13hcJhFdO6pU2kx16zRCOq/UDUJmYl\nUoU+1bUChW6RxVUoF6BtdbVyQ0Wp9kQyH4i6JVrYJnRpkBbCbLt91i2jQyPBElRWRiMoLIiiFTXh\nczhYHUtCV2BCxRozGdR0tD/VbSWhu4R0VYFh1eFB1YEUlRRIFdmQVtsqQF+ads6Lov52qDyy1Nb7\nmLY2FWCPQqEfoAhLykolRQtrqtVrN6fpwE3l8lMGVVmbVmVIdbQNBhdFhdKCKNIItIyOANV2kc4C\nSITosNEvt4m0FRBpF9q7JJovJmyP9GuahMEpBUTaTMgvRBnhbr7IchEQRLpEBhRGFdKK6khJpCuC\nUrBWplUYLQk9gshgmm7qiqLFBWhX6pq1QrUvMO3wX4vQHgmF0DaqkHaJLCqKdAoUqzA8orqtU2TR\nhmr1ziqUh1UHKuhol0TzOkR6yWjyPSJLSyI9apbINevS6j3LosLKLlM6qcBIVXWkKNJRhA4FBlU3\nbUnTNcsKhQNGVQeqUJknsqIKwynol7b1N929NRm0zKFOzGr+QeByVb22bkSfkTftpzAnqOXYestF\nGDNZj60Rg5lB9qVuMqrhfsx8OJqh/2bqa6I/w8xiKzFz2zsxi8MrnE8GM4zrDZgEPuYWioyXbyfG\nqP6Oab4PAD/FJvmjqZtkfhPoLsQmkoXYGloVMwllzUA/wNbo7nA+ecC7+BbqzD1bxqux9nxMuH42\ntnVpMNR/DbbT4EnAbwmOMo8slT7wxUVLtgGcvGH9AcevX/vvIsLp3T23/fe8+bd6F/8CW69qC3Xc\nH1snPO93I8NHvn7TxlOqthwhv162x7f2KRb7Qtm+TvO1v42Y09ivQtlra/M10+MDWDtXMGYCUAoW\nhI8FZ5WzqQtOvw7v57vhvddweKhjbZ3unPD9HMxH4L3AMzFHusux9bQbMT8MF+gfGD6/xtY9t/c9\n55MfBOeld4d8bwR+jLX1XsHDvtavX4Gtpd+M+VVsDm2yDOvbNdMi1M20tXf3UWyt9wTMn+EK6s42\nq9lJUNVzgXMbLHOHBM39RGysTQmv6+n9cvg5TN1Jbix8E9vZMBV0UHc8HcTWdGsYa0z/lbq/ztpQ\nrhSbK/Yfg848TNCrhHyni2ux8TseCtSFvGFszpgUNlSrbdeVRxeOqkY/HR464KbyaGlQtX1LmvZE\nkBaQNUWBCEkD8+08oFS88hntnUlRSI9p7/hGu8gxExLaEc2E3Juw3SnjoSacw0Pr2IH5IIAJZzWf\np1sx4boVKGC+LxHGQ27hod7vWWVgpjiAep1+j80XM0Ir19zfhHmh/gVzaOnDPCc7gTOms+Zeg7d9\nvns7n1zb5F47xtQqGIN/O6bhrhlri5d38dHA7c4nWyaqV/DevQ5jjv+OmdM/gXmf3gA82IyOt/3y\n5azXbWBM73Y+eW/mWhHozmrFYRvHfsCWsTzYvYtXZYSKZzif/KLh/mmYWf2OzLWnAQc7n3y2Rsf5\nZJN38VOA4sUD2/58cmfn2RcPDJz465HhA1/Q2fXrxVGh79mdnbdh29IATrijXD7h7kpl36tHRxbc\nVimXP7Rg4efO7+8/77wFC4/ABvNw2MbXhjHs52Nby1LMeWe0Zu73Lj4Z29L4Ve/is51Pakx4TASL\nznuB+c4nm4Nl4VLnE9+QLsLWWz/pfLIpc/3F2G6CPTDB4IrG7YiTRca69H3nkxu8bc8qZDx/G9M/\nExu8ik0avc36oXfxexr6yTOcT36REW5+UNs2tTPRYsvcO9nROnINNq7+igmiWeWj0Zq2XUANuJv6\n5DiCWeb+hO1y+BPm0f9zTKC7BHv/L28o0hDN96j/ABMcz2xybyIode3u+2Q885vgU5gyVNsJ8QvM\ni7oRNUfbbN6NuJO6YHMh5qSV9UWo4XaMOa7MXPsGdcE6i49g1tEjqAvVjfgtpkD8EHNEqzkK1rAV\n+BEmZD9A3Tn041i7PZfM3vIGfA7rF+08VAjahAnPNeHjr5gPxKnU39NIePbHmH9YFrVdV7Xvb2L9\n6XHhfs3CenMo32ex9/k4rM9k46nUnAd/iVm3wNrj79R3e42Fn5FxbNyl1tynVYgJmPtU4G0/66ed\nT8qtyG9XhXdx7wRbQZo9UwKKzidjmqe9i1d9f3Bw7beHBj6RKnJDebTtuPaOrZFQDear9M+jI0uX\nR4UHq+jAyZ1df39z77wfOZ98fpr1aMO2iYy1o2FG8Ba0Ytts7OUP+b8U07jumGz8gxbQ7MYEqElr\nbTOFiOyHMaetqnqsiPwZs1bsB3xdVb84zrNj7YYRbEvqozGm/iFgKHinC3UnpNdRZ+61ifMKTGAk\ns5f9CGzyvA5Y5nzyBV+P9bC388nqIMj+yfmkEmg8HdMiFzD2Etl5IfbC/piAWsIYQBGzCAnwb6FM\nrwjX9gvpVmBCyDmhXntju3NqWt85GOO9MNxfgDHFvTGm3BnyvAQTPk4PHtrHYVae7HvvBd6KMa9r\nMMH265jV4nTqFj0wq8QPQ52rgd5RmJXuKxizr2KM5gbM4nYXtpQZOZ9s8S4+DPinDP21mHJ3rPPJ\nx8P7XYox9xrtqzFLGaH8B2Hb+a7OxLJYRV0Q+RpBCMsyupD3ftR3G3zM+aTPh5gJoW73YVsCNwea\ne2EWwxLWtwrhs7/zyc98iDfhLZZG4nxS9S7+Z0wAeRD4DNY/3tRQloMyeZ8d0h0X/r8R+LLzyepM\nuQnpKtiYuh+Lavg755Mrs8vRuyRzD/GmL8C0k1+r6iXjpG0Zc88xM4QtP8PYBHPo9wcHj7ivWtlb\nzCksiswJcPilXd0/iUTOwTr+7ZPZF5xj90XGMrcA+DDwKoyJHgh8YzrMPQvv4qc4n/xujHudgbl2\nYozi75i1palw4y3AU7vzyd3N7jekXYZN2KsxbfIwLLjTPMzBbE0zOsEq1NYolGYtao3lz/7HJvaO\nsQT0MMF/3vlkjXfxMcAfsgJqWBIqZB1egxXwcS4E4glWpBGMcf9rEAq6gLSZMN0sz4kQ6nIM8OSQ\nf4Qto42E++3Ao51PrvYWSOfGiQR57+IzMAFlXWj3p2MCxc+bpH0dcL/LBNYJAlIfxrifDtztfDIt\nM7d38amYVn6N88lltTqPpSA19uOx0gYl9KKa1S77bkLbb3A++cx0ylzDbDL3VwBbVPVHIvJNVR1z\nvSxn7rsmvIv3wSToBzEJvbb/+gMAU5kEcswNiMh3VPWfReRb45nlG55ppWXu1cBNLkSanKvwNC9w\nTgAAIABJREFUFpVuUyssToHhLnQ+2Thh4t0MYfmzMouWvwOxpbOHLAnPFoJQVpqqZbYRsxmPfC/q\n0XUmMh8+MIvlyDFNuHoUNjDzXo4cjWjqMNuw5t6qWOs4n3x54lS7P1rJiIO5e84xdgDnk22znP8d\nE6dqOc1hxg68NmnMJnO/H2Pw19PEo7Bh8F/aLAMRuZKJty/NFCuYfeEip5HTmC10quoJs01ERBZj\nOzqOFJF3Af83Xnjp7G6YcfLMx3dOY3emszNoTH98q+qsfDDHjIswr8KXTTOPy2arfDmNnMbuTmNn\n0tldy57T+MejMZfqMhMas6a5q219m81QiDly5MiRI0eOJmjZee6zhHHNejmNnMY/OI2dSWc2MFfa\nIaexa9HYWXR2aRq7xD73HDly5MiRI0frsKtr7jly5MiRI0eOKSJn7jly5MiRI8ccw2xuhZs0RORo\n7ECNRFXfHq69jfqBMK/HDrH/MLZn/iJV/VVjGp3EGkM4d/5ELNKWqupLReRqLITofar6vyKyopHW\nFOvzKiw85WrgM6p642TqM0Uap2OH0SwCzlXV61tdjwytSUcbnEKep2LxnedhZyi/BQsXuU1V394K\nmiJyPPA+7KCKb2BbV56KxZd+Q0g2UxpPwU4fLGIxp7e2qh4h5OuZwHy1wDGnTVT+xjQ6zpkOOwv5\n+N61x3eg19IxPlfGd6Cze47x2Xbln4LL/0rgI+F3G3BJ+H0GdgDNu7HQkBHmZPCQNFOk92bgWeH3\nr7CQk68M/3egNY26/AvwHeAL2MlfE9ZnBu/tSODds1GPDI1XAKeE399scbsvxAb/d7FY2u9oFU0s\nvvPlwP/DDtH4drh+Usi/ZfXCYpq/bpbq8Z3wPWH5G9O0sq1m+H7y8T299zbr47tV/XSMfOfE+A55\n7FZjfKdr7iFcZWOoylc3/F+MHSYAJh3vFT6JqqbhONlmaSZFS1XXY6ctfTJcOyHk+00R+VETWlOu\nj6p+VUQOB/4LO01povpMmQZ2EtKbsAMiZlSPCTCVaINTxVnYQQvXh3J+LNS3FTR/p6q/EZHlwMeo\nn++8GjsSlhbQqOE07LjeL8xCPWqYTPmbpdlpyMf3bjm+YfbG+FwZ37CbjfGdztxV9UbsGMYdICIr\nM383AkvC772BvxEi3olI3zhpJkvrGOAqVU1DutppXpuxiFuNtKZcn4D12ElNk6nPlGiISAkz13xC\nVZOZ1mMCjBttcDoQm40+CFyuqtm4zbV3NmOaDe+jHTuJC3YMmTrjeonI3thpadlY0C2rRxNMpvxN\nw8LONvLxvVuOb2hxP51L4xt2zzG+q6y5H4QdgfcoEXmtql4oIteKyPlYo12AHTf4Qew0pS+q6miT\nNJPFqwnSsIgsBM7HYvluUlUvIl/K0ppGfV6LnTe8GDhnjLLuUJ+p0gDOw45LfEMI43lFq+uRwfeA\nT4vISdhZzK3AG7ETm+aLyAHYeduDWJ/8EHbM5IxoisgLsCMrFwCfBlY0CZnainq9BovGiIh8pVX1\nmGTI18a8xw0L+3AgH9+7/PiG1o/xuTS+YTcc4/k+9xw5cuTIkWOOId8KlyNHjhw5cswx5Mw9R44c\nOXLkmGPImXuOHDly5Mgxx5Az9xw5cuTIkWOOIWfuLYaIvEpELhOR80XkyyKyn4isEJG3NEm7SkQO\nnYUyvERELhKRozLXZrwR9uGAiHxnEmkk8/tAEXmniKwUkY9M8NxRIvL6VpQzxz8G8vHdWuTje/aw\nS2yFm4P4nKpeFrY5fBzbluNE5JHYlqD7gK8CxwDLRORSLILXkdi+yTOwkIS92DaXW1X1IhH5MNZm\nq4FPYVso2oEC8GZVrYpIDxb+8magXUT+gG2huCgMhj7gb6r6eRG5HovwtCe2neMJoezbw1iKyJnY\nHt5eLITkp0J93gV8AngEcArQAawKdToOuBtIVfX9mbyeBbwE25d6HnB8oDkv5Htqts7AHcAhInJW\nuH8OIEC/qp4lIjcBF2PbeG4PZF6PbU/pBJ4gIu8P9Tsd+CEW5esA4LOq+tcwKX9u/ObMkWMH5OM7\nH9+7PHLNfRahqhuxWNM1LMM6/tdV9Sbg98AFqvr7cL8MOGwSAPiWqv4X8EwROQQYVdW3qOrHsQhc\nK4EtQE94DlXdBvwa22v7AHCzqn4QeCHwZVU9A9t/CrBeVd8VynQlNrhPrRU2TFbHYrGUy8AjgbcD\n3wSuU9XbgH9R1dOxyaomJf9UVd8HNGotr8Oie71VVdcB/xHyXo/F0d6hzuG93Kyq52LRoTpDWfcT\nkTbgAVX9oKrenqFxYMgb4A5VPRO4BXg8NkmeD/wP9f2hw2IRrnLkmBLy8Z2P710ZueY+iwiS/Wjt\nfwiVeDfwGhF5NPWISgAvUtXnisjZQFe4VjsQQMInmz4C/qCqn2R8bM3k0RjUoBbZaiT8HsU0hSyN\nm1R1VaZOBwP9WOCILDTQaCx3Y5oshhryXtXk2dozEfBjVf1hJn2tblmkTa5l8yiw44S8GdMs1jU+\nlCPHeMjHdz6+d2XkzH128HoReQZmglpVuygiT8VMXIuAX2AS7TtE5MvAGhF5Jybh/qYxQ1W9SUQ6\ng+kuwST3z4X/C4A3qurwOGX6OvDxENVpUqdHBZqpiHwMk6o/ALwX+CfgTBF5HPA1Efk8NmG9D3jS\nOFl+EbhQRDZj8bi/JiIXAkPAj8d4piwi7wvpPy12QlObqv7nOOlrE8dBIvIBYA8s9vQoFuv6wFAX\nMI1o9bgvIkeOHZGP7+bIx/cuhDxCXY45BRF5IbBZVa9ocu87qvrPmf892Ell/7BONzly7E7Ix/fk\nkTP3HHMKIhJhp2f9YhJpDwKqqnrX7JcsR44cM0U+viePnLnnyJEjR44ccwy5t3yOHDly5Mgxx5Az\n9xw5cuTIkWOOIWfuOXLkyJEjxxxDztxz5MiRI0eOOYacuefIkSNHjhxzDDlzz5EjR44cOeYYcuae\nI0eOHDlyzDHkzD1Hjhw5cuSYY8iZe44cOXLkyDHHkDP3HDly5MiRY44hZ+45cuTIkSPHHEPO3HPk\nyJEjR445hpy558iRI0eOHHMMOXPPkSNHjhw55hhaytxFpFtErhGRk0XkNBH5goh8VUS6J3juwlaW\nw7u43bu4rZV55sjRSngXP/HhLsNUsQuNb/Eubm9lnq3Crly2HDsP3sWP8y4uPpxlaLXm/l/At8Lv\n56vqvwHfBF4wwXMrWlyOU4GTW5xnS+BdfLB38V4tyutprchnLsC7uNu7uLSTac7k/T+rZQXZedhV\nxvc+wLtanGer8EjgLa3IyLt4T+/ijlbkNYMyRN7Fez+cZXi4MMP3fxLQ28ryTBUtY+4i8gzgZmB9\nuKThezXwEGYWJP/LROQyJhj8oYMtnkJxSkC7d7HzLj5yMg94Fy/1Lu70Lu4Z4/5i7+JWvK+nAke0\nIB+AYwG8ixc+3FJiMwQtZslOIvdq4JjpPuxdvHeQtp85TppVDZeObbg/pkDpXTx/MtYk7+JHeRef\nFr73Dv1uaeb+E72LT5oon1ZjNsf3NFCo/fAuntfivGeKNqBVDPl1wKNalNd2hLnufyaZfD42tsbK\n6zmtKdX48C4+1Lv4NTN4XryL543XX7yLT2iwurwOE9Zq95/jXSxTIFtovBB42bO9i5d4Fx/d5H6H\nd/EJU6AxJlrJEI4HuoFDgCFgJFzfG7i/MbGqXgpcChAmgPHwHgDv4nOAx2AS0W+BJc4n68MLXxDo\n91MXWg4K6a8LE/PPnE/+FPJ6HfA155OBkPYM4O/AQu9iB7zX+SQNaVeFNJd7F/+ldn0iBKmv1/lk\nQ+ayAI/1Lo6dTz7rXbzc+WRdw3MO2OB8MprpTI8FTnI+WRXSxOF7aSj797yLnwxc4XxyR7gXOZ+k\nte9M/iXgTOAiIJlsfaYC7+I9sT7wH97F3wVSjDnsE64f43zy7Vo5gSOBKnC788mgd/ELgSuA/2TH\nttgXONn55FMNJAtAISOAqfOJMgZC+/8WeDHwPqzfPiHcWwD8AWNkz3E++cFYgl1oHwEOwNr1D8Bh\nwDrgpc4nq0KaNwE/B65qlkemrD0hrx6gHVgc0nzE+WQbsBw4wrt4DfDcWn/YCTieWRrf3sUHAC/P\n9O2S80l5nEc0pOsA3gqsCkL832vPBWG3Wnuv3sVvB74AnIa1wwbnk75w70iMiY06n/xx3LcwMSTk\neZzzyW/GqG8EiPNJNXNNgKOcT67JlAkgCgJyJ9YXljmf/HyiQoRnnuJ88v3w/2isL6/C+lVbQ3oJ\n10ecT9S7+Ejnk+sy9Xk58Fdgf+eTbHseDfyksc0yz9f+9wB7ZuamI4C/AYXGtm58NqATcBPVO5PH\nYmwMFYBN2Nzy1nDv/Zl+ki33U4AbqQuwUBdia3X9mXdxGvI9rFZO7+JHAfs4n/wkk/4hzD1cezyw\nBniOd/FxgcZHQ1+dDzzFu/i3wCHOJzdMts6NaBlzV9UzAUTkVcCDwDwR+SzWKGe0iMwewCnhdxE4\nJjCOfxoj/QGwfRIAeJZ38W3AALAnpkVfljE7PZJ6g7zHu/gibDKt4dnAs72LPwwcBazJdNYCZmq9\nHtNU+jDB4hHexZ/FrAlRJr/lNaHBu/hu55OvhnLuAbwqXP8A8D/AbcBguHYw1jn2C/k8Inz3hryf\n5F38eOAbwFnexV8FXu5dfLHzyT0h7Znh+1+Be4CveBcf5Hxy+xjvcUx4F88HOp1P1nozU9/pfLIa\nk3r/GpI1bR/v4j5gM/AcYAsmoN3iXXwNprHUtJb3eBd/AWubpcDiIKDsA9yN9bEF2OT3HmywXOJd\nfF9mEO8NPAO4xPlkGGv/+SH/fQmMPeCQ8PkocGRgpM8J+bQBK5xP7g1pI+DdwK3h/382qepxoezV\nUO69Ql6rgEuAZ3oX3421340hzxWYgFDD6cAnqAuuz232TmcRX8PqthC4A1ARuRljEjP1H3h57Yd3\n8cuAA8MYO8n55FuBUe8HPA6bqIdC8qyG/LzwWRX+nxXSfjL87wGehr3jVwCXAddknq3RvxPoC31k\nXASGeJrzySWZy7Xx/VTv4luBE5xPLm14tKasfAoTkhZgc+bJwDWBMdXKdBJwLTYvrCFjBfEuXo4x\nokcDFeeTv4d5aCHwHyHNH0Oao8L/4zEmRYNicTDwknD9fOB53sX3YX0XbC71Id1l3sWHEfqid/F/\nAx3exZ/BrFkHA23exTeGslSBA7G5849BODkVuAV4l3fxJ4ByRtF6HtDI3KuYoLMMm78j55N+xsaz\nsfmgE/g9cF/m3pnexXeEdjsTEw5r7dYoxKt38SLqfW0PbN4sYILPdaEfHAksCu+jppA9BlMIO0N5\nB8IzUBcaav4qzwJ+mqFfm6OnzdxFdUzlZqdBRC5T1YeYNMPk9zGCxIW9GMFeQJUdpaosapM9GNMY\nGiPd3cBvsMbaXhzqUm0Z0zgF6ArXRoAK1sHGciRKQhkbJc1NmDUBYGND+bP1zKZfNAaN6eBc4JVA\n3HD9x9gkcqXzye+mkqE3E/HjgPOBfwd+iQ2qx06zjBH2vhV799NFNeT1f5gAMaZpcRzcSRAQM7gE\neFnm/yj1/hJRFxiGgLEYRCcmmABsY0fN93bqfQTqwmYKvBd4PnBo7eZO1NwBEJEI+CrQrqovFJGT\ngEWqevE4zzQd3zVkLGPn81Dh6CpMmB0P2Tb5CcYMzwJ7P97FTwBObPLc37G22L/h+l8wZvAE4EtB\nk+3GltMc1sdfCfwRm5RvwoT6YaCZ6fhPzic/C3V9GvXlnIsxQQPgs8AbMEbUuLx0AybcZNdw30+d\nAdyNCS+fxfrOS5uUYTz8FZuPxlySyqAKnIf5FXROgUYfUDOJN6vjbcC9mFAdAx+mbuXbGzgcExRq\neAAT1AdgB2vk14C7gLPDsxo+zcp6LtZP7qauLF0IHOx88qvQL7+H+ZTchgktOyD0r+djAtam8HwR\neHtIch3WbrV55IPAfzcpC8A5mNLx2mz+Y6SdEC1j7iJyKsYg5gFfApZgmnE78AZVHRjn2aaD/1Gl\ntvsjobAwipZHdYkHgKpSLqNDAlERKQJUoVxBRztFeqIgARWQ4rDqNkCraFmQqFtkHiAC+mCa3l9G\nRxdF0bIICiWkrSA2WStUB1W3diBdJZEOhWoVLa+rVu8rIFFJaBOgXaRLQARhRHUISHtEFkZCBOiA\n6taKMrqsEK1MlXIkFMTKp6lS2aa6EaAkpCa5iAIUodgh0i2CtiEdERREYFuqa8voUKdITwEpClBB\nB1WpKFSHVPs6RDpcobC0TUQjKA+p3i9Q3rdYXFIS0U6R0QOKxTUAcaF4c1FkAKbembyLTyFoBTXc\nUynvOaqUtqTpIzen6fIUZXW1mg6rVtpFFhegtwpaVh0pq45WQbepblK00CnR/lVVUZAyOpwqqqBq\n7aujqkMKjKoOD6r2R9BRRrsAKSFtgCjo5jRdW0ErtTL1SrSgJ5KFABWlPIoOEvqUQNQp0iNIJJl+\nlqLV1AQ5wnVJlcooVoZtqW5JId2zUNgvgkIhWMJSSAtCqazbhUo1SUUrQ6p9nSK9JaSjgo60i/Rs\nSdM1AlGbSAdAqqRldLiEtPdEsiT0k2pSrd5SgGJPJAsAOiXq/cnQ4E5z2hGR52IM6GLgVFV9kYgc\nBpysqv/bkPY0zAQOsEJVH9OYX5g8P03QMqeImrBdDp8slHo7vpegKU8Rv8LmryzuBVYCX8QsKY3w\nmIYKJtxlJ9dLMQEjK6zfxUMFi2a4lbqFbjxcxI6KSishmPURbEy0fClvmvgGxnjPnsaz69jRMgvG\nu7IC2lpMW69h+7IfzZXLMsbz9g1phzALUS2P8dr8/aE82/vWLsHct2coshD4CDBvppL9VxYv2Qgw\noHrfhmr1Jq1PxjI/ivYRKKoxiW1qE/RiQQopqhvT9NaKUpkfiYuQoghtBWhXkBHVLf2q65dG0SN7\nJDpIgRHV9Vs1vUtB+9N0XQq6KIr2ieyZdFOa3qeQriwUnyVinXxUdatCWlWGR9C+InSUROYDlFWH\n+lXv70AWdEayZ0jf56vV6wCK1gFkQRStLCLtijJqjInUmBsKDKlurcBoWbUyojrUJtK1IIpWppau\nOqi6RaBQEulRkBL0RiJtqipr0+rG/lQHOkQWKbSP2rVKValsSqtdfaqlUdXCMe3tHFlq29YbRSMv\n7uo+ul1kM7ZufJXzyehY7Ra0kF7gyJ8ODT33jkp5n01p2vG1wYGuXpG0JMKSKNqkkC6JCvSIVCMh\n3ZbqRoG0U6QjEuiSqKtd6IgQyqrrBlXXFYWoDWkTjONGiJSg1CYm5PRG0R4FiAS0JLJRoVJVRqro\nSJfIHt0S7Y0x1RRQgcKmtHrDKGzrFtlDTCNOQ+9Py8rgoKbbrSkK0ibSLRDV8lDQTpHFERTbROa1\nIUsAymjfmmr1zwBBkKRdpLtoGr0ELiM9UeSiMEFuStM7h1S3LosKjyiJMaoR1a0plNtE5hWCNSCk\n69u/WDwRkAJ0jMLmEdVNZdXhs7duOWys9pktiMiPgYHA3Kc9vo/t6LgKYK9C8eAStAmICCImRJX7\nVNdEICWhXRBSNeGoN4oWzo8iR7DebUnTRICy6qBCdUEUrSiJtAO6Oa2u3pbqpgVRtKRNpKMEHd1R\ntBRAoLolTe8sCe0LpbCvCFGqWr6nWvm9KipifaED6WoT6RRgWLV/bVpdvX+xdHinKQn0p+naQdW+\nlcXi0W3IvEgoVlWHy7BVjca9KVo1mtYbIpAIir1RtEwgKkAxgjZBGNB0bb/qmg6R3vZgHazA4Kjq\nIJCOqA4WRKIlUXRAASlGQhUTAKqLo+jgCIoCkRg1WVKIuuZJREmkY2Wh0CPWI+8BfobNQzUBcTOm\nfT4BW38Gs2B8a1j1eF+tHAOkD1Sr29ZXq7dFSFdJWJACFWVkSNPNRZF57cjyFNUqlEdVt6XAljRd\nPaI6tKRQODpVbU8hTcP4rKiOjsKggqaqqqAjqoOb03QdQiFVlgEUhbbIlBnt13TTqJplbFNaXb9n\noRAvjKIVgKhS6VfdkJXweiNZUkI6MhqiKKQVGLJ3hQSFq7olTdcBWkFHFHRpVNinU2RBBIV2kflq\n84kMqK4J7aogmqpWN6Vp0inSu0ehcOiI6tYOkQW+Wr1Wbb7rDUVKh1S3RlDco1B4VFAc0qRava6s\nOtJpymLUHUWLLujvm4xQ1xSzwdw/ipnJ/rtFkn0W92KSczNcATw98z9r8mqGrJmohn6so5+HbfuZ\nCv4APLnhWtbck8VGgqMUNqAWNknTDJ/H1qoapc1p49Zyuedj/X1PT9HoNyMjh/1y6fIL4mKx5gD4\nGeeTDd7F/wV8zvlka/ZZbw6OewBy4oZ1r14cRWt6JOp7ekfH9S/s6r5jkkW4FlubquH72MSyjR3b\n+v2Y2X+8d7UaM+HVEIVPbX1sonXUKnUz+DbM1DkestakezEzYq1tr8GWJrJ5zgSd2EQMsJUghOws\ns7yInIWtyXYAl2N1fxphzV1VN4zzbFPmfvb8BQPYJCv3VyvXBElLU9CFUbRPAdoUGIVhhbQDFohI\nSVV5ME3v6td086KosFcJugQKxbDfvqJa2aLp/V0SLeoUWTyiOtApsqgvTdeoWdK2jKgOzY+iFYVg\nAdiSpuuGVPv3KhYOqygjCtIu0hsYhAyp9okJfL2gKNCX6hqBQrfIkipUClC6r1q5oapUuiJZIEC3\nRItLQidmgbIKZr5H0aERY9qUldEICvOjaM8gVOpIYI5tQicgEbJ9u+eo6vCQal8EvaNot5pla2RY\ndaCmHCiwsZqWyqhsTdPONpHRElI5qFTsPLzUlgqwZ6GwTYCiyOKKajWF6IFq9W9b0nTw5nL5CcOq\nsiGtojDSJjKwKIraFkVRau+aISDtEOksBNllRLU/VcptIu0Fe48dXSLzRUhVGenT9L4IkcgYqrSL\ntBeEYrBaSgQyL4r2DPfpENkGjKZQLauOtIn0FKCtDIOparVqAnE3wICmm7olWhwUue3MLYVqX5qu\n1YzGXYD2SCgoqOp24X1hyQS5YhVGR1W3dYgs3FCt3lGF8rDqQAUtd0o0r0Okx7qHAiI9IkvaRHpT\nNO1L9YH11ep9ywqFfbpEFgNSQYcrymhR6CgiHVhf3LQlTdctK0T7lZWhKlrplWiPCoyoanrRwLZp\nW+Za5lAnJg5+ELhcVa8V2T7vzcSb9pPACRgzWwJ8BfOQ3+Bt25Fga577YhpRVlh7IKQfBR4I62Y1\nhyCAX2DmnJWYdPp24OOYg8wQ5mSxDGMozQSB7ch4+XZgJuobsDWYB7CJ8ChMGq6Z134d6C7E1oUW\nYsyrgjnT7VPPne9jnsp3OJ+s8S6+hTpzPye8lzOAP2Pt+diwDrQKWwscwN7/WmzN8EmYyfE4IHpE\nqXTuhYsWbwM4ZcP6g45dv/bfI+D0nt7b3zVv/m3exb/AGEtbqONBmFD2od+NDB/52k0bTwkmdvn1\nsj2+vU+x2BfKdil14S2LjZhX86+8efrXnLH+jq0le8y8WME82QFqHq3nhzKsor4e/kuMyXzX+eTG\ncF8wZ7zjMAe8SnhXYKbRx2Pm2meGdrkcW6v8O2Y2W4Ex5f0xM+tvnU9uCOV9EPPW/1Hweq6ZfG/E\nnLQeBww7n1zmXfzjcO9lWF+7Betnm0I9l2LC0SWhPlBfm6utIX8UW298BnBlqO9Z4V7WSWhWoarn\nAuc2WOYOCZr7iZgwPSW8tqf3i+HnCMFhdBx8namvJbdjbSgh/3sy97ZS94/I4q/U/XXWhOdSzBei\n0f+ihl5MiKz54kwXf8H6z3iIqDt3DTM583gFKPalafGWcrl3FI0uGxraf3OalrZp2n59ebSngKQF\nuLcoaISkApUI9Int7T95ZkfH/UUkPaqt7dtFkeMmpLYjriY48GVwI2YZHA81/yp4aB3bqStN2+cm\nbIxtnmL5xkIBmxMijIfcTH1ZooZWCe5g/XTf8Pu3WH+bEVq55v4mzMnkL5hzSR+mgXUCZ0xnzb2G\nsK0jbrI9oua9XMLWOkrAOzANd+1YW6G8ix+Hbbna2ux+Q9qXYE4Ra7D1xk7Ma/kxGBPf1IxOcMAp\nZ83agemc6XxybuZaAeiubcsJ1xZiDb3F+eTuMcq1KiNUnOB8cmWTcl/jfHJn5tpTMWeRz9XoOJ9s\n9i4+Bih+aVv/H1/Y1X32Vwa2PedXI8MHvaCz6zdLosLWEzs7b8OYigBPvb1cfvrdlcq+V4+OLLi9\nUh796IJFn/1I/9YPfnjBoiPCuxp2PimHtjkY85b/HTZIr86+l+CQtzjsFni388n7xm0QtjP39wLz\nQ/nfCnzD+eSBhnQ15vtJ55NNmev/DHwXE5KOBX7pfPLgRHTHKQvA95xP/uZtn2zB+aQps/Iufjpm\n5QlKKb3N+qF38VkN/eQE55MrQx86G/g/55Prp1PmmaDFlrm3saOTWI25XYMJNNntWh8C3pn5/1t2\njDNwB/X17mGMyf4B21VQc+D6KSaIXIL1xUbL3iD1tfwsvo8Jymc1uTcRsgrHeDt7wJSZfah7yv+c\n5k5uHnPsy+bdiOz7uBBz0hqhbv2p4Vbqa8Q1jCVIfQSzjh5BXahuxG8xBeIHWF0vwBSkGvqAH2Le\n7A9QZ/IfD+U7ibEZ/2fDvQ4e6rC7AROea33iL5hS83ysrSPq9a85EGfxDcwy9U1Mqfg6JkDUnDlr\nFtaaEnIBxnOOwpZOjqTeHjXB8UpMOQV7zzczccCny7F3A+xia+7TKsQEzH0qCNr5BRPsk93t4V3c\nO8FWkGbPlIBisEyMlWbVdwcHHvjO0OAnVeGG8mj7ce0dfRFU0mDS+vPoyJLlUWFDFR04pbPrpv/s\nnfd/zidfHCvPCcrUhu35HZkw8fTy7wW2jbfnfYb5vwRbWrhzNuIFjEGzC9uPXJ0wcYsgIvthTG6r\nqh4rIn/GNLD9gK+r6pjtP9749hat8dEYUz8v2zfD9iQBXk+dudesaFcAf8i2q3fx4diI7w5nAAAg\nAElEQVTkeR22H/wLPsQQ8C7ey/nkfm9hf//ifFIJzzwdm6wXMPYS2XnOJ0PeYiykmIXsxUBbJo7B\n6aFMrwzXVmKMZE/M8nZOrRwhbRUTAld5F/8b8MVwfwG29LYSY8rtmNJ0MSa0/Ft45jiMeX8ps5d/\nHrbj5kfOJ3/1Ln43xrSGgddkLHpgjPWH2G6iFNueeRS2FPVVzOJapb69dxNm+ejGtnVt9S4+BHhR\ntjkxy8dxzicfD2WqWT/PwYTSq5xPLg/3DsasIeuAv2bqsYo6I/5KqP9DGF1oj1eGvx91PukP7/1V\noW73AW/E2vUnoY6/Du23nBAbA9u7/4tMvvsA9zufVIMicCjwoPPJp8O2uDdly+JdfCCwV8bD/tOY\nkPFrLL7FF51P7g9pa/1xVWiXH2JCyFuB32TyoFmdp4JZY+4h3vQFmHbya1W9ZJy0LWPuOWaGsCdz\nBNMeHv3ToaFDb6mU9xVzAIoikIVRNPzSru7LI5FzsK0rd84Wc86xayBjmVuAbVN6FcZEDwS+MV3m\nXoN38ZOdT/4wxr0O55PhsOxVCzZ1xVjCjXfxCqA9E9dhPLrLQp1WY7EMDge+jDG5LcC6ZnSCVajU\n2O+9i892Pjmn4VpHdt98qEcF6HAWmKhZuVYBnw9LcU/CttNlBZkCJhxkLYNFLBDOVRk6ta1k/xqY\neycW4Okh/ichz2gqilGg8WQsKNU5je8lWLIOcz65xrv4KOCmiWIIeBe/AWPG60K7Py3U9RdN0r4W\nY8Q/yVybh/nMRJh14R4X4pFMFd7Fp2KWir84n/y4Vuex6tDYj8dKG5TQrzifbAn/t7+b0PbrnU8u\nmE6Za5hN5v4KYIuq/khEvqmqLx4nbc7cd0F4i4JXxtbJT6UeVOb9mLY9pid9jrkJEfmOqv6ziHxr\nPLN8wzOttMz9KzYJXt2K/HZVBA1xcyssToHhLsguTc0VeIt8V5lIYJhB/gdgS2cPWRKeLQSBqDSW\n4DdZzGY88r0wsx2YeWc8PDDB/RwPA5xPkszfb4dPjhxZNHWYbVhzn0qwk3HhfHJRq/LaldFKRhyW\ni+YcYweYKQOcRP53Tpyq5TRHqId3njZmk7nfjzH462lyQE3D4G8Mz1hLcyVjR5drFVYw+8JFTiOn\nMVvoVNWWHDQxHsS287wfOFJE3gX833jhpbO7YcbJMx/fOY3dmc7OoDH98W0b/Fr/wRwvLsI8HF82\nzTwum63y5TRyGrs7jZ1JZ3cte07jH4/GXKrLTGjMmuautvVttkIh5siRI0eOHDnGQMvOc58ljGvW\ny2nkNP7BaexMOrOBudIOOY1di8bOorNL09gl9rnnyJEjR44cOVqHXV1zz5EjR44cOXJMETlzz5Ej\nR44cOeYYZnMr3KQhIkcDbwMSVX17uPY26gfCvB4L4/hhbM/8Rar6q8Y0Ook1hnDu/IlYVCpV1ZeK\nyNVYCNH7VPV/RWRFI60p1udVWFjG1cBnVPXGydRnijROxw5kWAScq6rXt7oeGVqTjjY4hTxPxeI7\nz8POUH4LFi5ym6q+vRU0ReR44H1YXPBvYFtXnoqFtXxDSDZTGk/BDnkpYtH6traqHiHk65nAfLXA\nMadNVP7GNDrOmQ47C/n43rXHd6DX0jE+V8Z3oLN7jvHZduWfgsv/SuAj4XcbcEn4fQZ2AM27sUMD\nIszJ4CFppkjvzcCzwu9fYSEnXxn+70BrGnX5F+A7wBeAZZOpzwze25HAu2ejHhkarwBOCb+/2eJ2\nX4gN/u9isbTf0Sqa2KlwlwP/D4th/e1w/aSQf8vqhUXwe90s1eM74XvC8jemaWVbzfD95ON7eu9t\n1sd3q/rpGPnOifEd8titxvhO19xDuMrGUJWvbvi/GDvlB0w63it8ElVNw3GyzdJMipaqrseO0Pxk\nuHZCyPebIvKjJrSmXB9V/aqIHI6dC/+RSdRnyjSwqFNvon7s6LTrMQGmEm1wqjgL+AxwfSjnx0J9\nW0Hzd6r6GxFZDnyM+vnOq7EY4rSARg2nAa8BvjAL9ahhMuVvlmanIR/fu+X4htkb43NlfMNuNsZ3\nOnNX1RuBh8SZFpGVmb8bsXPKwcJb/o0Q8U5E+sZJM1laxwBXqWoa0tVO89qMRdxqpDXl+gSsx460\nnEx9pkRDREqYueYTqprMtB4TYNxog9OB2Gz0QeByVb02c6v2zmZMs+F9tFM/FzobMnXG9RKRvbHT\n0rKn9LWsHk0wmfI3DQs728jH9245vqHF/XQujW/YPcf4rrLmfhB2HOCjROS1qnqhiFwrIudjjXYB\ncBfWWSrAF1V1tEmayeLVBGlYRBYC52PH721SVS8iX8rSmkZ9Xoud974YOGeMsu5Qn6nSAM4DDgLe\nEMJ4XtHqemTwPeDTInIS8KMZ5JPFG7GzoeeLyAHAE7HztIvY8Z43zJSmiLwAO7JyAXYM44omIVNb\nUa/XYNEYEZGvtKoekwz52pj3uGFhHw7k43uXH9/Q+jE+l8Y37IZjPN/nniNHjhw5cswx5FvhcuTI\nkSNHjjmGnLnnyJEjR44ccww5c8+RI0eOHDnmGHLmniNHjhw5cswx5My9xRCRV4nIZSJyvoh8WUT2\nE5EVIvKWJmlXicihs1CGl4jIRSJyVObajDfCPhwQke9MIo1kfh8oIu8UkZUi8pEJnjtKRF7finLm\n+MdAPr5bi3x8zx52ia1wcxCfU9XLwjaHj2PbcpyIPBLbEnQf8FXgGGCZiFyKRfA6Ets3eQYWkrAX\n2+Zyq6peJCIfxtpsNfApbAtFO1AA3qyqVRHpwcJf3gy0i8gfsC0UF4XB0Af8TVU/LyLXYxGe9sS2\nczwhlH17GEsRORPbw9uLhZD8VKjPu4BPAI8ATgE6gFWhTscBdwOpqr4/k9ezgJdg+1LPA44PNOeF\nfE/N1hm4AzhE5P+3d+5xdlXV4f+uc+88M3mTQHI4JgZBQKTysIhQhFoE5PHDqmixWioWtVRbWt9B\nDfKQViqCCggKiAhClVIFUcGCVauivBseESLkshMS8pqZZF733rN+f6x9mDOXO6/MnZCM+/v5zGfu\nOWef/Thn773WXnvvdeRsf/0cQIBuVT1bRJYD38K28azwyXwA257SBrxORM735Xsf8H3My9crgMtV\n9T7fKV8x8usMBIYQ2ndo3zs8YeQ+iajqBszXdMZ8rOLfqKrLgV8Al6nqL/z1MhBjnQDAzar6ceBN\nIrIvMKCqZ6nqxZgHrsXAZqDD34eqbgHuwfbargYeVdULgbcDV6vqmdj+U4B1qvpJn6efYo375Cyz\nvrM6AvOlXAb2AT4C3AQ8oKpPAO9R1fdhnVWmJf9IVc8Fakct78e8e/2zqq4F/sHHvQ7zoz2kzP65\nPKqq52Heodp8XpeISDOwWlUvVNUVuTT29HED/F5VlwKPAYdgneQlwKcY3B/aJ+bhKhAYF6F9h/a9\nIxNG7pOI1+wHsmPvKnElcLqI/AmDHpUATlHVk0Tks0C7P5d9EED8Xz58BPxSVS9lZDpzcdQ6Ncg8\nW/X73wPYSCGfxnJVXZYr0yuBbsxxRB71adTmuzZMnt6auJfVuTe7JwJuV9Xv58JnZcuT1jmXj6PA\n0A55EzayWFt7UyAwEqF9h/a9IxOE++TwARE5GjNBLctOishRmIlrDnAnptF+VESuBtaIyMcwDfdn\ntRGq6nIRafOmuxKmuV/hj2cBH1LVvhHydCNwsffqNKavR/k0UxH5IqZVXwB8DngrsFREXgtcLyJf\nwzqsc4HXjxDl14ErRWQT5o/7ehG5EugFbh/mnrKInOvDf0XsC03NqvqPI4TPOo69ROQCYDfM9/QA\n5ut6T18WsBHRqhEfRCAwlNC+6xPa9w5E8FAXmFKIyNuBTap6V51r31XVt+WOO7Avlf3RLroJBHYm\nQvseO0G4B6YUIhJhX8+6cwxh9wKqqvrU5OcsEAhMlNC+x04Q7oFAIBAITDHCavlAIBAIBKYYQbgH\nAoFAIDDFCMI9EAgEAoEpRhDugUAgEAhMMYJwDwQCgUBgihGEeyAQCAQCU4wg3AOBQCAQmGIE4R4I\nBAKBwBQjCPdAIBAIBKYYQbgHAoFAIDDFCMI9EAgEAoEpRhDugUAgEAhMMYJwDwQCgUBgihGEeyAQ\nCAQCU4yGCncRmSYivxORE0TkVBG5SkSuE5Fpo9x3ZSPz4eKkxcVJcyPjDAQaiYuTw17qPIyXHah9\ni4uT1kbG2Sh83lpe6nwEXlpcnBzi4qT4Uuah0SP3jwM3+99vUdW/A24C/nKU+xY2OB9vAU5scJwN\nwcXJ3i5OkgbFdXQj4pkKuDjp2N4KnYuTv5jA7Tvju9tR2vci4BMNjrNR7AP8cyMicnGy8KVWYlyc\nRC5OFr+UeXipmODzPw6Y3sj8jJeGCXcRORp4FFjnT6n/vwrYvU74U0XkNhG5jVEav69gu4wjO0Wg\n2cXJ7i5ODhrLDS5O5rs4aXNxMt3FidS5vouLk0Y8ryOBP2lAPACHAbg4mePipKlBcTYMP4qZt52S\ney/+eWwLLk5e5rXtY0cIs6zm1OE1108a4d5ZY1E+XJy8ysXJu1yc7OfiZJGvd/Nz11/v4mS7K66T\n2b63gUL2w8XJzAbHPVGagUaN3M8AXtWguF7Axck8FydnjzH4TOC0EeI6oSGZGgXfHv5uAveLi5OZ\nI9UXFydH1wjzMzBlLbt+Qj3ZMAKF2hNelh3v38Ehda63NmrQ1kizwZHANGBfoBfo9+dfBjxbG1hV\nbwBuAPAdwEh8BsDFyTnAwZhGdA8wL3altV7ozvLnu7CHKsCewIHAfb5j/knsSv/r4zoT+GbsSlt8\nGn+PdV6zgIUuTj4Xu1Lqwy7zYX7s4uTe2JWqY3kgLk7agOmxK63LnRbgYBcni2JX+qqLkwWxK62p\nuS8B1sWu1J+rTK8F3hy70jIfZrH/P9/n/T9dnBwO3Bm70gp/rRC7UjX7n4u/GfgUcC2wKitnI3Fx\nEmP14EwXJ7dgwmA58HKgDzg8dqWbsnwCBwAp8ETsSltdnLwD+Anwj8C5Wf5dnLwcOCl2pUtqkoyA\ngo8LII1dSRkGFycfBH4GnAKci3Wih/hrs4BfAs8BJ8audEsu3tp4xKf9CuBAFye/BPYH1gDvjF1p\nmQ/zYV+eX9eJI8q9g2nAHv5/CzDXh7nI19X5wGtcnKz2eVs2XBkbzJFMUvt2cbIXcGqubjfFrlQe\n4Rb14VqBs4BlXol/OLvPK7uVrA64OPkY8DXgr4EfA8/HrtTprx0EzAAGYlf65YhPYXTEx/nnsSv9\n9zDlLQDUtMkIeG3sSr/J5QmsTs8D2rC6MD92pR+Plgl/z5GxK/2HPz4EOM4/4xZq+n5fR1uBvtiV\n1MXJQbEr3Zcrz3uA3wF7xK70g9ytB7s4uR0o5t9Z7v7seDqwMHalJ/zxgcCDQKH2Xdfe62kDFoxW\n7lwcczF5UAA2YH3LWf7a+fl6kkv/MOAhrH/KyPchBwN3uDhJsee3f5ZPFyf7AYtjV8rX9Xp9RoT1\n5Q44zsXJkT6NL/i6OhM4zMXJPcB+sSs9MNYy19Iw4a6qSwFE5DRgPTBDRC7HXsqZDUpmIXC8/13A\nHsKtwMnDhN8DXhCyAG9ycbIC6AbmYR3WbS5OFvnrr2TwhXzGxcm1DK1QxwDHuDj5AvaCVucFKXAs\nVmEXYkrGwcCeLk6uwDR6AXb1cc3LlAYXJ0/HrnStz+cC4D3+/OeBTwIrgK3+3D7A6zDTJMDe/v90\nX6bDXJwcCnwbONvFyfXAu1ycXB+70lM+7Kf8/9OAp4FrXZzsHbvS48M8x2HxgrAtdqU1Lk7eCPw+\ndqVVwN8BWcXMzLZvrbn3OGCTf26dWMV+3MXJfZjGnGnNn3ZxchXQhAm32V5BWQw8iQmdWcAuwKf9\nPd92cfJM7EoDPq1FPp3rYlfqxd7DDB92D7xg9+zt//4d2N/FyRrs3WeK0e6xK630YSOfZvbsPlTn\nMR3lw1Wy+31cy4AbgaNdnDzt8/SgD7sQWJuL4wzgiwxa27b36P16TNGaDfweUBF5FKvXh04w7lOz\nHy5O3g3s4eLkIuCE2JW+42zu8hXAn2Idda8Pnh9lnej/lvnjpcBm4Ev+uB14I9ZG/hq4DRNW2b1Z\n+k8CXb6OjIgXiH8du9K3cqez/uIIFyePAkfXXAdfR12cfBkYwJ7pOsyU+xtnVsosT28G7gd2wxTG\nF6wgLk4WYAroazBF5hH/rOZgCj8uTn6O1aOD/fFRDCqx+YHFK4F3+vOXAie6OPkDVncBlgAlYC8f\nZn8G6+KngCYXJ5cDf4YpgJGLk4d9XqrY+zvWxcmvY1f6EXASpux/0sXJJT7/3T6+E4Fa4V71ce4G\nbMGUgk6G53j/XNuAX/i8Zyx1cfKUfy9LMeUwe2+11ln1ikJW1xYA78tdv8/XgwOxPiirFwAHAT90\ncdIORF45z+LPlIZMNh0H/DB3fan/v83CXVSHHdxsN0TkNlV9kXnHd34XYxpXNkIS/1dhqFaVpx17\nsQAbGewMalkJ/Bz4m3x2GHyRA1ilEkyAgI1YypiwHW4hUQlTEmrNkZuxxqGYApTPf1bOPN00dt7m\nPMx8XZuvH2GC7+7YlX42nghdnByPKTqXYB3Kf2MC+IBtzGNmdVHs2U+U/8IEwnu34d6n8ApijhvI\nCSMsj5lCGDFY73oYvt61YYoJWEeV73h+j1mcMjIFPAU+hymy+2UXt+PIHQARiYDrgBZVfbuIHA/M\nUdVaAZa/p277zshZxi7FLBx5focXTCNwI/BX/vePMMGwFOz5eGX3mDr3PYr1FYtrzv8Wm244FPh6\n7Eqpi5MOTIjuDtwF/C1m3XkT8BjWCfdRv579JnalO3xZ/4LB6ZzrMUUD4ArgA8CveLGy9BAmXPN9\nwQUMKukrMUX1q1j/8leMjweB5xnbOpAUuBD4CKbYjZV8X1avjE9gz3xfIAYuwgYw/dj72Z9BZRxg\nNXBDZnn1lpqlWF1YAXzW36tYu6yX1/P9Pc8wOFi6Etg3dqW7fL28BRucPIEpQEPw9eutwKsxWXMV\n1h98xAd50Jc760cuZPj1IudgysMZ+fiHCTsqDRPuInIypi3NAL6BjaKOwkxAH1TVrSPcW7fxv6qp\nuVQQirOjaLfcRIcKSBXKA6p9BSSKoOClfbmKlltFpgmIgBSQYp/qFoDUhLJME5mZSY8NafpsGR2Y\nHUXzCxAVkZai0ASoQtqr2tmCTCuKtCharSrldWl1VQGJmoRmAZpF2iKvcQ0ovSmaTpdolgiRAD2p\ndpbRgfmFwqKqUo6gUBAKCmlVqWxV3QhQhDQSEFBBKEKxVaRdgCaR1qIXIltU11bQnlaRjiJSjEDL\naK8qVYVKn2pXi0jrrCja1d9bFigVoDIriuYVRZgu0v/yYnGtgO7T1PxIm0gXjL8yOZv/HbKu4aGB\ngcW9qk3dmr6yM03np0CpUkn7odIMcwoiHamqlqG/rFqugm7VdLNC1C7RoqrVyagKA2rvQRXVVNEy\n9CloWbWvR3VLk9BcRFoBikILEKmi69PUVdCyerPiDInmdEQyVxUqMNCv2qNWRwCiNpEZ1Gjtfarp\nAFoF8PWJFKoDqn2CsEXTzlRJFxQKiyIoREIR0BSqRaS5gmbmPQVIlUqfanerSEdRaKkq/a0iHZvT\n9LlIKDT7cihUB9DeJqR1eiRz/L3VUrX6WEEoTJdotgBtIjNu6+3pGM/7mggichLwQeBbwMmqeoqI\nvBo4QVU/XxP2VAaVoIWqemBtfL7zvAw/yhxPVhhUrAf833Ccy6A1ZzzczeCoNeNpTNB8naGjt4w1\nDCpmPZggzLgRG4Xn53xXYdMao/E4gxa6kbgGUzomA2FwLUGZxijejeBmTLn67Dbcuw4biOS5mqEK\n2vOYtScjG3ykDH2/GVXMwrgH1p/0YApqZrFdiSlq9Tjfh3uhbu0Qwv2FCEVmY1rXjIlq9tfO3eV5\ngK2prno+rT4G1ksKyKwoehlQVNAB1S0CUavIXOzhpxvSdEVFqcyIZGEBaQKKRaFNgX7VTVtUn58b\nRftOl2hPgD7VdZ2aPqmKdmu6TkFnRdGiAjQr6MY0fQZIFxeLxwg0eQHTrVCtKn39aFcBWptFZipI\nVbWnU7XUJjK7XWSBz2fX6mr1QQWKXpOcHUUvL4i0qEIZlRRE8RINlV7VzQNQrqpW+lR7mkSmzY6i\nxQpUVCs9qp0ChaJIB0CzTYc0q6psUV3Xo9odwfw+1dYK0KtpX79SXVetdnRr2tSj2rRvU5Mc0NTc\n0xFFA++d1nFwi8gmTFP+VWbWrocfhUwDDviv3p6TV5TLizakaet/9fa07hIVqkVB50bReoV0flSg\nI5IqSLo1TdeJkLaKtBUQaROZ1moKkvarrtuquiaCQrNIs4BE3lrTJNLcDK0iUpguskCs8aT9qpuB\ntAr9VeibJrJrh0S7y6AFIBWINqbVhwdga4fIrgKRgIKkimoVenpUN/gbVIAOkfZmsdeRWjzaKrKL\nQLEJmdkiMl9RKrB5dbX6vwrS7xXJJmg3hXBQgegQiX1dZENafbJXtWt+obB3wUaPDKh2VqDcIjIz\nsjrGpjR9qle1a0mxeKxAFEFrv7KxX3VjGe09p3Pz/sO9n8lCRG4Htnrhvs3t+6jW1l8B7F4o7lMU\nWswkJyIgKVS60nSNgDSZ0kaqVHpVuzqiaPYMkYV4AdOpaQmgrPSmUJ0VRQuz9rUpTVd1a7ppVhTt\n0oS0Nou0tIvMwyvvG9P0D03QPCuKFkWW18qaauV/sjoA0CZRexO0CdCHdq+rps8uKRb3a/WDhK40\nXdOr2rmoWDykRWSGQCE1xbVLobo5TZ9OoZJpjlm8RZGmaSLzI4gKSLFgz0C3qj7XnaZrWkQ6Wny7\nrqj2lKFHIR1Q3RJBYX6hsKeYZSfdnKZPK1TnRtEezSKFosVPM0KL0NEkErWLNB3Q1NwRWZ3+A7YG\npJnBEfUmzMp1KGZeBxvV3tSZpm9cWakcqmh1TbXau7aaroyE1iaYqUAZ+vtUNzfB9BaReb5/Kg/A\nVl+PXb9q74JC4TUCLQppn6oqaKpa9mXzijxpWbVnc5qujaDQEUWzAIrQUvADnC7V5/tVewE2pum6\nhYXCojlRlIAp4F1pmp/SYnoUzWs2i5n4OimparVi1hbxynsEpJ1pulpBB9A+VXTXQmGPVpHZERSb\nbRCQCkRbVFenqhWfhKZQ3ZCmz7SLzNy1UNivX7WzVWTW6mr1vhTSVpHpApFC2qPp5iJS3K1QeLXP\nk5Yqld+WMYVfrL+Y+6XurrEodXWZDOH+79h87ycapNnnqWcmzbgLyG9N+iZDze21bMbPkdQ5N5Lp\nZDh+zmCDyBguv+sxywZYg5pdJ0w9Lse2+e02zrwNy8pKuf2S7u4jFaIf9/Ue9JN5u162qFh83l/+\nauxKz7s4+QRwee0cl7MFjguA6Jh1a0+Li4Vnp0vUeXRr60MntLWvrE1rGO7H5qsy/hN7jp0MfXbn\nYWbLkXZNPM1QE2s2lZNN54w0wgOb6snM4NkagOHIm+PzaW/AFj7di80R99OY1dOtDJoWu/HWgO1l\nlheRs7E52VbgDux5/rnP06Gq+vwI99YV7mfPnPXCtMWzlcrv/DBIU2BOFCWRL28FelNUW5FZkZhy\ntL5aXdmtunFOFMXNts++0ATtmVVvY5qWponMnSYyrx+2tInM6UzT1YBu0XRjv9I7J4riJlNOdbOm\na7am2rmwUDh4q2o/ELWITFc/nug1JTpqho6sx+zUdI0ghQ6ReSlUBYqrqpWHKkp1mrcETYtkdhPS\n7uPJ5uFe6HX7VXszIVVGByKkMDOSBYJEimq/6pYUtBlpB8Rbh8RboPq2atrVKtLhr9On2rdFtbuK\nSkWRKkpZ0TKqz1fTji5N2yJIFxWLsn9TczUC5heirQDNyOwKpFXVaE1afaIr1d7l5YHXDCh0acq8\nKOoqiGyZH0Xtc6OCYgKwB6i2iUyLIBJEezXtSmGgBWmJhEKrSNs0iWYBWkX7NlTTZwRoEUkjkGaR\ntoLpIqZfITJDZKEXhJRha1W1rFCtwEAztBdEmsuqvQppxSyC7YD2qG5qF5ldEMmb4bWqWu22Opo9\ney1CcwSFnFKhbSIziyKtBWjy1uGtrSIz16fVp6pmeeupoJV2iTpavIU4S2SayNxmkY4U0u40Xft8\nWi3Niwq7t4tZ36rQX0HLTUhLwfcJPaqbNqfp2vmFwpKyal8KlemRzK8o/Smk12/dss3Tsg1bUCci\nggnFO1T1fpEXyjyR1bSXYB3IAqxTvx6YG7vSehcnb/ZhqlinWmRoh/scZqYqA2v8CtAP4VcfAz/F\n5lFehpnH/gWb8zsldqU+bJHFPGwx4IgdfW6VbwvWoWcCazXWER4IvB5bQPIqbKX/ImyxST8m3M/z\n+X83Q802t2AL/1b4nQGPYqakCJujmQv8Azb/14ytuF3mFaP/wwTBasxkmOXjp8AbgOKSYtMFl8ye\n0wNw8vp1rzps3XN/D/D+aR1PfnrmrCddnPyInGDxC/reAXzhnr6+g07fuP54/9D1G3Pmfm/3YjFb\nFPNt4F11Htd6YHnsSnf7hUOv9+cfwrYIPou9twq2kj0FmmJXqgBf8XlYxuAc2E+xhVLfi13pEX9d\nsMV4R2CKUNU/K7A50kOx+es3YYsT78DMqf+Hmc0W+HuW+PM/j13pYZ/fDdiuhdv9CufP+Hj/D1uk\n9VqgN3alH7o4ucNfO9Xn93FsLm69fyfzfFrXY4ogDL77d/v/F2HzecdgCuzdDJqZ/1Dn+U4Kqnoe\ncF6NZW5fP3I/FjPVj4sPdkz/mv/Zx9AVyvWoXeswFloYfJ69mAKWMZxS/QiDbf05bG2NYu9vr2HS\n6cAvmMTMsNvKbxi6sLMe+fnjAeqbhmvpA1orqrI+TZurqnJTb8+SrWnavFW1+Q+VSkcBSQtAUUQj\nSJuR6m4FSQ9rmX7Tca1trgC6e7H4A148VTEav8baWJ4HsfUL20oLg0p8O4PPYw6144wAAA5DSURB\nVAX2XhtBAVsEGGHP+TFePG+fHwxMlCUMlum/sYXCE6KRc+4fxkbKv8VeXhc2AmsDztyWOfcMv1px\n99iVHqpzrRl7wBVsruOjwBWxKz03QnwHAU+OstoyC3sKJnjWAO/HNP2LsQVjDwOb6m25cnEyDSjn\nzdpe6HwidqXP584VgPbcStFsBfpioDN2pboduIuTZTml4qjYle6uuf4O4L7YlZ7MnTsK2Ct2pa9l\n6cSutNmZt7Qitqr0M5dv6T7hJ329e7+lrf3n00S2vrV92iOYUImANzxeLr9pZaX88nsHBmasrFT6\nL58956u39PZc8O5pHQdh774vdqWKX+TySuBt2LazFJt/euG5+AV5c2JX+paLk0/FrnTB8G9jsOyY\ncJ7h8/9PwM2xK62uCSfYXNylsSttzJ1/C3ArpiQdDtwTu9KG0dIdIS8At3gFoAVbyVu3k/fv4NeY\nwBgAOmJX6qoTbsizyN5xrky3xq704LbkeSI02DJ3FkOV5t9iytG92LRQfjX8vwEfyx3/D6a8ZeTn\npXsxofwLbAFiFvaH2Lz3tzHl7T01WRpukez3sM59rHvD86QMruW4GVNYh+NSTOn/f/74x9RfCJjN\n1ftZyrqsYFAZuRJbpNXL4OrsjMcZnCPOGE6Rugizjr6GF1tKM/4HU9hvxdp97ZqKLmyR67GY8vRq\nf/6L2EDnOIYX/Jdjg6MWXqwEPefLcqQ/vhdbqPpWBt9BH1anbgNq5U1W5mxx5g3Ytt1s4V+mDD7i\n8/wVrA4dgCl3BzI4uMwswHcyuEhxuc/fkB1DDK0f1OZth5pz36ZMjCLcx4MfnV8xyj7ZnR4XJx25\nPfpjvaeI7UcddpTk4mTZLT1bn72lt+fLKfDAwEDrES2t3RFUUm+6+s1A/9wFhcK6qrLlpLa25R+a\nPuM/Y1e6ZhvL0YRtE+kfNfC2xd8BbB1pz/sE4z8FWyn91GT4CxgmzXZMgdou6QGIyBJsyqRTVY8Q\nkV9jHd0S4EZV/foI9w7bvp35Q9gf66wvzNdNZz4cBFvElwn3LmzR7p21e9JdnLwa6zwfwPaDX5W7\ntjB2pdUuTl4L3J/zm/AXmMCYzfCj+QtjV+rz2ymzPc7vAFpzfgz+FrMi/a0/9zJMkCwAjsop4gsx\nQVvB2uIyFyenx670DX99JrZNbA9MKDf5uK/DhNMZ/p4jMKvS1bm9/DMw73jfj13pfmeOam7y952e\ns+iBCb7bMKtkiu2gOQibf78O2xWQYsrF/ZjAehpTfqLYlbpcnOyN3z6Xi/MB4M9i74fCDfrhOAdT\nSn+V7dN35t9gCbZV78FcOZYxOJ31wiLBWkHn30e2gPDfY1fq9s/9vZgCsQrbmroJuB2IY1f6me9z\n5mGCuIjt3b8rF2+CbXOuujh5G7Y7ZX3sSl9xcTIH+HA+Ly5OXpGLexkm+A/HFJ0PA1fFruR8WPFW\n5GWYMnkbZq38F/xupfx09A4p3L2/6cuw0ck9qvrtEcI2TLgHJoYzxyAD2M6Hg27t6dn/mWolwRZz\nSQTR7KjQ+8729h9HIudgI/OVIy28C+z85Cxzs4AvYD4SHsAEzHe2VbhnuDg5NHalXw1zrSU2h04t\n2BTUI8Bdwyk3fs9yS+xKT49WLj/1NhMTTG/Gpoa+gQmxTsyZ1ItWhvspmWJtvXdx8unYlc6tl//8\nMTbKa4ldqa5F03fwV8Su9JyLk9dh2+k0d72ACdlyzbkDY1f6bS6dAWyUnykdrYDWU6brxTkaPo1D\nMYF+bu1z8ZbV/byycQDw2EiDC3/P+7GpsrX+vR/p47yrTtgzgGdjV/ph7tx0THBG2Ej+6bwFczy4\nODkZsyTcm6VR+z5rwg+px8OF9YPQb8WutNkfv/BsXJx8BlMmLtuWPGdMpnB/N7BZVX8gIjep6jtG\nCBuE+w6IH1WVsf2bJzBoLjsfYKpbRwIvRkS+q6pvE5GbRzLL19zTSMvcacCjsSvd24j4dlRcnMwG\nNjfC4uQF7szYlRo1H73D4Kc/q6MpDBOIfw/My+h2mwLzClPTeC2ztUzmV2t2xzRsGH1P5OpRrgde\nAjJTkudW/xcI5Km7YLZmzr12nnebiV3p2kbFtSPTSEHsLRxTTrADDGf5aGD8T40equFp9jPo3nmb\nmUzh/iwm4DN3mkOoafw31ItARH7K8F6+GsVCJl+5CGmENCaLNlV942QnIuZD4nzgABH5JHDrSO6l\n87thRogztO+Qxs6czvZIY9vbt6pOyh82Z3UNtsLxXdsYx22Tlb+QRkhjZ09je6azs+Y9pPHHl8ZU\nKstE0pi0kbva1rfJcoUYCAQCgUBgGBr2PfdJYkSzXkgjpPFHnsb2TGcymCrvIaSxY6WxvdLZodPY\nIfa5BwKBQCAQaBw7+sg9EAgEAoHAOAnCPRAIBAKBKcZkboUbMyLyp5j7vZKqfsSf+xfMt28T9jWw\nBZhnrCpwjareXRtGxzDH4L87fyzmaUtV9a9E5F7MxeIzqvp5sc9JDklrnOU5DfMhvQr4qqo+Mpby\njDON92EfqZkDnKeqDza6HLm0xuxtcBxxnox5wZuBeQQ7C/u85BZV/Ugj0hSRI7HveS8HvoNtXTkK\nc2v5QR9somn8GfaBnCKwL+bVrCHl8C5flwIz1RzHnDpa/mvD6AjfdNhehPa9Y7dvn15D2/hUad8+\nnZ2zjU/2Uv5xLPlfDFzkfzcD3/a/z8Q+QPNpzGF/hC0yeFGYcab3T8Ax/vfdwNXA3/jjIWltQ1ne\nA3wXuAr7OMmo5ZnAczsA+PRklCOXxruBE/3vmxr83mdjjf97mC/tjzYqTezLd3cA12JfePoPf/54\nH3/DyoV9oOT9k1SO7/r/o+a/Nkwj39UEn09o39v23Ca9fTeqng4T75Ro3z6OnaqNb/eRu3dXWeuq\n8r01x3OB7PvQqzBnOLtjmn/qPydbL8yY0lLVddjXei71597o471JRH5QJ61xl0dVrxOR/YGPY19T\nGq08404Dcwv7YQY/O7rN5RiF8XgbHC9nA18FHvT5/KIvbyPS/Lmq/kxEdsW+OpWN/FZhHyuhAWlk\nnAqcDlw1CeXIGEv+64XZboT2vVO2b5i8Nj5V2jfsZG18uwt3VX2EF39uDxFZnDvcgH2/Hcy95cN4\nj3ci0jVCmLGmdTjwG1VNfbjsAxSbMI9btWmNuzyeddhXlsZSnnGlISJNmLnmS6pammg5RmFEb4Pb\nglhvdCFwh6ren7uUPbMJp1nzPFoY/PZ13mXqhMslIi/DvpbWnTvdsHLUYSz5r+sWdrIJ7XunbN/Q\n4Ho6ldo37JxtfEeZc98L+xzgq0TkDFW9UkTuF5FLsJd2GfAUVlkqwNdVdaBOmLHyXrw2LCKzgUuw\nzyJuVFUnIt/Ip7UN5TkD+77vXOCcYfI6pDzjTQP4V+x7zR/0bjzvanQ5ctwCfEVEjgd+MIF48nwI\n+x70TBF5BfZlqR6sTv4b8NBE0xSRv8Q+WTkL+wzjwjouUxtRrtMxb4yIyDcbVY4xunytjXtEt7Av\nBaF97/DtGxrfxqdS+4adsI2Hfe6BQCAQCEwxwla4QCAQCASmGEG4BwKBQCAwxQjCPRAIBAKBKUYQ\n7oFAIBAITDGCcG8wInKaiNwmIpeIyNUiskREForIWXXCLhOR/SYhD+8UkWtE5KDcuQlvhH0pEJHv\njiGM5H7vKSIfE5HFInLRKPcdJCIfaEQ+A38chPbdWEL7njx2iK1wU5ArVPU2v83hYmxbTiwi+2Bb\ngp4BrgMOB+aLyA2YB68DsH2TZ2IuCadj21weV9VrROQL2DtbBXwZ20LRAhSAf1LVqoh0YO4vHwVa\nROSX2BaKa3xj6AIeVtWviciDmIenBdh2jtf5vL/gxlJElmJ7eKdjLiS/7MvzSeBLwN7AiUArsMyX\n6Q3ASiBV1fNzcR0DvBPbl/qvwJE+zRk+3pPzZQZ+D+wrImf76+cAAnSr6tkishz4FraNZ4VP5gPY\n9pQ24HUicr4v3/uA72Nevl4BXK6q9/lO+YqRX2cgMITQvkP73uEJI/dJRFU3YL6mM+ZjFf9GVV0O\n/AK4TFV/4a+XgRjrBABuVtWPA28SkX2BAVU9S1UvxjxwLQY2Ax3+PlR1C3APttd2NfCoql4IvB24\nWlXPxPafAqxT1U/6PP0Ua9wnZ5n1ndURmC/lMrAP8BHgJuABVX0CeI+qvg/rrDIt+Ueqei5QO2p5\nP+bd659VdS3wDz7udZgf7SFl9s/lUVU9D/MO1ebzukREmoHVqnqhqq7IpbGnjxvg96q6FHgMOATr\nJC8BPsXg/tA+MQ9XgcC4CO07tO8dmTByn0S8Zj+QHXtXiSuB00XkTxj0qARwiqqeJCKfBdr9ueyD\nAOL/8uEj4Jeqeikj05mLo9apQebZqt//HsBGCvk0lqvqslyZXgl0Y44j8qhPozbftWHy9NbEvazO\nvdk9EXC7qn4/Fz4rW560zrl8HAWGdsibsJHF2tqbAoGRCO07tO8dmSDcJ4cPiMjRmAlqWXZSRI7C\nTFxzgDsxjfajInI1sEZEPoZpuD+rjVBVl4tImzfdlTDN/Qp/PAv4kKr2jZCnG4GLvVenMX09yqeZ\nisgXMa36AuBzwFuBpSLyWuB6Efka1mGdC7x+hCi/DlwpIpswf9zXi8iVQC9w+zD3lEXkXB/+K2Jf\naGpW1X8cIXzWcewlIhcAu2G+pwcwX9d7+rKAjYhWjfggAoGhhPZdn9C+dyCCh7rAlEJE3g5sUtW7\n6lz7rqq+LXfcgX2p7I920U0gsDMR2vfYCcI9MKUQkQj7etadYwi7F1BV1acmP2eBQGCihPY9doJw\nDwQCgUBgihFWywcCgUAgMMUIwj0QCAQCgSlGEO6BQCAQCEwxgnAPBAKBQGCKEYR7IBAIBAJTjP8P\n6UKCMPUVg4kAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f84c5a4a5c0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "size = 25\n",
    "\n",
    "listt = ['1mer_DNM_context_synonymous', '1mer_DNM_context_non_synonymous',\n",
    "        'CpGmer_DNM_context_synonymous', 'CpGmer_DNM_context_non_synonymous',\n",
    "        '3mer_DNM_context_synonymous', '3mer_DNM_context_non_synonymous',\n",
    "        '5mer_DNM_context_synonymous', '5mer_DNM_context_non_synonymous',\n",
    "        '7mer_DNM_context_synonymous', '7mer_DNM_context_non_synonymous']\n",
    "\n",
    "NROW=5\n",
    "NCOL=2\n",
    "\n",
    "axs = []\n",
    "\n",
    "fig = plt.figure(figsize=(7,4))\n",
    "\n",
    "for row in range(0, NROW):\n",
    "    for col in range(0, NCOL):\n",
    "        axs.append(plt.subplot2grid((NROW, NCOL), (row, col)))\n",
    "\n",
    "cnt=0\n",
    "for ele in listt:\n",
    "    plot_HD_BINNED(ele, axs[cnt], cnt, size)\n",
    "    cnt+=1\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.savefig(path.join('figures', 'internal_exon_centered_DNM_context_binned_selection.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exons centered analysis, Halldorsson 2019\n",
    "\n",
    "Mutation probabilities were computed genome-wide using DNM. They were used to compute the expected number of mutations based on a generalized k-mer model, with the method described in the **Exon-centered and intron-centered mutational analyses** section at Frigola et al. for the 3-mer case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get mutations\n",
    "mutations_file = mutations_folder + 'germinal_ultimate_dataset.bed.gz'\n",
    "mutations_df = pd.read_csv(mutations_file, sep=\"\\t\", header=None)\n",
    "mutations_df.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'class']\n",
    "mutations_df = mutations_df[['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type']]\n",
    "mutations_df = mutations_df[mutations_df['type'] == 'subs']\n",
    "mutations_df = mutations_df[mutations_df['sample'] == 'Halldorsson2019']\n",
    "mutations_df['ID'] = mutations_df.index + 1\n",
    "\n",
    "tumor_type = 'germinal'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 12895\n",
      "The total amount of mutations at the exonic sequences are 1257\n",
      "The category splice_acceptor_variant is found 1 times.\n",
      "The category None is found 3 times.\n",
      "The category splice_region_variant is found 18 times.\n",
      "The category synonymous_variant is found 295 times.\n",
      "The category stop_gained is found 46 times.\n",
      "The category missense_variant is found 893 times.\n",
      "The category initiator_codon_variant is found 1 times.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-c5999bcddf48>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n\u001b[0;32m----> 7\u001b[0;31m                             middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mexons_centered_results\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'results/exons_centered_'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'mer_DNM_context_golden_'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minclude\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.tsv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-18-feac5caaf26e>\u001b[0m in \u001b[0;36mobs_exp_muts_excluding_kmer\u001b[0;34m(mutations_df, exons_coords_df, signatures_file, tumor_type, middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\u001b[0m\n\u001b[1;32m    170\u001b[0m             \u001b[0mfirst_flank_region\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhg19\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_chr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmy_start\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m//\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_bases_first_flank\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m//\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m             first_flank_region_list = compute_intron_probs(first_flank_region, signatures_dict, 0, \n\u001b[0;32m--> 172\u001b[0;31m                                     middle_distance_threshold, tumor_type, k)\n\u001b[0m\u001b[1;32m    173\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    174\u001b[0m             \u001b[0msecond_flank_region\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhg19\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_chr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmy_end_exon\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m//\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_bases_second_flank\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m//\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-16-ae319fa11571>\u001b[0m in \u001b[0;36mcompute_intron_probs\u001b[0;34m(sequence, signatures_dict, rel_start, middle_distance_threshold, tumor_type, k)\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0;31m### Foreach alternative, each of the three possible changes given a kmer.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0malt_base\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmy_alt_bases\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m             \u001b[0;31m#### Built reference and alternate kmers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m             \u001b[0mtri_ref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprevious_base\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_ref_base\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mnext_base\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_Halldorsson.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_Halldorsson_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>115.149</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.851</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       115.149            5113        5124.851   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.13  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    10.291883\n",
       "dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    9.966747\n",
       " dtype: float64, 0    10.617018\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>299.403</td>\n",
       "      <td>5113</td>\n",
       "      <td>5171.597</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       299.403            5113        5171.597   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    19.57128\n",
       "dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    19.362319\n",
       " dtype: float64, 0    19.780241\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>131.129</td>\n",
       "      <td>5113</td>\n",
       "      <td>5108.871</td>\n",
       "      <td>0.679</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       131.129            5113        5108.871   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.679  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -3.148808\n",
       "dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.399596\n",
       " dtype: float64, 0   -2.898019\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>346.288</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.712</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       346.288            5113        5124.712   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.26  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.382156\n",
       "dtype: float64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.215567\n",
       " dtype: float64, 0    3.548745\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>130.631</td>\n",
       "      <td>5113</td>\n",
       "      <td>5109.369</td>\n",
       "      <td>0.644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       130.631            5113        5109.369   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.644  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.779585\n",
       "dtype: float64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.05383\n",
       " dtype: float64, 0   -2.50534\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>343.417</td>\n",
       "      <td>5113</td>\n",
       "      <td>5127.583</td>\n",
       "      <td>0.201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       343.417            5113        5127.583   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.201  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.246441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.080358\n",
       " dtype: float64, 0    4.412524\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>129.603</td>\n",
       "      <td>5113</td>\n",
       "      <td>5110.397</td>\n",
       "      <td>0.595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       129.603            5113        5110.397   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.595  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.008441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -2.276826\n",
       " dtype: float64, 0   -1.740057\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>342.903</td>\n",
       "      <td>5113</td>\n",
       "      <td>5128.097</td>\n",
       "      <td>0.193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       342.903            5113        5128.097   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.193  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.402703\n",
       "dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.233248\n",
       " dtype: float64, 0    4.572158\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>126.673</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113.327</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       126.673            5113        5113.327   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0               0.5  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.258145\n",
       "dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -0.018615\n",
       " dtype: float64, 0    0.534905\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>345.76</td>\n",
       "      <td>5113</td>\n",
       "      <td>5125.24</td>\n",
       "      <td>0.248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358        345.76            5113         5125.24   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.248  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.540028\n",
       "dtype: float64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.370396\n",
       " dtype: float64, 0    3.70966\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot results\n",
    "\n",
    "Binned average approach. Bin size is 25."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/ipykernel_launcher.py:20: MatplotlibDeprecationWarning: The set_axis_bgcolor function was deprecated in version 2.0. Use set_facecolor instead.\n",
      "/home/peluca/anaconda3/envs/intronexon/lib/python3.6/site-packages/matplotlib/font_manager.py:1297: UserWarning: findfont: Font family ['sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfcAAAEqCAYAAAAI1RnWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXecZFWV+L/nVXV1nOkJ5EcBgqIEA6tgWAQU1wSKIoIi\nmBDUVRFFQUGYGYyoa1gVMbu4ivx0XVcxrBFXXHdddVkVEwaY9iEMTGBCx6p3fn+c+6Zuva6uqu6u\n6unpud/Ppz71wr33nBvOTe+++0RVCQQCgUAgsHSIdrUCgUAgEAgEOkto3AOBQCAQWGKExj0QCAQC\ngSVGaNwDgUAgEFhihMY9EAgEAoElRmjcA4FAIBBYYoTGPRAIBAKBJUbTxl1Eetz/YSKyYmFUCgQC\ngUAgMB9ajdzfIiIx8BbggwugTyAQCAQCgXnSqnFfAZwGvB24s/vqBAKBQCAQmC+tGvebgJWq+gvg\ntu6rEwgEAoFAYL4UW9z/PqAi8nxgYgH0CQQCgUAgME9ajdzfDWwD7na/QCAQCAQCi5xWI/efqurX\nFkSTQCAQCAQCHUGaffJVRH4IbAVGAVXVMxdKsUAgEAgEAnOjaeMOICLLAFR124JoFAgEAoFAYF60\nGrm/CTjYnY6o6lULolUgEAgEAoE502pB3SpVPV9VzweWLYRCgUAgEAgE5kerBXUrROQF7nh1t5UJ\nBAKBQCAwf1pNy/cAT3Sn31LVqQXRKhAIBAKBwJyZsXEXkYuAAwBxl1RVL1koxQKBQCAQCMyNZo37\n/sAqYLu7NKSqty6UYoFAIBAIBObGjAvqVPWvwLNU9Q5VvQN43sKpFQgEAoFAYK7MuKBORL4AHCki\nR7tLYfvZQCAQCAR2A1otqDsqTMUHAoFAILB70epVuEeLyHucu4KqntR9lQKBQCAQCMyHVpvYPBT4\nsaqeDHx9AfQJBAKBQCAwT1o17luAkoicA/zNAugTCAQCgUBgnrR65r4MmAKeCvxeVX+1UIoFAoFA\nIBCYG61G7q9R1XFV/RJw9kIoFAgEAoFAYH6EV+ECgUAgEFhihFfhAoFAIBBYYrRq3K8HFBgChlX1\nxIVSLBAIBAKBwNxo+p67qj43O3YfkgkEAoFAILDIadq4i8hTPXfhVbhAIBAIBHYDWu1Qtzc2LT8B\nhM+9BgKBQCCwG9DqVbj/xj77egjw/G4pISIf7Ua4SVxelcTl+3Uj7EAg0B5dtO8oicvHdCPsQGB3\np1Xj/nbgv4DvAN9t5lBEDhWRT4jIF9352SLyMRG5TkQGW8g5oG2NZ8fDgGe2cpTE5ft3SX4gsCRY\npPbdC5zWylESl++fxGXpkg6BwKKkVeP+E+BXqvozVf1ZM4eq+idVPc+79ExVPR+4ATh9nnrOFQXa\nMepzuq1IILA7s4jtux3Oob16IBBYMrRq3PcHrhOR60Xkc7MMOzO89cCB+Zuu53+jiNxI93r27Tbu\ngV1AEpeHW9wvJXF5oI1wBpO43NM5zQJtsBjsezYUdrUCexpt2Le0cuPcFZO4PNQ5zfYMWjXuo6p6\nuqo+V1Xnuv3sQcBf8hdV9XOqeqqqngrcOcewAUjiciGJy6XctV5ax282Mvr9qb0kLvcmcblpheH8\n9HdKh6WEy6/XNLjup9fjgee1Edx5wGNayOvfHadmXQW4mMtQ1+0bppULnO31zSKIVrba43cQXZ3S\n28JPn+tY7nblai7MwYYa2rcXxopGbhrwCODlLXQrJXG51QLxRUm37LtV4/dIEblWRN4pIu9s5lBE\nVovItcAxIvJG4Msi8mHg2cCX5qpgmyOyE4EX5a6dDxwLSLujOrdAZ6ZK4FLqRyjnA8dlOs4g41L3\ny8tpqk/+fq7SmebX9WylmZtWfjrBLEfPM8n206sA9Ph5kunt0jwLo5h3l7nNhbuXdy/Khds0jRvR\nIJ9mDLOdsDL3OX8H0KAMLTS70r6TuLyc6WlwHPBSd7+QxOVWdVnk3M7UAJwGnOqdHwtckOk4g71c\nCLwe2Cenbyv7rtN3tmWvTTfShcbuUuDQDoQRu2OB+jxxNlTM2VKE2Xfe3vz4nQn8Xe6+n647828W\n9j0tz+dST2Rus3xvUF67Yt+tDOKF2KK6D7nfjKjqRlV9maoepqpvx3rrRwPjmKHMGpewl89w3Wdf\n7BGCzwBW6QtwuT+t4wp+PtMEGymemd1v4M5Prz5sQQ9Ox8tn0K2RzGlxynF5rofrh91IzquAI3z3\nXkFuFA+AFzPHfGmhd9MOg3c/3xA30hGs4vRnjc4HjsEM4iB3TbGydpQX3/2BN+XCKng6nIRVCL7u\npXwat4hPPr4nA2c4f3vh5XOTMrczLBfeoPtf5uucD2ehR4vdtO9G6dLE7jJ6gGzE8zzgb5v4Bygk\ncXkV8Cb/vueun5o9g9Ud2ULBy7ER5gNy4Wby83lyeZKbSczxAuBRM9l3Ym/5+J31RuH3eH7Iuwce\njs1oNXSTD3um6w38lpqVwZnKeM59fgDl2+mx2EDtKcDT3TUFSi7efji+vz6g19NtBfX17IuoldM6\nu8zHweMVwFGe29U5HWZr3+diM4xnA4+dSWin7Ltp466qd/i/WYatwHYs0adN27XJmhmuX+kSOqMI\nkMTlk71rQv0z99flws2H/WjgeOBg7/6jgDVJ4+dCKRDleuDSROdGMptxqZPtZ/KaGY6HsYbqmQ3u\nrwGegc1uZHquxUaELZ9nz4GHzHQjicsPwToVUBtJHZbE5X2By5g5fVZ4xyuxMlUETk/i8pOpbZHc\n78I4FcufPC9L4vIDgCuAZc6PzznUVzx/gxlklmaN8Hvvy7AOxxrgcTl3F+HFL4nLh7hrM3Gx+88v\nGjvBhbOrt4LupH3n66HTmJ5+JHH54iQuPzmJy0+hfuZnGXByUnvrZQ1mO364z8FG2v79wzwdDqM+\nrdOcXoNYXH2b9MvYM9y9rDw0ewwwCDwReHUSlw91cSt4uvS4sLIwnsb0+iNr4NckcflBWdySuJyV\nv1OpDXhOS2ztypokLq9I4vKZSVzOOkMvAx7UQMdM3kWJDYwy2Wdhnak12HT5TnJxyMgWW67BOubQ\nfB3UALDc/R+RxOU3UJ8vdengGnGcm4dhnfY11HfUAPbDPl2e+csed61x56925cFnObUOHlgHI9Mh\nb7uPYHrc8+d7Ob1W0KB8Oz1WOn8XNro/G5p9FW5YVe+bR9g/VNUfiMi+wHvIPTsVkbOpjcjaWnCT\nxOVHYj06sJ7Yu7Pg3P9jk7j8WOBad55/Dl+Mk5GKd77Wu73K/fvP8Q52/9lzoRclcRng/VjGn0B9\nGmYFZS3woxZxeX2cjLwrictXYpsEvR84EttTwOcReV2z4yQuXwJ82HP7ENeA5uP3UOAnSVxeQ71h\nneQqwBOArwNPjJORtzTTOxeHNcBb4mSk6horgNWe3Ktp/FhirXd6sJPv338S1tnKyML8EJan2eOR\nYawDlpEZ9M6On+sEvspzcxZWaT8I6Hfh7nD3DnLhk8TlJ2I9/bryEicja91x9gpWLzDpjv2OzVE5\nP8MN4p6Fe4Z3mm/o/BGZ7/ekJC4vi5ORr+bDWyA6ad8RUE1stuJi7/r3cu6WYfl9GzDqXd/b/S/P\nuT8/sRkcmD6zB24a10vXI9zxl7HGN5/mOxsqVw/4+j/UHWeNwaXA2iQuvxIYA34K/B54LbDJuVlB\nbf+QK7zwsufLZyZx+Y/UyhdJXPYblWzk+Bx370ysDsFzvxb7omfWkT8Ds59tXtyW+e6zMu4Ypn5g\nBJB1opYlcfko7NFMXmaGXx9k5eDJSVzer4Gfm7BO61a8ETPwZGZGcvKyTu+zXLhPw2Yxqu48k3sg\nNoOSsRI4N4nL6+JkRF1HsUB9/f5C79i357fhBgoz2PcR7nAIq0/8qX1/cOj7XZnE5UvjZOTqhrFu\ngxk/HCMiVwObVHXOgbtwSsDnVPWMJm5udAtv6miQUP4imnEsw4RaYk1RG63nr/kswzItBebTgckj\n1Ixo0sluxji1+PwZyDbc6aHWMdmR99Qlfo6NVO+hVlmui5ORaQUkicsXA+8FrsQ6BU/Nu3GMUZu6\nbMRyLE8rwLbcvRK1Efs2F9YU9SNlsIo1kzHG9BH7HdQ6ae2wGTP0hNqzwR3U9+D9eH0oTkbuaTKy\nb8UvaDLb4bGd6TMN5CriBafD9g21dJ6i1qj9DngwZu9Vdy2zD/9aRj81u9pC+6/MtcMAZucVrFMO\nVj6Xzeij5mYUe4TYqB7rNvfirTlpgV9/VjCbKlFbW7AdS9cqzWcollOL573U22aBWodrlFqHp5Eu\nWX00xvR6Yrb8Dngg8FVsRgTq6948N8XJyE1JXL6Q2gBwNnwIm95vxVamd07nZd/NGvd/wHpMt+Km\nuFX1zIaOG/s/HXgSVkF/WFVvauK2ofE/tFRaX0AKgyIrBKQo0hthllpBJ7anunkokhUlpB9gXHX7\nxjS9c1UU7Tsgkk2la8UqCR1X3Q7o8kj2riqVglBUpZpCWoWpCjpZQnpLIgOKpgoqoApaUSYVNJPb\nK9I/KLJSQFOoVtFKr0h/CelPoSoQVdBJQRjTdGtVmRqIZDhCIoBRTe+rKJN9IgORM5AKTCqaLpdo\nb0sXIic/dXmAgEwoO6roVBHpiaxxkzHVrZPouDiHAoWSSJ83DBSAXpG0hGhR0CKiPSJ6VE/P6H5R\nYWpVFI0+uFT6mpcFV8fJyFh2ksTlU6lNxWWN4C3Aw/5SqRx/e7VyvzHV6OaJicnNaboNV8mmaDFC\nVqaoREjBxTeKoFCFqQiK6irgKjpVVSolkYEUqgpaRas7UvWNXyIoCtAnMlwU6VU0BZhQxjA/UwAV\npVJFK0Wkpyg2so+QHgHJypEqqVo6o5Bu13RLVan0igxGEPWI9JbEOm2pUq1ARdlpOJk/LSKlXmEw\ntbgo7vrWNN1YUSZXRtF+BaFHQSaV0RStplCtKJXI8qtfd3ozply5S9HqmOr2CCn2inUsqlD9ycRE\n3WKuhaIT9n1kTykBGIpkhUBUQIoFKwspoPdpem8v0jcQyTCWwdW70+r6QZGh4SjaB9eRrygTAkyo\njlVgalkkqwCNIFLQVK0cVdDxCCn0CgPOtsHlX1WZUkgV0q2abixAcZlEqyJBVEkrMFWEnl6RQYVq\nBAVXrzClOjaJjveJDBaREsCk6ui46o4ekd4i9IhYWVSo9ossLwo9zk6j1DV67lwqykTFld8IEUCm\n0PEJ1VHJkhSkJNIndpxds9R24bq6gYFIqn0iaYTo4cXiWEmEQ4vFfe9XKEokyH5R4c+RyAQwAtwy\noXrixrR6nCrpLVOT0XbViW1p2ndbpTI1pVpV6NmYpndVUUmhoErUI7IMiBQVV3iliJSsDrS8dTaW\nTqmOO1uMqjBZREpb0/TeWhx21n1RD1IqCn0pVItIaVJ1DKyeyMKrKtVIiHqQXgUpQk8kVqdOKRMp\nWtFaXuuE6uikMtUr9BWQoghRCUtLV052Dgj91lEg6kX6vbAAmEB3jKvuGBRZURLpd2VysopOKlBR\npiy/6BNE1Au1CpWsfI6r7lCgT6Q/q59+Pjmxcia7akXL77lj77GiqrPqMbldq67BDOAmVf1sE7cN\njf9jq1bfA8gO1fUTqpuqkO5I9d6CUNq/UDgWKCikd1Wr/1OAnnKheEoKlQiKSbXy7R2qW/pFhiMo\n9ImsLImsBBhTvXdzmt4uUCyJDApIv8gqsfB0a5omqfVKI4GoKFIqwSBIcWUUHSlmNLopTX9TQcd6\nRIYiM1K2pOn6qupUSaQ/sspg+YDIPgAVmNiepnf1iiwflGh/Bami4+OqmyIoFt1OX5Oq2zam6Z8V\npGQNZNaAUBTpGRTZG6xyn1IdLYkMDYkckNbKomKFeFvqZg8y/+Oq0aRqVEFlSpH70rTnV1NTw4rK\nbyuVFceXeneURKrPGRj401P7B27B5P8W+By1adaHAQcklcrQ1du2Hj6lyjfHxwr7FArbh0QqR/T0\n9B5UKFYLVlluK4qkAvdVVLcqpFM2WqGiOp5CGoEURIoC0isy5Ap2dUeabo6Ewn5R4UjZafeqzqin\n1Br/dFOa/jmF6qDIyggKBaRYcA15yeWNQjquugWoVFWnqjBZgJ6SyIDLT7FyEO3VL7I/kFbRsUll\nEyg7VP9ahfESDInTNatEcf8Cul31rorqpN0T+kVWDkfyAEGooNvuqlb/ryhSGhRZDdCDDEZiHZVx\n1U0VZULEKvkIiiVkSAQZEDkgcrMFk8qGCjq6Iooe+sKN9y7owrqMTtj3dav32iAgFdixNU1/p8AO\n1XtUSfcrFI4rCAOqsFnTX42m6cZDij2nRzYSlE1p+l+b0uofe0WGCkhfjzDg8o2KMr4hrf5CQPpE\nlgP0iqwsuMc2o6r3TKqOuWSWSCiUkGWgsjyKDu9Blivodk3/NKZ6TxHpz8rT1jT964Tq9qJIqQf6\nCiJ9QyJl10hXtqZ6eyT0Dkt0mFi5q45qeg+gPSJDmLupDdXqbxTSAmQdzazzHg1F0d64hkaVVITi\nkMiBuM4Kzs4ryugUOoprtLCBhmInaRWdBNicpn1ViDanad+d1cpACtFvpqaWT6gyBb2TqqXIdYbA\nhuoDImMKleEoGj2wUNgskJYLxWQwkslyoXhkD1IoCmkEaQFJFd0xpnp3hKSIDYoqquNTbnYjG8AM\nWD6UFNiuepdCdZ8oOrpgjaI1qNZzrqSug3Nfmv55Csb6RJb1QL8ghZLYTEmE9EZiswyTqluqyoTC\n1Ljqlkgo9ovs5WxTXF0+NCCSzeRVxlQTIJ1Q3TKl7CgK/Vk5kZ15IjsHVuOqmybR7S7MqEdkcIVE\nR7i8qW5Iqz9WSIdE9sPk9RXc7MWksnUS3erCwtV3KwWkF9m7R1gJMAX3TapuWi7R0WdvvKfZwsym\ntGrcr6C2InlEVa9qO2CRc4EtqvpVEblBVc9q4rbVtN3tTH8W7VPBptmzd9tTalNmc2GU1ovNvknj\nZ0FXYdPV7dJsams78P+oLULrFpNA6eaJ8aN+P1WJk2pl+OM7th+pTrHH9vbtKNhIRxD49dTUYFKt\nIqCPLPX+6dG9vb8eFhl93uDQ6/pEXtKGvPtwz6xa4D8iaEV+ARTAb4AjGrjtFB8EXjlHv1+hthp4\nriyPk5HXzjOMOdFh+26HX2CdyiyPK03ctuJ72JsxzdhC/ULOjHXMbmFsU+JkZG2DdNhA7vW6bpGq\nslW1bu1VBCyPogrwDuANwG1xMvJZaDvP/EdaX8YWG3aCduuNjIZT3Tma1b8/w57XZ9xO83bIZ7aP\nAxsxFCcj+fUObdPqPciVbotJRORdswz7QOCX7ng+z5T+k9qzxsyw1mELR56NLVzbB7iR+g0RfCP8\nX6wyBShji/H8jsq+2KrRP2AF8yiscf9XaivQx7EFYkdgK9OngHdiX8t7M7YgJluMAbZAbos778EW\nv/wA67Sd4MlXL05/j/e8G8D5943lW1ihe6Nz8zxsgctn3fEXsUUzABudzNOBt2L5cCi17XazNPo/\noHR8b9+643tZA9z8sFLpx3/X13/pzRPjR/65Utk/mypV0JN6+0af2T/w0wjoF6lG1rH9zzgZ+Yu3\n0OifXf70Ys/bPuTJ+hq2Oh4sf7MNaLYA3wZ+7eXNK7HXrh7szm93+fQEz80vsDLwEWwh0xeBP2Er\nZLPG/fPY8zacHtuwRU7ZqtUvY52606h/vj4C/DpORn7sFr+oF8YUtWe62fTZNXEyssFbUZ2Vwa8D\n/0M9WeN+E/Yq19tyfta5/3OxfLsJS5u/d3G8lV1Hp+z7C9QWZH0Ee3f9WmwR2BrgZuD4OBn5klss\nmj23fS9Wbh7pzrO0inC2mJOzxsl6ErWOQVYBZ4+Y1nn+t2M2vIZah9237/8Avp8Lf6vTK/MD9fb9\nKCcf4NNOPtQ//yVORq5xjehV2MJP/3GoX699B1vl34etmcm2CPcHF9dhddcFzu/FWF35OxcXWSGS\nhXeV0zcCroiTkfEkLl8VJyON3jzJ0uSv2LNzX6+Pe8d3YjN/D8Lycn+sbsj4W6z+OsS79i5s/4CM\nX2GvXb4PW+i4DKubU0/O74HrPT9rnB5Z5/cOLM3XYHaYvRb3H1hbci31PBcrAzc52Xtha29+HCcj\n/97AvqG+zhasLO9H7Zl+Viaf4OL9QayOPgqrs+9zcVwD/DGXTrOm1cj9k1jjoMBJqtr2CNL17Der\n6o0i8nlVfU7uvr+atl9VT54WSCAQWJQE+w4EFjetGvcerKckwLdUdXJGx9P9DmI9k3Hg5mbP5JqE\n8V1shWQ3OYAObI8ZZAQZu0AG7KKGM9h3kLGIZSyUnIWQMXf7VtVF+wNuDDKCjCBj18vZXXUPMvY8\nGUspLvOR0XSHOpGdbyYEAoFAIBDYTWi1t/yu/mDFbD8zG2QEGXuSjIWU0w2WSj4EGYtLxkLJWdQy\nWj1zvxFbjX4fgKpeM1dBgUAgEAgEFoZWr8L5r7/N3AsIBAKBQCCwaGg1LX8gcKqq/oDae8aBQCAQ\nCAQWMa1G7o/GdkuC9nfmmTUichy2ucKIqr7OXbsYe/G/B9tgZn9sJqEKfEpVv593o82eMdRkPQPb\nWW4Ftl/+c0XkJ9gmEHeo6ttF5IC8rFnG54XYxhPrgQ+p6i/bic8sZbwEOA77mMFbVPWWTsfDk9X2\nVqOzCPMZwCnYDlKfwDYgugPYrqqv64RMETkJ22DoVmwTmwOwTWt6qX15a74yHottHlTENsO4r1Px\nEJFDsc2PhlX1DPfueFP9825UdaE+PDQjwb4Xt307eR218aVi307O7mnjLZbhvw/bhWgY+EyXl/wf\nArzbHZeAz7rjV2Aftr8Cmz2IsEUG09zMUt5FwJPc8feBTwIvcOd1suYQl+djO4h9DNs5rWV85pFu\nxwBXdCMenoxzgae54xs6nO8rMeP/F+CjwOs7JRP7/OM3sJ2p7g98wV0/xYXfsXhhW2y+tEvx+KL7\nb6l/3k0n82qe6RPse27p1nX77lQ5nSHcJWHfLozdysZbjdw/jW3fdy221eq8EZEHA2/PXc7vfLca\n21ccrHd8oPuNqGrq3tBr5KYtWaq6AfuW8z+6aye7cG8Qka82kDXr+KjqdSLyEOyNg3e3EZ9Zy8A+\nk3ghte0m5xyPFnRqq9FGvAnbmvYWp+d7XHw7ITP/zfFs5Lee2mdWOxWvs4HzgI91IR4Z7ejfyM2C\nEex7t7Rv6J6NLxX7ht3Mxls17ndT2+f7rtnp2BhV/SUw7SMSInKId7qR2reHD8L2Dv8LcKCIbG3i\npl1ZxwP/rWqfCc3+sT2m+xvImnV8HBuwfZDbic+sZLjdA68B3qeqI/ONRwv+ghXgW2i9TqMt3B4K\n7wC+oao/925laTZvmbn06KW2L/lBLnzmKwNARA4C7tP6Lyd2LB4NaEd/301DRKRHVadE5DBgo9pX\n8+ZFsO/d0r6hw+V0Kdk37J423qpxfy+1FfPvwTbT7zgicji2Wf5RInKBqn5URH4uIu/HMu0abCP9\nd2Afffi4qk42cNMuL8b1hsU+A/t+bBvNTaqaiMgnfFlziM8FwN9go491M+haF5/ZysA+YnM48HK3\njed3Oh0Pjy8BHxSRU7CPXHSCV2EfUBgWkftj6ztGsTL5TuwDM/OSKfXfHP8gcICIfBirGF/hnHUi\nXucBn3Iy/6lT8RD7LOxbgWNE5I3Al9vQv5GbmXiLiPwj8BZsVHBOC/dzItj3ordv6LyNLyX7ht3Q\nxlu9575OVde446tUdTafMg0EAosYEfkIVjHdDJyjqpfsYpUCgUCHmHHkLiJfAPZxvUahc9MNgUBg\ncXATcKiq/kJEbtvVygQCgc7RdOQeCASWLiKyH3ASttobVb1ulyoUCAQ6RtNn7iLyD0A5O1fVM7uu\nUSAQWCjeDVyPPbMNBAJLiFYL6qqhQQ8Eliw/VdWv7WolAoFA52m1oO6V2OYTWwFU9aqFUSsQCHQb\nEfkhZtuj2G5uoSMfCCwRWo3cT8Bek5lYAF0CgcACoqqPFZFl7nhbK/eBQGD3oVXj/hvsRf1ObJIQ\nCAQWESLyJuBgdzwSZuYCgaVDq8b9IOBt2KtwyvRtJAOBwO7LKlU9H0BE3tXKcSAQ2H1o1biv9Y7D\nO3OBwNJihYi8wB2v3qWaBAKBjtJqQd3bsUZ9CHiwqj5uoRQLBALdxe1f/kR3+i1VndqV+gQCgc7R\n9iY2IvIaVX1vl/UJBAILgIhchH3/OvucmIbtZwOBpUOrTWzehY3ce4DlC6JRIBBYCG4AVgHb3fnQ\nLtQlEAh0mFbT8ge7wwlV7cgnXwOBwOJARK7MVsiLyNtU9bJdrVMgEOgMzT4cswZvEZ2IhE1sAoEl\ngvsw1JEicrS7dPeu1CcQCHSWGUfuIvJAd9gHXAisUNVnLZRigUCgu4jIUap6667WIxAIdJ5W0/Ln\nAqcAH1DVHy2YVoFAoOuIyEuAZ2MzeAVVPWnXahQIBDrFjN9oF5GfAo8DPgMMi8hTF0yrQCCwEDwU\n+LGqngx8fVcrEwgEOkez1fIfcP97LYQigUBgwdkClETkHOBvdrUygUCgc7T9nnsgEFhauI/GTAFP\nBX6vqr/axSoFAoEOMeO0fCAQWPK8RlXHVfVLwNm7WplAINA5Wu0tHwgEliDhVbhAYGkTpuUDgT2U\n8CpcILB0CY17ILCHIiLXU/sw1LCqnriLVQoEAh0iTMsHAnsoqvrc7Nh9SCYQCCwRQuMeCOyheHtX\nFAmvwgUCS4rQuAcCey57Y9PyE0D43GsgsIRYFK/CichHuxFuEpdXJ3H5sG6EHQgsAf4b++zrIcDz\nuyWki/YdJXH54d0IOxDY3elY4y4ih4rIJ0Tki+78bBH5mIhcJyKDLbwf0Ck9cjwUOK2VoyQuP7CV\nm0BgCfJ24L+A7wDfbeZwkdp3L/C0Vo6SuPzAJC5Ll3QIBBYlHWvcVfVPqnqed+mZqno+cANweqfk\nzFYtoB2jfm5rJ4HAkuMnwK9U9Weq+rNmDhexfbfDc2mvHggElgzdnJbPDG89cGD+puv53ygiN9K9\nnn27jXtgF5DE5RUt7peSuDzQRjiDSVzu6Zxmewz7A9eJyPUi8rlZ+l0M9j0bCrtagT2NNuxbWrlx\n7opJXB5foi9kAAAgAElEQVTqnGZ7BgvxzP0g4C/5i6r6OVU9VVVPBe6cjwCX+b25a310MH5JXB7w\np/aSuNyXxOWmFYbz07Jx2hNJ4nIJmPb6VS69Hg+c00ZwLwH+toW8gd1xatZVgN0qQ6OqerqqPldV\n57r9bNftG6aVC5zt9c8iiFa2WvI7iElcLrg6pJmf/iQuD+2O5WouzMGGGtq3F8aKRm4acCzw9y10\nKyVxebdcIN4t++7kM/fVInItcIyIvBH4soh8GPte9JfmGm6bI7ITgBfnrp2PFQppd1TnFujMVAlc\nQv0I5SXAIzMdZ5BxCQ1WIbfSJ38/V+lM8+vkSzM3DfwUk7jc0c7dLEfPM1USfnoVgB7faJ3ekotz\n5q7gucvn+yV4XzjM53WrNG5Eg3yaMcx2wsrc5/zFdG8l+yNF5FoReaeIvLOZw11p30lcXs70NDgO\neKm7305ZjlrIOw041Ts/Frgg85O3MceFwOuAfXL6trLvOn1nW/badBN1obG7BJjvAuVLsDINrg7I\nxT9y6VPwbEmAYrN6ETgLeOJM9/00n4195/N8LvWEF1bRi59fXrti35185r5RVV+mqoep6tux3vrR\nwDhmKLPGJcDlDa7njWxfYN/c9QHsVT8BLk/i8jLff4NME+AJwFkuA8T9/DTyj/uAkju+PNNzpp5t\nTua0OOW4PNfD9cNuJOdC4EjffXbsxcNv/AU4jznmSwu9Z+zZ5/QoeLo0zBPH3tR/1OQC7J3sNwAH\nOz+KlbWjvXD2Y3o6Fz2Zj8MqBF/3Uj6Nff3aiO/JwLNd+dnLl9+kzO0My4U36P6Xu+tR3k+TtJot\nL8QW1X3I/WakS/ad2ea0dMldi/zr7l4Js0GA5wHHN/EP1jisdvIiL1+z9O3DFujtdI/VIWB58xrg\n8Fy4mfx8nlye2MzUTDwfeMxM9p3YWz519urr7IffqHy6OD0cG4CQj2u+XDcqS3mZnpueZmVwpjLe\nyPY9fDs9DhuoPQU4zbPvkov3TPVoL9Dr6bYyd//F1MppnV020tvxCqyMZ27z5Sdv39NsNRfHc7AZ\nxrOxAWlDOmXf3X7mvh0zgGnTdm1y5QzX17iEzsh6qCc30CFLpIt9/+7n8xj3O8jJXQM8GrgyicvD\nDXRIgSip74FLg3CbyWzGJVg8/UxeM8PxMmwEdXqD+2uAZwIn5u7tz+ymNdvloU3uPQTrVECt7B2a\nxOV9scI/U/r46b8CM+QCFt8nYfk8iJW1NdgK6rRBOC9N4vLhzs2Q8+NzLvUVz8PdNZro5vfel1Er\nP4/LuXutH0YSl++HNRgz8doZrp/owjmpid+2UNU7/N9svTN/+87SI18PPYPp6Qdmw0/GKn6fIeDx\nSVx+gBfumlxl+xzgVe74Snf/MGp1zGHUL9BLc3oNYOXOt0m/jD3T3cvKQ7PHAIPYQOKiJC4fCjsf\nM2TpUXRhZWE8nVqdlHGZa+DXJHH5QcCrXThZ+TsF6+CCNZIDWF22AjiT2mOslwMPaqDjGhfOaxN7\n3p3JPgtrpNYAj/A95OKQ8SwvvAvccbOGqx+zoz7gCKwT7+dLXT2a1J7ZK1b3PM7dz3eu9sUrN64B\n7ffCujCJy/fP+VlGrYOHF+aVTLfdRzC9vcqnxV4ujGFmsF/XKVmDy8/50M1nFD9U1R+IyL7Ae7De\n9U5E5GxqI7K2FtwkcflRmHGDNRLZVGJWWI5P4vLxwLXuvC6Dk7hcjJORine+1ru90v37z9kOcv9Z\nRr4oicsA7weWY70vv3Jf44X74xZxuTRORq52bieA9wJHYe8c+xyb1zU7TuLyG6gfcT04icsPbhC/\nhwA/ycUX4CRnkI8FvgE8KU5Grmqmdy4Oa4E3x8lI1TVWAKs8OVcDl87gL+Ngcr3YJC4/GXiUd2m1\n83MNlqfZ45HlOXfZyGuVF9ZewCs9N9lo/UFAvwt3zF0ru/BJ4vKTsLSf8vWOk5G17vgZ7nIfMOmO\nH+LJOSrnZ1mDuGfhnumd5hs6f0Tm+z0xicvL4mTkK/nw2kFEhlX1vrn4dXTSviOg6mYr/E7N93Lu\nhrBHYbcBO7zre7v/ZTn3L3UdR6g1dD5PhLp0PcIdf7nBPah1nte4esDXP8v7rMN4KbA2icsXAqPA\nT4HfYdP49zo3w9T2F7jCC+/l7v85SVz+AzY7gtPHb1Quy9y5e2dhDSKe+7XABmqN1LOxafGtXtz8\nWc2dZdy79zrqyabmlyVx+SgXZl5mhl8f7O/+n+Lli+/nB1jndSveiJland8IycnL6pIzXLhPx2b6\nqu4806GMzVxlrATOSeLyujgZUddRLFBfv7/IO17u6f02rGzOZN9HusMhoOKHmRsc+n5XJHH5DXEy\n8o6GsW6Drn84RkRKwOdU9Ywmbm50C2/qaJBQBWqFdAxLqIhaYk1hvels2s6/5rMc69ikwJZ249IG\ngstkrMGebOIWdz/T8w6socNdyxqqbR3Urxk/w0aqG4FsVmRdnIxMKyBJXL4EeDfWU/0G00dSGRPU\nT3XmGcbytArkG5peah2urVgFWWF6hzSiVqFmZcJnBDPkdrnP6ZVQezY4Tn2nz4/Xh+Jk5J5GRt0m\nvwQe3NKVxW3aTEuuIm4bEbka2KSqV8/FvxdOJ+0bag3NJJbOAH/AOktVavmbpb9/LWOAWn5tofEs\nzlwZxMrcFLVGdwfTZ4HybMfycG/q67FRXMPTZTbhdXpb4NefFUy/Ejb6BYvLZqbPbuQZxsqsAvdQ\nnw8Fap2+UazemUmXfdz/KFYXzIffA4cDX6W2R4Jf9+a5KU5Gbkri8kXYrOFsuYYWiwEd26m1HTuZ\nq31DF0fuIpJNma4APjjHYA7CjOYRmK5ZJqdYYflvrEHKKu/bgC9gC2MeghWqFCvYKZaxVWxKZCtW\nkYxjFclWrIDt4+ROOj8pVsA3uuMNWENYBo5xcsewRvhAzAAyY9/k7v/WhX8EtUrp1+7+wd61TU7W\nY6g9Z1KsIhH3KwK3u/CWU6vEbnO6Ze56vfTKRn/5f//4FJcW/mYmfdRGtSRx+WlYekNt2ijrCT+S\nmoGsxwxmGEvjXuBQzKB7MYMvYp2ybdQKderOx1y4WcU3CfwvVpFk8el354djldakO7/Lpdl97n8U\nM5zlWKdFnE4Fd/9eahVYlte3OPf3w/JgL6xMqNNtK7WydUoSl1N3vsL5qbjzzM1/YuXnFE/2HU63\n05yuA1j50dxvE7WG5NdYucpmSfzR62wpAs8XkYe7NFFVPbOFn5100L4jbPalH0vzZdTK3Dewmaxs\n5DWBzRIcg6VlBcuf7Fv0f8YanROx8tDn4rbDnSdY+h1KzbazdN7iwp8A/sO5O8HpV8HyYQVwf6w8\nDTpZYOX9TndvLyfzT+63L7XHSlnZPtpdU6fjztkhJ+9eJ6/gziPMttdTK/8FbDaiSM0u/P+iCzuL\no/+f2U0W1u0u3iOYnZ2AzVqlTs4OzEY3O3fLgJs83YouvD4vzAizt7sxm1/hwht3abWPC3sz1uH5\nXs5vrwt3b6wu2OHScsTFY7NLtymXH73UOgursfIkWH2wjVp+V12+rMbK37Dze5Bzn+V1Nqh5ehKX\nszLSR33ZyTort2GzMo/CymvBxXsT9lhik4vLQUx/XLPV6VdxYaRYOSq447XMka6N3N2uVddgRnWT\nqn62iduZevZ3YAl+K5ZJU1jB6MOMO0uAr2EZ9Crnpgf4KJbAe7nz/ak1RCPAL7AEX+FkHEhtNP8b\nrBBkhXcAKwQ9WMWRufshljkr3b0U+BU1A+h18rNprG1YQdgXMwawCn7E+c9GqhuBn1NriPxKaAAr\nQOLSdgvWuGXTcb7bezFjyjLZz+z8NcWM+q8u7nc5vfqwRuV6apv9HIONavuwgohLk6+6OD/UpVsW\njmKV0wanX9Y47vDSudfFaTW1kdGdLl0e7675jd6Ec1NxaVXB8rfk/Ay68FZRKyd3Umsox6lVOlk+\nFzADfICn553u+A9O35VORr4yzSqm27A8zc73xxYJgVVI/45VPNkjn6xyAWt8fL891GZRDqfWCVrv\ndHtCnIzMeeGNiBzlwkJVZzVD1GH73gr8D5avCVYmnoyb+gRuxvLhIndNgH/DOmHDmE0sx+xMsMr+\nB9TyXzCby0bL6zFbbJTOj3THqQt/BEv3zO8fsHzs864fiZWdSaxe6cc6wQWsXN7u4pZ1LieAH7l4\n9lI/+i1Qs++sw9njyfAb6q1Yx7BRA55iZalRwy9YHXOXS7f93fXHUxtpfw5bS5EtTFZqs4hPcnpn\njWVmK3+m3kZ3OB2yeIF1SLIZqNtd+hzv9PDrrilqMze/dLJXUpt1zWYhBqjNMmyg1pG7x6XbgdR3\nklbgHpk5Gbc5uXe5tPSfs4v3y87/itW5BS+8R1HL66+4NLkftRmazG43Unskg/O/r3MXU5sZudfJ\neWycjMx5XVQ3G/dzgS2q+lURuUFVz2rittW03e1Mfxbtk03XlrAES2k9Jd6MUeoXUjTimzR+FnQV\nMy8EbESjqeaMrcAXmf6aX6fJHg/shxmQ35N9ILX3lLMGcH/g+5hxJFhlVwU+jFuh24LN1DoyzdhA\n7jWjJjSaIvwNuWeQHeaD1D/Pnw3/RhtbI7dgKE5G8s9D20ZErqDWyRhR1bbXW3TYvtvh/4CHUato\n5zPV/j2sIWvGFhpPw65jdgtjmxInI2sbpMNdNF4jsNC8A1vQ9rs4Gbke2s6zv1BbF/NlbIFkJ5gp\nT2Yie8TWjCozL37MHlVm3E7zdshnNm5nYjBORl4/V8/dXC1/INbrhfk9U/oR7nls9vzB/X/B3b8Z\n6zG+F2ukxoHJ3LOKn7vzdcAngNQ7X0dtAd5t2HRTNt35r14YY87PDe58ktqCvqsyveJkJKt03oc9\ns14LvMVduwkbUeDJfqt3viET5sJ6T5yMrKd+NfK3sNeXsnS4zV3PRk5f8NzeC/yLO36rk/EZX4Y7\nvAWbjns51iB+AHtT4N+Af8K2Kf0vbKbie8AnsQWDP8V6mOPYsylfz89QeyZ5T07WNZ67H3nHm53+\n67C0u8bF4Zeemz8D36ae/wPudjLGnOx12PRqxvVenoN1nL7v3f9XLA2358JeD3wz85sLY8rpvMlz\n/6Fc2cr4mh9GnIz8r3fv+9TK7E4/zt1a4I+euyztvjCfht2xUlXPd9vI5hejtaJT9u2X14+4/2up\npcMPAeJk5F+pH5W+ByuTAOql3U5bpFaO1jp3N2B1SfaM/nb3v8nz82Z3bZvnb11231uD8gOml4f7\ncn78MrAOGxBkfNpz+1U/QeJk5FrPzw3U45epb2NTzHdis5f5MgdwHbV0XYeNgK/3ypbvPotPlobj\n7tr1TCfzd6dzm8WFOBn5uOcuweoUsLz8A/X58m3Mpn3yey780oX7PuwR16+Aq3J1/O8a2N0nvPu3\ne+5/4l3/AdaZWpf7/Q6rC76PzQLgzv9zBvsmZ9+fxupGsvh5/m521z/grn3RnW/xwvwD83z/vdsj\n982qeqOIfF5Vn5O776+m7VfV/GtsgUCgi4jIJ7HKTYGTVLXtGaJg34HA4qbbz9w/iI3ebm72TK5J\nGN/FW9DVJQ6gA9tjBhlBxi6QAfNoOEWkB3vlS4BvqWrbj7KCfQcZi1jGQslZCBlz7xir6qL9ATcG\nGUFGkLHr5eyuugcZe56MpRSX+chYiA/HBAKBRYiIzHmlfSAQWNws9sZ9tp+hDDKCjD1JxnzlTNs9\ncIFZKvkQZCwuGQslZ1HL6PoOdYFAYHHivrW+820UVb2muY9AILC7sFt+/zYQCHSEd3nHoZcfCCwh\nFvu0fCAQ6B4HAqeq6g9ob3/7QCCwm7AoRu4ichz2OccRVX2du3YxtoVfD/AybFe0d2EbZnxKVb+f\nd6NtPGMQkWdgO8utwPbTfq6I/ATbwvQOVX27iByQlzXL+LwQ+6zieuBDqvrLduIzSxkvwbY1XQW8\nRVVv6XQ8PFltbzU6izCfgW0hvBzbbOI12H7r21X1dZ2QKSInYZuS3Ap8Hnt15XHYDnzZl7fmK+Ox\n2BfRitgWofd1Kh4icij2KdxhVT3DvTveVP+8G1Vttv/8o6ltnHRIu3rNlmDfi9u+nbyO2vhSsW8n\nZ/e08W4v5Z/Fkv9DgHe74xLwWXf8CuyTpFdgo4sIW2Qwzc0s5V0EPMkdfx/bde0F7rxO1hzi8nxs\n16GPYduntozPPNLtGOCKbsTDk3Eu8DR3fEOH830lZvz/gn0P4PWdkol9B+AbwKexPfC/4K6f4sLv\nWLywLTZf2qV4fNH9t9Q/76ZFuO/DtkoeBj7TyXxtICvY99zSrev23alyOkO4S8K+XRi7lY0v+Mhd\nRB6M2z7VI78z1mpqW/6tx6YPD8R6/ql7g6eRm7ZkqeoG4O+Af3TXTnbh3iAiX20ga9bxUdXrROQh\n2Irkd7cRn1nLwLbNvJDaXvZzjkcLDqS2BWynP0/5Juyb9Lc4Pd/j4tsJmflvjmcjv/XUvr/dqXid\nDZwHfKwL8choR/9Gbmbi09j3069l+pafcyLY925p39A9G18q9g27mY0veOOuqr/EPslah4gc4p1u\nxL6mBvZhi1/gPkYgIlubuGlX1vHAf6tq6txl+8Fvxr5YlJc16/g4NmB7drcTn1nJcLuLXQO8T1VH\n5huPFmQfgriFDq3TcO9YvwP4hqr+3LuVpdm8ZebSI/uKFVgeZPvgzzteInIQcJ/Wf1mtY/FoQDv6\n+25m4m7si39g+2vPm2Dfu6V9Q4fL6VKyb9g9bXyxPHM/HPvS0lEicoGqflREfi4i78cy7Rrs4xnv\nwD768HFVnWzgpl1ejOsNi8hK4P3YNpqbVDURkU/4suYQnwuwb1CvBtbNoGtdfGYrA7ga+wzoy902\nnt/pdDw8vgR8UEROIfeRi3nwKuAJwLCI3B97/juKlcl3Yh+DmZdMmf7N8QNE5MNYxfgK56wT8ToP\n+JST+U+dioeIrMY++HOMiLwR+HIb+jdyMxPvpbZi/j3UPunbUYJ9L3r7hs7b+FKyb9gNbTy85x4I\n7KGIyDpVXeOOr1LV2XyqOBAILGIWxcg9EAgsLCLyBWAfNyoUwmuxgcCSIozcA4FAIBBYYoSReyCw\nhyIi/wCUs3NVPXMXqhMIBDpIaNwDgT2XamjQA4GlSXjOFgjsuawXkXeLyJUiEhbTtYGIvFBEbhSR\n94vIJ0XkUBE5QERe08DtWhE5ugs6PEdEPiUiD/eu7Zaf7xWRL7bhRrzjB4jIJSJyiIi8u4W/h4vI\nyzqh5+5IGLkHAnsuJwBvASZ2tSK7Gdeq6o3uNab3Yq/dxSJyBPbK3x3AdcDx2KLFz2E79B2DvRf9\nCmzL0WXYa2y/VdVPici7sDp5PfAB7BWpXqAAXKSqVREZwra3/TXQKyI/wl6R+pRr7LYCv1DVj4jI\nLdgObvtjr2s9yum+c5taEbkce0d/GbZF7AdcfN6I7WD4IOBpQB+w1sXpROBPQKqqb/XCehLwHOy9\n86uBk5zM5S7cZ/hxBm4DjhSRN7n767DFndtU9U0icivwGew1vd87MS/DXj/rBx4lIm918XsJ8BVs\nF7/7Ax9W1Z+5Tte1zbNzaRJG7oHAnstvsI04RtwvMAtUdSO2l3zGPljDdr2q3grcDFyjqje7+1NA\njDXyAP9PVS8FnigiRwKTqvoaVX0vtsPeIcAWYMj5Q1W3Azdh79LfCfxaVd8BPBv4pKq+Anu/HGCD\nqr7R6fRdrPF+Rqas64ycgO2VPgUcAbwOuAH4X1X9HfB8VX0J1hnJRsHfVNU3A/lZiZdiu/e9VlXv\nBl7pwt6A7ZNfF2eXLr9W1bdgu7/1O10PFZEScKeqvkNVf+/JeIALG+A2Vb0cK8ePxDpB7wcuo/b+\n97jbwW6PI4zcA4E9l4OAt2GjJWX6NrGBJriR+2R27rZC/RNwnog8lNqOaQBnqurTRWQNMOCuZR/8\nEPfz3UfAj1T1H2nOfV4Y+Vefsp3rJtzxJDYT4Mu4VVXXenF6ILAN2xjGR52MvN55Nz5jubDXNvCb\n+YmAr6nqVzz3Wdx80gbX/DAK1He4NmMzB3fnPS11QuMeCOy5rPWOwzux7fMyEfk7bIp5bXZRRB6H\nTWGvAr6NjVhfLyKfBP4qIpdgI9gf5ANU1VtFpN9NzY9gI/Nr3fkK4FWqOt5Ep+uB97pd29r6OpyT\nmYrIe7BR89uAq4BnAZeLyLHAP4vIR7AOyZuBxzQJ8uPAR0VkM7bf/j+LyEeBMeBrM/iZEpE3O/cf\nFPsCW0lVX93EfdYxOFxE3gbsh+2wOIntZf8AFxewGY/1TRNiiRLecw8E9lBE5O1Yoz4EPFhVH7eL\nVQoEmiIizwY2q+p3Gtz7oqqe4Z0PYV8i3CMX1YXGPRAIICLZs95AYNEiIhH2dbxvt+H2cOx1zz92\nX7PFR2jcA4E9FDflq9gzyuWqet4uVikQCHSI0LgHAnsoInKwO5xQ1Y588jUQCCwOQuMeCOyBuFXb\ndcavqlftInUCgUCHCavlA4E9k8+7/z7gQqa/+hQIBHZjwsg9ENhDEZFzgVOAD6jqj3a1PoFAoHOE\nxj0Q2AMRkZ8CvwD+BTc9r6pf36VKBQKBjhGm5QOBPZMPuP+9dqkWgUCgK4SReyAQCAQCS4zw4ZhA\nIBAIBJYYoXEPBAKBQGCJERr3QCAQCASWGKFxDwQCgUBgiREa90AgEAgElhihcQ8EAoFAYIkRGvdA\nIBAIBJYYi6JxF5GPdiPcJC6vTuLy/bsRdiAQaI8u2neUxOVjuxF2ILC707HGXUQOFZFPiMgX3fnZ\nIvIxEblORAZbeD+gU3rkeCjw9FaOkrh8RJfkBwJLgkVq373Y3vhNSeLyEUlcli7pEAgsSjrWuKvq\nn1T1PO/SM1X1fOAG4PROyZmtWkA7Rn1WtxUJBHZnFrF9t8NZLJJZykBgoehmgc8Mbz1wYP6m6/nf\nKCI30r2efbuNe2CBSeKyJHF5ZQs3vUlcbjUqJInLQ0lcLnVOu0AbLAb7ng2hcV9g2rDvlnWAc9eT\nxOVlndNsz2AhCvxBwF/yF1X1c6p6qqqeCtw5HwFJXC4mcbkvd62PDsYvicuD/tReEpf7krhcaMNP\ny8ZpD6UEvDp/MZdejwfOaSOslwCPaeYgn3+7C64CXMxlqOv2DdPKBc72BmYRRCtbLfkdxCQuF/J1\nSgM//UlcXrY7lqu5kMTlgVnGtaF9e2GsaOSmAY8A/r6Fbr1JXO6ZhW6Lhm7Zdyefua8WkWuBY0Tk\njcCXReTDwLOBL8013DZHZCcCL85duwA4DpB2R3XOoGf6Ut7rqR+hnA88MtNxBhmvd7+8nKb65O/n\nKp1pfl3PVpq5mcFPRzt3HRo9++kVAUU/TzK9XZqL567H72y5RtHX5/XA3t79KBdu0zRuRIN8KswU\nZjthZe5z/mIalKGFZlfadxKXlzM9DR6J2XjWuW9VlqMW8p4BnOqdHwe8NPOTK28ZrwYuBvbN6dvK\nvuv0nW3Za9NN1IXG7hLgsHmG8XqsTIObVc3FP3I2XvDsWbB6YMZ6EXv08ncz3ffru1nYd08+z+dS\nT2Rus3xvUF67Yt+dfOa+UVVfpqqHqerbsd760cA4MKcVrS5zL2twPa/3PsA+uYzox3rrAlzmKoid\n/vNhOL9PAM50BStyDYTf4/f99GKjT5yOl82gWyOZ0+KU4zLXS65z784byXk1cLQX/8u8gpzFw2/8\nI2y02+mVxpc169k7PTK9C9k1T89Gabc3cLZ3fgHwN8AbgEOcf8XK2tFZfIH9mJ7OmcwIeBz1ay0u\nc71/oZbG4qdjG/E9GXi2M969ffkzlDn//DIX3qD7H3bXp5VTL44LRpfsO8Irq951ycWvUVnuAbKR\n9TnACS3KUjGJy3s5eQUvXzP77nNhZhSwOgQsb14DHJ4Ltzen8057bVHxvwD425nsO7G3fPzGX3yd\nnZtezw859wUsT16Sc1PwwvPTalpZctcLvjt3q6dZGZyhXs3nSX4WxbfT47CB2lOA05wfxfLmsib1\naAno9XRbmbv/Yhd2nb98PZnT65VYGc/c5suPH06+rWhk3+cCxwPPA05iBprUhbOi28/ct2NGM23a\nrk2umOH6lS6hM7LR0skNdMgy7LW+f/fz+Vvg0cDBTu6V7vyKJC6vaKBDCtRlgisc+XCbyWzGJVg8\n/QJ35QzHQ8Cz3C9//0pswdNJuXv7Uqu8OsnDmtx7CJAtysrS7bAkLu8HXM7M6TOcOy45/6cDT8by\neRAra1cCT8PyJ89Lk7h8uHMzxPRp3XOpr3ge7q7RRDe/QRgCysCbsM6Dz2v9MJK4fD+swZiJ7F5+\n0diJLpyTmvhdCDph31l65CvVZ2KPZPJcjFX6T81dH8TSI3vt9UrMdvz67TlYZQ3OvhN7TTarYw7N\nhZlSXz/2Y425b5NZGVOsLF6Z1EbLzR4DDGB11UVJXD4Mdja8WXoUXVhZGKdRq5My3ug6EFcm9rbP\nq104Wf31FGozCqclcXmAWl12FtbIALwcaPS20JUunIuTuDzkyT4Ls4kryXXqcnHIOMML7wJ33Kxj\n2o/ZUZ/T6w3U20BdPZrUntkr9nbU4939fOdqH6yuyPxJEpf7vbAuZPqsxCD1dUSWt1cw3XaPZXp7\nlU+L1S6M5cAJNMDF50rae1zRlJmmoDvBD1X1ByKyL/AerLeyExE5m9qIrK0FN0lcfjTwJHd6HnB1\nFpz7Pz6Jy8cD17rz/DROMU5GKt75Wu921oD3etcOdv8Xuf8XJXEZ4B+pZZAvY40X7n+1iMsb4mTk\nHc7tFPAPWC/xkJzTY/O6ZsdJXL4M+IDn9ugkLh+ddwc8GPhJLr4AJ7nK6G+BbwJPjpORvJtmcVgL\nvDlORqpJXM4qx5WenKuBS2fwl3EQuYKexOWn4B55OFY7P9dg6V1215fl3GV5t8oLay9qlTpYJQ9W\ncfS5cCfctQNd+CRx+clY2k/6emfpk8TlbIV4n+fmIZ6cI3N+hhrEPQvXn0HId7j9GRff74lJXB6O\nkyTbIC0AACAASURBVJEv58NbIDpp3xFQTWx2ze+EfzfnbhAbfd2GdSwyssct+UVXL0vi8j7ueL8G\ncp8Idel6hDv+twb3oPZWwBpXD2QUMBvLdAQr92uTuHwRMAr8D/BbrNO+wblZTq3z6DcML3f/z03i\n8h+AsexGEpcv9txlI8ez3L3nAg/0FXP6b6DWSJ2J5cd9XtyGfPe5OmAIeB31ZLY+5OqbM/ybuTTz\n64MsD57q5Yvv5z+wumAr3ogZr1FuRE7eY93/me7eacAxQNWd7+/uHwS80PO3AjgnictXxclI6gYB\nReo77/7bIss82W/D5fsM9n2kOxwCKn6YucGh73c4icuXxcnI2xpGug1Etd23SeYoQKQEfE5Vz2ji\n5ka38KaOBglVpFZIx7BGMaJWqU9Q63Hnr/mscGGlwKZ249IGQq2CGcdrGGagSq2Hvx4rcGC6Z/pv\no/1XfubDz7CR6mYg6w2vi5ORabKTuPwG4J1YD/ObzGx8U9QbR56VWPwrwJbcvT5qjfR9wA6mj6Zw\n/rMKddSF5fMXGqzmbsJWrNJNqD0bnKS+E1eh1jH+UJyM3NPIqNvkVuCoNtxNUN/xBGA2nbFu0GH7\n9u1nErMhgNuxztgUtfzNZp38axmD3v1NNJ7FmStDWBmsYOUN999qcd8OrM7ai/p6rFGZ7Qa+XbdC\nqD3ymMTqqV5qjfM2LF1bvY20AounYh2MqnevSM2+dgD3NtFlP/e/A6sL5sNtwAOAG6mtsRihNmDI\nc1OcjNyUxOXXYvXCbLmGFosBHQ3L0Hzsu2uNu4icjo2yVwAfVtWbmrhtaPxH9pTujKAwEMnyCKSI\n9GSlqYJOjqpuHRRZ1oP0AUyioxvT9K6VUbTPgMhQFrNUmQJ0XHUHoEORrEqhUkAKiqpCmkKlokz2\nCL0lpE+tQCqgavImBJhSJnZouqVXpH9QomEARatVqJSQ/h6hN4VKBMVUmUSQMdWtVaUyILJMoCAC\nY6pbK8pEnzAgSCGLk0K6XKK9FFKBSCy6iou3CNGE6vaqMlUUejK/Y6pbp1QnxJlaBFFJpD+zvBKi\nBUGLCAXQgogWQYsiPLBYHNsrKlRWF6LRR5R6v0KtM/HOOBnJKjC/Fwxm4MuAW4CHra9UTry9Urnf\nmGr048mJ8S1puj2CPoFeheIkuipVRISCmGELIBVlKoIexBK6CpUqWiki/VV0Si3/dIfqFgVxiSER\nFBSkV2SoAMXUNfxTqhMuvyqAVqFaVdLI0r3k0qaYpWsVplR35jUK6ajq9hSt9Fn6SREplcTKmEK1\namHjpZMCFJCeXqHfC0sF2J7qpgo6NRxF+xSRoiur46pUFNIKWikghZLsrFB3GmVFmVRIFdIx1W0R\nFPpFBgCqUP3PifG6EdBC0Qn7fmBPz10CDEi0PIKoIBQjiFIr+2xLdVNJ6OuXaAhQRfXuajoyFMnQ\nMolWY+kbVdFJgEllrIpWhiRamUI1goIAKVQVtIKOR1DotfTbmXeujE2pHafbNN0YQWGZRKuyclJB\np4pIsVdkoGr5VayiU4IwqTo2iY73iQz2ICVAJlV3jKvu6BHpLbryXbX8rvaLLCtYp1cj2TnIsLQy\nm5ioolMgiJV5mULHJ1RHnT4IQknoc3UEYqaRuUcgEtnZEVaACNKDCsXJXhGWR9FePUCviAxKtCFF\np0ZVN25M0ztWRdERAyL3SyHdUK0WxlUrU1D6a7U6nqJphPRsSdO7LUiz435Xt6nTT0EKUJpEx4Go\niJQUSNF0UnWsgPQ6O50qIr1bNd2Y+UMRxOq+ApQE6a2i1SJSmkInYGdapoqqy+uoxzqaFKAnQgpY\n+JNVpeKqFwCdQMemVCd7RfoiKBaQqEfoFUQU1Yp1GLMSoiJ2GEFUQrJOo7o2gkl0dFx1dFBkOKtz\nK8pkBZ1yuk4iSK/5resMVdFKap07HXPtU5+VTwH4ycREux2yaXRzWv7fsWdFW6n10GbFG5cvF0C3\nq/7fhOpmheqOVDcVhNK+UeE4tcK14+60+j8R9JQLxdPURr+FO6uVr4+qbukVWR5BoU9kZUlkFcC4\n6m83pekdERRKIgMKUb/Ialex6H1p+hdrZ4gEoiL09ogMChRXRtFR2LMT3Zym/zuljPcIgxEUUtAt\naTqSQqUIfZFIsQ9Z3h/JfgBV1Q3bVO/qFRkeENnfrrF5TNONEVLsEQYVmFT9y6Y0vV2BHmtUNLWi\nqQWR0qDIPgApVCZVR0siQ0MisUJJvYZlUvXeFKZSq2ykClRUpQJSdf/b0rT42dEdwwryy6mpw48t\nlR5YQtJnDQzcflJv35Fb94+Ly6PoF8DncQ37hOrDx1QPvKtaHXrvtq1XTKry3Ynx6IBCYesyiSoP\n6ulZFRcKaQRpFbYVQHtF/pRarzutqI4BVNysSgQU3OLHfpEhERGU6pimmyORwr5R4ajI6jOVrMNl\nlXEVqG7R9I+pajoQRcNFKBagWPz/7J15mCVVebjfr+7W+6wMMsUdAdkRDS5o3ElckqAJigZC4goi\nhhgx4jKyzQgoLo+IomwKioqAuEQwapRNJYnmJ6AERBGEuRbb7Etv996q7/fHOTX3dPXt7tvd9/b0\ntOd9npm+VXXqfGf7vnPqrCIlAYrG6AQCWlM2gNZjqMYwGkCpgPQKBGobUl0iK7pEwgQ0QQerqhsT\n0CHVqK6MFIV+TOMgNZwB1gADbFd9tK5apWH0liwKgoMAasqDjyfxXTko9ojsgcnf3gDyCjKiurEO\no6rGv0DIF5E+BekRWZkT+tWk/4aaMrQ4CCab39BpZq3f5wwsVgFieGxLktwPMKzJxhiSPXO55+bM\nl8zg5iS5fUh10z65/OsCMcNWm5Lkjs1J8mABenMipQL0dEuwEiBGK+vj5B5FpSTSD1ASWZJDSgCD\nmqyvwXCjoiRfRPoFZCCQg/IiSwDdkST3D6luyIt054xusSNJHh+FwZwpZ90FkVKfBKtARWHj1kQf\nDoTiIgmeZsvV6JAm6wEKIn2AJMrj65P4ftuAL6TW3jZeg74g2AMQVVTRJBDJ94jsLUZH1LZEta66\nqQ5D9jqx/6UfJEkMVW2UTRlULT4Wxz1bkiR4qF5/bIcmI3Wlt4ouCoAlQXBwHjlMgI1J8odR1cFF\nQVAfCKQqoEuD3MYcJCtzucMCJC+QWJ1M6qpPDKquT/VTTMN5NFatiogGNq17RBYVkCKCDibJeiBe\nkcsdkoduEBX7rkJdoSagQ5o8FMNwUaS3gPkYsnpIznzYFQCq6OZYdTSB2ojqlgDyPRLsYXU7ECHI\nQ1+PBKuAkqL1IdV1QDKqurkGQ3noyiFdNk/ENgjsdxUyrLqxhg7aRlSQF+lbLMEhNq/XP57E/5Mo\ncV8gTxEIctCVsx8GVdVtVdXtafkXkKLIEjEfKcsLyGKAOrp1VNk0EIg7NDFtOvnl/kZgi6reKCLX\nqeqEu8C10G33MOPHol3S7t88xtgqaetrZrTSzTZRd/SHmd7EuabdrZatwA2MHevpBKNA6Rejowf+\nvl4LH4vjRV8c3HFYbAxC7pB8oWoVTgBZF9cLddVYRPSFxdJvn1cs3r84yA2/tqf7g90SvL0FeZtw\nxsUn4XGaj5U2o1mX/X3Yse8OcTFjx/Onw79jKsfZ0BtGlV2yTK7N+t0Kd2Mmazp14Yy5heYT9ly2\n0JiH47IWO7emHYRRZU2TdHgM2KuJ87nmAsyEtvvDqHIttJxn7hDjdzDLDNtBq3YjpZWhCHdoNEs6\nVJnyMJPXQy4PMX6S5nTpCaPK+2f6cidny++NGcuAsWMt0+Vn2PHYdPzB/r3ePv8p8AfgQkz3RhWo\nWTepAfilvV4LfAGIneu1wCXW3W+B22hM1nHX7w7Zd66111XMuDPWD8KosiaMKmkX26cxY9ZrgPPs\nvVuB25131mIUKL1+PBVm/bowjCoVGukI5ovpo046/M7e/6r9e73jdj3wTfv7fCvjK64M+/NXwF1H\nlkr/eEJv363vHVj0b/ftFfb9dq/wvJv32PPq9w8M3PL+gYGbPzAw8J8fHBj4/leWLv/6gyv3Pu/3\ne4Xnf3nZ8hv+uX/g/07o7f3S/o9GkSP7ahpjkk84su6ikd4Adzi/N9nwr7Xxv9TG4deOm4dsGrjc\nDTxuZQxZ2WsxZSPl606eg2k43eI8/zbwNcxwg8sjwPfTdzN+1GyY3fHCz2XKVsr3XD/CqHKX8+wW\nYDTjd1oG1gC/d9x9zv6+fldV7JZ26ff1NPQ0nQh7CY10+AmAM3EwHUL5FPDf9l6STXP32il712Fs\nSdrw/4P9u9F558P23jbnPVe/07DexvjysCXzztrM9fedeH/JefZdN0HCqHKZ8851jMUtU/8JPIiZ\nH3IT48scwJeByzBpthbT03JNWrYyY7pr3TQIo8qIvXct40nlRNbtTn/CqHKl4y7CNLLB5OUDjE2b\n/8TotMvHM9e/tv5+Bvgv4B7Gpi2YBsgaxqbBVc7zPzjuf+7cvx3TmFqb+Xc/xhbcAjxh3W4H7sjY\ngp1k9PtqGps3PeQ+p2GXPmvvfcNeb3bC/gBNJiNPh05/uW9W1ZtE5FpVPT7z3J1N262q2WVsHo9n\nnuL12+OZ33Sycu/FdFuOAD9T1a/NwI+bcZaBdIiVtGF7TC/Dy9gFMmAXVZxev72MeSxjruTMhYwZ\n63fHJtSpmfn31ll6M9xsrK6dTDQe6GV4GfNdRiqn0zKa4fXby5ivMuZKzlzJmOm7nRxz93g8Ho/H\nswuY75X7NV6Gl+FlzAs5nWCh5IOXMb9kzJWceS2j4zvUeTwej8fjmVvm+5e7x+PxeDyeaeIrd4/H\n4/F4Fhid3H62ZUTkSMxxjhVVPd3eey+wL2bnuVMwOzZ9ArNhxlWqemvWjbYwxiAix2B2llsMqKr+\ng4j8ArgTeERVPyoiK7Oyphmft2BOJVoHfE5V72klPtOUcRLmhKylwHmqene74+HI6sUcgFAFbpvJ\nsqcmfh4DHI05jOGLmCMUHwF2qOrp7ZApIi8DzsUcznItZunKUZgdAdOTt2Yr48WYE9HymN3wtrYr\nHiKyH+Yo3EWq+nq7dnzS8Gfd2FntuxSv3/Nbv628tur4QtFvK2f31HFVnRf/MNv6fdL+LgJfs79P\nxRzjdxbmWMUAM8lgnJtpyjsNeJX9fStwJfBmez1G1gzi8ibMtrFXYM4RnjI+s0i3I4CzOhEPR8Yb\ngdfY39e1Od+XYJT/m8DlwPvaJRNz9vn3gS9hzvr+hr1/tPW/bfHCbLH5jg7F4wb7d8rwZ920M69m\nmT5ev2eWbh3X73aV0wn8XRD6bf3YrXR8zr/cReRw7PapDm/LXC/DbDsKpnW8t/1XUdVEzAFIzdy0\nJEtVnwRegTmXHeAvrb/XiciNTWRNOz6qerWIPAOzheAnW4jPtGVgtj79Vxp72c84HlOwN2bLR5jd\nVqPNOBOzperdNpyfsvFth8zsmePpl986GmevtyteJ2DOALiiA/FIaSX8zdzMGV6/d0v9hs7p+ELR\nb9jNdHzOK3dVvYfGObo7EZF9nMuNmHOPwRxA8Gvsudwism0SN63KehHwc1VNrLt0P/jNmHOgs7Km\nHR/Lk5hjUVuJz7RkiEgB013zaVWtzDYeU5CeiX43bZqnIcYaXQB8X1XvdB6laTZrmZn0KNE4WnOV\n9Z/ZygAQkVXAVnVOfKKN8WhCK+F33cwZXr93S/2GNpfThaTfsHvq+HwZcz8Qc9LSYSJysqpeLiJ3\nishFmEz7POaAhAswh8N8Qc1Rglk3rfI2bGtYzPGOF2G20dykqpGIfNGVNYP4nAw8C/P1sXaCsI6J\nz3RlAB8DDgTeabfx/HG74+HwLeBiETkauHEW/ri8C3g5sEhE9gf+HHPoSx5zcMSvZitTxp45fjGw\nUkQuwRjGU62zdsTrROwhFSLy5XbFQ0SWYQ78OUJEVgPfaSH8zdzsUrx+z3v9hvbr+ELSb9gNddyv\nc/d4PB6PZ4Hhl8J5PB6Px7PA8JW7x+PxeDwLDF+5ezwej8ezwPCVu8fj8Xg8CwxfubcZEXmLiNwk\nIheJyJUisp+IrBSR9zRxu0ZEnt6BMBwvIleJyLOde7NeCLsrEJEbWnAjzu8DROT9IrKPiHxyivee\nLSKntCOcnj8NvH63F6/fnWNeLIVbgFyqqjfZZQ4XYpblhCJyCGZJ0CPA1cCLgBUicg1mB68jMOsm\nT8VsSdiPWeZyv6peJSKfwOTZOuCzmCUUJSAHnKaqsYj0Yba/vA8oicgdmCUUV1ll2Ab8WlUvE5G7\nMTs87YVZzvF8G/ad21iKyBmYNbz9mC0kP2vjsxr4NHAw8BqgC1hj4/RS4CEgUdXzHb9eBRyPWZf6\nMeBlVuaA9fcYN87AA8ChInKmfb4WEGC7qp4pIvcCX8Es4/mdFXMKZnlKN/B8ETnfxu8k4LuYXb72\nBy5R1V9ao3zp5Nnp8YzB67fX73mP/3LvIKq6EbPXdMoKTMH/uqreC/wM+Lyq/sw+rwEhxggAXK+q\nHwBeKSKHAlVVfY+qXojZgWsfYAvQZ99DVXcAt2HW2j4K3KeqFwBvAK5U1VMx608BnlTV1TZMN2OU\n+5g0sNZYvQSzl3INOAQ4HbgOuEtVfwu8SVVPwhirtJX8A1U9F8h+tbwDs7vXv6nqE8C/WL+fxOyj\nPSbONl3uU9XzMLtDdduw7iciReBRVb1AVX/nyDjA+g3wgKqeAfwGeB7GSF4EfIjG+tARMTtceTzT\nwuu31+/5jP9y7yC2ZV9Nr+1WiQ8BJ4rIM2nsqATw96r6tyJyDtBj76UHAoj957oPgDtU9TNMzlbH\nj+ymBunOVqP2dxXzpeDKuFdV1zhxOgjYjtk4wkWtjGy4s25chjN+r2nybvpOAHxPVb/ruE/j5pI0\nuef6kWOsQd6M+bJ4IvuSxzMZXr+9fs9nfOXeGU4RkVdguqDWpDdF5ChMF9dS4EeYFu37RORK4DER\neT+mhXt71kNVvVdEum3XXQXTcr/UXi8G3qWqI5OE6evAhXZXp5ZOj7IyExH5FKZV/RHgw8CxwBki\n8lzgqyJyGcZgnQu8YBIvvwBcLiKbMftxf1VELgeGge9N8E5NRM617i8Wc0JTUVXfPYn71HAcKCIf\nAZ6C2Xu6itnr+gAbFzBfROsmTQiPZyxev5vj9Xse4Xeo8ywoROQNwGZV/XGTZzeo6uud6z7MSWV/\nspNuPJ7dCa/freMrd8+CQkQCzOlZP2rB7YFArKoPdj5kHo9ntnj9bh1fuXs8Ho/Hs8Dws+U9Ho/H\n41lg+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3D0ej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H\n41lg+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3D0ej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H\n41lg+Mrd4/F4PJ4Fxryo3EXk8k74G4Xl5VFYPqATfns8ntbooH4HUVh+Xif89nh2d9pWuYvIfiLy\nRRG5wV6fICJXiMjVItI7xesr2xWODM8AXjOVoygsH9Yh+R7PgmCe6ncJ+OupHEVh+bAoLEuHwuDx\nzEvaVrmr6kOqeqJz67Wq+nbgOuB17ZIz3WABrSj1GzodEI9nd2Ye63crvIF50kvp8cwVnSzwqeKt\nA/bOPrQt/5tE5CY617JvtXL3zDFRWJYoLC+dwk0pCstTfRUSheX+KCwX2xc6TwvMB/2eDr5yn2Na\n0O8pbYB1V4jC8kD7QvanwVwU+FXAH7M3VfUaVX21qr4aeHQ2Amzmd2XudQO52fib8a/P7dqLwnJ3\nFJbzLbzT164wLDAKwL9mb2bS6y+Af2rBrxOBF07mIJt/uwvWAM7nMtRx/YZx5QKrez3T8GJSWxCF\n5aLbQIzCcs7akMne6YnC8sDuWK5mQhSWe6cZ16b67fixuJmbJjwH+OcpwlaKwnJhGmGbN3RKv9s5\n5r5MRC4FjhCR1cB3ROQSTJfYt2bqbxSWSy04ewlwUubeycBzAWnRj1ShJ6qwTwfKzvXbgeelYZxA\nxun2X1bOpOHJPnevm71rDZNM5qbJO4UoLLe1cddqOlsmMhJuegVA3lXaNNw2zVM/BCi4eWcrRTc8\npwN7OM9zGfeTpnEzmuTThH624lfqPvNeSJMyNNfsSv22X23ZNDgSeId93kpZDqaQdwzw6gn8L2V1\nzPJu4N+APTPhnUq/x4R3umWvRTdBByq79wH7z9KP02n09AiMi39g0ydri3OT2UXgOOCVEz1303wa\n+j0uz2diJ1K3TrwKUVh2G5sd0e92jrlvVNVTVPVpqvpRTGv96cAIppKdNjZzVze5nw33CmB5JiO6\nMK11AVZHYXmR+37WD/vuK4HjorCct24kkwmu/0XM1yc2jKsnCFszmePilGG1bSWPcW+vm8l5N3C4\n694pyEEaFzcsmMbJkVOEY7qsnqxlb9MzDWMuveeGs8lrewAnONfvAJ4FfBDYx5F3GHCYVR4BnsL4\ndM6nsoCjgOMzYU8bDGkai5uOLcT35cAbrPLu4cq38cs515Lxc7X1r8/+XWzvjyunThznjA7pdw6n\nrDr3g0z83LIs9lkBM6EOTO/OS6coS/k0T1L9tm7TCqSLhj6DKZ9pb+BqTCV+UMZf90vfzd/V0eTD\nRG8GXuTEc4x+R2F5WabyFzfM1k3JeYeM+zxGt9+ecbOz/GfSalxZcirZrP9p43rC95pdZ/Ikmzeu\nnj4PeBvwN8DfuXJtvCeyo0WglOpVFJaXZJ6fCDw/+14mfNn4/AtwuON2OWPLj+uPZBoj2bRYDbwR\n08P4j8DLmIBJbOG06PSY+w6MgozrtmuRMye4f7ZN6JQ0UV/eJAxphr3Hfd/+c3kRpmCtsnLPBl4A\nnOUYWpcEGJMJtnCcPYHhbSYzLRTN3L+viV9nT/C7FzOp6XVNnp9t7x+VebYCa7zaXFH82STPnkGj\nhyVNt/2jsPwUGmnejIHM7yImX4/FzJZWTBp0AWcBf4vJnywnR2H5ICunF8h2u76Jsd23z7H3mCRs\nrhHvxfTunIEZUnB5rw1byr6YCiNLOpZ9WuY65WXWn6PYtbRDv9P0yJa/Y4C/bOL+dIzRPzpzvwd4\nKZAuez0bozuufTsBONX+PtM+35+Gjdkv42eq32nYujB57epkWsYUo2NnRY2v5Z2N1yb61Y0pH+9J\n5dqGQVrG8lZOWhaPYbx+pA2Is6OwfCi2vERh+anW7V9hdBxMJdkDnGlt2XEYewemu/tQxnM2phy/\n1zY4U9l/j6mkzibzcZCJQ8qxjn/vsL8nszfdQB+m4XYwplJ0dWCMHbWVONbNMzDpejZj9RLMR8Ir\nnfckMsMuqV+nMb5XooexQz9p3p7JeN09kvH1VTYtllo/+oEX0wQbn7MZW1/NiEnHjGfJT1X1dhHZ\nE/gUprWyExE5gcYXWUsTbqKw/AIaGXQScEHqnf37wigsvxC4zF6PyeAoLOfDqFJ3rtc4j9Mve7er\nZZX9mxrat0ZhGeAzmErmJRn356R/o7D88yni8qEwqnzEvhNHYfmTmK/PfTNOj8yGNf0dheUzgYsc\nt0+PwvLTm8Tv6cAvMvEFeJk1EC+IwvIPgVeFUSXrZrI4rAHODaNKHIXl1DguceR8DPjABO+lrCJT\n0KOw/DeMNRzL7Dufx+RpOjzSl3GX5sUSGl1+e9Aw6gD/YP8eimnprwFq9l5o/ScKy3+N+SIddcOd\npk8UllOjVXLcPMORc0jmnd4mcU/9dXsQsg1ut8fFffclUVheFEaVb2f9myPaqd8BRgcGGGs0f5xx\n14PJkweA7c79dLglO3b5z85HwArG8yoYk66H2N/fpWFnznHcp43nc6wdSMlhdAxsPmPK/RqMkR6y\n9uB+e/8J66YfU1HC2IbfO+3fE6Kw/HtgMH0QheX3Oe4+ZP/+vX12AnCgGzAbn/U0KqnjMb1aW5x4\n97ruMzagl/HdxqmN6o3C8uE0KnBXZoprD9Lhi6OtXmbf+SnGFmyjkZ5gGioTkpGX2pLj7LPXAs8E\n6vY6LYurgLc67w0A/xiF5Q+HUSWJwvKBmPrRrSPdod8+R/ZHmFy/D3PeqeP0EGU+Dt13+6OwfGYY\nVc5rHuupEdVWV5PMUIBIEbhGVV8/iZub7MSbMTRJqDwNBR7EGOUcDaM+CsQYY9GVueey1PqVABta\njUsLCI1GwjBOxTABbs/CH2mMRXXRCP9WWl/yMxvuxHR3b6URh7VhVBknOwrLH8I0rM4Gfog1kk2I\nmXwi0zL7vA5syjzrxuQTwBYcA5chhzGS0CgTLo8yvdnaOzBlLMJU9tjwuUqe0KiEPxdGlfXNlLpF\n7qP5l1OWGmO7jQGYTmOsE7RZv139GcXokGBm5B+ESYM0f9PKyr2X0k+jV2YDzXtxZsoAJu9rNMrk\nCA19nYhBTJxS25O1Y53G1eupcO1nFVP+u4C97L1twMYW/FlKIx8eZ6wdztNopA8CT07gh9DQ30GM\nLZgNv8d8od9EY46Fq+tZbgujym1RWD6d8Y3HVriERmNtMpqWodnod8cqdxF5HcboLwYuUdXbJnHb\nVPkPKRQeD5CgR2RABPJIUVEFNIH6kOq2bpH+IlIEqKIjG+PkiaVBsLxLpE9BA5AYagpUVQcVtC+Q\nJbFSywl5QFWNf3W0mkdKRZESoIoiiCqqdaUKUEdHB1W3FpHuvkAGABIlTqBeFLpySDGBOAf5xCit\njKhuj9FatwT9gVEcGVbdVkerXSI9YiuKWKna8C0FEoFAkABjnMS2AqSmOliHeh7ygZjKc0R1e01N\nY0LMv1zJxAMByQtJDkGAnEAJSfqDIAbYP58fWR7k4qVBMPS8Uuk7NBoTHw+jylCaH04rGIyi9QJ3\nAUc8Uq8f9Yd6fd9hVfmv6kh1c5JsD5AugZJCPkaX1BURyAWmQg4UqEMtgLzsvNZaosR5oStWagkm\nwwc12ZpGTYEAySkqJZHeAMknaCIQVG0axmhdTd4kMRoHSD4nFMy7Zi6GmjSvqYmvAiiqJr+IL2ja\npgAAIABJREFUSyJdAQR5pFAQukyZIEmUum2V2XdQm675okgPYysTHUx0Sx2tLg6CFVY2dWUkgVgh\nidF6ALmiSJfY+Np807oJX6KQDGuyPUDyXSLdAAka//fo6JgvoLmiHfp9YKHwBCA9Iv1i0jknQpAo\nSSDI9iTZVBTp6hbpVZMuuj6O/9grQV9/IEuNfksQozWAqjIco/VekcVGByUQCBI0BpKaMhoIuRIm\n/Wy+I6Ax1BVNFJLtiW4OhNyABGkPkFrbUCgK3QnUc0g+Ruti5dZUR0oivQUxPT9VZWhEdagoFHNI\nASCBeoLG3SL9OSQHEJhKTnd2z4ixA3W0KiBi1b6uOjqKDjszSKUoUkpth3Vrktte27KWFicNBC3n\n8tWiCAMiy4siFEXoE1lv7enGjXH8yNJc7uAekafGSvJkEhdGVOs11cLjSTySKHEgFLcmyXp2mhqk\nW6RfkJyCqE3WAClUVYcFcjmhmOpjFR3JISUFravWCiKlbUmySXdGDcHaybx5r2htQjG1w8amq6qp\nC2Krp0WAvFCwtpMaWk2UGEdXq6rDNbRaFOnKQT5ApGj0GwWNlbqTbikaCLki0q1Ompq81qERdKhX\nZFER6QJj12K0pqCJmnqgKNKdHY+I0bq1AzqiOghoCekRW+7+tzq6hBnSycq9F9ONWgVuU9WvTeK2\nqfJ/ddkejwIMafLwiOrmBOJh1c0BFFbkckcaxUXXx/H/BlDYO58/RhUVCKI4vmlYk60lU+iCkrC0\nJLIMYFR1w6YkeSSAoGCMsXSLLBfIKejWJFmn9utMIMiJFIvQK5BfHARPN4VYdUuS3FuD4QL0iVFW\n3apJpa5az4uUclAoivT3iuwFUFOGd2jyWElkUZ9IqEAMoyOqG4B8AekDGEW3b06SRwDypnLUtEDl\noNgbBCsASVTrVRgsQl9vEOwNoGoqFgWtotsTpZag1Ix7YpAE2JEk+WHVYFg1/7t6bUCBX9Vqezyr\nUBwpiSSv6+555Kiurt92i8iiILgbs1nJcTb9njOoyarH47j3wu3b9q0q3Do6Invnclv7JIgPKRSK\n5VwuDkSSWHV7TkQLsCUxre6krjqMMahVTCOGnBhj1y3SH5iKtz6iuiWA3J653GHGeKG28lM1Dac6\nEG9Jkj8kEPeILMpBPieSK5hWsBRF+myDSqvKhsRU+tUEqnkoFEwjUMTmdUlkzxKyEqCODlaVjYAO\naRLVYKSA9FuDvNMApf9EkG1JEtVhVEwcpFtk6SIJDrINlx2Px/EvnTwkD72B/SIfUt0Ym687UWOc\n80WRflXoDSQMoEtBR1U31GF4sQRHnLhpwy5ZhtUO/b5u+R4RQKwMbdXkfoBh1Y2JoitywXNySI+C\nbEmS/xvSZONT8/ljA9tLszlJfrYpSR4siPTkoZhDentEVmIa88Prk/j/AIpGp6QksjgndKkiQ6pP\n1NFhnHQuifQryCIJDsoLAwq6I9EHh1U35IWuwH5VDao+NqI6lIN83lQOpX4JVmEaEfWtqg/nobgo\nCPZXCFSJh1WfVNCimLkjMdQ3xPFvFZIAyZlvEIOC9AXBCmvbEvuBkusTKWMawIrVgboyXEeHEuce\npjJRUK2bspRW/sEO1cLjcdxTR4OhREd3qI7UVHtr6OIAYUkQrMhDLhDYECd/GEUHBySoLQqCqoAu\nDYJNOUj2yuUOy4nkMbqbYCrEoUFN1mN1FEhiGE2gJk7jpUdkUUGkKCYt16PEe+SCg/NIt31PRVCg\nrkpNBB1K9A8xOlwU6S2Yii9XFFMO7AdEHqCKbjGNZq2Nqm4JkHx3IHtgvmcECApIX7fIU8WkdTyk\nSQQko8qmGjqYh56c/SjC0e00f0ZUN1ZVB8X6WRDpXRQEh4rN68eT+L8V4j6RvQSCAOnKi+m9qKpu\nHVW20fA8KAlLQIIuYXlBZDFATdlWVd00EMgz3rhxw3RWII2hk5X7G4EtqnqjiFynqsdN4naqbruH\ngX0mEZd2Weawhhw7xjJDhph6De0PaD4W9GEmnnzVjMm69DZjlhmdOMHzdjECdN1ZHd3/gXo9fDSO\nF31hx/bDY5AdqrmD8/mqIPbzVqUSxwXbgtcXFku/eV6p9JslQTB8XE/vGdgZulOwAVg+pSt4jEZX\n4FS4XeUp92LmMXSKizEzamfCdzATpWZDTxhV3j9LP2ZEm/W7FdJho3ZwC+MnPGbZgumVyLKWsePw\nsyKMKmuapMNk3cRzyQWYVSm/CaPKddBynj0CPNX+/jbw2jaFp1W7kbKJxtDeRGSH3Fx+CTzbuX6Y\nyeshlweBp7XodiK6w6gybs5Sq3RytvzeQMX+zo55T4efYsdZ0vEH+/d6+/wnwEPAhVZODahbN6nc\n/2ev1wJXOM/X2n+XWHf3A7fSmKzzTSccg/adr9vrUcyEMawfhFFlTRhV0i7ZT2PGrNcA6aSIW4Hb\nnHfWZvx4LBVm/boojCoVzJhjyg+Ajzrp8Ft7/6v27/WO2yeBG+zv862Mq10Z9uevgDufVSy98bie\n3tvf0z/w3nv3Cvvu3ys875Y99vzSBwcW3fzBgYEffWhg4AdnDiz+3jXLln/t9yvD8363V3j+VcuW\nf+uUvv7fHNfT+5UwqkSO7C/TGJN83JF1J40JjwA/c35vtOFfa+N/mY3Drxw3D2LG+V3uAh6zMgZt\nHNdm/L7GyXMwZeoW5/m3MGm4jbE8DPxH+m7Gj5oN83rH/cWZspVyk+tHGFXudp7dDIxk/E7LwBrM\nBLLU3cX29/W7qmK3tEu/r8N+/QGX2r+X0EiH2wHCqPLdzHufAv4rlZ9Nc/faKXvXYhrM6Rj3Q/bv\nBuedD9t7W533XP1Ov4ZuY3x52Jx5Z23m+j+c8F/lPPt3N2JhVLnCeefaTLzdMvVDzBjyH4EbGV/m\nwOjhZZg0XosZe78mLVuZMd21bhqEUWXE3ruO8aRyKtbtTn/CqHKV4+5RTCMbTF7+jrFp858YnXb5\nWOb619bfi4E77PXaTNh/00Tv3HA85Lj/H+f+7TaMazP/7sfYgpsx8wXA1As/y9iCnWT0+yuYRhqY\nPHLL5E/t/c/ae9+w1xudsP8O07CaMZ3+ct+sqjeJyLWqenzmuTubtltVmy198Xg88xCv3x7P/KbT\nY+4XY7p8fzbZmNwkftyMmTHbSVbShu0xvQwvYxfIgF1UcXr99jLmsYy5kjMXMmas3x1b565m5t9b\nZ+nNcLOxunYy0Xigl+FlzHcZqZxOy2iG128vY77KmCs5cyVjpu/6k5I8Ho/H41lgzPfK/Rovw8vw\nMuaFnE6wUPLBy5hfMuZKzryW0fEd6jwej8fj8cwt8/3L3ePxeDwezzTxlbvH4/F4PAuMTp4K1zIi\nciTmSMyKqp5u770Xc/pQATgFs1PZJzAbZlylqrdm3WgLYwwicgxmZ7nFgKrqP4jILzCbqzyiqh8V\ns43lGFnTjM9bMCc1rQM+p6r3tBKfaco4CXMi2lLgPFW9u93xcGS1vNXoNPw8BnN05wDwRczpWY8A\nO1T19HbIFJGXAediNtG4FrN05SjMQUPpYQ6zlfFizIloecwBMFvbFQ8R2Q9z7OYiVX29XTs+afiz\nbuys9l2K1+/5rd9WXlt1fKHot5Wze+q4qs6Lf5ht/T5pfxeBr9nfp2KO8TsLOBzT23BNMzfTlHca\n8Cr7+1bgSuDN9nqMrBnE5U2YneGuwBw1OWV8ZpFuRwBndSIejow3Aq+xv69rc74vwSj/N4HLgfe1\nSybmjO/vA1/CnAT1DXv/aOt/2+KF2Ur2HR2Kxw3275Thz7ppZ17NMn28fs8s3Tqu3+0qpxP4uyD0\n2/qxW+n4nH+5i8jh2O1THd6WuV5GY0vPdZitLvfGtPwTMQclNXPTkixVfRJ4BeZcdoC/tP5eJyI3\nNpE17fio6tUi8gzM+c2fbCE+05aB2Tv5X2nsZT/jeEzB3sA99vdsthptxpnA54C7bTg/ZePbDpnZ\nM8fTL791NM5eb1e8TsCcAXBFB+KR0kr4m7mZM7x+75b6DZ3T8YWi37Cb6ficV+6qeg+Nc3R3IiL7\nOJcbaRwQsAqzl/Afgb1FZNskblqV9SLg56qaWHfpfvCbMecPZ2VNOz6WJzGnWLUSn2nJEJECprvm\n06pamW08piA9a/5u2jRPQ4w1ugD4vqre6TxK02zWMjPpUaJxFOsq6z+zlQEgIquAraq63bndtng0\noZXwu27mDK/fu6V+Q5vL6ULSb9g9dXy+jLkfiDlp6TAROVlVLxeRO0XkIkymfR5zuMAFmFN8vqCq\n1SZuWuVt2NawiCwBLsJso7lJVSMR+aIrawbxORlzgtUyYO0EYR0Tn+nKwByucCDwTruN54/bHQ+H\nbwEXi8jRmAMq2sG7gJcDi0Rkf+DPMafx5YGPYw6LmZVMGXvm+MXAShG5BGMYT7XO2hGvE7GHVIjI\nl9sVDzFHFJ8PHCEiq4HvtBD+Zm52KV6/571+Q/t1fCHpN+yGOu7XuXs8Ho/Hs8DwS+E8Ho/H41lg\n+Mrd4/F4PJ4Fhq/cPR6Px+NZYPjK3ePxeDyeBYav3NuMiLxFRG4SkYtE5EoR2U9EVorIe5q4XSMi\nT+9AGI4XkatE5NnOvVkvhN0ViMgNLbgR5/cBIvJ+EdlHRD45xXvPFpFT2hFOz58GXr/bi9fvzjEv\nlsItQC5V1ZvsMocLMctyQhE5BLMk6BHgauBFwAoRuQazg9cRmHWTp2K2JOzHLHO5X1WvEpFPYPJs\nHfBZzBKKEpADTlPVWET6MNtf3geUROQOzBKKq6wybAN+raqXicjdmB2e9sIs53i+DfvObSxF5AzM\nGt5+zBaSn7XxWQ18GjgYeA3QBayxcXop8BCQqOr5jl+vAo7HrEv9GPAyK3PA+nuMG2fgAeBQETnT\nPl8LCLBdVc8UkXuBr2CW8fzOijkFszylG3i+iJxv43cS8F3MLl/7A5eo6i+tUb508uz0eMbg9dvr\n97zHf7l3EFXdiNlrOmUFpuB/XVXvBX4GfF5Vf2af14AQYwQArlfVDwCvFJFDgaqqvkdVL8TswLUP\nsAXos++hqjuA2zBrbR8F7lPVC4A3AFeq6qmY9acAT6rqahummzHKfUwaWGusXoLZS7kGHAKcDlwH\n3KWqvwXepKonYYxV2kr+gaqeC2S/Wt6B2d3r31T1CeBfrN9PYvbRHhNnmy73qep5mN2hum1Y9xOR\nIvCoql6gqr9zZBxg/QZ4QFXPAH4DPA9jJC8CPkRjfeiImB2uPJ5p4fXb6/d8xn+5dxDbsq+m13ar\nxIeAE0XkmTR2VAL4e1X9WxE5B+ix99IDAcT+c90HwB2q+hkmZ6vjR3ZTg3Rnq1H7u4r5UnBl3Kuq\na5w4HQRsx2wc4aJWRjbcWTcuwxm/1zR5N30nAL6nqt913Kdxc0ma3HP9yDHWIG/GfFk8kX3J45kM\nr99ev+czvnLvDKeIyCswXVBr0psichSmi2sp8CNMi/Z9InIl8JiIvB/Twr0966Gq3isi3bbrroJp\nuV9qrxcD71LVkUnC9HXgQrurU0unR1mZiYh8CtOq/gjwYeBY4AwReS7wVRG5DGOwzgVeMImXXwAu\nF5HNmP24vyoilwPDwPcmeKcmIuda9xeLOaGpqKrvnsR9ajgOFJGPAE/B7D1dxex1fYCNC5gvonWT\nJoTHMxav383x+j2P8DvUeRYUIvIGYLOq/rjJsxtU9fXOdR/mpLI/2Uk3Hs/uhNfv1vGVu2dBISIB\n5vSsH7Xg9kAgVtUHOx8yj8czW7x+t46v3D0ej8fjWWD42fIej8fj8SwwfOXu8Xg8Hs8Cw1fuHo/H\n4/EsMHzl7vF4PB7PAsNX7h6Px+PxLDB85e7xeDwezwLDV+4ej8fj8SwwfOXu8Xg8Hs8Cw1fuHo/H\n4/EsMHzl7vF4PB7PAsNX7h6Px+PxLDB85e7xeDwezwLDV+4ej8fj8SwwfOXu8Xg8Hs8CY15U7iJy\neSf8jcLyHlFYPqgTfns8ntbooH4HUVj+80747fHs7rStcheR/UTkiyJyg70+QUSuEJGrRaR3itdX\ntiscGQ4Hjp7KURSWD++QfI9nQTBP9bsEvGoqR1FYPjwKy9KhMHg885K2Ve6q+pCqnujceq2qvh24\nDnhdu+RMN1hAK0p9bKcD4vHszsxj/W6FY5knvZQez1zRyQKfKt46YO/sQ9vyv0lEbqJzLftWK3fP\nHBOFZYnC8rIp3HRFYbmvBb/6o7BcbF/oPC0wH/R7OvjKfY5pQb+ntAHWXSEKywPtC9mfBnNR4FcB\nf8zeVNVrVPXVqvpq4NHZCLCZ35251wPkZuNvxr8+t2svCss9UVjOT/FOfxSW+9sVhgVGAXhX9mYm\nvf4C+KcW/DoJeOFkDmxe7HYNPWsAp2zg7ELmQr8lq0dW96YaDnCZ1BZEYbkUheWS67+1IZO90xuF\n5UW7Y7maCVkb2ALN9Nv1Y0kzN014LvDPU4Sta3dt4HeqjmjnmPsyEbkUOEJEVgPfEZFLgDcA35qp\nv67CTcJLMQbe5WTgSEBa9IMoLOeisFyY4PHpQNm5fjvwvDSME8h4r/2XlTNpeLLPM0Zn3LtWvkzm\npsk7xSgst7Vx12o6WyYyEu/NuBmTJ2m4s3EGCm5jy1YIbnjeC+zhPM/6O2kaN6NJPk3oZyt+pe4z\n7+2NKXu7lF2s3wOM16MjgXdYPwpRWJ6qIR9MIe8Y4NUZ/09O32lS3gD+FXgPsKd7swX9HqN70y17\nLboJOlDZnQ7s3wY/xvT0ZOIf2PTJO/qsGDuQ1bcu5/I44BWT+Lszzaej39k8n4mdcPwqpDYiU17H\n1RHtoJ1j7htV9RRVfZqqfhTTWn86MIJpeU0bayhXN7mfVeQVwLJMZVXCxE+A1VFYXuS+n/XDZuJf\nAcc5mSCZr/MxlYn9hw3j6gnC1kzmuDhlWG1buGPc2+tmct4NPMOJ/+r0uZUbZIxJjkbjp52snqxl\nb9MzDUcaPnHC2Szt9gD+0bl+B/BsTDrs6/h3GPB0axQCYC/Gp3MhlYXpFTg+E/YuG540jXeGd4Kw\nZeP7cuANVpFXuPKz8bN+u36utv712b9L7P2scREbxzn9WuyQfucxcQ0y94PMPbeMpM8KQFp5vRF4\n6RRlqZDmiWtcncZYF+DqemDvgcmbfwMOyvi7s/K04UrfXz1Fxfpm4MWpjSGj31FYXtakrLgVwuq0\nYnPslOu+gPnweLvj78642rC6aTWuLNn7boWY+p82rid8L3O9833nWbbecfX0+cDbgL8GjnHeKdh4\nu/580HmvAJTStIjC8tKMvycC6cqKMXo5STr8C2ZSdup2OWPLj+uPRJkPwyb6/SbgRRh7dhQTMIkt\nnBadHnPfgVGQcd12LXLGBPfPsgmdkibEy5uEIc2w97jv238uL8YYqVVW7lmYrt4zo7C8uEkYFMhW\nmmLD1szwNpOZFopm7k9v4tdZE/zuAV7L2ImBZzl/j2VsYToLWI41Xm2uKI6Y5NkzsQaHRtnbPwrL\nTwHOpEn6WNzxtn4ajapjMUZAgT5Mg+5M4G+BuIk/b4/C8sFWTg8N453yJsZ23z4XY4iZJGyuEe/B\n9O6sZrzyvjfjx76YCiNLOpb97sx1ylGYOP7FBOGZK9qh32fav9ny91rgL5u4Px2T39kVMN3AS4AD\n7PVZGN1x7dsJNLp2z7DPD6BhY/bN+Jnqdxq2EiavXZ1MHPfHYmxFWjZ3Nl6b6FcXJh/fAzzNusvR\nKB95Kycti6+lYZNSPmgbEGcBhwCnWX/2sW5fRaOn6u8iM8RwhrVlx2PsHTZNDmU8ZwEfAt4XmW7j\nVPYbMI2ps8h8HGTigOM+9e8d9vdk9qYLM9xSAg7G6JKrA2PsqNMIVuAZmHJzFg0bkbIc58ve5ku3\n49dpjO+V6LH/UlI/z2C87h7J+PoqmxZLMHnbj6nkx2EbJWcxtr6aEZOOGc+Sn6rq7SKyJ/Apxn59\nISInYBQOWpxwE4XlFwCvtJcnAx9JvbN/X2DdXGavx7Seo7CcD6NK3ble4zweaPLOKvv3NPv3rVFY\nBvgMJoNewtgK4pz0bxSWfzFFXM4Ko8q59h2NwvLHMV9C+2Wcpl3/a5x319i/ZwOfdtweFoXlw5rE\n7zDgF5n4ArwsMl1Lfx6F5f8EXhlGlaybyeKwBjg3jCpxFJafZm8vduR8DPjABO+lrKJhaNLnrwae\n49xaat+5BJM/ab70MvarMe0mW0Lja28P4FTHTfq1fiimpb8GSMvESus/UVj+G+v3iBvuNH2isPx6\nR+ao/f1MR84hmXd6msQ99fcE5zLb4HaHW9x3XxyF5UVhVJlxl/gsaad+B0AcmUlTrtH8Ueqd/duN\nyZMHgG2Ou7QSy85NODVqTNjag/G8Csak6yH29400KoJzHPfpqoBzrB1ww5/qXToP4APAGkyjbjAK\nyz8HfmPvP+GEN51T4lYE77R//ykKyw9gGlHYsLr69CH79+/ts38iU0HZ+KynUUkdDzwF2OzEu9d1\nn7EB3YzvNk4bQr2RWUY8ZrVRppwucq5X2L+vznycpe/cgfmg2oaxhSl/xSRk5KWV5vH22eswlX7d\nXqdlcRXwVue9fuAfo7D84TCqJJHZHyXH2DryZOd3ryP7I5nrbPjSuPTZcLhDd+7HoftuXxSWzw6j\nyoebx3pqRLXV1SQzFCBSBK5R1ddP4uYmO/FmDE0SqoDJBDAFvorJgHQy3Qgm8Zrdc1mGybQEU/Db\nRQCkX/lDOBVDC0RAaH930wj/Zlpf8jMb7gSehVGstKGzNowq42RHYflMTIE+G/ghE681nmq1wnJM\nXsXAhsyzbvscTBrsoDl5J7xpmXB5DNNF3ypDGEPo5kfCxL1cnwujyvpmSj0BaXqk6Xo/5gtlKmKa\nTAqbTmOsE7RZvwXTMAOjO0P2XgQcCNQwDSmhUSFVGZ/nAzT0ZwPNe3FmymJMWajSKJOjNBqWEzFk\n31nMWDu2HROvTuPqNYzVy6yOBzQaA6OY8HXRaKRtY7y+NmOZ9UeBxzH5kMrK02ikD9Jo8GQJrFyx\n7ja1IHcyHsT0ltxEY47FZDbitjCq3BaF5fcz9it+IrL6fQmNxtpkVMl8jMLs9LtjlbuIvA5j9BcD\nl6jqbZO4bar8BxUKTwRIrkekD5C8UFDbFaZKfUh1e7dIf0FMS6iqjGxOkicXB8HyLpEeAAGJ0bqA\njipDimqfBItjqAWQF0AhSSCuo9UCUizITkVVWztpXakqaIxWB1W3FpHu3kAGABIlVqgXhK4cUlC0\nnkMKMdQFGFHdHqP1bpH+AAmw9+potYj05MRUHHWlCmhfIEsSJQmEIIBAQcWUGRGQqupgDLUcFOy7\nMqy6va47vyDJQVAQ6bJpsJMeES2IJHnQvAg50IMKhZHlQVBfGgTDLyh1fYtGwfx4GFWG0nedVjDA\nMMaA3gUc8Yd67S8erNf3HVEN/mt0tLo5SXYEQkmglEA+VhbHqARITiAnIApaR2s5pCCQU9AErcdK\nXBAp1dGaqgnLkOpWtVERkw45gC6RPvtuEgi5muqoArFSN/4Rx2icQ/J5oagmbfIY+dh8TXRnfqsO\nqm6Plbgk0i0Q5IVCAekGsP7V0wRKw6egOaFQRLoVtIZSMwHWIZP/tQEJluSEnDTkxqbsaT1AckWT\nXmleI6a81dSUw3hEdUcOyXUJ3YKQoPX/GR1t9kXacdqh3wcUCk8C0iPSF0AuQHIBSAKxQLBDk81F\nka4ukR6bzro+iaMeCfr6A1mikASQi5UaglRVhxOo94osSv2w+hMraE0ZzUGuKCaPwKQxQIzWE2Nb\nku2JbslBri+QxdY+aKzUc0K+JNIVK/VAyCdKXUCq6HBNGS0JPQXT2KGqOjSiOlQUKdnyRmw+MpIu\nE99AgTyS14a+EYDU0WpdqQkgYnS+poyOqg4LO/VZCiKlwNqERlKbC4HA/BPUmDECRMNcrloSoV9k\naVGEogh9IhtjqA+pbtqUJJUlQXBgj8jesWryZJIUdmgSx0p+SHV4VDWpo6VNSbLJBM/I7RbpE2Pb\ndupoDvJVdCRAgpytuKxujOSRotX/ahHp2q7Jzkrb6ncASAEp5oRiAvUC0lVDR7H5YfVWY7QeQFBA\nuhTIC0VrZ6WGjibWFmDt+ajqUA0dLSLdOSEvEBTF6DeK1huNrZ35oigBEpREehRTa6b5NmryerBX\ngkVFMY3KGGp1pYa1cWLi0u3kFam7GI0BHVYdBLQk0pO6+8XoaLMh4ZboZLf8D4G/w7TywincNmXN\nwOIaUB9U/c2I6hZF42HVzQFSXJELjsQY6C1Pxsn/5oRCmMu9DtMalEfj+MZh1S1FpE+EfBeytCSy\nVFFGlHu2JskjQK4gdGEK5x5pJbE1SSoJJDaBc3mRYhHpA3JLguDpAv0JJFuS5Od1dLhgZOQU2JIk\n6xIlzgtFgUKXyECPyFMUqMMfB5Pk8aLIQK9IiLm3fUR1QwD5vJFBFX1wkwkfeSjagT21FVOhLwj2\nBIhV61UYKkFfbxDsvbOCMj/qVfSxWLUGpiAqMKiaq6tKDaSuGmxXzX9raGiRgtxVqz7tzwqF/Usi\nyeu7e9cd1dV1xNa9QhYFwV3A9diKfUT1uYNJsurxJO791PZt59QUvWV0RPbJ5bb0BUH90Hyh/6BC\nYXFgGk3bAxgtiNyfqG4BkioMiwl/1TRcIBCjjN22IQfEw6rbAghWBLlDRQhQ1NgqNEFrQJxAvDlJ\nHsYYzf4A8nmRfA66BLQoMmANRTKiukmhnkC1rlrNiRRLSI9AgLFM0iWyoktkLzEGZLiKPimgg4k+\nGqPDRWOYC+LYUTFGVQAGk+SPdagWRZIiol3CsoEgOFSBmurjT8TJ/xOh0COyp4LkoU8gr6aBtqFm\nGmiixt9CUehXky5hgHSZ8quP1WBosQTPnoletYlZ6/f5ixZXAYmVdds0uV+BYZNHukeQe05gv5S2\navJ/w4luKudzx+aQMsDmJPnppiR+qCDSnYdSXqSn2+pUrDy4IYnvVaBod8/rElkS2CE5XfxHAAAN\nbUlEQVS0IdUn6qojtrEoAeRKIgMKMhAEBxWQRYomO1R/Pay6Pg/dOaRk9Cd5vKo6FJgyVsohpf5A\nnioQJFDfmiQP5ZDSokAOwFTg24dVnwQ0j/Qbd1rbkCT3J6bsB6khtx8SQZ/IniBiK5E4gEJvEOyN\nadBq2qCsoZvqyhBGt1VR7MePKiR10+jBln8Gk6T4RJJ0V1Vz65LkoR2ajFSV3prqkkBgWRDsm0f2\n3xDHbEiSB0ZVBxcFQXV5EIwUROJFIiNFiPfN5w/uFgls5ZaYSpqHB5PkibQSBajDaKy6s9dUEO0W\nWZQXbFrqelV0eS44KA/dqf2y4Y8TqAno1iRZV4PRInTnRboD00DrwyRIKbDd3KPKlhgdTaA2oro1\ngFyPBMsFAhHT2ClAb48EZaCoUB/WZJ2CVpXNNXQwbyr8UqYS3qnnw6obq6qDAkEgEhShZ3EQHCIm\nr594LI5/rhD3SbCnCEEeunKYj6wqur2mus3xVEoiS+zfZQWRJQB11Y1VZdOiQJ7BLOjkl/sbgS2q\neqOIXKeqx03idqpuu4eBfSYRl3Zp7Gw5MruuuLRrdjJ+QPOxoLWMHaebikEmXq+7Efg245f5tZsR\noOvuavVpD9RrK6M4XnTFju3PjEF2qOYOzheqaQtdUdbFcSGBegD6wmLpN88vle5bGgQjb+jpPYOx\n41ITsZ7mY6BZ3K7xqWjWdX4PzmzXDnAxZkbtTPg2ZqLUbOgKo8oHp3bWftqs363wS8wqiXZwC1NP\nSNxCY4jNZbr6PSlhVFnTJB3+SJONgXYBF2BmpN8bRpVvQMt59jANe92Ocp7Sqt1I2UBjaG8iaoyf\nfJeSLXMPM3k95PIAjQmeM2VW+t3JL/e9McYVZlfR/gQzEWGfVBHs30Mwaxt/gpmo8R84k3GsmzMx\ncfzfMKp8DyAKyyHw5jCqpJPxiMwSmXdixj8fw0y26gG+SWOyyI4wqnwyCssHYiYKjWAnjLnhsv4B\nXBhGla32Oo+ZGXwLpvFxlONWgHOsHydjx7XcsZYoLD8CPNVe/gDTFb7avnM8Zsz2K5hZrNfZdAEz\njvUTzIzV88KoUo/C8r7YGeCOYbkbKBy9/on0+vbT+gfuAD7021rtqevi+goxPRkJkAxIMPjcUun3\njB1T/0kYVR51Jhp9CXi9zbvHwqhymfX7lzYO6czSn9KYULcBuDmMKr9x4v5OzLKrdBb+7+0/t2F1\nJ7AijCpfiMLy6cA3w6jyBzsLP63crwmjyu+sn2sw4/h30piV/S1Mo+41wCLH7z8A94VR5X+de6kf\nVYzBSWisdb44jCobrBuhURHcGEaVX2b8SI3ej4EXhlHlY47fOGXkBMyY848xZfRfgOvcdNoFtEu/\nr8PomABXYBqHlwFPYtLuNuBlYVS5MQrLrqH9FGbZ1AuAWhhVzk8fNJkUlqbp1zFDCekYfTr+uh7Y\nw+pDgJlLsiWMKp927E1Wv28No8rtGf83hVHlM1n5jh/PpTHT/6owqjxif/87phcEAFuO03cOAv7B\niYrbuPghZgJuF3B3Wr4yFfCXMLbq7WFUOTcKy6cB3wujygOZsLvlLQDODqPKCGZS4Dgc27EOWOU2\nUsKo8iUnDBHwf5gJcrcBe4VR5euO7D/H5IE7ETA7EfdXwDPDqPK5KCy/HOgPo8q3M3G9L4wq19t7\nqd59mcZkwAfDqPIV6/6/aSyLux04JIwqV7jxi8LyccBW4GbM5NiVmF6qn4ZR5eZs2rnp5zw7CaMn\nDwAHOOn7F5jJ2J8Jo8qmKCwfipkUuSGMKhdbP3/rptNM6PSX+2ZVvUlErlXV4zPP3dm03arabOmL\nx+OZh3j99njmN52s3Hsx3ZYjwM9U9Wsz8ONmzMStTrKSWW6P6WV4GbtIBuyiitPrt5cxj2XMlZy5\nkDFj/e5Yt7yamX9vndLh5Aw3G6trJxONB3oZXsZ8l5HK6bSMZnj99jLmq4y5kjNXMmb6rj8pyePx\neDyeBcZ8r9yv8TK8DC9jXsjpBAslH7yM+SVjruTMaxkd36HO4/F4PB7P3DLfv9w9Ho/H4/FME1+5\nezwej8ezwOjkJjYtIyJHYjYbqKjq6fbeezGnDxWAUzAb+38Cs2HGVap6a9aNtjDGICLHYDZAWQyo\nqv6DiPwCs6HJI6r6URFZmZU1zfi8BbMpwTrgc6p6TyvxmaaMkzDHDC4FzlPVu9sdD0dWL/B5zOYf\nt81k2VMTP4/BbOgxAHwRc8ThI8AOVT29HTJF5GXAucC9wLWYpStHYQ75SA9zmK2MF2NORMtjNj/a\n2q54iMh+mM1+Fqnq6+3a8UnDn3VjZ7XvUrx+z2/9tvLaquMLRb+tnN1Tx1V1XvzDbOv3Sfu7CHzN\n/j4Vs4PZWZjdxgLMJINxbqYp7zTgVfb3rcCVwJvt9RhZM4jLm4AbMDturWglPrNItyOAszoRD0fG\nG4HX2N/XtTnfl2CU/5vA5cD72iUTeCnwfcwuXfsD37D3j7b+ty1ewDGY86o7EY8b7N8pw5910868\nmmX6eP2eWbp1XL/bVU4n8HdB6Lf1Y7fS8Tn/cheRw4GPZm6/LXO9jMZRrOswW/jtjWn5J2LOImrm\npiVZag5yeAXmXHaAv7T+XiciNzaRNe34qOrVIvIMzDaKn2whPtOWgTn+8F8x22XOKh5T0K6tRptx\nJvA54G4bzk/Z+LZDZvbM8fTLbx2N0+3aFa8TgBOBKzoQj5RWwt/MzZzh9Xu31G/onI4vFP2G3UzH\n57xyV9V7aJyjuxMR2ce53Ehjw/9VwK+xhymIyLZJ3LQq60XAz1XVHh+r9uA1NmOOMc3KmnZ8LE9i\nzm1uJT7TkiEiBUx3zadVtTLbeExBepDF3bRpnoYYa3QB8H1VvdN5lKbZrGVm0qOEPS4Ykwd/tL9n\nHS8RWQVsVdXtzu22xaMJrYTfdTNneP3eLfUb2lxOF5J+w+6p4/NlzP1AzEb/h4nIyap6uYjcKSIX\nYTLt85hDHi7AnIn8BVWtNnHTKm/DtobFHLN3EWYbzU2qGonIF11ZM4jPycCzMF8faycI65j4TFcG\n5nCFA4F32m08f9zueDh8C7hYRI4GbpyFPy7vAl4OLBKR/TEHOQxhyuTHMYdFzEqmjD1z/GJgpYhc\ngjGMp1pn7YjXicBVVuaX2xUPEVkGnA8cISKrge+0EP5mbnYpXr/nvX5D+3V8Iek37IY67te5ezwe\nj8ezwPBL4Twej8fjWWD4yt3j8Xg8ngWGr9w9Ho/H41lg+Mrd4/F4PJ4Fhq/c24yIvEVEbhKRi0Tk\nShHZT0RWish7mrhdIyJP70AYjheRq0Tk2c69WS+E3RWIyA0tuBHn9wEi8n4R2UdEPjnFe88WkVPa\nEU7PnwZev9uL1+/OMS+Wwi1ALlXVm+wyhwsxy3JCETkEsyToEeBq4EXAChG5BrOD1xGYdZOnYrYk\n7Mcsc7lfVa8SkU9g8mwd8FnMEooSkANOU9VYRPow21/eB5RE5A7MEoqrrDJsA36tqpeJyN2YHZ72\nwizneL4N+85tLEXkDMwa3n7MFpKftfFZDXwaOBh4DdAFrLFxeinwEJCo6vmOX68CjsesS/0Y8DIr\nc8D6e4wbZ+AB4FAROdM+XwsIsF1VzxSRe4GvYJbx/M6KOeX/t3c/ITaFYRzHv88oNbKQhZSNxIiN\nlchCWbCU8mdpQ1FI5E/MwmSQlYnEkKymRFZKKRTFUlZDiAVFpjREDTPxs3jekzO3cRtpaly/T03N\nufc973nPnXneP6d7nkPentIOLI+I4+X8tgI3yCxf84Hzkh6VTrm3+Z/TbBTHt+N70vPKfQJJ+kDm\nmq7MIv/xr0jqBx4A5yQ9KO+PAHPITgDgmqSDwJqIWAwMS9ojqYfMwDUX+AhML/sh6Qtwj7zX9i3w\nRNJJYCNwWdIO8v5TgAFJh0qb7pLBva5qbOmsVpK5lEeARcA+4CrwWNIzYLOkrWRnVc2Sb0nqBhpX\nLdvI7F57Jb0Hdpa6B8g82qPOuXwuTyQdI7NDtZe2zouIqcBbSSclPa8dY0GpG+CFpE7gKbCM7CRP\nA4f5dX/o18gMV2Z/xPHt+J7MvHKfQGVmP1xtl1SJr4AtEbGEXxmVADZJWhsRR4Bp5bXqgQBRfurl\n24CHks7Q3KdaHY1JDarMVt/K78PkSqF+jH5JXbVzWgh8JhNH1Kkco7HdjWXqhhrq7hpj32qfNuCm\npBu18tW51f0Y47V6HVMY3SEPkiuL9407mTXj+HZ8T2Ye3CfG9ohYTV6C6qpejIhV5CWumcBtcka7\nPyIuA+8i4gA5w73fWKGk/ohoL5fu3pAz996yPQPYJelrkzZdAXpKVqdxPT2qHPNHRJwiZ9UngKPA\neqAzIpYCfRFxgeywuoEVTaq8BFyMiEEyH3dfRFwEhoCbv9lnJCK6S/mzkU9omippd5PyVcfREREn\ngNlk7ulhMtf1gnIukCui100/CLPRHN9jc3xPIs5QZy0lIjYCg5LujPHedUkbatvTySeV/bdfujH7\nlzi+x8+Du7WUiGgjn551exxlO4Dvkl5OfMvM7G85vsfPg7uZmVmL8bflzczMWowHdzMzsxbjwd3M\nzKzFeHA3MzNrMR7czczMWsxP0fz66rnw5RIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f84c5cb8358>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "size = 25\n",
    "\n",
    "listt = ['1mer_DNM_context_golden_synonymous', '1mer_DNM_context_golden_non_synonymous',\n",
    "        'CpGmer_DNM_context_golden_synonymous', 'CpGmer_DNM_context_golden_non_synonymous',\n",
    "        '3mer_DNM_context_golden_synonymous', '3mer_DNM_context_golden_non_synonymous',\n",
    "        '5mer_DNM_context_golden_synonymous', '5mer_DNM_context_golden_non_synonymous',\n",
    "        '7mer_DNM_context_golden_synonymous', '7mer_DNM_context_golden_non_synonymous']\n",
    "\n",
    "NROW=5\n",
    "NCOL=2\n",
    "\n",
    "axs = []\n",
    "\n",
    "fig = plt.figure(figsize=(7,4))\n",
    "\n",
    "for row in range(0, NROW):\n",
    "    for col in range(0, NCOL):\n",
    "        axs.append(plt.subplot2grid((NROW, NCOL), (row, col)))\n",
    "\n",
    "cnt=0\n",
    "for ele in listt:\n",
    "    plot_HD_BINNED(ele, axs[cnt], cnt, size)\n",
    "    cnt+=1\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.savefig(path.join('figures', 'internal_exon_centered_DNM_context_binned_selection.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exons centered analysis, Goldmann 2018\n",
    "\n",
    "Mutation probabilities were computed genome-wide using DNM. They were used to compute the expected number of mutations based on a generalized k-mer model, with the method described in the **Exon-centered and intron-centered mutational analyses** section at Frigola et al. for the 3-mer case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get mutations\n",
    "mutations_file = mutations_folder + 'germinal_ultimate_dataset.bed.gz'\n",
    "mutations_df = pd.read_csv(mutations_file, sep=\"\\t\", header=None)\n",
    "mutations_df.columns = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type', 'class']\n",
    "mutations_df = mutations_df[['chr', 'start', 'end', 'ref', 'alt', 'sample', 'type']]\n",
    "mutations_df = mutations_df[mutations_df['type'] == 'subs']\n",
    "mutations_df = mutations_df[mutations_df['sample'] == 'Goldmann2018']\n",
    "mutations_df['ID'] = mutations_df.index + 1\n",
    "\n",
    "tumor_type = 'germinal'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>115.149</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.851</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       115.149            5113        5124.851   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.13  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    10.291883\n",
       "dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    9.966747\n",
       " dtype: float64, 0    10.617018\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 1\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>299.403</td>\n",
       "      <td>5113</td>\n",
       "      <td>5171.597</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       299.403            5113        5171.597   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.001  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 1\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    19.57128\n",
       "dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    19.362319\n",
       " dtype: float64, 0    19.780241\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>131.129</td>\n",
       "      <td>5113</td>\n",
       "      <td>5108.871</td>\n",
       "      <td>0.679</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       131.129            5113        5108.871   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.679  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -3.148808\n",
       "dtype: float64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.399596\n",
       " dtype: float64, 0   -2.898019\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-mer context dependency, 9 classes with CpG, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_CpG(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_CpGmer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/CpG_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding_CpG(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>346.288</td>\n",
       "      <td>5113</td>\n",
       "      <td>5124.712</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       346.288            5113        5124.712   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0              0.26  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_CpGmer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.382156\n",
       "dtype: float64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.215567\n",
       " dtype: float64, 0    3.548745\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>130.631</td>\n",
       "      <td>5113</td>\n",
       "      <td>5109.369</td>\n",
       "      <td>0.644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       130.631            5113        5109.369   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.644  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.779585\n",
       "dtype: float64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -3.05383\n",
       " dtype: float64, 0   -2.50534\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 3\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_signatures_DNM_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>343.417</td>\n",
       "      <td>5113</td>\n",
       "      <td>5127.583</td>\n",
       "      <td>0.201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       343.417            5113        5127.583   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.201  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 3\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.246441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.080358\n",
       " dtype: float64, 0    4.412524\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>129.603</td>\n",
       "      <td>5113</td>\n",
       "      <td>5110.397</td>\n",
       "      <td>0.595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       129.603            5113        5110.397   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.595  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -2.008441\n",
       "dtype: float64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -2.276826\n",
       " dtype: float64, 0   -1.740057\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 5\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>342.903</td>\n",
       "      <td>5113</td>\n",
       "      <td>5128.097</td>\n",
       "      <td>0.193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358       342.903            5113        5128.097   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.193  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 5\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.402703\n",
       "dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    4.233248\n",
       " dtype: float64, 0    4.572158\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 127\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>127</td>\n",
       "      <td>126.673</td>\n",
       "      <td>5113</td>\n",
       "      <td>5113.327</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           127       126.673            5113        5113.327   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0               0.5  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.258145\n",
       "dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0   -0.018615\n",
       " dtype: float64, 0    0.534905\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7-mer context dependency, non_synonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of mutations at the middle exon-centered sequences is 5598\n",
      "The total amount of mutations at the exonic sequences are 494\n",
      "The category initiator_codon_variant is found 1 times.\n",
      "The category missense_variant is found 336 times.\n",
      "The category splice_region_variant is found 2 times.\n",
      "The category stop_gained is found 25 times.\n",
      "The category synonymous_variant is found 130 times.\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "include = 'non_synonymous'\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "\n",
    "exons_centered_results = obs_exp_muts_excluding_kmer(mutations_df, exons_coords_df, signatures_file, tumor_type, \n",
    "                            middle_distance_threshold, k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results.to_csv('results/exons_centered_' + str(k) + 'mer_DNM_context_golden_' + str(include) + '.tsv', sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Significance through permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of mutations at the middle exon-centered sequences is 5598\n",
      "Observed at middle exons: 358\n",
      "Observed at flanking introns: 5113\n"
     ]
    }
   ],
   "source": [
    "k = 7\n",
    "middle_distance_threshold = 1000\n",
    "signatures_file = 'results/' + str(k) + 'mer_DNM_signatures_decomp_golden.txt'\n",
    "include = 'non_synonymous'\n",
    "\n",
    "(exons_centered_results, exonic_obs, intronic_obs) = stacked_sequence_level_permutation_excluding(mutations_df,\n",
    "                                    exons_coords_df, signatures_file, tumor_type, middle_distance_threshold,\n",
    "                                    k, include, tb_consequence_type, consequence_rank_dict)\n",
    "\n",
    "exons_centered_results = pd.DataFrame.transpose(exons_centered_results)\n",
    "exons_centered_results.to_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'),\n",
    "                              sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_exon_cnt</th>\n",
       "      <th>exp_exon_cnt</th>\n",
       "      <th>obs_intron_cnt</th>\n",
       "      <th>exp_intron_cnt</th>\n",
       "      <th>empirical_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>358</td>\n",
       "      <td>345.76</td>\n",
       "      <td>5113</td>\n",
       "      <td>5125.24</td>\n",
       "      <td>0.248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   obs_exon_cnt  exp_exon_cnt  obs_intron_cnt  exp_intron_cnt  \\\n",
       "0           358        345.76            5113         5125.24   \n",
       "\n",
       "   empirical_pvalue  \n",
       "0             0.248  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k = 7\n",
    "results = []\n",
    "RANDOMIZATION = 1000\n",
    "include = 'non_synonymous'\n",
    "    \n",
    "# Read the expected exon and intron mutation counts from permutations\n",
    "rand_output = pd.read_csv(path.join('results', 'exons_centered_' + str(k) + 'mer_permut_' + include + '_golden.tsv'), sep=\"\\t\")\n",
    "    \n",
    "# count the number of times the expected exon mutations is higher than the observed\n",
    "count = len(rand_output[rand_output['exon_Exp'] >= exonic_obs])\n",
    "    \n",
    "if count == 0:\n",
    "    count = 1 \n",
    "    \n",
    "results.append([exonic_obs, rand_output['exon_Exp'].mean(), intronic_obs, rand_output['intron_Exp'].mean(), count/RANDOMIZATION])\n",
    "    \n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.columns = ['obs_exon_cnt', 'exp_exon_cnt','obs_intron_cnt', 'exp_intron_cnt', 'empirical_pvalue']\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    3.540028\n",
       "dtype: float64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exonic_increase = (results_df[\"obs_exon_cnt\"]/results_df[\"exp_exon_cnt\"]-1)*100\n",
    "exonic_increase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0    3.370396\n",
       " dtype: float64, 0    3.70966\n",
       " dtype: float64)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rand_output['increase'] = (exonic_obs/rand_output['exon_Exp']-1)*100\n",
    "error = (st.stdev(rand_output['increase'].tolist())/np.sqrt(1000))\n",
    "(exonic_increase-error, exonic_increase+error)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
